{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "930f199b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/stirunag/falconframes_env/lib/python3.10/site-packages/spacy/util.py:910: UserWarning: [W095] Model 'en_core_sci_sm' (0.5.4) was trained with spaCy v3.7.4 and may not be 100% compatible with the current version (3.7.0). If you see errors or degraded performance, download a newer compatible model or retrain your custom model with the current spaCy version. For more details and available updates, run: python -m spacy validate\n",
      "  warnings.warn(warn_msg)\n",
      "/home/stirunag/falconframes_env/lib/python3.10/site-packages/spacy/language.py:2170: FutureWarning: Possible set union at position 6328\n",
      "  deserializers[\"tokenizer\"] = lambda p: self.tokenizer.from_disk(  # type: ignore[union-attr]\n"
     ]
    }
   ],
   "source": [
    "import argparse\n",
    "import sys\n",
    "import regex as re\n",
    "import io\n",
    "import gzip\n",
    "import json\n",
    "from bs4 import BeautifulSoup\n",
    "import lxml\n",
    "from collections import defaultdict\n",
    "from tqdm import tqdm\n",
    "import spacy\n",
    "import scispacy\n",
    "\n",
    "# Load SciSpacy model\n",
    "nlp = spacy.load(\"en_core_sci_sm\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f2dec99d-c14c-4a67-a394-c9d38cb4b215",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Precompile regex patterns for section tagging\n",
    "titleMapsBody = {\n",
    "    'INTRO': ['introduction', 'background', 'related literature', 'literature review', 'objective', 'aim ', 'purpose of this study', 'study (purpose|aim|aims)', r'\\d+\\. (purpose|aims|aim)', '(aims|aim|purpose) of the study', '(the|drug|systematic|book) review', 'review of literature', 'related work', 'recent advance'],\n",
    "    'METHODS': ['supplement', 'methods and materials', 'method', 'material', 'experimental procedure', 'implementation', 'methodology', 'treatment', 'statistical analysis', \"experimental\", r'\\d+\\. experimental$', 'experimental (section|evaluation|design|approach|protocol|setting|set up|investigation|detail|part|perspective|tool)', \"the study\", r'\\d+\\. the study$', \"protocol\", \"protocols\", 'study protocol', 'construction and content', r'experiment \\d+', '^experiments$', 'analysis', 'utility', 'design', r'\\d+\\. theory$', \"theory\", 'theory and ', 'theory of '],\n",
    "    'RESULTS': ['result', 'finding', 'diagnosis'],\n",
    "    'DISCUSS': ['discussion', 'management of', r'\\d+\\. management', 'safety and tolerability', 'limitations', 'perspective', 'commentary', r'\\d+\\. comment'],\n",
    "    'CONCL': ['conclusion', 'key message', 'future', 'summary', 'recommendation', 'implications for clinical practice', 'concluding remark'],\n",
    "    'CASE': ['case study report', 'case report', 'case presentation', 'case description', r'case \\d+', r'\\d+\\. case', 'case summary', 'case history'],\n",
    "    'ACK_FUND': ['funding', 'acknowledgement', 'acknowledgment', 'financial disclosure'],\n",
    "    'AUTH_CONT': [\"author contribution\", \"authors' contribution\", \"author's contribution\"],\n",
    "    'COMP_INT': ['competing interest', 'conflict of interest', 'conflicts of interest', 'disclosure', 'declaration'],\n",
    "    'ABBR': ['abbreviation'],\n",
    "    'SUPPL': ['supplemental data', 'supplementary file', 'supplemental file', 'supplementary data', 'supplementary figure', 'supplemental figure', 'supporting information', 'supplemental file', 'supplemental material', 'supplementary material', 'supplement material', 'additional data files', 'supplemental information', 'supplementary information', 'supplemental information', 'supporting information', 'supplemental table', 'supplementary table', 'supplement table', 'supplementary material', 'supplemental material', 'supplement material', 'supplementary video']\n",
    "}\n",
    "\n",
    "titleExactMapsBody = {\n",
    "    'INTRO': [\"aim\", \"aims\", \"purpose\", \"purposes\", \"purpose/aim\", \"purpose of study\", \"review\", \"reviews\", \"minireview\"],\n",
    "    'METHODS': [\"experimental\", \"the study\", \"protocol\", \"protocols\"],\n",
    "    'DISCUSS': [\"management\", \"comment\", \"comments\"],\n",
    "    'CASE': [\"case\", \"cases\"]\n",
    "}\n",
    "\n",
    "titleMapsBack = {\n",
    "    'REF': ['reference', 'literature cited', 'references', 'bibliography'],\n",
    "    'ACK_FUND': ['funding', 'acknowledgement', 'acknowledgment', 'acknowlegement', 'acknowlegement', 'open access', 'financial support', 'grant', 'author note', 'financial disclosure'],\n",
    "    'ABBR': ['abbreviation', 'glossary'],\n",
    "    'COMP_INT': ['competing interest', 'conflict of interest', 'conflicts of interest', 'disclosure', 'declaration', 'conflicts', 'interest'],\n",
    "    'SUPPL': ['supplementary', 'supporting information', 'supplemental', 'web extra material'],\n",
    "    'APPENDIX': ['appendix', 'appendices'],\n",
    "    'AUTH_CONT': ['author', 'contribution']\n",
    "}\n",
    "\n",
    "# Precompile regex patterns\n",
    "compiled_titleMapsBody = {\n",
    "    key: [re.compile(pattern, re.IGNORECASE) for pattern in patterns]\n",
    "    for key, patterns in titleMapsBody.items()\n",
    "}\n",
    "\n",
    "compiled_titleExactMapsBody = {\n",
    "    key: [pattern.lower() for pattern in patterns]\n",
    "    for key, patterns in titleExactMapsBody.items()\n",
    "}\n",
    "\n",
    "compiled_titleMapsBack = {\n",
    "    key: [re.compile(pattern, re.IGNORECASE) for pattern in patterns]\n",
    "    for key, patterns in titleMapsBack.items()\n",
    "}\n",
    "\n",
    "def createSecTag(soup, secType):\n",
    "    secTag = soup.new_tag('SecTag')\n",
    "    secTag['type'] = secType\n",
    "    return secTag\n",
    "\n",
    "# Function to read XML or GZ files and split into individual articles\n",
    "def getfileblocks(file_path, document_flag):\n",
    "    sub_file_blocks = []\n",
    "    if file_path.endswith('.gz'):\n",
    "        open_func = lambda x: gzip.open(x, 'rt', encoding='utf8')\n",
    "    else:\n",
    "        open_func = lambda x: open(x, 'r', encoding='utf8')\n",
    "\n",
    "    try:\n",
    "        with open_func(file_path) as fh:\n",
    "            content = fh.read()\n",
    "            if document_flag in ['f', 'a']:\n",
    "                # Split content by <!DOCTYPE article ...> or <article ...> tags\n",
    "                articles = re.split(r'(?=<!DOCTYPE article|<article(?![\\w-]))', content)\n",
    "                sub_file_blocks = [article.strip() for article in articles if article.strip() and '<!DOCTYPE' not in article]\n",
    "            else:\n",
    "                print('ERROR: unknown document type :' + document_flag)\n",
    "    except Exception as e:\n",
    "        print('Error processing file: ' + str(file_path))\n",
    "        print(e)\n",
    "\n",
    "    return sub_file_blocks\n",
    "\n",
    "# Function to split text into sentences using spaCy\n",
    "def sentence_split(text, sent_id):\n",
    "    sentences = []\n",
    "    doc = nlp(text)\n",
    "    for sent in doc.sents:\n",
    "        sentences.append(sent.text.strip())\n",
    "    return sent_id + len(sentences), sentences\n",
    "\n",
    "# Function to process nested tags and collect sentences\n",
    "def call_sentence_tags(ch, sent_id):\n",
    "    sentences = []\n",
    "    for gch in ch.children:\n",
    "        if isinstance(gch, str):\n",
    "            continue  # Skip strings directly under ch\n",
    "        if gch.name in ['article-title', 'title', 'subtitle', 'trans-title', 'trans-subtitle', 'alt-title', 'label', 'td', 'th']:\n",
    "            if gch.find('p', recursive=False):\n",
    "                sent_id, sub_sentences = call_sentence_tags(gch, sent_id)\n",
    "                sentences.extend(sub_sentences)\n",
    "            else:\n",
    "                text = gch.get_text(separator=' ', strip=True)\n",
    "                if text:\n",
    "                    sent_id, sents = sentence_split(text, sent_id)\n",
    "                    sentences.extend(sents)\n",
    "        elif gch.name in [\"sec\", \"fig\", \"statement\", \"div\", \"boxed-text\", \"list\", \"list-item\", \"disp-quote\", \"speech\",\n",
    "                          \"fn-group\", \"fn\", \"def-list\", \"def-item\", \"def\", \"ack\", \"array\", \"table-wrap\", \"table\",\n",
    "                          \"tbody\", \"thead\", \"tr\", \"caption\", \"answer\", \"sec-meta\", \"glossary\", \"question\", \"question-wrap\"]:\n",
    "            sent_id, sub_sentences = call_sentence_tags(gch, sent_id)\n",
    "            sentences.extend(sub_sentences)\n",
    "        elif gch.name == 'p':\n",
    "            sent_id, sub_sentences = process_p_tag(gch, sent_id)\n",
    "            sentences.extend(sub_sentences)\n",
    "        else:\n",
    "            text = gch.get_text(separator=' ', strip=True)\n",
    "            if text:\n",
    "                sent_id, sents = sentence_split(text, sent_id)\n",
    "                sentences.extend(sents)\n",
    "    return sent_id, sentences\n",
    "\n",
    "# Function to process paragraph tags\n",
    "def process_p_tag(gch, sent_id):\n",
    "    sentences = []\n",
    "    if not (len(gch.contents) == 1 and (not gch.contents[0].string) and (gch.contents[0].name in [\"ext-link\", \"e-mail\", \"uri\", \"inline-supplementary-material\", \"related-article\", \"related-object\", \"address\", \"alternatives\", \"array\", \"funding-source\", \"inline-graphic\"])):\n",
    "        text = gch.get_text(separator=' ', strip=True)\n",
    "        if text:\n",
    "            sent_id, sents = sentence_split(text, sent_id)\n",
    "            sentences.extend(sents)\n",
    "    return sent_id, sentences\n",
    "\n",
    "# Function to process the front section\n",
    "def process_front(front):\n",
    "    sent_id = 1\n",
    "    sections = {}\n",
    "    keywords = []\n",
    "    \n",
    "    if front.find('article-meta'):\n",
    "        art_meta = front.find('article-meta')\n",
    "        \n",
    "        for ch in art_meta.find_all(recursive=False):\n",
    "            if ch.name in ['title-group', 'supplement', 'supplementary-material', 'abstract', 'trans-abstract', 'kwd-group', 'funding-group']:\n",
    "                section_title = ch.name.upper()\n",
    "                \n",
    "                if section_title == 'KWD-GROUP':\n",
    "                    # Extract keywords as a list from kwd-group\n",
    "                    keywords = [kwd.text.strip() for kwd in ch.find_all('kwd')]\n",
    "                else:\n",
    "                    sent_id, sentences = call_sentence_tags(ch, sent_id)\n",
    "                    if sentences:\n",
    "                        sections.setdefault(section_title, []).extend(sentences)\n",
    "            else:\n",
    "                pass  # Ignore other tags\n",
    "    \n",
    "    return sent_id, sections, keywords\n",
    "\n",
    "# Function to process the back section\n",
    "def process_back(back, sent_id):\n",
    "    sent_id = 1\n",
    "    sections = {}\n",
    "    for ch in back.find_all(recursive=False):\n",
    "        if ch.name in ['sec', 'p', 'ack', 'alternatives', 'array', 'preformat', 'fig', 'fig-group', 'question-wrap',\n",
    "                 'question-wrap-group', 'list', 'table-wrap-group', 'table-wrap', 'display-formula',\n",
    "                 'display-formula-group', 'def-list', 'list', 'supplementary-material', 'kwd-group',\n",
    "                 'funding-group', 'statement', 'ref-list', 'glossary']:\n",
    "            # Sections with titles\n",
    "            if ch.name == 'ref-list':\n",
    "                sent_id, sentences = reference_sents(ch, sent_id)\n",
    "                if sentences:\n",
    "                    sections.setdefault('REF', []).extend(sentences)\n",
    "            else:\n",
    "                title = ch.find('title')\n",
    "                if title:\n",
    "                    section_title = title.get_text(separator=' ', strip=True).strip().upper()\n",
    "                else:\n",
    "                    section_title = ch.name.upper()\n",
    "                sent_id, sentences = call_sentence_tags(ch, sent_id)\n",
    "                if sentences:\n",
    "                    sections.setdefault(section_title, []).extend(sentences)\n",
    "        else:\n",
    "            pass  # Ignore other tags\n",
    "    return sent_id, sections\n",
    "\n",
    "# Function to process reference sentences\n",
    "def reference_sents(ref_list, sent_id):\n",
    "    sentences = []\n",
    "    for ch in ref_list.children:\n",
    "        if isinstance(ch, str):\n",
    "            continue  # Skip strings directly under ref_list\n",
    "        if ch.name == 'ref':\n",
    "            sub_text = ''\n",
    "            for gch in ch.children:\n",
    "                if isinstance(gch, str):\n",
    "                    continue\n",
    "                sub_text += \" \" + \" \".join([d.string for d in gch.descendants if d.string])\n",
    "            if sub_text:\n",
    "                sent_id, sents = sentence_split(sub_text, sent_id)\n",
    "                sentences.extend(sents)\n",
    "        elif ch.name in [\"sec\", \"fig\", \"statement\", \"div\", \"boxed-text\", \"list\", \"list-item\", \"disp-quote\", \"speech\",\n",
    "                         \"fn-group\", \"fn\", \"def-list\", \"def-item\", \"def\", \"ack\", \"array\", \"table-wrap\", \"table\",\n",
    "                         \"tbody\", \"caption\", \"answer\", \"sec-meta\", \"glossary\", \"question\", \"question-wrap\"]:\n",
    "            sent_id, sub_sentences = call_sentence_tags(ch, sent_id)\n",
    "            sentences.extend(sub_sentences)\n",
    "        else:\n",
    "            pass  # Ignore other tags\n",
    "    return sent_id, sentences\n",
    "\n",
    "# Function to match section titles to predefined section types\n",
    "def titleMatch(title, secFlag):\n",
    "    matchKeys = []\n",
    "    # Check if the flag is 'body' or 'back' and apply the respective mappings\n",
    "    if secFlag == 'body':\n",
    "        titleMaps = compiled_titleMapsBody\n",
    "        exactMaps = compiled_titleExactMapsBody\n",
    "    else:\n",
    "        titleMaps = compiled_titleMapsBack\n",
    "        exactMaps = {}\n",
    "\n",
    "    title_lower = title.lower().strip()\n",
    "    # Check exact matches first\n",
    "    for key, patterns in exactMaps.items():\n",
    "        if title_lower in patterns:\n",
    "            matchKeys.append(key)\n",
    "            break  # If exact match found, no need to check further\n",
    "\n",
    "    # If no exact match, check regex patterns\n",
    "    if not matchKeys:\n",
    "        for key, patterns in titleMaps.items():\n",
    "            if any(pattern.search(title_lower) for pattern in patterns):\n",
    "                matchKeys.append(key)\n",
    "\n",
    "    return ','.join(matchKeys) if matchKeys else None\n",
    "\n",
    "# Function to apply section tagging to the soup object\n",
    "def section_tag(soup):\n",
    "    # Add Figure sections\n",
    "    for fig in soup.find_all('fig', recursive=True):\n",
    "        if not fig.find_all('fig', recursive=True):\n",
    "            fig_tag = createSecTag(soup, 'FIG')\n",
    "            fig.wrap(fig_tag)\n",
    "    \n",
    "    # Add Table sections\n",
    "    for table in soup.find_all('table-wrap', recursive=True):\n",
    "        if not table.find_all('table-wrap', recursive=True):\n",
    "            table_tag = createSecTag(soup, 'TABLE')\n",
    "            table.wrap(table_tag)\n",
    "\n",
    "    # Process front section\n",
    "    if soup.front:\n",
    "        if soup.front.abstract:\n",
    "            secAbs = createSecTag(soup, 'ABSTRACT')\n",
    "            soup.front.abstract.wrap(secAbs)\n",
    "        if soup.front.find('kwd-group'):\n",
    "            secKwd = createSecTag(soup, 'KEYWORD')\n",
    "            soup.front.find('kwd-group').wrap(secKwd)\n",
    "\n",
    "    # Process body section\n",
    "    if soup.body:\n",
    "        for sec in soup.body.find_all('sec', recursive=False):\n",
    "            title = sec.find('title')\n",
    "            if title:\n",
    "                title_text = title.get_text(separator=' ', strip=True)\n",
    "                mappedTitle = titleMatch(title_text, 'body')\n",
    "                if mappedTitle:\n",
    "                    secBody = createSecTag(soup, mappedTitle)\n",
    "                    sec.wrap(secBody)\n",
    "    # Process back sections\n",
    "    if soup.back:\n",
    "        for sec in soup.back.find_all(['sec', 'ref-list', 'app-group', 'ack', 'glossary', 'notes', 'fn-group'], recursive=False):\n",
    "            if sec.name == 'ref-list':\n",
    "                secRef = createSecTag(soup, 'REF')\n",
    "                sec.wrap(secRef)\n",
    "            else:\n",
    "                title = sec.find('title')\n",
    "                if title:\n",
    "                    title_text = title.get_text(separator=' ', strip=True)\n",
    "                    mappedTitle = titleMatch(title_text, 'back')\n",
    "                    if mappedTitle:\n",
    "                        secBack = createSecTag(soup, mappedTitle)\n",
    "                        sec.wrap(secBack)\n",
    "\n",
    "# Function to process the body section\n",
    "def process_body(body, sent_id):\n",
    "    sent_id = 1\n",
    "    sections = {}\n",
    "    for ch in body.find_all(recursive=False):\n",
    "        if ch.name == 'p':\n",
    "            sent_id, sentences = process_p_tag(ch, sent_id)\n",
    "            sections.setdefault('BODY', []).extend(sentences)\n",
    "        elif ch.name in ['sec', 'ack', 'alternatives', 'array', 'preformat', 'fig', 'fig-group', 'question-wrap', 'list', 'table-wrap-group', 'table-wrap', 'display-formula', 'display-formula-group', 'def-list', 'list', 'supplementary-material', 'kwd-group', 'funding-group', 'statement']:\n",
    "            title = ch.find('title')\n",
    "            if title:\n",
    "                section_title = title.get_text(separator=' ', strip=True).strip().upper()\n",
    "            else:\n",
    "                section_title = ch.name.upper()\n",
    "            sent_id, sentences = call_sentence_tags(ch, sent_id)\n",
    "            if sentences:\n",
    "                sections.setdefault(section_title, []).extend(sentences)\n",
    "    return sent_id, sections\n",
    "\n",
    "# Main function to process each article and collect data\n",
    "def process_full_text(each_file):\n",
    "    # Replace body tag with orig_body to prevent BeautifulSoup from removing it\n",
    "    each_file = re.sub(r'<body(\\s[^>]*)?>', '<orig_body\\\\1>', each_file)\n",
    "    each_file = each_file.replace('</body>', '</orig_body>')\n",
    "    try:\n",
    "        xml_soup = BeautifulSoup(each_file, 'lxml')\n",
    "        # Remove extra html and body tags added by BeautifulSoup\n",
    "        if xml_soup.html:\n",
    "            xml_soup.html.unwrap()\n",
    "        if xml_soup.body:\n",
    "            xml_soup.body.unwrap()\n",
    "        if xml_soup.find('orig_body'):\n",
    "            xml_soup.find('orig_body').name = 'body'\n",
    "\n",
    "        # Extract attributes from the <article> tag\n",
    "        article_tag = xml_soup.find('article')\n",
    "        if article_tag:\n",
    "            open_status = article_tag.get('open-status', '')\n",
    "            article_type = article_tag.get('article-type', '')\n",
    "        else:\n",
    "            open_status = ''\n",
    "            article_type = ''\n",
    "\n",
    "        # Extract article IDs\n",
    "        article_ids = {}\n",
    "        for id_tag in xml_soup.find_all('article-id'):\n",
    "            id_type = id_tag.get('pub-id-type', 'unknown')\n",
    "            article_ids[id_type] = id_tag.text.strip()\n",
    "        if not article_ids:\n",
    "            print('No article IDs found')\n",
    "            return None\n",
    "\n",
    "        # Apply section tagging\n",
    "        section_tag(xml_soup)\n",
    "        \n",
    "        sent_id = 1\n",
    "        sections = {}\n",
    "        keywords = []\n",
    "\n",
    "        # Process sections under SecTag\n",
    "        for sec_tag in xml_soup.find_all('SecTag'):\n",
    "            sec_type = sec_tag.get('type', 'unknown').strip().upper()\n",
    "            if sec_type == 'KEYWORD':\n",
    "                # Extract keywords\n",
    "                keywords = [kwd.text.strip() for kwd in sec_tag.find_all('kwd')]\n",
    "                continue  # Skip further processing of keywords here\n",
    "            if sec_type not in sections:\n",
    "                sections[sec_type] = []\n",
    "            # Exclude nested 'SecTag's to avoid duplicate text\n",
    "            for nested_sec in sec_tag.find_all('SecTag', recursive=True):\n",
    "                nested_sec.extract()\n",
    "            sent_id, sentences = call_sentence_tags(sec_tag, sent_id)\n",
    "            sections[sec_type].extend(sentences)\n",
    "\n",
    "        # Process front section if not already processed\n",
    "        if xml_soup.article.find('front'):\n",
    "            sent_id, front_sections, front_keywords = process_front(xml_soup.article.find('front'))\n",
    "            for k, v in front_sections.items():\n",
    "                sections.setdefault(k, []).extend(v)\n",
    "            if front_keywords:\n",
    "                keywords.extend(front_keywords)\n",
    "\n",
    "        # Process body section if not already processed\n",
    "        if xml_soup.article.find('body'):\n",
    "            sent_id, body_sections = process_body(xml_soup.article.find('body'), sent_id)\n",
    "            for k, v in body_sections.items():\n",
    "                sections.setdefault(k, []).extend(v)\n",
    "        \n",
    "        # Process back section if not already processed\n",
    "        if xml_soup.article.find('back'):\n",
    "            sent_id, back_sections = process_back(xml_soup.article.find('back'), sent_id)\n",
    "            for k, v in back_sections.items():\n",
    "                sections.setdefault(k, []).extend(v)\n",
    "\n",
    "        # Remove empty sections\n",
    "        sections = {k: v for k, v in sections.items() if v}\n",
    "\n",
    "        return {\n",
    "            'article_ids': article_ids,\n",
    "            'open_status': open_status,\n",
    "            'article_type': article_type,\n",
    "            'keywords': keywords,\n",
    "            'sections': sections\n",
    "        }\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Error processing article: {e}\")\n",
    "        return None\n",
    "\n",
    "# Function to process each article and write to output file\n",
    "def process_each_article(each_file_path, out_file, document_flag):\n",
    "    files_list = getfileblocks(each_file_path, document_flag)\n",
    "    with open(out_file, 'w', encoding='utf-8') as out:\n",
    "        for each_file in tqdm(files_list, desc=\"Processing Articles\", disable=False):\n",
    "            if document_flag == 'f':\n",
    "                data = process_full_text(each_file)\n",
    "            else:\n",
    "                print('Document type not supported.')\n",
    "                continue\n",
    "            if data:\n",
    "                out.write(json.dumps(data) + '\\n')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "83c6382c-eec5-4af7-91d3-e261b6f4cbc4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output1.jsonl  patch-07-10-2024-0.xml.gz   sentenciser-Copy2.ipynb\n",
      "output2.jsonl  patch-28-01-2023-21.xml.gz  sentenciser-Copy3.ipynb\n",
      "outputa.jsonl  profile_stats\t\t   sentenciser.ipynb\n",
      "output.jsonl   sentenciser-Copy1.ipynb\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d3f67204",
   "metadata": {},
   "outputs": [],
   "source": [
    "# input_file =  'patch-28-01-2023-21.xml.gz' #'patch-07-10-2024-0.xml.gz'\n",
    "# output_file ='output.jsonl' \n",
    "# document_flag = 'f'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6f7ec2de",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Articles:   0%|                             | 0/1000 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Articles:   3%|▋                   | 32/1000 [00:02<01:16, 12.72it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error processing article: local variable 'sent_id' referenced before assignment\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Articles:   3%|▋                   | 34/1000 [00:03<01:48,  8.91it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n",
      "Error processing article: local variable 'sent_id' referenced before assignment\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Articles:   6%|█▎                  | 65/1000 [00:04<00:51, 18.33it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error processing article: local variable 'sent_id' referenced before assignment\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Articles:   6%|█▎                  | 65/1000 [00:04<01:06, 14.12it/s]\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mprocess_each_article\u001b[49m\u001b[43m(\u001b[49m\u001b[43minput_file\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moutput_file\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdocument_flag\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[2], line 388\u001b[0m, in \u001b[0;36mprocess_each_article\u001b[0;34m(each_file_path, out_file, document_flag)\u001b[0m\n\u001b[1;32m    386\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m each_file \u001b[38;5;129;01min\u001b[39;00m tqdm(files_list, desc\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mProcessing Articles\u001b[39m\u001b[38;5;124m\"\u001b[39m, disable\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m):\n\u001b[1;32m    387\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m document_flag \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[0;32m--> 388\u001b[0m         data \u001b[38;5;241m=\u001b[39m \u001b[43mprocess_full_text\u001b[49m\u001b[43m(\u001b[49m\u001b[43meach_file\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    389\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    390\u001b[0m         \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mDocument type not supported.\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "Cell \u001b[0;32mIn[2], line 344\u001b[0m, in \u001b[0;36mprocess_full_text\u001b[0;34m(each_file)\u001b[0m\n\u001b[1;32m    342\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m nested_sec \u001b[38;5;129;01min\u001b[39;00m sec_tag\u001b[38;5;241m.\u001b[39mfind_all(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSecTag\u001b[39m\u001b[38;5;124m'\u001b[39m, recursive\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m):\n\u001b[1;32m    343\u001b[0m         nested_sec\u001b[38;5;241m.\u001b[39mextract()\n\u001b[0;32m--> 344\u001b[0m     sent_id, sentences \u001b[38;5;241m=\u001b[39m \u001b[43mcall_sentence_tags\u001b[49m\u001b[43m(\u001b[49m\u001b[43msec_tag\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msent_id\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    345\u001b[0m     sections[sec_type]\u001b[38;5;241m.\u001b[39mextend(sentences)\n\u001b[1;32m    347\u001b[0m \u001b[38;5;66;03m# Process front section if not already processed\u001b[39;00m\n",
      "Cell \u001b[0;32mIn[2], line 103\u001b[0m, in \u001b[0;36mcall_sentence_tags\u001b[0;34m(ch, sent_id)\u001b[0m\n\u001b[1;32m     99\u001b[0m             sentences\u001b[38;5;241m.\u001b[39mextend(sents)\n\u001b[1;32m    100\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m gch\u001b[38;5;241m.\u001b[39mname \u001b[38;5;129;01min\u001b[39;00m [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msec\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfig\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstatement\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdiv\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mboxed-text\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlist\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlist-item\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdisp-quote\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mspeech\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    101\u001b[0m                   \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfn-group\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfn\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdef-list\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdef-item\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdef\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mack\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124marray\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtable-wrap\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtable\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    102\u001b[0m                   \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtbody\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mthead\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtr\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcaption\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124manswer\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msec-meta\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mglossary\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mquestion\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mquestion-wrap\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n\u001b[0;32m--> 103\u001b[0m     sent_id, sub_sentences \u001b[38;5;241m=\u001b[39m \u001b[43mcall_sentence_tags\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgch\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msent_id\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    104\u001b[0m     sentences\u001b[38;5;241m.\u001b[39mextend(sub_sentences)\n\u001b[1;32m    105\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m gch\u001b[38;5;241m.\u001b[39mname \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mp\u001b[39m\u001b[38;5;124m'\u001b[39m:\n",
      "Cell \u001b[0;32mIn[2], line 103\u001b[0m, in \u001b[0;36mcall_sentence_tags\u001b[0;34m(ch, sent_id)\u001b[0m\n\u001b[1;32m     99\u001b[0m             sentences\u001b[38;5;241m.\u001b[39mextend(sents)\n\u001b[1;32m    100\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m gch\u001b[38;5;241m.\u001b[39mname \u001b[38;5;129;01min\u001b[39;00m [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msec\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfig\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstatement\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdiv\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mboxed-text\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlist\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlist-item\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdisp-quote\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mspeech\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    101\u001b[0m                   \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfn-group\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfn\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdef-list\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdef-item\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdef\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mack\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124marray\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtable-wrap\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtable\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    102\u001b[0m                   \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtbody\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mthead\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtr\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcaption\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124manswer\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msec-meta\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mglossary\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mquestion\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mquestion-wrap\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n\u001b[0;32m--> 103\u001b[0m     sent_id, sub_sentences \u001b[38;5;241m=\u001b[39m \u001b[43mcall_sentence_tags\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgch\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msent_id\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    104\u001b[0m     sentences\u001b[38;5;241m.\u001b[39mextend(sub_sentences)\n\u001b[1;32m    105\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m gch\u001b[38;5;241m.\u001b[39mname \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mp\u001b[39m\u001b[38;5;124m'\u001b[39m:\n",
      "Cell \u001b[0;32mIn[2], line 106\u001b[0m, in \u001b[0;36mcall_sentence_tags\u001b[0;34m(ch, sent_id)\u001b[0m\n\u001b[1;32m    104\u001b[0m     sentences\u001b[38;5;241m.\u001b[39mextend(sub_sentences)\n\u001b[1;32m    105\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m gch\u001b[38;5;241m.\u001b[39mname \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mp\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[0;32m--> 106\u001b[0m     sent_id, sub_sentences \u001b[38;5;241m=\u001b[39m \u001b[43mprocess_p_tag\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgch\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msent_id\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    107\u001b[0m     sentences\u001b[38;5;241m.\u001b[39mextend(sub_sentences)\n\u001b[1;32m    108\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "Cell \u001b[0;32mIn[2], line 121\u001b[0m, in \u001b[0;36mprocess_p_tag\u001b[0;34m(gch, sent_id)\u001b[0m\n\u001b[1;32m    119\u001b[0m     text \u001b[38;5;241m=\u001b[39m gch\u001b[38;5;241m.\u001b[39mget_text(separator\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m \u001b[39m\u001b[38;5;124m'\u001b[39m, strip\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m    120\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m text:\n\u001b[0;32m--> 121\u001b[0m         sent_id, sents \u001b[38;5;241m=\u001b[39m \u001b[43msentence_split\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtext\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msent_id\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    122\u001b[0m         sentences\u001b[38;5;241m.\u001b[39mextend(sents)\n\u001b[1;32m    123\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m sent_id, sentences\n",
      "Cell \u001b[0;32mIn[2], line 80\u001b[0m, in \u001b[0;36msentence_split\u001b[0;34m(text, sent_id)\u001b[0m\n\u001b[1;32m     78\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21msentence_split\u001b[39m(text, sent_id):\n\u001b[1;32m     79\u001b[0m     sentences \u001b[38;5;241m=\u001b[39m []\n\u001b[0;32m---> 80\u001b[0m     doc \u001b[38;5;241m=\u001b[39m \u001b[43mnlp\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtext\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     81\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m sent \u001b[38;5;129;01min\u001b[39;00m doc\u001b[38;5;241m.\u001b[39msents:\n\u001b[1;32m     82\u001b[0m         sentences\u001b[38;5;241m.\u001b[39mappend(sent\u001b[38;5;241m.\u001b[39mtext\u001b[38;5;241m.\u001b[39mstrip())\n",
      "File \u001b[0;32m~/falconframes_env/lib/python3.10/site-packages/spacy/language.py:1049\u001b[0m, in \u001b[0;36mLanguage.__call__\u001b[0;34m(self, text, disable, component_cfg)\u001b[0m\n\u001b[1;32m   1047\u001b[0m     error_handler \u001b[38;5;241m=\u001b[39m proc\u001b[38;5;241m.\u001b[39mget_error_handler()\n\u001b[1;32m   1048\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1049\u001b[0m     doc \u001b[38;5;241m=\u001b[39m \u001b[43mproc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdoc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mcomponent_cfg\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m{\u001b[49m\u001b[43m}\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m  \u001b[38;5;66;03m# type: ignore[call-arg]\u001b[39;00m\n\u001b[1;32m   1050\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m   1051\u001b[0m     \u001b[38;5;66;03m# This typically happens if a component is not initialized\u001b[39;00m\n\u001b[1;32m   1052\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(Errors\u001b[38;5;241m.\u001b[39mE109\u001b[38;5;241m.\u001b[39mformat(name\u001b[38;5;241m=\u001b[39mname)) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01me\u001b[39;00m\n",
      "File \u001b[0;32m~/falconframes_env/lib/python3.10/site-packages/spacy/pipeline/lemmatizer.py:135\u001b[0m, in \u001b[0;36mLemmatizer.__call__\u001b[0;34m(self, doc)\u001b[0m\n\u001b[1;32m    133\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m token \u001b[38;5;129;01min\u001b[39;00m doc:\n\u001b[1;32m    134\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moverwrite \u001b[38;5;129;01mor\u001b[39;00m token\u001b[38;5;241m.\u001b[39mlemma \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m--> 135\u001b[0m             token\u001b[38;5;241m.\u001b[39mlemma_ \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlemmatize\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtoken\u001b[49m\u001b[43m)\u001b[49m[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m    136\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m doc\n\u001b[1;32m    137\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/falconframes_env/lib/python3.10/site-packages/spacy/pipeline/lemmatizer.py:243\u001b[0m, in \u001b[0;36mLemmatizer.rule_lemmatize\u001b[0;34m(self, token)\u001b[0m\n\u001b[1;32m    241\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m form:\n\u001b[1;32m    242\u001b[0m     \u001b[38;5;28;01mpass\u001b[39;00m\n\u001b[0;32m--> 243\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m form \u001b[38;5;129;01min\u001b[39;00m index \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[43mform\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43misalpha\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[1;32m    244\u001b[0m     forms\u001b[38;5;241m.\u001b[39mappend(form)\n\u001b[1;32m    245\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "process_each_article(input_file, output_file, document_flag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "aaaebc67",
   "metadata": {},
   "outputs": [],
   "source": [
    "ss = getfileblocks(input_file, document_flag)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "42eb2ac5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1000"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2a8cce26-b5b4-4f64-bf2a-a36047b8e5cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'article_ids': {'pmcid': '9878535',\n",
       "  'publisher-id': 'JIAPS-27-741',\n",
       "  'doi': '10.4103/jiaps.jiaps_84_22'},\n",
       " 'open_status': 'O',\n",
       " 'article_type': 'research-article',\n",
       " 'keywords': ['Angiotensin II receptor type 2 gene',\n",
       "  'congenital anomalies of the kidney and urinary tract',\n",
       "  'nephrogenic genes'],\n",
       " 'sections': {'ABSTRACT': ['Background: Congenital anomalies of the kidney and urinary tract (CAKUT) are a common cause of end-stage renal disease in children.',\n",
       "   'While certain nephrogenic genes have been incriminated in these malformations, data to identify the frequency of gene polymorphisms in Asian Indian children with CAKUT are scarce.',\n",
       "   'This study was done to identify the effect of polymorphisms in paired-box gene 2 (PAX2), bone morphogenetic protein (BMP)-4, angiotensin-converting enzyme (ACE), and angiotensin II receptor Type 2 (AGTR2) nephrogenic genes on the development of CAKUT.',\n",
       "   'Materials and Methods: In this prospective cohort study, 158 children <12 years old (86 cases with CAKUT and 72 age-matched controls) were analyzed.',\n",
       "   'DNA from both sets was extracted from peripheral blood using the Keygen DNA extraction kit, and single-nucleotide gene polymorphisms (SNPs) in PAX2, BMP-4, ACE, and AGTR2 nephrogenic genes were detected by polymerase chain reaction (PCR) using previously published primers and PCR conditions.',\n",
       "   'Results: The presence of A allele SNP for AGTR2 gene at rs3736556 was found to be significantly correlated with the development of ureteropelvic junction obstruction and vesicoureteral reflux (VUR) with the TT allelic genotype having a lower incidence of pelviureteric junction obstruction (odds ratio [OR] 0.18 [95% confidence interval [CI], 0.06–0.55], P = 0.01) and VUR (OR 0.31 [95% CI, 0.11–0.91], P = 0.03).',\n",
       "   'Furthermore, on substratification of the patients with the presence of the A allele of AGTR2, 24 out of 27 patients with scarring were found to harbor the D allele of the ACE gene, thus predisposing them to further renal damage.',\n",
       "   'Conclusion: This study points to early evidence in the implication of nephrogenic genes in development as well as predisposition to renal injury in Asian Indian patients with CAKUT.'],\n",
       "  'TABLE': ['Table 1',\n",
       "   'List of primer sequences used in the study',\n",
       "   'GENE and SNP code',\n",
       "   'Primer type',\n",
       "   \"DNA sequence 5'to 3'\",\n",
       "   'PAX2 (SNP code: rs4244341)',\n",
       "   'Normal primer',\n",
       "   'TCTGAGGGGTCAAAGGGACG',\n",
       "   'Mutant Primer',\n",
       "   'TCTGAGGGGTCAAAGGGACT',\n",
       "   'Common primer',\n",
       "   'AGGCTCGGCTGACTGTTTAT',\n",
       "   'BMP4 (SNP code: rs17563)',\n",
       "   'Normal primer',\n",
       "   'AGCATCCCTGAGAACGAGGC',\n",
       "   'Mutant Primer',\n",
       "   'AGCATCCCTGAGAACGAGGT',\n",
       "   'Common primer',\n",
       "   'TCGGCTAATCCTGACATGCT',\n",
       "   'AT2R (SNP code: rs3736556)',\n",
       "   'Normal primer',\n",
       "   'TATGTTAATTTGTTAGGTCA',\n",
       "   'Mutant primer',\n",
       "   'TATGTTAATTTGTTAGGTCT',\n",
       "   'Common primer',\n",
       "   'TGAAATGAGAAATGAACATC',\n",
       "   'SNP: Single-nucleotide gene polymorphism, BMP4: Bone morphogenetic protein 4, PAX2: Paired Box gene 2',\n",
       "   'Table 2',\n",
       "   'Demographic characteristics of patients with congenital anomalies of the kidney and urinary tract',\n",
       "   'Uropathy',\n",
       "   'Number of patients',\n",
       "   'Median age (months)',\n",
       "   'Sex',\n",
       "   'Male',\n",
       "   'Female',\n",
       "   'VUR',\n",
       "   '24',\n",
       "   '34 (22.5-69)',\n",
       "   '12 (50)',\n",
       "   '12 (50)',\n",
       "   'PUJO',\n",
       "   '27',\n",
       "   '29 (22-48)',\n",
       "   '19 (70.3)',\n",
       "   '8 (29.7)',\n",
       "   'PUV',\n",
       "   '30',\n",
       "   '24.5 (9.75-52.5)',\n",
       "   '30 (100)',\n",
       "   '0',\n",
       "   'EB',\n",
       "   '3',\n",
       "   '25 (5-84)',\n",
       "   '1 (33.3)',\n",
       "   '2 (66.7)',\n",
       "   'MCKD',\n",
       "   '2',\n",
       "   '48 (36-60)',\n",
       "   '2 (100)',\n",
       "   '0',\n",
       "   'Total',\n",
       "   '86',\n",
       "   '32 (13.5-60)',\n",
       "   '64 (74.4)',\n",
       "   '22 (25.6)',\n",
       "   'Data are expressed as n, median (IQR), or n (%).',\n",
       "   'VUR: Vesicoureteric reflux, PUJO: Pelviureteric junction obstruction, PUV: Posterior urethral valve, EB: Exstrophy bladder, MCDK: Multicystic dysplastic kidney, IQR: Interquartile range',\n",
       "   'Table 3',\n",
       "   'Study gene genotypes and analysis with congenital anomalies of the kidney and urinary tract cases versus controls',\n",
       "   'Gene',\n",
       "   'Allele',\n",
       "   'Case (%)',\n",
       "   'Control (%)',\n",
       "   'P',\n",
       "   'OR (95% CI)',\n",
       "   'P',\n",
       "   'BMP4',\n",
       "   'GG',\n",
       "   '17',\n",
       "   '20',\n",
       "   '0.69',\n",
       "   '1.00',\n",
       "   '0.32',\n",
       "   'GA',\n",
       "   '73',\n",
       "   '64',\n",
       "   '0.78 (0.34-1.77)',\n",
       "   '0.64',\n",
       "   'AA',\n",
       "   '10',\n",
       "   '16',\n",
       "   '1.60 (0.50-5.09)',\n",
       "   '0.42',\n",
       "   'PAX2',\n",
       "   'GG',\n",
       "   '29',\n",
       "   '38',\n",
       "   '0.77',\n",
       "   '1.00',\n",
       "   '0.48',\n",
       "   'GT',\n",
       "   '63',\n",
       "   '57',\n",
       "   '0.70 (0.36-1.39)',\n",
       "   '0.31',\n",
       "   'TT',\n",
       "   '8',\n",
       "   '5',\n",
       "   '0.53 (0.14-2.02)',\n",
       "   '0.35',\n",
       "   'AGTR2',\n",
       "   'AA',\n",
       "   '27',\n",
       "   '38',\n",
       "   '0.95',\n",
       "   '1.00',\n",
       "   '0.17',\n",
       "   'AT',\n",
       "   '51',\n",
       "   '38',\n",
       "   '1.17 (0.67-2.04)',\n",
       "   '0.57',\n",
       "   'TT',\n",
       "   '22',\n",
       "   '24',\n",
       "   '0.70 (0.20-2.44)',\n",
       "   '0.58',\n",
       "   'ACE',\n",
       "   'II',\n",
       "   '59',\n",
       "   '40',\n",
       "   '0.79',\n",
       "   '1.00',\n",
       "   '0.20',\n",
       "   'ID',\n",
       "   '29',\n",
       "   '54',\n",
       "   '0.52 (0.25-1.09)',\n",
       "   '0.08',\n",
       "   'DD',\n",
       "   '12',\n",
       "   '6',\n",
       "   '0.81 (0.34-1.90)',\n",
       "   '0.62',\n",
       "   'BMP4: Bone morphogenetic protein 4, PAX2: Paired Box gene 2, OR: Odds ratio, CI: Confidence interval, ACE: Angiotensin-converting enzyme, AGTR2: Angiotensin II receptor type 2',\n",
       "   'Table 4',\n",
       "   'Angiotensin II receptor type 2 genotype analysis and associations with different variants of congenital anomalies of the kidney and urinary tract patients versus controls',\n",
       "   'Phenotype',\n",
       "   'Allele',\n",
       "   'Case (%)',\n",
       "   'Control (%)',\n",
       "   'P',\n",
       "   'OR',\n",
       "   'P',\n",
       "   'PUJO',\n",
       "   'AA',\n",
       "   '74',\n",
       "   '40',\n",
       "   '0.04*',\n",
       "   '1.00',\n",
       "   '0.06',\n",
       "   'AT',\n",
       "   '8',\n",
       "   '6',\n",
       "   '0.72 (0.12-4.34)',\n",
       "   '0.72',\n",
       "   'TT',\n",
       "   '18',\n",
       "   '54',\n",
       "   '0.18 (0.06-0.55)',\n",
       "   '0.01*',\n",
       "   'PUV',\n",
       "   'AA',\n",
       "   '50',\n",
       "   '40',\n",
       "   '0.33',\n",
       "   '1.00',\n",
       "   '0.38',\n",
       "   'AT',\n",
       "   '10',\n",
       "   '6',\n",
       "   '0.60 (0.24-1.46)',\n",
       "   '0.25',\n",
       "   'TT',\n",
       "   '40',\n",
       "   '54',\n",
       "   '1.45 (0.28-7.33)',\n",
       "   '0.65',\n",
       "   'VUR',\n",
       "   'AA',\n",
       "   '58',\n",
       "   '40',\n",
       "   '0.03*',\n",
       "   '1.00',\n",
       "   '0.06',\n",
       "   'AT',\n",
       "   '17',\n",
       "   '6',\n",
       "   '2.07 (0.45-9.52)',\n",
       "   '0.35',\n",
       "   'TT',\n",
       "   '25',\n",
       "   '54',\n",
       "   '0.31 (0.11-0.92)',\n",
       "   '0.03*',\n",
       "   'OR: Odds ratio, PUJO: Pelviureteric junction obstruction, PUV: Posterior urethral valve, VUR: Vesicoureteric reflux',\n",
       "   'Table 5',\n",
       "   'Sub-stratification of all patients with the presence of into A allele in angiotensin II receptor type 2 (76), i.e., all patients with proposed increased predisposition to development of congenital anomalies of the kidney and urinary tract and presence or absence of angiotensin-converting enzyme insertion/deletion polymorphism',\n",
       "   'No Scar (49) (%)',\n",
       "   'Scars (27) (%)',\n",
       "   'D allele (ACE I/D) present',\n",
       "   '36 (73.5)',\n",
       "   '24 (88.9)',\n",
       "   'D allele (ACE I/D) absent',\n",
       "   '13 (26.5)',\n",
       "   '3 (11.1)',\n",
       "   'ACE I/D: Angiotensin-converting enzyme insertion/deletion'],\n",
       "  'ACK_FUND': ['Acknowledgment',\n",
       "   \"The authors are grateful to Prof. Balram Bhargava-Secretary (Director General-Indian Council of Medical Research (ICMR)) and Dr. R S Dhaliwal (Scientist 'G at ICMR) for providing support for this project.\",\n",
       "   'The authors are thankful to ICMR for funding the project.'],\n",
       "  'REF': ['R EFERENCES 1.',\n",
       "   'Woolf AS Lopes FM Ranjzad P Roberts NA Congenital disorders of the human urinary tract: Recent insights from genetic and molecular studies Front Pediatr 2019 7 136 31032239 2.',\n",
       "   'Dastgiri S Stone DH Le-Ha C Gilmour WH Prevalence and secular trend of congenital anomalies in Glasgow, UK Arch Dis Child 2002 86 257 63 11919098 3.',\n",
       "   'Harambat J van Stralen KJ Kim JJ Tizard EJ Epidemiology of chronic kidney disease in children Pediatr Nephrol 2012 27 363 73 21713524 4.',\n",
       "   'Izquierdo L Porteous M Paramo PG Connor JM Evidence for genetic heterogeneity in hereditary hydronephrosis caused by pelvi-ureteric junction obstruction, with one locus assigned to chromosome 6p Hum Genet 1992 89 557 60 1634233 5.',\n",
       "   'Fryns JP Kleczkowska A Moerman P Vandenberghe K Hereditary hydronephrosis and the short arm of chromosome 6 Hum Genet 1993 91 514 5 6.',\n",
       "   'Woolf AS Environmental influences on renal tract development: A focus on maternal diet and the glucocorticoid hypothesis Klin Padiatr 2011 223 Suppl 1 S10 7 21472672 7.',\n",
       "   'Li ZY Chen YM Qiu LQ Chen DQ Hu CG Xu JY Prevalence, types, and malformations in congenital anomalies of the kidney and urinary tract in newborns: A retrospective hospital-based study Ital J Pediatr 2019 45 50 30999930 8.',\n",
       "   'Harris J Robert E Källén B Epidemiologic characteristics of kidney malformations Eur J Epidemiol 2000 16 985 92 11421481 9.',\n",
       "   'Raatikainen-Ahokas A Hytönen M Tenhunen A Sainio K Sariola H BMP-4 affects the differentiation of metanephric mesenchyme and reveals an early anterior-posterior axis of the embryonic kidney Dev Dyn 2000 217 146 58 10706139 10.',\n",
       "   'Michos O Gonçalves A Lopez-Rios J Tiecke E Naillat F Beier K Reduction of BMP4 activity by gremlin 1 enables ureteric bud outgrowth and GDNF/WNT11 feedback signalling during kidney branching morphogenesis Development 2007 134 2397 405 17522159 11.',\n",
       "   'Miyazaki Y Oshima K Fogo A Ichikawa I Evidence that bone morphogenetic protein 4 has multiple biological functions during kidney and urinary tract development Kidney Int 2003 63 835 44 12631064 12.',\n",
       "   'Reis GS Simões E Silva AC Freitas IS Heilbuth TR Marco LA Oliveira EA Study of the association between the BMP4 gene and congenital anomalies of the kidney and urinary tract J Pediatr (Rio J) 2014 90 58 64 24131739 13.',\n",
       "   'Quinlan J Lemire M Hudson T Qu H Benjamin A Roy A A common variant of the PAX2 gene is associated with reduced newborn kidney size J Am Soc Nephrol 2007 18 1915 21 17513325 14.',\n",
       "   'de Miranda DM Dos Santos Júnior AC Dos Reis GS Freitas IS Carvalho TG de Marco LA PAX2 polymorphisms and congenital abnormalities of the kidney and urinary tract in a Brazilian pediatric population: Evidence for a role in vesicoureteral reflux Mol Diagn Ther 2014 18 451 7 24633556 15.',\n",
       "   'Wang F Fang Q Yu N Zhao D Zhang Y Wang J Association between genetic polymorphism of the angiotensin-converting enzyme and diabetic nephropathy: A meta-analysis comprising 26,580 subjects J Renin Angiotensin Aldosterone Syst 2012 13 161 74 21810896 16.',\n",
       "   'Yu ZY Chen LS Zhang LC Zhou TB Meta-analysis of the relationship between ACE I/D gene polymorphism and end-stage renal disease in patients with diabetic nephropathy Nephrology (Carlton) 2012 17 480 7 22385293 17.',\n",
       "   'Qin YH Zhou TB Su LN Lei FY Huang WF Zhao YJ Association between ACE polymorphism and risk of IgA nephropathy: A meta-analysis J Renin Angiotensin Aldosterone Syst 2011 12 215 23 21357308 18.',\n",
       "   'Yong D Qing WQ Hua L Kan JJ Xi CJ Jin QQ Association of angiotensin I-converting enzyme gene insertion/deletion polymorphism and IgA nephropathy: A meta-analysis Am J Nephrol 2006 26 511 8 17124384 19.',\n",
       "   'Zhou TB Qin YH Su LN Lei FY Huang WF Zhao YJ The association between angiotensin-converting enzyme insertion/deletion gene variant and risk of focal segmental glomerulosclerosis: A systematic review and meta-analysis J Renin Angiotensin Aldosterone Syst 2011 12 624 33 21652690 20.',\n",
       "   'Zhou TB Ou C Qin YH Su LN Lei FY Huang WF Association of angiotensin converting enzyme insertion/deletion gene polymorphism with idiopathic nephrotic syndrome susceptibility in children: A meta-analysis J Renin Angiotensin Aldosterone Syst 2011 12 601 10 21946394 21.',\n",
       "   'Rigat B Hubert C Alhenc-Gelas F Cambien F Corvol P Soubrier F An insertion/deletion polymorphism in the angiotensin I-converting enzyme gene accounting for half the variance of serum enzyme levels J Clin Invest 1990 86 1343 6 1976655 22.',\n",
       "   'Rigoli L Chimenz R di Bella C Cavallaro E Caruso R Briuglia S Angiotensin-converting enzyme and angiotensin type 2 receptor gene genotype distributions in Italian children with congenital uropathies Pediatr Res 2004 56 988 93 15470205 23.',\n",
       "   'Bajpai M Pratap A Somitesh C Tyagi J Angiotensin converting enzyme gene polymorphism in Asian Indian children with congenital uropathies J Urol 2004 171 838 40 14713838 24.',\n",
       "   'Nishimura H Yerkes E Hohenfellner K Miyazaki Y Ma J Hunley TE Role of the angiotensin type 2 receptor gene in congenital anomalies of the kidney and urinary tract, CAKUT, of mice and men Mol Cell 1999 3 1 10 10024874 25.',\n",
       "   'Hiraoka M Taniguchi T Nakai H Kino M Okada Y Tanizawa A No evidence for AT2R gene derangement in human urinary tract anomalies Kidney Int 2001 59 1244 9 11260384 26.',\n",
       "   'Miranda DM Dos Santos AC Jr. Sarubi HC Bastos-Rodrigues L Rosa DV Freitas IS Association of angiotensin type 2 receptor gene polymorphisms with ureteropelvic junction obstruction in Brazilian patients Nephrology (Carlton) 2014 19 714 20 24995698'],\n",
       "  'TITLE-GROUP': ['Early Evidence on Genetic Polymorphisms in Conferring A “Two-Hit” Propensity to Renal Injury in Asian Indian Children'],\n",
       "  'I NTRODUCTION': ['I NTRODUCTION',\n",
       "   'Congenital anomalies of the kidney and urinary tract (CAKUT) are considered a major cause of chronic renal failure in young children and infants.',\n",
       "   '[ 1 ] This disease constellation constitutes approximately 20%–30% of all anomalies identified in the prenatal period and has a reported incidence of 0.3–1.6/1000 live and stillborn infants.',\n",
       "   '[ 2 ] Because CAKUTs play a causative role in 30%–50% of cases of end-stage renal disease (ESRD) in children, it is vital to ensure early diagnosis and initiation of therapy to minimize renal parenchymal damage and potentially delay the onset of ESRD.',\n",
       "   '[ 3 ] CAKUT has also been hypothesized to have a common underlying genetic cause.',\n",
       "   '[ 4 5 ] It has also been reported that those children with CAKUT and polymorphism in nephrogenic genes have a propensity to develop type II diabetes and metabolic syndromes later in life.',\n",
       "   '[ 6 ]',\n",
       "   'The aim was to study the prevalence of Bone morphogenetic protein 4 (BMP-4), Paired Box gene 2 (PAX2), Angiotensin-converting enzyme (ACE), and Angiotensin II Receptor Type 2 (AGTR2) gene polymorphism in Indian children with congenital uropathies and their association with renal outcomes.',\n",
       "   'The presence of both an allele predisposing to the development of a certain urological anomaly and also an allele conferring renal damage, as shown by scarring, could be useful as a pointer to the patients who are in need of a closer follow-up right from the inception of the disease.',\n",
       "   'Data from genetic studies to identify the frequency of gene polymorphisms in Indian children with CAKUT are scarce, and this study was done to identify gene polymorphisms as risk factors for renal parenchymal damage in patients with congenital urological abnormalities.'],\n",
       "  'M ATERIALS AND M ETHODS': ['M ATERIALS AND M ETHODS',\n",
       "   'In this prospective cohort study, medical records of 86 children <12 years old with congenital uropathies who presented to the pediatric urology clinic at our institute were reviewed.',\n",
       "   'Patients not willing for follow-up, syndromic children with other congenital anomalies, and those who refused consent were excluded from the study.',\n",
       "   'A standardized data collection form developed for reviewing records was used.',\n",
       "   'Patient history and investigations, including serum creatinine, renal ultrasound, dimercaptosuccinic acid (DMSA) radionuclide scan, and renal dynamic scan were recorded.',\n",
       "   \"The study was done according to the ethical committee guidelines, informed consent was obtained from participants' legal guardians, and assent was also taken from children.\",\n",
       "   'The patients were telephonically contacted and asked to make a targeted visit, and blood samples for these patients were obtained.',\n",
       "   'Blood samples from sex- and age-matched children from the same geographical area with no urological complaints who presented to our clinic were also obtained as controls.',\n",
       "   'All controls were normotensive and had serum creatinine within the normal range at the time of blood sampling.',\n",
       "   'Family history of renal disease, urinary tract malformation, or urinary tract infection was ruled out.',\n",
       "   'After informed consent, all cases and controls were submitted to an intravenous puncture to collect 10 mL of peripheral blood samples.',\n",
       "   'DNA was extracted from peripheral blood using the Keygen DNA extraction kit.',\n",
       "   'The amplification-refractory mutation system (ARMS) polymerase chain reaction (PCR) primers to detect point mutations, as described in the table, were used [ Table 1 ].',\n",
       "   \"Table 1 List of primer sequences used in the study GENE and SNP code Primer type DNA sequence 5'to 3' PAX2 (SNP code: rs4244341) Normal primer TCTGAGGGGTCAAAGGGACG Mutant Primer TCTGAGGGGTCAAAGGGACT Common primer AGGCTCGGCTGACTGTTTAT BMP4 (SNP code: rs17563) Normal primer AGCATCCCTGAGAACGAGGC Mutant Primer AGCATCCCTGAGAACGAGGT Common primer TCGGCTAATCCTGACATGCT AT2R (SNP code: rs3736556) Normal primer TATGTTAATTTGTTAGGTCA Mutant primer TATGTTAATTTGTTAGGTCT Common primer TGAAATGAGAAATGAACATC SNP: Single-nucleotide gene polymorphism, BMP4: Bone morphogenetic protein 4, PAX2: Paired Box gene 2\",\n",
       "   \"The primer sequences were 5'CTG AGA CCA CTC CCA TC 3' and 5'GAT GTG GCC ATC ACA TTC GTC AGA T 3' for ACE polymorphism (ACE gene I/D polymorphism is characterized by the presence or absence of a 287 bp Alu repeat sequence within intron 16).\",\n",
       "   'Amplification with this primer pair produces 490 bp and 190 bp products corresponding to the I allele and D allele, respectively.',\n",
       "   'Normal and mutant ARMS reaction premixes were prepared according to previously published ARMS PCR guidelines, and 40 μL aliquots of the premix were dispensed into reaction tubes suitable for use in a thermal cycler.',\n",
       "   'After adding Taq DNA polymerase, a conventional PCR device was used according to the following protocol: samples were exposed to 94°C for 2 min, then 35 cycles at 94°C for 15 s, 59°C for 10 s, and 72°C for 30 s. Case and control samples were randomly arranged in well plates with at least 20% of genotypes retyped as quality control.',\n",
       "   'The PCR products were run on a 2% agarose gel containing ethidium bromide and visualized by ultraviolet illumination.',\n",
       "   'Statistical comparisons of genotype distributions and allelic frequencies were compared between groups by the Chi-square test.',\n",
       "   'The odds ratios (OR) with the 95% confidence interval (CI) were calculated by comparing the allelic distributions in the study groups.',\n",
       "   'A P < 0.05 was considered statistically significant.',\n",
       "   'Data analysis was carried out using Stata 12.0 (College Station, Texas, USA).'],\n",
       "  'R ESULTS': ['R ESULTS',\n",
       "   'The sample for analysis consisted of 158 patients, 86 cases and 72 controls without the urological disease.',\n",
       "   'Within the CAKUT group, there were 64 (74.4%) males and 22 (25.6%) females with a median age of 32 months.',\n",
       "   'Demographic characteristics of patients within the CAKUT group, as well as the distribution of patients by their presenting uropathy, are displayed in Table 2 .',\n",
       "   'Table 2 Demographic characteristics of patients with congenital anomalies of the kidney and urinary tract Uropathy Number of patients Median age (months) Sex Male Female VUR 24 34 (22.5-69) 12 (50) 12 (50) PUJO 27 29 (22-48) 19 (70.3) 8 (29.7) PUV 30 24.5 (9.75-52.5) 30 (100) 0 EB 3 25 (5-84) 1 (33.3) 2 (66.7) MCKD 2 48 (36-60) 2 (100) 0 Total 86 32 (13.5-60) 64 (74.4) 22 (25.6) Data are expressed as n, median (IQR), or n (%).',\n",
       "   'VUR: Vesicoureteric reflux, PUJO: Pelviureteric junction obstruction, PUV: Posterior urethral valve, EB: Exstrophy bladder, MCDK: Multicystic dysplastic kidney, IQR: Interquartile range',\n",
       "   'For the purpose of this study, the following genes and their alleles were analyzed: BMP4 (GG, GA, AA); PAX2 (GG, GT, TT); AGTR2 (AA, AT, TT), and ACE (II, ID, DD).',\n",
       "   'Initial analysis showed that the expression of these genes did not signify an increased risk for the development of CAKUT in the patient group compared to the control group [ Table 3 ].',\n",
       "   'Table 3 Study gene genotypes and analysis with congenital anomalies of the kidney and urinary tract cases versus controls Gene Allele Case (%) Control (%) P OR (95% CI) P BMP4 GG 17 20 0.69 1.00 0.32 GA 73 64 0.78 (0.34-1.77) 0.64 AA 10 16 1.60 (0.50-5.09) 0.42 PAX2 GG 29 38 0.77 1.00 0.48 GT 63 57 0.70 (0.36-1.39) 0.31 TT 8 5 0.53 (0.14-2.02) 0.35 AGTR2 AA 27 38 0.95 1.00 0.17 AT 51 38 1.17 (0.67-2.04) 0.57 TT 22 24 0.70 (0.20-2.44) 0.58 ACE II 59 40 0.79 1.00 0.20 ID 29 54 0.52 (0.25-1.09) 0.08 DD 12 6 0.81 (0.34-1.90) 0.62 BMP4: Bone morphogenetic protein 4, PAX2: Paired Box gene 2, OR: Odds ratio, CI: Confidence interval, ACE: Angiotensin-converting enzyme, AGTR2: Angiotensin II receptor type 2',\n",
       "   'On further analysis, when polymorphisms of the AGTR2 gene were evaluated based on their presenting phenotype (pelviureteric junction obstruction [PUJO], vesicoureteral reflux [VUR], PUV, etc.), children expressing the TT allele genotype of AGTR2 gene were seen to have a lower incidence of PUJO (OR 0.18 [95% CI, 0.06–0.55], P = 0.01) as well as VUR (OR 0.31 [95% CI, 0.11–0.91), P = 0.03) [ Table 4 ].',\n",
       "   'Table 4 Angiotensin II receptor type 2 genotype analysis and associations with different variants of congenital anomalies of the kidney and urinary tract patients versus controls Phenotype Allele Case (%) Control (%) P OR P PUJO AA 74 40 0.04* 1.00 0.06 AT 8 6 0.72 (0.12-4.34) 0.72 TT 18 54 0.18 (0.06-0.55) 0.01* PUV AA 50 40 0.33 1.00 0.38 AT 10 6 0.60 (0.24-1.46) 0.25 TT 40 54 1.45 (0.28-7.33) 0.65 VUR AA 58 40 0.03* 1.00 0.06 AT 17 6 2.07 (0.45-9.52) 0.35 TT 25 54 0.31 (0.11-0.92) 0.03* OR: Odds ratio, PUJO: Pelviureteric junction obstruction, PUV: Posterior urethral valve, VUR: Vesicoureteric reflux',\n",
       "   'Polymorphisms for the genes, i.e., BMP-4, AGTR2, and PAX2, were not found to be associated with an increased incidence of renal scarring in the CAKUT group.',\n",
       "   'The D allele of the ACE I/D gene was present in 25 out of 30 of the patients who developed scarring.',\n",
       "   'Thus, the presence of D allele was found to be significantly associated with the development of renal scarring.',\n",
       "   'Since expression of the D allele of the ACE gene was seen to increase the propensity of renal parenchymal damage, we decided to perform subgroup analysis of patients with CAKUT expressing the A allele of the AGTR2 gene.',\n",
       "   'This analysis revealed a significant association with expression of the D allele of ACE in the subgroup who demonstrated renal parenchymal scarring (24/27 patients with renal scarring, 36/49 patients without renal scarring-[ Table 5 ]).',\n",
       "   'Table 5 Sub-stratification of all patients with the presence of into A allele in angiotensin II receptor type 2 (76), i.e., all patients with proposed increased predisposition to development of congenital anomalies of the kidney and urinary tract and presence or absence of angiotensin-converting enzyme insertion/deletion polymorphism No Scar (49) (%) Scars (27) (%) D allele (ACE I/D) present 36 (73.5) 24 (88.9) D allele (ACE I/D) absent 13 (26.5) 3 (11.1) ACE I/D: Angiotensin-converting enzyme insertion/deletion'],\n",
       "  'D ISCUSSION': ['D ISCUSSION',\n",
       "   'CAKUT is one of the common causes of end-stage renal disease in children, leading to considerable morbidity and mortality.',\n",
       "   'In our study, of the 86 children presenting to the outpatient clinic and diagnosed with CAKUT, PUV was the most commonly witnessed pathology, with an incidence of 30 (34.8%) patients.',\n",
       "   'Male children (74.4%) had a significantly higher incidence of CAKUT than females (25.6%).',\n",
       "   'Evidence of the male gender is an important predisposing factor in the development of CAKUTs has been documented in previous studies.',\n",
       "   'Li et al . found that male neonates were more likely to present with CAKUTs than female neonates.',\n",
       "   '[ 7 ] Other studies have shown that male children are between 1.3 and 1.9 times more likely to be affected by CAKUT than female children.',\n",
       "   '[ 8 ]',\n",
       "   'The genetic polymorphisms chosen for this study have been shown to be associated with certain CAKUT pathologies in previous studies.',\n",
       "   'It is well known that the BMP4 family takes part in the ureteric bud development.',\n",
       "   '[ 9 10 ] Miyazaki et al . demonstrated that mice with reduced expression of BMP4 demonstrated three different types of malformations – hydronephrosis with hypo/dysplastic kidneys, ureterovesical junction obstruction, and duplex kidney with the bifid ureter, which led to the hypothesis that BMP4 is a fine-tuning protein that modulates the number of functional nephrons and ureteric branching.',\n",
       "   '[ 11 ] In our study, the presence of BMP4 polymorphisms was not seen to correlate with an increased risk of developing CAKUT, which is in contrast to a previous study comprising a sample of Brazilian patients conducted in 2014, wherein Reis et al . found the presence of allele A at rs17563 to be a risk factor for anomalies of CAKUT.',\n",
       "   '[ 12 ] Reis et al . found that when the AA genotype for BMP-4 was present, the risk of developing congenital uropathies was 2.49 times higher than when the GG genotype was present.',\n",
       "   'PAX2 polymorphism is associated with reduced kidney size in neonates.',\n",
       "   '[ 13 ] In this study, PAX2 polymorphisms of the alleles studied rs4244351 (G/T alleles) did not show a significant association with the development of CAKUT.',\n",
       "   'This is in contrast to de Miranda et al ., who studied PAX2 polymorphisms in 241 cases versus 259 controls in a Brazilian pediatric population.',\n",
       "   '[ 14 ] The authors evaluated associations with five single-nucleotide polymorphisms covering the entire PAX2 gene.',\n",
       "   'In the subgroup of patients with VUR, frequencies of monozygotic ancestral alleles significantly differed at the markers rs11190693 (A/T alleles) and rs4244351 (G/T alleles) in comparison with controls, whereas no changes were detected in cases of PUJO or MCDK.',\n",
       "   'ACE insertion/deletion (ACE I/D) gene polymorphism is a risk factor for renal parenchymal damage in patients with congenital urologic abnormalities and seems particularly relevant in children with VUR, where it is an independent predisposing factor.',\n",
       "   'Previous studies have indicated that the ACE I/D polymorphisms may be associated with diabetic nephropathy, IgA nephropathy, nephritic syndrome, or focal segmental glomerulosclerosis.',\n",
       "   '[ 15 16 17 18 19 20 ] The serum ACE level has been found to vary with I versus D polymorphism of the ACE gene.',\n",
       "   'It has been reported that the DD genotype of this polymorphism is associated with higher levels of plasma and tissue ACE activity, as described by Rigat et al .',\n",
       "   '[ 21 ]',\n",
       "   'In our study, the analysis showed that expression of ACE gene polymorphisms did not signify an increased risk for the development of CAKUT in the patient group as compared to the control group though the presence of the D polymorphism (DD or ID allele) was seen to be associated with a significantly increased incidence of renal parenchymal damage, as assessed by renal scarring (on DMSA scan) in children with CAKUT.',\n",
       "   'This agrees with the previous study conducted in Italian children by Rigoli et al ., who compared the incidence of the ACE I/D polymorphism in 102 children with CAKUT, leading to varying degrees of renal impairment and in 92 healthy controls with a similar ethnic background and found that ACE genotype distribution did not differ between the CAKUT patients and normal control subjects.',\n",
       "   '[ 22 ] The findings of our study are also in keeping with the findings of a previous study by Bajpai et al . who evaluated the role of ACE I/D polymorphism as a risk factor for progressive renal damage in Indian children with congenital uropathies.',\n",
       "   '[ 23 ]',\n",
       "   'Studies of the angiotensin II receptor 2 gene (AGTR2) in patients with CAKUT have historically shown controversial results.',\n",
       "   'Nishimura et al . observed an increased A–G transition in the AT2R gene in male Caucasian American and German patients with multicystic dysplastic kidney (MCDK) and PUJO,[ 24 ] while Hiraoka et al . reported that there is no evidence for this AT2R gene derangement in human urinary tract anomalies in Japanese patients.',\n",
       "   '[ 25 ]',\n",
       "   'In our study, the presence of the A allele at rs3736556 of the AGTR2 gene was significantly correlated with the development of PUJO and VUR but not PUV.',\n",
       "   'Our study has a few findings similar to that by Miranda et al ., where the authors studied 290 pediatric patients with CAKUT and 262 healthy controls from the same geographic area for AGTR2 gene polymorphisms at rs1403543, rs3736556, rs35474657, rs5193, and rs5194 and found that the diagnosis of PUJO was significantly associated with AGTR2 gene polymorphisms at rs3736556 and rs5194.',\n",
       "   'On the other hand, the diagnoses of VUR and of multicystic dysplastic kidney were not associated with AGTR2 gene polymorphisms.',\n",
       "   '[ 26 ]',\n",
       "   'On further stratification of the patients of CAKUT who expressed the A allele (AGTR2), a significant correlation was found between the expression of the D allele (ACE I/D) and renal parenchymal damage.',\n",
       "   'Although limited by the sample size in this study, further studies confirming this trend would be beneficial in establishing the association between multiple mutations and worsening renal injury.',\n",
       "   'This would, in turn, assist in closer follow-up of children with the presence of at least one polymorphism, as inheriting additional mutations or “second hits” would confer them with increased propensity to renal injury.',\n",
       "   'Besides close clinical monitoring of high-risk children, this information would aid in genetic counseling.',\n",
       "   'Although being a pilot study in Asian Indian children, the primary limitation of our study is the small sample size rendering definitive conclusions regarding the general population difficult.',\n",
       "   'Further, adequately powered studies are needed to explore the association between gene polymorphisms and the development of congenital uropathies in this patient population.'],\n",
       "  'C ONCLUSION': ['C ONCLUSION',\n",
       "   'In our study, a significant association was found between the presence of the A allele of the AGTR2 gene at rs3736556 with the development of PUJO and VUR.',\n",
       "   'Furthermore, these children were more likely to develop renal scarring if they were found to harbor the D allele (ACE I/D polymorphism) as well.',\n",
       "   'As congenital uropathies are quite common and also known to lead to end-stage renal disease, early identification of children at risk and also segregation of them into those who would need closer follow-up due to additional genetic polymorphisms could lead to significant improvement in their management.',\n",
       "   'Financial support and sponsorship',\n",
       "   'Indian Council of Medical Research.',\n",
       "   'Conflicts of interest',\n",
       "   'There are no conflicts of interest.']}}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "process_full_text(ss[999])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06f31a47-ce68-46fe-9fca-7e501b74fa14",
   "metadata": {},
   "outputs": [],
   "source": [
    "ss[194]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d83ba89-8992-4750-a830-813f3de5d4c3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48218ad2-b0c4-483b-a9de-93ec3407d554",
   "metadata": {},
   "outputs": [],
   "source": [
    "#fix keywords\n",
    "\n",
    "#<kwd-group><kwd>astrocyte</kwd><kwd>microglia</kwd><kwd>excitability</kwd><kwd>hydrogen sulfide</kwd><kwd>BDNF</kwd><kwd>polyamine</kwd><kwd>dexmedetomidine</kwd><kwd>astrocyte-microglia co-culture</kwd></kwd-group>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "080d58e0-ba2d-42eb-8a94-f06ab9a44bf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "<title>Author contributions</title> # titles are repeating, take care"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edd26a39-2dbe-44bb-bfe6-267e59a87dd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to process the front section\n",
    "def process_front(front):\n",
    "    sent_id = 1\n",
    "    sections = {}\n",
    "    \n",
    "    if front.find('article-meta'):\n",
    "        art_meta = front.find('article-meta')\n",
    "        \n",
    "        for ch in art_meta.find_all(recursive=False):\n",
    "            if ch.name in ['title-group', 'supplement', 'supplementary-material', 'abstract', 'trans-abstract', 'kwd-group', 'funding-group']:\n",
    "                section_title = ch.name\n",
    "                \n",
    "                if section_title == 'kwd-group':\n",
    "                    # Extract keywords as a list from kwd-group\n",
    "                    keywords = [kwd.text.strip() for kwd in ch.find_all('kwd')]\n",
    "                    sections['KEYWORD'] = keywords  # Store keywords as a list\n",
    "                else:\n",
    "                    sent_id, sentences = call_sentence_tags(ch, sent_id)\n",
    "                    if sentences:\n",
    "                        sections[section_title] = sentences\n",
    "            else:\n",
    "                pass  # Ignore other tags\n",
    "                \n",
    "    return sent_id, sections\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cce7202d-c3a1-4397-960d-b7be4c38938f",
   "metadata": {},
   "outputs": [],
   "source": [
    "each_file = ss[194].replace('<body>', '<orig_body>').replace('</body>', '</orig_body>')\n",
    "\n",
    "xml_soup = BeautifulSoup(each_file, 'lxml')\n",
    "# Unwrap additional tags to keep the main structure intact\n",
    "if xml_soup.html:\n",
    "    xml_soup.html.unwrap()\n",
    "if xml_soup.body:\n",
    "    xml_soup.body.unwrap()\n",
    "if xml_soup.find('orig_body'):\n",
    "    xml_soup.find('orig_body').name = 'body'\n",
    "            \n",
    "sent_id, front_sections = process_front(xml_soup.article.find('front'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9337b142-dee0-4b79-873d-b4a77a9a4e43",
   "metadata": {},
   "outputs": [],
   "source": [
    "front_sections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea8c04e6-b14d-4b03-9e0a-3fe048d57acd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sections={}\n",
    "# Process front, body, and back sections\n",
    "if 'ABSTRACT' not in sections and xml_soup.article.find('front'):\n",
    "    sent_id, front_sections = process_front(xml_soup.article.find('front'))\n",
    "    for k, v in front_sections.items():\n",
    "        sections.setdefault(k, []).extend(v)  # Ensure that 'KEYWORD' remains a list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b91d83c-b72e-45bc-89b3-8f8a07c55bbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "527023de-0b71-4b88-ae58-7111c7339cc1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
