{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "96521a4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "from requests.compat import urljoin\n",
    "import requests\n",
    "import glob\n",
    "import os\n",
    "import re\n",
    "from bs4 import BeautifulSoup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1d2f151d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import spacy\n",
    "nlp = spacy.load(\"en_core_sci_sm\")\n",
    "# nlp = spacy.load(\"en_core_web_sm\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "76e98eaa",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_132314/2228744081.py:1: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  AN_RS_PMCIDS_df =  pd.read_csv('../data/accession_data.csv')\n"
     ]
    }
   ],
   "source": [
    "AN_RS_PMCIDS_df =  pd.read_csv('../data/accession_data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4d573c4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['accession_type', 'source', 'extid', 'pmcid'], dtype='object')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AN_RS_PMCIDS_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e7cdcdc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession_type</th>\n",
       "      <th>pmcid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>alphafold</td>\n",
       "      <td>PMC10550955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>alphafold</td>\n",
       "      <td>PMC10526293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>alphafold</td>\n",
       "      <td>PMC10508174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>alphafold</td>\n",
       "      <td>PMC10068818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>alphafold</td>\n",
       "      <td>PMC10518620</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  accession_type        pmcid\n",
       "0      alphafold  PMC10550955\n",
       "1      alphafold  PMC10526293\n",
       "2      alphafold  PMC10508174\n",
       "3      alphafold  PMC10068818\n",
       "4      alphafold  PMC10518620"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AN_RS_PMCIDS_df = AN_RS_PMCIDS_df[['accession_type', 'pmcid']]\n",
    "AN_RS_PMCIDS_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e3be62e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "size_ = 5\n",
    "\n",
    "grouped = AN_RS_PMCIDS_df.groupby('accession_type')\n",
    "\n",
    "# Empty dictionary to store the separate dataframes\n",
    "dfs = {}\n",
    "\n",
    "# Loop through each group\n",
    "for name, group in grouped:\n",
    "    # Select up to 100 pmcids from each group\n",
    "    selected = group.head(size_)\n",
    "    # Store the resulting DataFrame in the dictionary\n",
    "    dfs[name] = selected\n",
    "    \n",
    "combined_df = pd.concat(dfs.values())\n",
    "combined_df_clean = combined_df.dropna()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "822342f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['alphafold', 'arrayexpress', 'bia', 'biomodels', 'bioproject',\n",
       "       'biosamples', 'biostudies', 'brenda', 'cath', 'cellosaurus',\n",
       "       'chebi', 'chembl', 'complexportal', 'dbgap', 'doi', 'ebisc', 'efo',\n",
       "       'ega', 'emdb', 'empiar', 'ena', 'ensembl', 'eudract', 'gca', 'geo',\n",
       "       'gisaid', 'go', 'hgnc', 'hipsci', 'hpa', 'igsr', 'intact',\n",
       "       'interpro', 'metabolights', 'metagenomics', 'mint', 'nct', 'omim',\n",
       "       'orphadata', 'pdb', 'pfam', 'prode', 'reactome', 'refseq',\n",
       "       'refsnp', 'rfam', 'rhea', 'rnacentral', 'rrid', 'treefam',\n",
       "       'uniparc', 'uniprot'], dtype=object)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df_clean['accession_type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "90a9b317",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession_type</th>\n",
       "      <th>pmcid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>657684</th>\n",
       "      <td>metabolights</td>\n",
       "      <td>PMC10051505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9596</th>\n",
       "      <td>bioproject</td>\n",
       "      <td>PMC5525643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9192</th>\n",
       "      <td>biomodels</td>\n",
       "      <td>PMC4360671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71002</th>\n",
       "      <td>dbgap</td>\n",
       "      <td>PMC5120758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>606017</th>\n",
       "      <td>go</td>\n",
       "      <td>PMC10404263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70934</th>\n",
       "      <td>complexportal</td>\n",
       "      <td>PMC3026084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>974803</th>\n",
       "      <td>orphadata</td>\n",
       "      <td>PMC9335361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630219</th>\n",
       "      <td>gisaid</td>\n",
       "      <td>PMC8560824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>635640</th>\n",
       "      <td>hpa</td>\n",
       "      <td>PMC9907840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2634333</th>\n",
       "      <td>rnacentral</td>\n",
       "      <td>PMC8190129</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession_type        pmcid\n",
       "657684    metabolights  PMC10051505\n",
       "9596        bioproject   PMC5525643\n",
       "9192         biomodels   PMC4360671\n",
       "71002            dbgap   PMC5120758\n",
       "606017              go  PMC10404263\n",
       "70934    complexportal   PMC3026084\n",
       "974803       orphadata   PMC9335361\n",
       "630219          gisaid   PMC8560824\n",
       "635640             hpa   PMC9907840\n",
       "2634333     rnacentral   PMC8190129"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df_clean.sample(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4437e856",
   "metadata": {},
   "outputs": [],
   "source": [
    "pmcids = combined_df_clean['pmcid'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f622a585",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from urllib.parse import urljoin\n",
    "import csv\n",
    "from tqdm import tqdm\n",
    "\n",
    "def get_Json_through_PMCIDs(pmcids, type_):\n",
    "    base_url = \"https://www.ebi.ac.uk/europepmc/annotations_api/\"\n",
    "    pmcid_query = ','.join([f\"PMC%3A{id[3:]}\" for id in pmcids])  # Prepare PMCIDs for the query\n",
    "    article_url = urljoin(base_url, f\"annotationsByArticleIds?articleIds={pmcid_query}&type={type_}&format=JSON\")\n",
    "    r = requests.get(article_url)\n",
    "\n",
    "    if r.status_code == 200:\n",
    "        return r\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "def get_epmc_annotations_to_file(PMCids):\n",
    "    with open('../data/AN_RS_annotations_from_api.csv', 'w', newline='\\n') as f1:\n",
    "        test_writer = csv.writer(f1, delimiter='\\t', lineterminator='\\n')\n",
    "\n",
    "        # tqdm is now tracking the overall progress\n",
    "        for i in tqdm(range(0, len(PMCids), 8)):\n",
    "            batch_ids = PMCids[i:i+8]\n",
    "\n",
    "            for ann_type in ['Accession Numbers', 'Resources']:\n",
    "                if ann_type == 'Accession Numbers':\n",
    "                    ent_type = 'AN'\n",
    "                elif ann_type == 'Resources':\n",
    "                    ent_type = 'RS'\n",
    "\n",
    "                json_annotations = get_Json_through_PMCIDs(batch_ids, ann_type)\n",
    "                if json_annotations:\n",
    "                    json_results = json_annotations.json()\n",
    "                    for article in json_results:\n",
    "                        try:\n",
    "                            pmc_id = article['pmcid']\n",
    "                            for each_annotation in article['annotations']:\n",
    "                                try:\n",
    "                                    exact = each_annotation['prefix'] + each_annotation['exact'] + each_annotation['postfix']\n",
    "                                except KeyError:\n",
    "                                    exact = each_annotation.get('exact', '')\n",
    "                                token = each_annotation['tags'][0]['name']\n",
    "                                row = [pmc_id, exact, token, ent_type]\n",
    "                                test_writer.writerow(row)\n",
    "                        except (IndexError, KeyError):\n",
    "                            pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8fd337b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████| 32/32 [00:37<00:00,  1.17s/it]\n"
     ]
    }
   ],
   "source": [
    "get_epmc_annotations_to_file(pmcids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7e7f046c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmcid</th>\n",
       "      <th>partial_sentence</th>\n",
       "      <th>token</th>\n",
       "      <th>ner</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>80311</th>\n",
       "      <td>PMC2954584</td>\n",
       "      <td>617105</td>\n",
       "      <td>617105</td>\n",
       "      <td>AN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128247</th>\n",
       "      <td>PMC6909262</td>\n",
       "      <td>13snp_rs6548616_0w_rrs6548616AIMS793995750.3290.1</td>\n",
       "      <td>rs6548616</td>\n",
       "      <td>AN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56556</th>\n",
       "      <td>PMC1501050</td>\n",
       "      <td>7242);neurogenesis (GO:0007399);protein bindin...</td>\n",
       "      <td>GO:0007399</td>\n",
       "      <td>AN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30259</th>\n",
       "      <td>PMC2848993</td>\n",
       "      <td>A*9212——TBC63782AB292417M SatakeA*02:113NA*9</td>\n",
       "      <td>AB292417</td>\n",
       "      <td>AN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106273</th>\n",
       "      <td>PMC4309170</td>\n",
       "      <td>2OM5 2OO0 2OO9 2OOA 2OOQ 2OP7 2OPG 2OPU 2OPV</td>\n",
       "      <td>2OOQ</td>\n",
       "      <td>AN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37456</th>\n",
       "      <td>PMC5525643</td>\n",
       "      <td>S1040153SAMEA3733004ERX1298490ERR1226278ITSITS...</td>\n",
       "      <td>ERX1298490</td>\n",
       "      <td>AN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101486</th>\n",
       "      <td>PMC2848993</td>\n",
       "      <td>�DRB1*1401V2JCB15932AB049831DRB1*14:40:01DRB1*14</td>\n",
       "      <td>AB049831</td>\n",
       "      <td>AN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16000</th>\n",
       "      <td>PMC8035907</td>\n",
       "      <td>which is available via NCBI BioSample SAMN1581...</td>\n",
       "      <td>SAMN15816786</td>\n",
       "      <td>AN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49664</th>\n",
       "      <td>PMC10190287</td>\n",
       "      <td>84687 42,9887,96565 SRR22784686 5,261,38135,12...</td>\n",
       "      <td>SRR22784686</td>\n",
       "      <td>AN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66601</th>\n",
       "      <td>PMC10404263</td>\n",
       "      <td>iron ion transport|GO:0055076, transition met...</td>\n",
       "      <td>GO:0055076</td>\n",
       "      <td>AN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              pmcid                                   partial_sentence  \\\n",
       "80311    PMC2954584                                             617105   \n",
       "128247   PMC6909262  13snp_rs6548616_0w_rrs6548616AIMS793995750.3290.1   \n",
       "56556    PMC1501050  7242);neurogenesis (GO:0007399);protein bindin...   \n",
       "30259    PMC2848993       A*9212——TBC63782AB292417M SatakeA*02:113NA*9   \n",
       "106273   PMC4309170       2OM5 2OO0 2OO9 2OOA 2OOQ 2OP7 2OPG 2OPU 2OPV   \n",
       "37456    PMC5525643  S1040153SAMEA3733004ERX1298490ERR1226278ITSITS...   \n",
       "101486   PMC2848993   �DRB1*1401V2JCB15932AB049831DRB1*14:40:01DRB1*14   \n",
       "16000    PMC8035907  which is available via NCBI BioSample SAMN1581...   \n",
       "49664   PMC10190287  84687 42,9887,96565 SRR22784686 5,261,38135,12...   \n",
       "66601   PMC10404263   iron ion transport|GO:0055076, transition met...   \n",
       "\n",
       "               token ner  \n",
       "80311         617105  AN  \n",
       "128247     rs6548616  AN  \n",
       "56556     GO:0007399  AN  \n",
       "30259       AB292417  AN  \n",
       "106273          2OOQ  AN  \n",
       "37456     ERX1298490  AN  \n",
       "101486      AB049831  AN  \n",
       "16000   SAMN15816786  AN  \n",
       "49664    SRR22784686  AN  \n",
       "66601     GO:0055076  AN  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotations_df = pd.read_csv('../data/AN_RS_annotations_from_api.csv', sep='\\t', names=['pmcid', 'partial_sentence', 'token', 'ner'])\n",
    "annotations_df.sample(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0468c00a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['AN', 'RS'], dtype=object)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotations_df['ner'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "660a73c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Gene Ontology',\n",
       " 'ArrayExpress',\n",
       " 'Uniprot',\n",
       " 'UniProt',\n",
       " 'Protein DataBank',\n",
       " 'Ensembl',\n",
       " 'PDB',\n",
       " 'BioModels',\n",
       " 'BioStudies',\n",
       " '1000 Genomes',\n",
       " 'Electron Microscopy Public Image Archive',\n",
       " 'EMPIAR',\n",
       " 'Electron Microscopy Data Bank',\n",
       " 'Protein Data Bank',\n",
       " 'European Nucleotide Archive',\n",
       " 'ENA',\n",
       " 'EGA',\n",
       " 'PDBe',\n",
       " 'ChEMBL',\n",
       " 'MetaboLights',\n",
       " 'Rfam',\n",
       " 'Protein Data Bank in Europe',\n",
       " 'PRIDE',\n",
       " 'European Variation Archive',\n",
       " 'EVA',\n",
       " 'BioSamples',\n",
       " 'Europe PubMed Central',\n",
       " 'GWAS Catalog',\n",
       " 'Experimental Factor Ontology',\n",
       " 'identifiers.org',\n",
       " 'Swiss-Prot',\n",
       " 'UniprotKB',\n",
       " 'HGNC',\n",
       " 'InterProScan',\n",
       " 'InterPro',\n",
       " 'ChEBI',\n",
       " 'Chemical Entities of Biological Interest',\n",
       " 'SILVA rRNA database',\n",
       " 'SILVA',\n",
       " 'Swissprot',\n",
       " 'Pfam',\n",
       " 'CATH',\n",
       " 'database STRING',\n",
       " 'BRENDA',\n",
       " 'Cellosaurus',\n",
       " 'cellosaurus',\n",
       " 'Reactome',\n",
       " 'IntAct',\n",
       " 'IntAct database',\n",
       " 'uniprot',\n",
       " 'STRING database',\n",
       " 'UniProtKB',\n",
       " 'STRING',\n",
       " 'Complex Portal',\n",
       " 'RNAcentral',\n",
       " 'molecular interaction database',\n",
       " 'Expression Atlas',\n",
       " 'Rhea knowledgebase',\n",
       " 'Rhea',\n",
       " 'http://string-db.org',\n",
       " 'Molecular Interaction Database',\n",
       " 'Array Express',\n",
       " 'European Genome-phenome Archive',\n",
       " 'Ontology Lookup Service',\n",
       " 'OLS',\n",
       " 'EFO',\n",
       " 'SwissProt',\n",
       " 'TrEMBL',\n",
       " 'EMDataBank',\n",
       " 'HUGO Gene Nomenclature Committee',\n",
       " 'UNIPROT',\n",
       " 'Human Protein Atlas',\n",
       " 'HPA',\n",
       " 'Protein Atlas',\n",
       " 'proteinatlas.org',\n",
       " 'protein atlas',\n",
       " 'string-db.org',\n",
       " 'https://string-db.org',\n",
       " 'SWISS-PROT',\n",
       " 'uniprotkb',\n",
       " 'SILVA database',\n",
       " 'Silva',\n",
       " 'Protein Databank',\n",
       " 'RCSB PDB',\n",
       " 'Ensembl Genomes',\n",
       " 'Rhea database',\n",
       " 'WormBase',\n",
       " 'Rhea reaction',\n",
       " 'STRING network',\n",
       " 'UniParc']"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(annotations_df[annotations_df['ner']=='RS'].token.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ffd11fd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get unique PMCIDs\n",
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "import re\n",
    "\n",
    "def get_full_text_xml(pmcid):\n",
    "    segmented_sentences = []\n",
    "    url = f\"https://www.ebi.ac.uk/europepmc/webservices/rest/{pmcid}/fullTextXML\"\n",
    "    response = requests.get(url)\n",
    "    if response.status_code == 200:\n",
    "        soup = BeautifulSoup(response.content, 'xml')\n",
    "        p_tags = soup.find_all('p')\n",
    "        p_texts = [tag.get_text() for tag in p_tags]\n",
    "        # Using nlp.pipe for batch processing\n",
    "        for doc in nlp.pipe(p_texts):\n",
    "            sentences = [sent.text.strip() for sent in doc.sents]\n",
    "            segmented_sentences.extend(sentences)\n",
    "        return segmented_sentences\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a09a2e45",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_pmcids = annotations_df['pmcid'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "bf5eb240",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "241"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(unique_pmcids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3dfe5245",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████| 241/241 [14:21<00:00,  3.57s/it]\n"
     ]
    }
   ],
   "source": [
    "def find_sentence_with_substring(string_list, substring):\n",
    "    for text in string_list:\n",
    "        sentences = re.split(r'(?<=[.!?])\\s+', text)\n",
    "        for sentence in sentences:\n",
    "            if substring in sentence:\n",
    "                return sentence\n",
    "    return None\n",
    "\n",
    "def process_pmcid(df, pmcid, p_texts):\n",
    "    sentences_data = {}\n",
    "    for _, row in df[df['pmcid'] == pmcid].iterrows():\n",
    "        sentence = find_sentence_with_substring(p_texts, row['partial_sentence'])\n",
    "        if sentence:\n",
    "            if sentence not in sentences_data:\n",
    "                sentences_data[sentence] = set()\n",
    "\n",
    "            sentences_data[sentence].add((row['token'], row['ner']))\n",
    "\n",
    "    return [[pmcid, sentence, list(ner_tags)] for sentence, ner_tags in sentences_data.items()]\n",
    "\n",
    "final_data = []\n",
    "\n",
    "for pmcid in tqdm(unique_pmcids):\n",
    "    p_texts = get_full_text_xml(pmcid)\n",
    "    if p_texts:\n",
    "        processed_data = process_pmcid(annotations_df, pmcid, p_texts)\n",
    "        final_data.extend(processed_data)\n",
    "\n",
    "# Convert to DataFrame\n",
    "final_df = pd.DataFrame(final_data, columns=['pmcid', 'sentence', 'ner'])\n",
    "\n",
    "# Save as TSV\n",
    "final_df.to_csv('../data/AN-RS_xxx7.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1b521d1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['PMC8080698',\n",
       "  'Table 4RNA-seq or gene-chip data from loss or gain of function of each PcG subunit proteina, b.UnitData IDTarget cell (cell line)Functional alteration (method)EZH1E-GEOD-36288LSK (Lin−Sca-1+c-kit+) cellsLoss of function (KO)E-GEOD-59090CD34+ hematopoietic stem and progenitor cell (HSPC)-derived proerythroblastsLoss of function (shRNA)EZH1/ EZH2E-GEOD-62198MLL-AF9 induced leukemia progenitorLoss of function (inhibitor, UNC1999)EZH2E-GEOD-59090CD34+ HSPC-derived proerythroblastsLoss of function (shRNA)E-GEOD-71870Ovarian tumor (OC8 cell line)Loss of function (shRNA)E-GEOD-71671Monocyte (THP-1 cell line)Loss of function (inhibitor, GSK126)E-MTAB-3227Gastric cancer (MKN45 cell line)Loss of function (siRNA)E-GEOD-82072Megakaryocyte-erythrocyte precursorLoss of function (cKO; scl-cre)E-GEOD-82073Long-term-hematopoietic stem cell (LT-HSC)Loss of function (cKO; scl-cre)E-MTAB-2893Chronic myeloid leukemia CD34+ cellsLoss of function (inhibitor, GSK343)E-MTAB-3552Nonchronic myeloid leukemia CD34+ cellsLoss of function (inhibitor, GSK343)E-MTAB-5766Acute myeloid leukemia cell (F-36P, MOLM-13, and OCI-M2 cell line)Loss of function (shRNA)E-MTAB-7739Macroglobulinemia cell (RPCI-WM1 cell line)Loss of function (inhibitor, Tazemetostat)GSE101316Mouse bone marrow-derived macrophageLoss of function (KO)EEDE-GEOD-12982Embryonic stem cell (ES cell)Loss of function (KO)E-GEOD-49305ES cellLoss of function (KO)E-GEOD-62198MLL-AF9 induced leukemia progenitorLoss of function (shRNA)E-GEOD-53508ES cellLoss of function (KO)E-GEOD-53506ES cellLoss of function (KO)E-GEOD-59090CD34+ HSPC-derived proerythroblastsLoss of function (shRNA)SUZ12E-GEOD-31354ES cellLoss of function (KO)E-GEOD-53508ES cellLoss of function (Genetrap)E-GEOD-59090CD34+ HSPC-derived proerythroblastsLoss of function (shRNA)E-GEOD-60808HSPCLoss of function (shRNA)Jarid2E-GEOD-60808HSPCLoss of function (shRNA)RING1AE-MTAB-5661ES cellLoss of function (KO)RING1A/BE-MTAB-5661ES cellLoss of function (KO, cKO)RING1BE-GEOD-67868ES cellLoss of function (KO)E-GEOD-69824ES cellLoss of function (KO)E-GEOD-71007Ewing’s sarcoma (A4573, A673, ES1, and TC71 cell line)Loss of function (shRNA)Mel18E-GEOD-67868Embryonic stem cellLoss of function (shRNA)Bmi1E-GEOD-21912Multiple myeloma(RPMI-8226 cell line)Loss of function (shRNA)E-GEOD-19796HSPCLoss of function (KO)E-GEOD-20958ES cell and HSCGain of function (overexpression)E-GEOD-31086Common myeloid progenitorLoss of function (KO)E-GEOD-54262Erythroleukemia (K562 cell line) and Chronic myeloid leukemiaLoss of function (shRNA)E-GEOD-71007Ewing’s sarcoma (A673 and TC71 cell line)Loss of function (shRNA)Cbx7E-GEOD-34191ES cellLoss of function (shRNA)PCGF1E-GEOD-33280HSPCLoss of function (shRNA)PCGF3/5E-MTAB-5642ES cellLoss of function (KO)KO knockout mice, cKO conditional knockout mice.aInformation in Table 4 was acquired from https://www.ebi.ac.uk/arrayexpress/.bEach reference for Table 4 is contained within contents of each Data ID.Table 5ChIP-seq data of each PcG subunit proteina, b.UnitData IDTarget cell (cell line)EZH1E-GEOD-59090CD34+ hematopoietic stem and progenitor (HSPC)-derived proerythroblastsEZH2E-GEOD-18776Mouse embryonic stem cell (ES cell)E-MTAB-1305Human ES cellE-GEOD-51079In vitro cultured Th1 and Th2 cellE-GEOD-42706mouse resting B cell (CD43- B cell)E-GEOD-49178ES cellE-GEOD-52300Human liver cancer (HepG2 cell line)E-GEOD-48518induced pluripotent stem cellE-GEOD-46536ES cellE-GEOD-53495Human embryonic kidney cell (293T Rex cell line)E-GEOD-57632Multiple myelomaE-GEOD-61586Neural stem cell and glioma (SF7761 cell line)E-MTAB-2002Mouse ES cellE-GEOD-47082Mouse ES cellE-GEOD-59090CD34+ HSPC-derived proerythroblastsE-GEOD-70440Mammary glandE-GEOD-60160Mouse ES cellE-MTAB-6410Chronic lymphocytic leukemiaGSE101320Mouse bone marrow-derived macrophagesEEDE-GEOD-61902SpermatocyteE-GEOD-59090CD34+ HSPC-derived proerythroblastsE-MTAB-6165Mouse ES cellSUZ12E-GEOD-34483Mouse ES cellE-GEOD-42616Mouse ES cellE-GEOD-44286Mouse ES cellE-GEOD-52300Human liver cancer (HepG2 cell line)E-MTAB-2481Mouse ES cellE-GEOD-55698ES cellE-GEOD-52619Mouse ES cellE-GEOD-47528Primary CD4+ helper T cellE-GEOD-58023ES cellE-GEOD-62437MLL-AF9 transformed leukemiaE-GEOD-43915Mouse E13.5 brainE-GEOD-59090CD34+ HSPC-derived proerythroblastsE-GEOD-57926Mouse heart and embryonic fibroblast cellE-GEOD-61148Mouse thymocyte and thymic T cell tumor (ILC87 cell line)E-GEOD-74330Mouse ES cellE-GEOD-83082Mouse ES cellJarid2E-GEOD-19708Mouse ES cellRING1BE-GEOD-23716Mouse ES cellE-GEOD-55698ES cellE-GEOD-43915Mouse E13.5 brainE-GEOD-67868Mouse ES cellE-GEOD-72164Mouse ES cellE-GEOD-74330Mouse ES cellMEL18E-GEOD-67868Mouse ES cellE-GEOD-74330Mouse ES cellCBX2E-GEOD-29611Erythroleukemia (K562 cell line)CBX3E-GEOD-29611Erythroleukemia (K562 cell line)E-GEOD-32465Human colon cancer (HCT116 cell line) and erythroleukemia (K562 cell line)E-GEOD-28115Human colon cancer (HCT116 cell line)E-GEOD-44242Mouse ES cell and induced pluripotent stem cellCBX7E-GEOD-23716Mouse ES cellE-GEOD-42466Mouse ES cellE-GEOD-42706mouse resting B cell (CD43- B cell)CBX8E-GEOD-29611Erythroleukemia (K562 cell line)E-GEOD-54052Mouse ES cellaInformation in Table 5 was acquired from https://www.ebi.ac.uk/arrayexpress/.bEach reference for Table 5 is contained within contents of each Data ID.',\n",
       "  [('GSE101316', 'AN'),\n",
       "   ('E-GEOD-57632', 'AN'),\n",
       "   ('E-GEOD-49178', 'AN'),\n",
       "   ('E-GEOD-53495', 'AN'),\n",
       "   ('E-GEOD-23716', 'AN'),\n",
       "   ('E-GEOD-42706', 'AN'),\n",
       "   ('E-GEOD-59090', 'AN'),\n",
       "   ('E-GEOD-60160', 'AN'),\n",
       "   ('E-GEOD-47528', 'AN'),\n",
       "   ('E-GEOD-29611', 'AN'),\n",
       "   ('E-GEOD-18776', 'AN'),\n",
       "   ('E-GEOD-48518', 'AN'),\n",
       "   ('E-GEOD-42466', 'AN'),\n",
       "   ('E-MTAB-6165', 'AN'),\n",
       "   ('E-GEOD-70440', 'AN'),\n",
       "   ('E-GEOD-61148', 'AN'),\n",
       "   ('E-GEOD-83082', 'AN'),\n",
       "   ('E-GEOD-51079', 'AN'),\n",
       "   ('E-GEOD-57926', 'AN'),\n",
       "   ('E-GEOD-52619', 'AN'),\n",
       "   ('E-GEOD-44242', 'AN'),\n",
       "   ('E-MTAB-6410', 'AN'),\n",
       "   ('E-GEOD-47082', 'AN'),\n",
       "   ('E-GEOD-58023', 'AN'),\n",
       "   ('E-GEOD-61902', 'AN'),\n",
       "   ('E-MTAB-2481', 'AN'),\n",
       "   ('E-GEOD-61586', 'AN'),\n",
       "   ('E-GEOD-55698', 'AN'),\n",
       "   ('E-GEOD-44286', 'AN'),\n",
       "   ('E-GEOD-67868', 'AN'),\n",
       "   ('E-GEOD-52300', 'AN'),\n",
       "   ('E-GEOD-34483', 'AN'),\n",
       "   ('E-MTAB-2002', 'AN'),\n",
       "   ('E-GEOD-42616', 'AN'),\n",
       "   ('E-GEOD-54052', 'AN'),\n",
       "   ('E-GEOD-43915', 'AN'),\n",
       "   ('E-GEOD-28115', 'AN'),\n",
       "   ('E-GEOD-46536', 'AN'),\n",
       "   ('GSE101320', 'AN'),\n",
       "   ('E-GEOD-32465', 'AN'),\n",
       "   ('E-GEOD-74330', 'AN'),\n",
       "   ('E-GEOD-19708', 'AN'),\n",
       "   ('E-GEOD-72164', 'AN'),\n",
       "   ('E-MTAB-1305', 'AN'),\n",
       "   ('E-GEOD-62437', 'AN')]],\n",
       " ['PMC8080698',\n",
       "  'Target polycomb subunitAgentMode of actionNCT IDPhaseTarget tumor typesStatusEZH2Tazemetostat (EPZ-6438)S-adenosyl- L-methionine (SAM) competitive inhibitorNCT02875548IIDiffuse large B cell lymphomaRecruitingNCT02601950IIINI (hSNF5; SMARCB1)-negative tumor and relapsed/refractory synovial sarcomaRecruitingNCT01897571I/IIAdvanced solid tumor and B cell lymphomaActive, not recruitingNCT02601937IINI (hSNF5; SMARCB1)-negative tumor and synovial sarcomaRecruitingNCT02860286IIMalignant mesotheliomaCompletedNCT03213665IIRelapsed/refractory advanced solid tumor and non-Hodgkin’s lymphomaRecruitingNCT02889523Ib/IIDiffuse large B cell lymphomaSuspendedNCT03217253IMetastatic or unresectable solid tumorActive,not recruitingNCT03348631IIRecurrent ovarian cancerSuspendedGSK2816126SAM competitive inhibitorNCT02082977IRelapsed/refractory diffuse large B cell lymphoma and transformed follicular lymphomaTerminatedCPI-1205SAM competitive inhibitorNCT02395601IB cell lymphomaCompletedCPI-0209Second generation inhibitor of EZH2NCT04104776I/IIAdvanced solid tumorRecruitingPF-06821497SAM competitive inhibitorNCT03460977IRelapsed/refractory small cell lung cancer,castration-resistant prostate cancer, and follicular lymphomaRecruitingSHR2554SAM competitive inhibitorNCT03603951IRelapsed/refractory mature lymphoid neoplasmRecruitingEZH1 and EZH2DS-3201b (Valemetostat tosylate)SAM competitive inhibitorNCT04102150IIRelapsed/refractory adult T cell leukemia/lymphomaRecruitingEEDMAK683Binds to EED and change overall shape of PRC2NCT02900651I/IIDiffuse large B cell lymphomaRecruitingBMI1PTC-596Phosphorylation of BMI-1 at two N-terminal sites which leads to the degradation of BMI-1NCT03605550IbHigh grade glioma and diffuse intrinsic pontine gliomaRecruitingNCT03206645IOvarian cancerRecruitingNCT02404480IAdvanced solid cancerCompletedNCT03761095ILeiomyosarcomaRecruitingaInformation in Table 6 was acquired from https://clinicaltrials.gov/.',\n",
       "  [('NCT02395601', 'AN'),\n",
       "   ('NCT03348631', 'AN'),\n",
       "   ('NCT03761095', 'AN'),\n",
       "   ('NCT03217253', 'AN'),\n",
       "   ('NCT02875548', 'AN'),\n",
       "   ('NCT03206645', 'AN'),\n",
       "   ('NCT03603951', 'AN'),\n",
       "   ('NCT02404480', 'AN'),\n",
       "   ('NCT02860286', 'AN'),\n",
       "   ('NCT02889523', 'AN'),\n",
       "   ('NCT02900651', 'AN'),\n",
       "   ('NCT03460977', 'AN'),\n",
       "   ('NCT02601950', 'AN'),\n",
       "   ('NCT01897571', 'AN'),\n",
       "   ('NCT02082977', 'AN'),\n",
       "   ('NCT04104776', 'AN'),\n",
       "   ('NCT04102150', 'AN'),\n",
       "   ('NCT02601937', 'AN'),\n",
       "   ('NCT03213665', 'AN'),\n",
       "   ('NCT03605550', 'AN')]],\n",
       " ['PMC10068818',\n",
       "  '(B) Close-up of the boxed region from (A) showing the rigid-body docked crystal structure of KIT(T417I, ∆418-419) D4D5 fragment dimers (PDB ID 4PGZ, protomer A in beige, protomer B in blue).',\n",
       "  [('4PGZ', 'AN')]],\n",
       " ['PMC10068818',\n",
       "  '(C) Cryo-EM map resulting from a masked refinement of the extracellular domain with two copies of the crystal structure of monomeric KIT (PDB ID 2EC8) docked into the map.',\n",
       "  [('2EC8', 'AN')]],\n",
       " ['PMC10068818',\n",
       "  'As starting model for the cryo-EM model of the ECD of wild-type KIT:SCF, the crystal structure of the truncated ECD of KIT:SCF (PDB ID 2E9W) was docked into the cryo-EM map using UCSF Chimera.',\n",
       "  [('2E9W', 'AN')]],\n",
       " ['PMC10068818',\n",
       "  'Due to a lack of high-resolution structural data for D5, AlphaFold (40) structure predictions were used for PDGFRα (UniProt P16234, AF-P16234-F1-model_v2), PDGFRβ (UniProt P09619, AF-P09619-F1-model_v2), CSF1R (UniProt P07333, AF-P07333-F1-model_v2), and FLT3 (UniProt P36888, AF-P36888-F1-model_v2).',\n",
       "  [('P16234', 'AN'),\n",
       "   ('UniProt', 'RS'),\n",
       "   ('AF-P16234-F1', 'AN'),\n",
       "   ('P36888', 'AN'),\n",
       "   ('P09619', 'AN'),\n",
       "   ('AF-P09619-F1', 'AN'),\n",
       "   ('AF-P07333-F1', 'AN'),\n",
       "   ('P07333', 'AN')]],\n",
       " ['PMC10508174',\n",
       "  'X-ray intensities were processed using the HKL2000 package.66 An initial electron density map was determined by molecular replacement using Phenix.68 The DDX39B (PDB ID 1XTJ) structure was used as a molecular replacement model.',\n",
       "  [('1XTJ', 'AN')]],\n",
       " ['PMC10508174',\n",
       "  'The DDX39B/Tho1/RNA structure, colored as in Figure 2A, is overlayed on the Sub2/Yra1/RNA structure (PDB: 5SUP), colored in gray except for Yra1 (orange).',\n",
       "  [('5SUP', 'AN')]],\n",
       " ['PMC10508174',\n",
       "  '(B) THO complex (PDB: 7LUV) and DIM bind to overlapping regions (indicated by an oval) on DDX39B-CTD.',\n",
       "  [('7LUV', 'AN')]],\n",
       " ['PMC10508174',\n",
       "  'RNA Features (mRNA length, pre-mRNA length, exon count, mean exon length, mean intron length, %GC content of full-length mRNA, 5′UTR, CDS, and 3′UTR) of export inhibited and not affected transcripts were extracted based on Ensembl GTF annotations (GRCh38, release version 104) for each Ensembl Canonical transcript.',\n",
       "  [('Ensembl', 'RS')]]]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_data[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "b323b253",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "c1dfb7af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "b'<!DOCTYPE article PUBLIC \"-//NLM//DTD JATS (Z39.96) Journal Archiving and Interchange DTD with MathML3 v1.2 20190208//EN\" \"JATS-archivearticle1-mathml3.dtd\"> \\n<article xmlns:mml=\"http://www.w3.org/1998/Math/MathML\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" xmlns:ali=\"http://www.niso.org/schemas/ali/1.0/\" article-type=\"review-article\"><?properties open_access?><?DTDIdentifier.IdentifierValue -//Springer-Verlag//DTD A++ V2.4//EN?><?DTDIdentifier.IdentifierType public?><?SourceDTD.DTDName A++V2.4.dtd?><?SourceDTD.Version 2.4?><?ConverterInfo.XSLTName springer2nlmx2.xsl?><?ConverterInfo.Version 1?><front><journal-meta><journal-id journal-id-type=\"nlm-ta\">Exp Mol Med</journal-id><journal-id journal-id-type=\"iso-abbrev\">Exp Mol Med</journal-id><journal-title-group><journal-title>Experimental &#x00026; Molecular Medicine</journal-title></journal-title-group><issn pub-type=\"ppub\">1226-3613</issn><issn pub-type=\"epub\">2092-6413</issn><publisher><publisher-name>Nature Publishing Group UK</publisher-name><publisher-loc>London</publisher-loc></publisher></journal-meta><article-meta><article-id pub-id-type=\"pmcid\">8080698</article-id><article-id pub-id-type=\"pmid\">32636442</article-id><article-id pub-id-type=\"publisher-id\">462</article-id><article-id pub-id-type=\"doi\">10.1038/s12276-020-0462-5</article-id><article-categories><subj-group subj-group-type=\"heading\"><subject>Review Article</subject></subj-group></article-categories><title-group><article-title>Structural heterogeneity of the mammalian polycomb repressor complex in immune regulation</article-title></title-group><contrib-group><contrib contrib-type=\"author\"><name><surname>Kang</surname><given-names>Seok-Jin</given-names></name><xref ref-type=\"aff\" rid=\"Aff1\"/></contrib><contrib contrib-type=\"author\" corresp=\"yes\"><contrib-id contrib-id-type=\"orcid\">http://orcid.org/0000-0002-5940-8620</contrib-id><name><surname>Chun</surname><given-names>Taehoon</given-names></name><address><email>tchun@korea.ac.kr</email></address><xref ref-type=\"aff\" rid=\"Aff1\"/></contrib><aff id=\"Aff1\"><institution-wrap><institution-id institution-id-type=\"GRID\">grid.222754.4</institution-id><institution-id institution-id-type=\"ISNI\">0000 0001 0840 2678</institution-id><institution>Department of Biotechnology, College of Life Sciences and Biotechnology, </institution><institution>Korea University, </institution></institution-wrap>Seoul, 02841 Republic of Korea </aff></contrib-group><pub-date pub-type=\"epub\"><day>7</day><month>7</month><year>2020</year></pub-date><pub-date pub-type=\"pmc-release\"><day>7</day><month>7</month><year>2020</year></pub-date><pub-date pub-type=\"collection\"><month>7</month><year>2020</year></pub-date><volume>52</volume><issue>7</issue><fpage>1004</fpage><lpage>1015</lpage><history><date date-type=\"received\"><day>3</day><month>1</month><year>2020</year></date><date date-type=\"rev-recd\"><day>21</day><month>4</month><year>2020</year></date><date date-type=\"accepted\"><day>25</day><month>5</month><year>2020</year></date></history><permissions><copyright-statement>&#x000a9; The Author(s) 2020</copyright-statement><license><ali:license_ref specific-use=\"textmining\" content-type=\"ccbylicense\">https://creativecommons.org/licenses/by/4.0/</ali:license_ref><license-p><bold>Open Access</bold> This article is licensed under a Creative Commons Attribution 4.0 International License, which permits use, sharing, adaptation, distribution and reproduction in any medium or format, as long as you give appropriate credit to the original author(s) and the source, provide a link to the Creative Commons license, and indicate if changes were made. The images or other third party material in this article are included in the article&#x02019;s Creative Commons license, unless indicated otherwise in a credit line to the material. If material is not included in the article&#x02019;s Creative Commons license and your intended use is not permitted by statutory regulation or exceeds the permitted use, you will need to obtain permission directly from the copyright holder. To view a copy of this license, visit <ext-link ext-link-type=\"uri\" xlink:href=\"https://creativecommons.org/licenses/by/4.0/\">http://creativecommons.org/licenses/by/4.0/</ext-link>.</license-p></license></permissions><abstract id=\"Abs1\"><p id=\"Par1\">Epigenetic regulation is mainly mediated by enzymes that can modify the structure of chromatin by altering the structure of DNA or histones. Proteins involved in epigenetic processes have been identified to study the detailed molecular mechanisms involved in the regulation of specific mRNA expression. Evolutionarily well-conserved polycomb group (PcG) proteins can function as transcriptional repressors by the trimethylation of histone H3 at the lysine 27 residue (H3K27me3) and the monoubiquitination of histone H2A at the lysine 119 residue (H2AK119ub). PcG proteins form two functionally distinct protein complexes: polycomb repressor complex 1 (PRC1) and PRC2. In mammals, the structural heterogeneity of each PRC complex is dramatically increased by several paralogs of its subunit proteins. Genetic studies with transgenic mice along with RNA-seq and chromatin immunoprecipitation (ChIP)-seq analyses might be helpful for defining the cell-specific functions of paralogs of PcG proteins. Here, we summarize current knowledge about the immune regulatory role of PcG proteins related to the compositional diversity of each PRC complex and introduce therapeutic drugs that target PcG proteins in hematopoietic malignancy.</p></abstract><abstract id=\"Abs2\" abstract-type=\"LongSummary\"><title>Immunology: Gene repressors regulating immunity offer cancer drug targets</title><p id=\"Par2\">Protein complexes that suppress gene activity by remodeling chromatin, the substance that contains most of a cell&#x02019;s DNA, play a critical role in regulating the immune system and provide a therapeutic target for treating blood cancers. Seok-Jin Kang and Taehoon Chun from Korea University in Seoul, South Korea, review how polycomb group proteins, best known for their function in embryonic development, also contribute to the formation of immune cells from blood stem cell precursors. Studies with stem cells and cancer cells have begun to reveal many targets of these proteins, and drug companies are evaluating candidate agents directed against some polycomb group proteins in patients with lymphoma and other cancers. More comprehensive profiling of protein function across a broad range of immune cell types could reveal new targets for additional diseases associated with immune dysfunction.</p></abstract><kwd-group kwd-group-type=\"npg-subject\"><title>Subject terms</title><kwd>Immunogenetics</kwd><kwd>Haematopoiesis</kwd></kwd-group><funding-group><award-group><funding-source><institution-wrap><institution-id institution-id-type=\"FundRef\">https://doi.org/10.13039/501100003725</institution-id><institution>National Research Foundation of Korea (NRF)</institution></institution-wrap></funding-source><award-id>2019R1A2C2088046</award-id><award-id>2019R1A2C2088046</award-id><principal-award-recipient><name><surname>Kang</surname><given-names>Seok-Jin</given-names></name><name><surname>Chun</surname><given-names>Taehoon</given-names></name></principal-award-recipient></award-group></funding-group><custom-meta-group><custom-meta><meta-name>issue-copyright-statement</meta-name><meta-value>&#x000a9; The Author(s) 2020</meta-value></custom-meta></custom-meta-group></article-meta></front><body><sec id=\"Sec1\" sec-type=\"introduction\"><title>Introduction</title><p id=\"Par3\">In eukaryotes, the alteration of chromatin structure is one of the main methods for modifying cell phenotypes by regulating specific DNA replication and mRNA transcription<sup><xref ref-type=\"bibr\" rid=\"CR1\">1</xref></sup>. In addition to DNA methylation, changing the properties of certain amino acid residues at histones is a major method for modifying the structure of chromatin. The enzymes involved in the acetylation, methylation, ubiquitination, and phosphorylation of histones have been identified and extensively studied to define the biological function of each enzyme<sup><xref ref-type=\"bibr\" rid=\"CR2\">2</xref></sup>. Many studies have provided evidence that histone modification plays a decisive role in cell fates such as carcinogenesis, differentiation, proliferation, and senescence<sup><xref ref-type=\"bibr\" rid=\"CR3\">3</xref></sup>.</p><p id=\"Par4\">Polycomb group (PcG) proteins were originally identified from fruit flies. They are well conserved from invertebrates to mammals during evolution. PcG proteins can act as transcriptional repressors by inhibiting the mRNA transcription of specific gene loci through the trimethylation or monoubiquitination of histones H3 and H2A<sup><xref ref-type=\"bibr\" rid=\"CR4\">4</xref></sup>. To initiate and maintain such chromatin modification, two distinct protein complexes, polycomb repressor complex 1 (PRC1) and PRC2, work in coordination with each other. PRC2 exhibits methyltransferase activity to add methyl functional groups to specific amino acid residues of histone H3, while PRC1 exhibits E3 ubiquitin-ligase activity to modify the structure of histone H2A<sup><xref ref-type=\"bibr\" rid=\"CR4\">4</xref>,<xref ref-type=\"bibr\" rid=\"CR5\">5</xref></sup>. Mammalian PRC complexes display structural plasticity because the existence of several paralogs of PcG subunit proteins<sup><xref ref-type=\"bibr\" rid=\"CR6\">6</xref></sup>. In particular, more than 100 different types of mammalian PRC1 complexes may exist based on a simple combinatorial algorithm<sup><xref ref-type=\"bibr\" rid=\"CR7\">7</xref></sup>.</p><p id=\"Par5\">Although recent progress in biochemical and molecular analyses involving transgenic animal techniques has revealed the functional importance of the core subunit of the PcG proteins that regulate mRNA expression through histone modifications, how each of the paralogs of PcG subunit protein interact with each other to orchestrate the fine tuning of chromatin structure remains elusive. In this review, we summarize current knowledge about the immune regulatory role of PcG proteins related to the compositional diversity of each PRC complex. We also introduce therapeutic drugs that target PcG proteins.</p></sec><sec id=\"Sec2\"><title>Structural heterogeneity related to the function of PcG proteins</title><p id=\"Par6\">PcG genes were initially identified as genes involved in the regulation of homeotic gene expression, critical for the body axis plan and segment development in fruit flies<sup><xref ref-type=\"bibr\" rid=\"CR8\">8</xref></sup>. PcG proteins are present in plants, nematodes, and metazoan species from flies to mammals, indicating that these proteins are well-conserved transcriptional repressors via the modification of chromatin structure during evolution<sup><xref ref-type=\"bibr\" rid=\"CR9\">9</xref></sup>. Each PcG protein is a subunit of multiprotein complexes categorized by two different functional groups: PRC1 and PRC2<sup><xref ref-type=\"bibr\" rid=\"CR10\">10</xref></sup>.</p><p id=\"Par7\">Embryonic ectoderm development (EED), suppressor of zeste (SUZ)12, and enhancer of zeste homolog (EZH) are the catalytic core subunits of PRC2. Since EZH has two paralogs (EZH1 and EZH2), two structural variants are found in the catalytic core of PRC2 (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1a</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR11\">11</xref></sup>. EZH2 is the enzymatic subunit of the PRC2 complex, which acts as an S-adenosyl-<sc>l</sc>-methionine (SAM)-dependent histone methyltransferase via the mono-, di-, or trimethylation of lysine 27 residue at histone H3 (H3K27me1, H3K27me2, or H3K27me3) (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1</xref> and Table <xref rid=\"Tab1\" ref-type=\"table\">1</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR11\">11</xref>&#x02013;<xref ref-type=\"bibr\" rid=\"CR16\">16</xref></sup>. EZH1 also acts as a methyltransferase with reduced enzyme activity compared to EZH2<sup><xref ref-type=\"bibr\" rid=\"CR10\">10</xref></sup>. The SET domain of EZH1 or EZH2, which contains the catalytic core and SAM-binding site, is indispensable for their methyltransferase activity. However, purified EZH1 or EZH2 monomers alone are unable to efficiently exert enzyme activity in vitro because they must bind with two other noncatalytic subunit proteins, SUZ12 and EED (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1</xref>, Tables <xref rid=\"Tab1\" ref-type=\"table\">1</xref> and <xref rid=\"Tab2\" ref-type=\"table\">2</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR9\">9</xref>,<xref ref-type=\"bibr\" rid=\"CR11\">11</xref>,<xref ref-type=\"bibr\" rid=\"CR12\">12</xref>,<xref ref-type=\"bibr\" rid=\"CR17\">17</xref>&#x02013;<xref ref-type=\"bibr\" rid=\"CR21\">21</xref></sup>. SUZ12 contains a zinc-finger domain that can bind to DNA or RNA and facilitate protein&#x02013;protein interactions<sup><xref ref-type=\"bibr\" rid=\"CR22\">22</xref></sup>. EED contains WD40 repeats that can putatively bind to H3K27me3 (Table <xref rid=\"Tab2\" ref-type=\"table\">2</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR23\">23</xref></sup>. The fourth member of the PRC2 core subunit is retinoblastoma-binding protein 4 (RBBP4) (NURF55) or RBBP7 (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1</xref>, Tables <xref rid=\"Tab1\" ref-type=\"table\">1</xref> and <xref rid=\"Tab2\" ref-type=\"table\">2</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR9\">9</xref>,<xref ref-type=\"bibr\" rid=\"CR12\">12</xref>,<xref ref-type=\"bibr\" rid=\"CR18\">18</xref>,<xref ref-type=\"bibr\" rid=\"CR19\">19</xref>,<xref ref-type=\"bibr\" rid=\"CR24\">24</xref>,<xref ref-type=\"bibr\" rid=\"CR25\">25</xref></sup>. Whether RBBP 4/7 is included in the catalytic core of PRC2 is still controversial because RBBP 4/7 activity is not required for the catalytic activity of PRC2 in vitro<sup><xref ref-type=\"bibr\" rid=\"CR26\">26</xref></sup>. However, RBBP 4/7 also contains WD40 domains that can bind to histones and facilitate the catalytic activity of PRC2 in vivo<sup><xref ref-type=\"bibr\" rid=\"CR26\">26</xref></sup>.<fig id=\"Fig1\"><label>Fig. 1</label><caption><title>The repressive mechanism of specific mRNA transcription by PcG proteins through the modification of chromatin structure.</title><p>Schematic representation of transcriptional repression by PcG proteins according to the &#x02018;hierarchical repressive model&#x02019; (a) and the &#x02018;reverse-hierarchical repressive model&#x02019; (<bold>b</bold>). <bold>a</bold> Core subunits of PRC2 (EED, EZH, SUZ12, RBBP) recognize and repress a target locus by introducing H3K27me3. The CBX subunit of canonical PRC1 (PRC1.2 and PRC1.4) then recognizes the H3K27me3 tag, and canonical PRC1 further represses the target locus by introducing H2AK119. <bold>b</bold> The KDM2B subunit of noncanonical PRC1 (PRC1.1) recognizes CpG, and PRC1.1 represses the target locus by introducing H2AK119. The JARID2 subunit of PRC2.2 then recognizes the H2AK119 tag, and PRC2.2 further represses the target locus by introducing H3K27me3.</p></caption><graphic xlink:href=\"12276_2020_462_Fig1_HTML\" id=\"d32e434\"/></fig><table-wrap id=\"Tab1\"><label>Table 1</label><caption><p>Each subunit of canonical and noncanonical polycomb complexes in mammals.</p></caption><table frame=\"hsides\" rules=\"groups\"><thead><tr><th>PRC</th><th>Core subunits</th><th>Classification</th><th>Other subunits</th><th>Reference</th></tr></thead><tbody><tr><td>PRC2.1</td><td>EZH, EED, SUZ12, RBBP</td><td>N.D.</td><td>PCL, EPOP (C17orf96), LCOR (C10orf12)</td><td rowspan=\"2\"><sup><xref ref-type=\"bibr\" rid=\"CR11\">11</xref></sup></td></tr><tr><td>PRC2.2</td><td>EZH, EED, SUZ12, RBBP</td><td>N.D.</td><td>AEBP2, JARID2</td></tr><tr><td rowspan=\"3\">PRC1.1</td><td>PCGF1 (RNF68, NSPC1), RING1</td><td>Noncanonical</td><td>RYBP, KDM2B (FBXL10), BCOR, SKP1</td><td rowspan=\"11\"><sup><xref ref-type=\"bibr\" rid=\"CR12\">12</xref></sup></td></tr><tr><td>PCGF1 (RNF68, NSPC1), RING1</td><td>Noncanonical</td><td>YAF2, KDM2B (FBXL10), BCOR, SKP1</td></tr><tr><td>PCGF1 (RNF68, NSPC1), RING1</td><td>Noncanonical</td><td>CBX8, KDM2B (FBXL10), BCOR, SKP1</td></tr><tr><td>PRC1.2</td><td>PCGF2 (MEL-18), RING1</td><td>Canonical</td><td>CBX, PHC, SCMH</td></tr><tr><td rowspan=\"2\">PRC1.3</td><td>PCGF3 (RNF3), RING1</td><td>Noncanonical</td><td>RYBP, AUTS2, FBRS, CKII (CSNK2A1)</td></tr><tr><td>PCGF3 (RNF3), RING1</td><td>Noncanonical</td><td>YAF2, AUTS2, FBRS, CKII (CSNK2A1)</td></tr><tr><td>PRC1.4</td><td>PCGF4 (BMI-1), RING1</td><td>Canonical</td><td>CBX, PHC, SCMH</td></tr><tr><td rowspan=\"2\">PRC1.5</td><td>PCGF5 (RNF159), RING1</td><td>Noncanonical</td><td>RYBP, AUTS2, FBRS, CKII (CSNK2A1)</td></tr><tr><td>PCGF5 (RNF159), RING1</td><td>Noncanonical</td><td>YAF2, AUTS2, FBRS, CKII (CSNK2A1)</td></tr><tr><td rowspan=\"2\">PRC1.6</td><td>PCGF6 (RNF134, MBLR), RING1</td><td>Noncanonical</td><td>RYBP, WDR5, L3MBTL2, ESF6-TDP1, MGA-MAX, CBX3 (HP1&#x003b3;)</td></tr><tr><td>PCGF6 (RNF134, MBLR), RING1</td><td>Noncanonical</td><td>YAF2, WDR5, L3MBTL2, ESF6-TDP1, MGA-MAX, CBX3 (HP1&#x003b3;)</td></tr></tbody></table><table-wrap-foot><p><italic>N.D.</italic> not determined.</p></table-wrap-foot></table-wrap><table-wrap id=\"Tab2\"><label>Table 2</label><caption><p>Paralogs of each subunit of PRC2 and canonical PRC1 in mammals.</p></caption><table frame=\"hsides\" rules=\"groups\"><thead><tr><th>Core subunit of PRC2</th><th>Homologs in fruit flies</th><th>Paralogs in mammals</th><th>Protein domain</th><th>Function</th><th>References</th></tr></thead><tbody><tr><td>EED</td><td>ESC</td><td>none</td><td>WD40 repeat</td><td>Binding to H3K27me3</td><td><sup><xref ref-type=\"bibr\" rid=\"CR9\">9</xref></sup></td></tr><tr><td rowspan=\"2\">EZH</td><td rowspan=\"2\">E(Z)</td><td>EZH1</td><td rowspan=\"2\">SET domain</td><td rowspan=\"2\">H3K27 methyltransferase</td><td rowspan=\"2\"><sup><xref ref-type=\"bibr\" rid=\"CR17\">17</xref></sup></td></tr><tr><td>EZH2</td></tr><tr><td>SUZ12</td><td>SU(Z)12</td><td>none</td><td>Zinc-finger domain</td><td>DNA/RNA binding and protein&#x02013;protein interaction</td><td><sup><xref ref-type=\"bibr\" rid=\"CR17\">17</xref>,<xref ref-type=\"bibr\" rid=\"CR18\">18</xref></sup></td></tr><tr><td rowspan=\"2\">RBBP</td><td rowspan=\"2\">NURF55</td><td>RBBP4</td><td rowspan=\"2\">WD40 repeat</td><td rowspan=\"2\">Binding to unmodified nucleosomes</td><td rowspan=\"2\"><sup><xref ref-type=\"bibr\" rid=\"CR18\">18</xref>,<xref ref-type=\"bibr\" rid=\"CR19\">19</xref></sup></td></tr><tr><td>RBBP7</td></tr></tbody></table><table frame=\"hsides\" rules=\"groups\"><thead><tr><th>Core subunit of canonical PRC1</th><th>Fruit fly</th><th>Mammalian</th><th>Protein domain</th><th>Function</th><th>References</th></tr></thead><tbody><tr><td rowspan=\"2\">RING1</td><td rowspan=\"2\">RING</td><td>RING1A (RING1)</td><td rowspan=\"2\">RING finger domain</td><td rowspan=\"2\">Nucleosome binding and H2AK119 mono-ubiquitin ligase</td><td rowspan=\"2\"><sup><xref ref-type=\"bibr\" rid=\"CR20\">20</xref></sup></td></tr><tr><td>RING1B (RNF2)</td></tr><tr><td rowspan=\"6\">PCGF</td><td rowspan=\"6\">PSC</td><td>PCGF1</td><td rowspan=\"6\">RING finger domain and RAWUL domain</td><td rowspan=\"6\">H2AK119 mono-ubiquitin ligase and protein&#x02013;protein interaction</td><td rowspan=\"6\"><sup><xref ref-type=\"bibr\" rid=\"CR21\">21</xref></sup></td></tr><tr><td>MEL18 (PCGF2)</td></tr><tr><td>PCGF3</td></tr><tr><td>BMI1 (PCGF4)</td></tr><tr><td>PCGF5</td></tr><tr><td>PCGF6</td></tr></tbody></table></table-wrap></p><p id=\"Par8\">In addition to the core subunits of PRC2, several other proteins can bind to these subunits and modulate the enzyme activity of PRC2. Two different types of PRC2 complexes (PRC 2.1 and PRC 2.2) have been identified based on their noncore subunit proteins in humans (Table <xref rid=\"Tab1\" ref-type=\"table\">1</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR11\">11</xref>,<xref ref-type=\"bibr\" rid=\"CR12\">12</xref>,<xref ref-type=\"bibr\" rid=\"CR27\">27</xref></sup>. PRC2.1 contains three other subunits, including polycomb-like protein (PCL), PRC2-associated LCOR isoform (PALI), elongin B/C and PRC2-associated protein (EPOP) (Table <xref rid=\"Tab1\" ref-type=\"table\">1</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR11\">11</xref>,<xref ref-type=\"bibr\" rid=\"CR12\">12</xref>,<xref ref-type=\"bibr\" rid=\"CR28\">28</xref>&#x02013;<xref ref-type=\"bibr\" rid=\"CR30\">30</xref></sup>. PCL has three paralogs: PCL1, PCL2, and PCL3. They are also known as PHF1 (PCL1), MTF2 (PCL2), and PHF19 (PCL3), respectively. PALI, also known as C10ORF12, has two paralogs: PALI1 and PALI2<sup><xref ref-type=\"bibr\" rid=\"CR28\">28</xref>&#x02013;<xref ref-type=\"bibr\" rid=\"CR30\">30</xref></sup>. Three noncore subunit proteins (PCL, PALI and EPOP) can act as enhancers to facilitate the catalytic activity of PRC 2.1. The function of PCL is essential for H3K27me3 by PRC 2.1 because the recognition of H3K36me2/3 by the TUDOR domain of PCL is a prerequisite for PRC 2.1 to introduce H3K27me3 marks<sup><xref ref-type=\"bibr\" rid=\"CR31\">31</xref></sup>. PCL is also required for the recognition of unmethylated CpG islands of DNA by PRC 2.1<sup><xref ref-type=\"bibr\" rid=\"CR32\">32</xref>,<xref ref-type=\"bibr\" rid=\"CR33\">33</xref></sup>. PALI1 can facilitate the catalytic activity of PRC2 both in vitro and in vivo<sup><xref ref-type=\"bibr\" rid=\"CR34\">34</xref></sup>. Similar to the phenotype of EZH2-deficient mice, PALI1-deficient mice exhibit embryonic lethality<sup><xref ref-type=\"bibr\" rid=\"CR34\">34</xref></sup>. EPOP can mediate the interaction between PRC2.1 and elongin B/C, which is important for maintaining the transcriptional repression of PRC2&#x02019;s target locus<sup><xref ref-type=\"bibr\" rid=\"CR35\">35</xref></sup>.</p><p id=\"Par9\">Adipocyte enhancer-binding protein 2 (AEBP2) and Jumonji AT-rich interactive domain 2 (JARID2) are additional subunits that for PRC 2.2 along with the PRC2 core subunits (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1</xref> and Table <xref rid=\"Tab1\" ref-type=\"table\">1</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR11\">11</xref>,<xref ref-type=\"bibr\" rid=\"CR12\">12</xref>,<xref ref-type=\"bibr\" rid=\"CR27\">27</xref>,<xref ref-type=\"bibr\" rid=\"CR36\">36</xref></sup>. Both AEBP2 and JARID2 are required to recruit PRC 2.2 to chromatin by specifically binding to the CpG-rich region of DNA<sup><xref ref-type=\"bibr\" rid=\"CR36\">36</xref></sup>. Recent studies have indicated that Jarid2-containing PRC 2.2 can specifically recognize and bind to the mono-ubiquitinated lysine 119 residue at histone H2A (H2AK119Ub) tagged by the PRC1.1 (noncanonical PRC 1) complex (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1b</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR37\">37</xref></sup>. The binding of H2AK119Ub by Jarid2 can further facilitate the methyltransferase activity of PRC 2.2 (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1b</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR37\">37</xref></sup>.</p><p id=\"Par10\">The subunits of PRC1 complexes are much more diverse than those of PRC2 (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1b</xref>, Table <xref rid=\"Tab1\" ref-type=\"table\">1</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR11\">11</xref>,<xref ref-type=\"bibr\" rid=\"CR12\">12</xref></sup>. There are two groups of PRC1 complexes categorized based on the original findings in fruit flies. Canonical PRC1 complexes are composed of subunit proteins conserved from flies to mammals, whereas the subunit proteins of noncanonical PRC1 complexes are less conserved in flies<sup><xref ref-type=\"bibr\" rid=\"CR38\">38</xref></sup>. Really interesting new gene 1 (RING) and polycomb group ring finger (PCGF) have been found in both canonical and noncanonical PRC1 complexes, suggesting that these proteins are structurally and functionally essential components<sup><xref ref-type=\"bibr\" rid=\"CR38\">38</xref></sup>. RING proteins exhibit two paralogs (RING1A and RING1B) that possess E3 ubiquitin ligase activity when they are combined with PCGF proteins (H2AK119Ub activity) (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1</xref>, Tables <xref rid=\"Tab1\" ref-type=\"table\">1</xref> and <xref rid=\"Tab2\" ref-type=\"table\">2</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR9\">9</xref>,<xref ref-type=\"bibr\" rid=\"CR11\">11</xref>,<xref ref-type=\"bibr\" rid=\"CR12\">12</xref>,<xref ref-type=\"bibr\" rid=\"CR17\">17</xref>&#x02013;<xref ref-type=\"bibr\" rid=\"CR21\">21</xref>,<xref ref-type=\"bibr\" rid=\"CR38\">38</xref>,<xref ref-type=\"bibr\" rid=\"CR39\">39</xref></sup>. PCGF proteins exhibit six paralogs (PCGF1&#x02013;PCGF6)<sup><xref ref-type=\"bibr\" rid=\"CR20\">20</xref></sup>. Upon interaction with RING proteins, PCGF proteins can increase ubiquitin ligase activity by acting as cofactors<sup><xref ref-type=\"bibr\" rid=\"CR39\">39</xref>,<xref ref-type=\"bibr\" rid=\"CR40\">40</xref></sup>. Each PCGF paralog (PCGF1 through PCGF6) can be a subunit of different types of PRC1 complexes (PRC1.1 through PRC 1.6) (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1a</xref>, Tables <xref rid=\"Tab1\" ref-type=\"table\">1</xref> and <xref rid=\"Tab2\" ref-type=\"table\">2</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR9\">9</xref>,<xref ref-type=\"bibr\" rid=\"CR11\">11</xref>,<xref ref-type=\"bibr\" rid=\"CR12\">12</xref></sup>.</p><p id=\"Par11\">Among the six different types of PRC1 complexes (PRC1.1&#x02013;PRC 1.6), the PCGF2 (MEL-18)-containing PRC1.2 and PCGF4 (BMI-1)-containing PRC1.4 complexes are classified as canonical PRC1 complexes<sup><xref ref-type=\"bibr\" rid=\"CR41\">41</xref></sup>. Chromobox homologs (CBX) can form canonical PRC1 complexes with RING proteins and PCGF2 (MEL-18) or PCGF4 (BMI-1) (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1a</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR41\">41</xref></sup>. Five CBX paralogs (CBX2, CBX4, CBX6, CBX7, CBX8) have been found to act as subunits of the canonical PRC1 complex in mammals (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1a</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR41\">41</xref></sup>. The proposed role of CBX in the canonical PRC1 complex is to recruit PRC1 to H3K27me3 tags because CBX proteins contain chromodomains that recognize the H3K27me3 tag introduced by PRC2 (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1a</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR42\">42</xref>,<xref ref-type=\"bibr\" rid=\"CR43\">43</xref></sup>. Additionally, polyhomeotic homolog (PHC) and sex comb on midleg homolog (SCMH) can interact with core proteins (RING and PCGF) to form canonical PRC1 complexes (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1a</xref>, Table <xref rid=\"Tab1\" ref-type=\"table\">1</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR44\">44</xref>,<xref ref-type=\"bibr\" rid=\"CR45\">45</xref></sup>. PHC proteins exhibit three paralogs (PHC1-PHC3)<sup><xref ref-type=\"bibr\" rid=\"CR3\">3</xref>,<xref ref-type=\"bibr\" rid=\"CR7\">7</xref></sup>. SCMH proteins exhibit two paralogs (SCMH1 and SCMH2)<sup><xref ref-type=\"bibr\" rid=\"CR46\">46</xref></sup>. Both types of proteins contain a sterile &#x003b1; motif domain that allows them to bind to other canonical PRC1 complex proteins and participate in the recruitment of PRC1 to chromatin (Table <xref rid=\"Tab1\" ref-type=\"table\">1</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR44\">44</xref>,<xref ref-type=\"bibr\" rid=\"CR45\">45</xref></sup>. PHC proteins also have zinc-finger domains that facilitate nucleic acid binding and chromatin compaction<sup><xref ref-type=\"bibr\" rid=\"CR46\">46</xref></sup>.</p><p id=\"Par12\">The noncanonical PRC1 complex is composed of more protein subunits (Table <xref rid=\"Tab1\" ref-type=\"table\">1</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR11\">11</xref>,<xref ref-type=\"bibr\" rid=\"CR12\">12</xref></sup>. In the noncanonical PRC1 complex, the core subunits (RING1 and PCGF) can interact with ring and YY1 binding protein (RYBP) or YY1-associated factor 2 (YAF2) or CBX8 (Table <xref rid=\"Tab1\" ref-type=\"table\">1</xref>) via <italic>C</italic>-terminal ring finger and WD40 ubiquitin-like (RAWUL) domains<sup><xref ref-type=\"bibr\" rid=\"CR12\">12</xref>,<xref ref-type=\"bibr\" rid=\"CR47\">47</xref></sup>. Previous observations have indicated that RYBP can compete with CBX for the binding site of RING1B<sup><xref ref-type=\"bibr\" rid=\"CR48\">48</xref></sup>. YAF2 and RYBP occur in the noncanonical PRC1 complex in a mutually exclusive manner, since YAF2 is a homolog of RYBP (Table <xref rid=\"Tab1\" ref-type=\"table\">1</xref>).</p><p id=\"Par13\">The function of the noncanonical PRC1 complex is clearly different from that of the canonical PRC1 complex (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1</xref>). According to the &#x02018;hierarchical repressive model&#x02019;, PRC2 can repress a target locus via an H3K27me3 tag. The canonical PRC1 complex can recognize this methylation tag through CBX and further repress a target locus by introducing a H2AK119 mark (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1a</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR43\">43</xref></sup>. Recently, the RYBP-containing noncanonical PRC1 complex has been found to show higher E3 ligase activity than PCGF4-RING1B containing canonical PRC1 complex<sup><xref ref-type=\"bibr\" rid=\"CR49\">49</xref></sup>. This finding suggests that another pathway for transcriptional repression exists in addition to the &#x02018;hierarchical repressive model&#x02019;. Indeed, the CxxC DNA-binding domain of KDM2B in the PRC1.1 complex can specifically recognize CpG DNA sequences and recruit PRC1.1 to a target locus<sup><xref ref-type=\"bibr\" rid=\"CR50\">50</xref>,<xref ref-type=\"bibr\" rid=\"CR51\">51</xref></sup>. PRC1.1 then suppresses specific mRNA transcription via an H2AK119ub tag (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1b</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR50\">50</xref>,<xref ref-type=\"bibr\" rid=\"CR51\">51</xref></sup>. Thereafter, PRC 2.2-containing Jarid2 can specifically recognize and bind the H2AK119Ub tag and further modify the structure of chromatin by introducing an H3K27me3 tag (Fig. <xref rid=\"Fig1\" ref-type=\"fig\">1b</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR37\">37</xref>,<xref ref-type=\"bibr\" rid=\"CR50\">50</xref>,<xref ref-type=\"bibr\" rid=\"CR51\">51</xref></sup>. This model is known as the &#x02018;reverse hierarchical repressive model&#x02019; because PRC1.1 first represses the specific transcription of mRNA instead of PRC2.2.</p><p id=\"Par14\">In fruit flies, putative DNA regions recognized by PRCs have been identified, validated, and designated as PcG/trithorax-group response elements (PREs)<sup><xref ref-type=\"bibr\" rid=\"CR2\">2</xref>&#x02013;<xref ref-type=\"bibr\" rid=\"CR7\">7</xref></sup>. The existence of vertebrate PRE sites around CpG-rich sequences has also been suggested<sup><xref ref-type=\"bibr\" rid=\"CR36\">36</xref>,<xref ref-type=\"bibr\" rid=\"CR52\">52</xref></sup>. However, the conserved DNA-binding motif of mammalian PRCs and the detailed mechanism by which mammalian PRCs recognize specific DNA regions remain elusive. In fruit flies, it has been suggested that the pleiohomeotic (Pho) protein can recognize PREs and guide the core subunits of PRC1 and PRC2 to PREs since the core subunits of PRC2 or PRC1 do not directly bind to DNA<sup><xref ref-type=\"bibr\" rid=\"CR53\">53</xref></sup>. In vertebrates, YinYang1 (YY1), a Pho homolog, can bind to a conserved DNA region and interact with PRC1 subunits<sup><xref ref-type=\"bibr\" rid=\"CR54\">54</xref></sup>. Therefore, YY1 may recognize PRE sites and guide noncanonical PRC1 by interacting with RYBP or YAF2<sup><xref ref-type=\"bibr\" rid=\"CR55\">55</xref></sup>.</p></sec><sec id=\"Sec3\"><title>The role of PcG proteins in immune regulation</title><p id=\"Par15\">A knockout (KO) mouse model and the cell type-specific deletion of PcG genes generated in a conditional knockout (cKO) mouse model using the cre-lox system have been used in most studies to study the function of PcG proteins in immune regulation (Table <xref rid=\"Tab3\" ref-type=\"table\">3</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR56\">56</xref>&#x02013;<xref ref-type=\"bibr\" rid=\"CR82\">82</xref></sup>. Except gene encoding RBBP, mice deficient in the genes encoding each core subunit of PRC2 have been generated and characterized (Table <xref rid=\"Tab3\" ref-type=\"table\">3</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR56\">56</xref>&#x02013;<xref ref-type=\"bibr\" rid=\"CR82\">82</xref></sup>. Based on animal studies, Ezh1 can partially replace the function of EZh2 in specific cell types<sup><xref ref-type=\"bibr\" rid=\"CR83\">83</xref>,<xref ref-type=\"bibr\" rid=\"CR84\">84</xref></sup>. For example, Ezh2 is not required for the self-renewal activity of long-term hematopoietic stem cells (LT-HSCs) in adult bone marrow<sup><xref ref-type=\"bibr\" rid=\"CR64\">64</xref></sup>. However, Ezh1-deficient mice exhibit immunodeficiency due to a significant loss of the self-renewal activity of HSCs<sup><xref ref-type=\"bibr\" rid=\"CR56\">56</xref></sup>. Because the <italic>INK4a</italic>/<italic>Arf</italic> locus, encoding <italic>p16INK4a</italic> and <italic>p19Arf</italic>, which can suppress cell cycle progression, is a target of PcG-mediated repression, the deficiency of certain core subunits of PRC2 and canonical PRC1 can cause the loss of self-renewal activity of HSCs<sup><xref ref-type=\"bibr\" rid=\"CR85\">85</xref></sup>. In addition to EZH1 deficiency, insufficiency of other subunits of PRC2, including EED or SUZ12, can lead to the loss of the self-renewal activity of HSCs<sup><xref ref-type=\"bibr\" rid=\"CR64\">64</xref>,<xref ref-type=\"bibr\" rid=\"CR66\">66</xref></sup>. The deficiency of some canonical subunits of PRC1, including BMI-1 and PHC1, can also cause the loss of self-renewal activity of HSCs<sup><xref ref-type=\"bibr\" rid=\"CR67\">67</xref>,<xref ref-type=\"bibr\" rid=\"CR68\">68</xref>,<xref ref-type=\"bibr\" rid=\"CR75\">75</xref></sup>. However, other canonical subunits of PRC1, including MEL18, CBX2, CBX8, and PHC2, do not influence the self-renewal activity of HSCs<sup><xref ref-type=\"bibr\" rid=\"CR72\">72</xref>,<xref ref-type=\"bibr\" rid=\"CR76\">76</xref>,<xref ref-type=\"bibr\" rid=\"CR77\">77</xref>,<xref ref-type=\"bibr\" rid=\"CR79\">79</xref></sup>. These phenotypic variations observed in each of the mice deficient in different PcG subunits reflect structural heterogeneity depending on the specific stage of cells or tissues due to the redundancy or paralogs of each PcG subunit (Table <xref rid=\"Tab3\" ref-type=\"table\">3</xref>). Cell type-specific roles of various PRC1 and PRC2 complexes have already suggested (Fig. <xref rid=\"Fig2\" ref-type=\"fig\">2</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR12\">12</xref></sup>. In support of these ideas, EZH2 expression in LT-HSCs peaks on embryonic day 14.5 and gradually decreases thereafter until 10 months postnatal<sup><xref ref-type=\"bibr\" rid=\"CR64\">64</xref></sup>. However, EZH1 expression in LT-HSCs gradually increases from embryonic day 14.5 to 10 months after birth<sup><xref ref-type=\"bibr\" rid=\"CR64\">64</xref></sup>. BMI-1 and MEL18 expression patterns also follow the paradigm of EZH1/2 expression. BMI-1 is mainly expressed in specific lineage precursors of immune cells, whereas the expression of MEL-18 is correlated with mature immune cell populations<sup><xref ref-type=\"bibr\" rid=\"CR86\">86</xref></sup>. In addition to contributing to the self-renewal activity of HSCs, PcG proteins participate in the differentiation of hematopoietic progenitor cells (HPCs) into specific lineages of immune cells. The contributions of PRC2 and canonical PRC1 to immune cell differentiation according to the &#x02018;hierarchical repressive model&#x02019; are summarized in Table <xref rid=\"Tab3\" ref-type=\"table\">3</xref> and Fig. <xref rid=\"Fig2\" ref-type=\"fig\">2</xref>.<table-wrap id=\"Tab3\"><label>Table 3</label><caption><p>Hematopoietic cell fate decision by PRC2 and canonical PRC1 according to the &#x02018;hierarchical repressive model&#x02019; in mice.</p></caption><table frame=\"hsides\" rules=\"groups\"><thead><tr><th>Complex</th><th>Gene</th><th>mouse type</th><th>Abnormality of specific gene expression</th><th>Phenotype in mice</th><th>Reference</th></tr></thead><tbody><tr><td rowspan=\"14\">PRC2</td><td>Ezh1</td><td>KO</td><td><italic>Gata3, Runx1, Meis1, Myb Dntt, Flk2, Igh6</italic>, and <italic>Ikaros</italic></td><td>Loss of self-renewal activity of HSC and defective in B cell development</td><td><sup><xref ref-type=\"bibr\" rid=\"CR56\">56</xref></sup></td></tr><tr><td rowspan=\"8\">Ezh2</td><td>KO</td><td>None</td><td>Embryonic lethality at embryonic day 7.5</td><td><sup><xref ref-type=\"bibr\" rid=\"CR57\">57</xref></sup></td></tr><tr><td>cKO (Tie2-Cre)</td><td><italic>Gata2</italic></td><td>Embryonic lethality due to fatal anemia</td><td><sup><xref ref-type=\"bibr\" rid=\"CR58\">58</xref></sup></td></tr><tr><td>cKO (ERT-Cre)</td><td><italic>Gata2</italic></td><td>Defects in B cell development</td><td><sup><xref ref-type=\"bibr\" rid=\"CR58\">58</xref></sup></td></tr><tr><td>cKO (CD4-Cre)</td><td><italic>T-bet, Eomes</italic>, and <italic>Gata3</italic></td><td>Enhanced Th cell plasticity (spontaneous Th<sub>1</sub> or Th<sub>2</sub> polarization without stimulation)</td><td><sup><xref ref-type=\"bibr\" rid=\"CR59\">59</xref></sup></td></tr><tr><td>cKO (CD4-Cre)</td><td><italic>Ifng</italic>, <italic>Gata3</italic>, and <italic>Il10</italic></td><td>Increased apoptosis of effector Th<sub>0</sub> cells</td><td><sup><xref ref-type=\"bibr\" rid=\"CR60\">60</xref></sup></td></tr><tr><td>cKO (GzmB-Cre)</td><td><italic>Foxo1</italic></td><td>Fewer antigen-specific CD8<sup>+</sup> T cells</td><td><sup><xref ref-type=\"bibr\" rid=\"CR61\">61</xref></sup></td></tr><tr><td>cKO (Vav1-Cre)</td><td><italic>Klrk1, Il2ra, Il7r, Cxcr3, Ccr7</italic>, and <italic>Xcr1</italic></td><td>Increased number of NK cells</td><td><sup><xref ref-type=\"bibr\" rid=\"CR62\">62</xref></sup></td></tr><tr><td>cKO (CD4-Cre)</td><td><italic>Cd160, Zbtb16, Ccr4, Il4, Ifng</italic>, and <italic>Il12</italic></td><td>Increased number of <italic>invariant</italic> NKT cells</td><td><sup><xref ref-type=\"bibr\" rid=\"CR63\">63</xref></sup></td></tr><tr><td rowspan=\"4\">Eed</td><td>KO</td><td><italic>Ckdn2a, Cdkn1a, Cdkn2b, HoxC4, p21</italic>, and <italic>Wig1</italic></td><td>Severe anemia and leukopenia</td><td><sup><xref ref-type=\"bibr\" rid=\"CR64\">64</xref></sup></td></tr><tr><td>cKO (Vav1-Cre)</td><td><italic>Ckdn2a, Cdkn1a, Cdkn2b, HoxC4, p21</italic>, and <italic>Wig1</italic></td><td>Pancytopenia without defect of hematopoietic stem cell (HSC)</td><td><sup><xref ref-type=\"bibr\" rid=\"CR64\">64</xref></sup></td></tr><tr><td>cKO (Mx1-Cre)</td><td>None</td><td>Higher frequency of spontaneous T cell leukemia</td><td><sup><xref ref-type=\"bibr\" rid=\"CR65\">65</xref></sup></td></tr><tr><td>cKO (CD4-Cre)</td><td><italic>Foxo1</italic></td><td>Fewer antigen-specific CD8<sup>+</sup> T cells</td><td><sup><xref ref-type=\"bibr\" rid=\"CR61\">61</xref></sup></td></tr><tr><td>Suz12</td><td>cKO (Vav1-Cre)</td><td>None</td><td>Loss of HSC function and defects in lymphocyte development</td><td><sup><xref ref-type=\"bibr\" rid=\"CR66\">66</xref></sup></td></tr><tr><td rowspan=\"19\">Canonical PRC1</td><td rowspan=\"5\">Bmi1</td><td>KO</td><td>None</td><td>Less proliferative activity of leukemic stem and progenitor cells</td><td><sup><xref ref-type=\"bibr\" rid=\"CR67\">67</xref></sup></td></tr><tr><td>KO</td><td><italic>p16, p19, Wig1, Tjp1</italic>, and <italic>Hoxa9</italic></td><td>Loss of self-renewal activity of HSC</td><td><sup><xref ref-type=\"bibr\" rid=\"CR68\">68</xref></sup></td></tr><tr><td>KO x OTII</td><td><italic>Ink4a/Arf, Bax, Puma, Noxa, Bad</italic>, and <italic>Fas</italic></td><td>Increased apoptosis of memory Th<sub>2</sub> cells</td><td><sup><xref ref-type=\"bibr\" rid=\"CR69\">69</xref></sup></td></tr><tr><td>KO</td><td><italic>Cdkn2a</italic></td><td>Impaired cell expansion of double negative (DN) thymocytes</td><td><sup><xref ref-type=\"bibr\" rid=\"CR70\">70</xref></sup></td></tr><tr><td>KO</td><td><italic>Cdkn2a</italic></td><td>Decreased Th<sub>2</sub> cell polarization</td><td><sup><xref ref-type=\"bibr\" rid=\"CR71\">71</xref></sup></td></tr><tr><td rowspan=\"2\">Mel18</td><td>KO</td><td>None</td><td>No significant defect of HSC function</td><td><sup><xref ref-type=\"bibr\" rid=\"CR72\">72</xref></sup></td></tr><tr><td>KO</td><td><italic>Gata3</italic></td><td>Impaired Th<sub>2</sub> cell polarization</td><td><sup><xref ref-type=\"bibr\" rid=\"CR73\">73</xref></sup></td></tr><tr><td rowspan=\"2\">Phc1</td><td>KO</td><td>None</td><td>Perinatal lethality</td><td><sup><xref ref-type=\"bibr\" rid=\"CR74\">74</xref></sup></td></tr><tr><td>Reconstitution of E14.5 HSC with KO</td><td>None</td><td>Reduced number of lymphocytes</td><td><sup><xref ref-type=\"bibr\" rid=\"CR75\">75</xref></sup></td></tr><tr><td>Phc2</td><td>KO</td><td><italic>Vcam1</italic></td><td>Defective mobilization of hematopoietic stem and progenitor cell (HSPC)</td><td><sup><xref ref-type=\"bibr\" rid=\"CR76\">76</xref></sup></td></tr><tr><td>Cbx2</td><td>KO</td><td>None</td><td>Reduced number of thymocytes and splenocytes without defect of HSC and defective in T cell development</td><td><sup><xref ref-type=\"bibr\" rid=\"CR77\">77</xref></sup></td></tr><tr><td rowspan=\"3\">Cbx4</td><td>KO</td><td>None</td><td>Thymic hypoplasia</td><td rowspan=\"3\"><sup><xref ref-type=\"bibr\" rid=\"CR78\">78</xref></sup></td></tr><tr><td>cKO (Foxn1-Cre)</td><td><italic>K14, Cd80</italic>, and <italic>Cd86</italic></td><td>Thymic hypoplasia</td></tr><tr><td>cKO (Lck-Cre)</td><td>None</td><td>No apparent defects</td></tr><tr><td rowspan=\"2\">Cbx8</td><td>cKO (ERT-Cre)</td><td><italic>Hoxa9</italic></td><td>No apparent defects</td><td><sup><xref ref-type=\"bibr\" rid=\"CR79\">79</xref></sup></td></tr><tr><td>cKO (C&#x003b3;1-Cre)</td><td><italic>Cdkn1a, Prdm1</italic>, and <italic>Irf4</italic></td><td>Defects in germinal center formation</td><td><sup><xref ref-type=\"bibr\" rid=\"CR80\">80</xref></sup></td></tr><tr><td rowspan=\"2\">Ring1B</td><td>cKO (Mx1-Cre)</td><td><italic>Cdkn2a</italic> and <italic>Ccnd2</italic></td><td>Hypocellular bone marrow</td><td rowspan=\"2\"><sup><xref ref-type=\"bibr\" rid=\"CR81\">81</xref></sup></td></tr><tr><td>cKO (ER<sup>T2</sup>-Cre)</td><td><italic>Cdkn2a</italic> and <italic>Ccnd2</italic></td><td>Hypocellular bone marrow</td></tr><tr><td>Rng1A and Ring1B</td><td>Double cKO (Lck-Cre)</td><td><italic>Pax5, Ebf1, Irf4, and Irf8</italic></td><td>Defects in B cell development</td><td><sup><xref ref-type=\"bibr\" rid=\"CR82\">82</xref></sup></td></tr></tbody></table><table-wrap-foot><p><italic>KO</italic> knockout mice, <italic>cKO</italic> conditional knockout mice.</p></table-wrap-foot></table-wrap><fig id=\"Fig2\"><label>Fig. 2</label><caption><title>Functional contribution of PcG proteins during immune cell differentiation.</title><p>Schematic representation of particular PRC2 or PRC1 complexes involved in hematopoiesis according to the &#x02018;hierarchical repressive model&#x02019;.</p></caption><graphic xlink:href=\"12276_2020_462_Fig2_HTML\" id=\"d32e2023\"/></fig></p><p id=\"Par16\">Studies on the importance of PcG proteins in immune cell function are much less common than studies on the influence of PcG proteins during the differentiation of immune cells (Table <xref rid=\"Tab3\" ref-type=\"table\">3</xref>). Most studies on the functional contribution of PcG proteins to immune cell function have focused on T cell function (Table <xref rid=\"Tab3\" ref-type=\"table\">3</xref>). The CD8<sup>+</sup> T cell-specific deletion of <italic>Ezh2</italic> or <italic>Eed</italic> using the CD4-Cre or granzyme B-Cre system revealed that the antigen-specific activation of CD8<sup>+</sup> T cells requires the function of the PRC2 complex (Table <xref rid=\"Tab3\" ref-type=\"table\">3</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR61\">61</xref></sup>. Interestingly, the contribution of PcG proteins to CD4<sup>+</sup> T cell function is controversial because the phenotypes of each of the PcG protein-deficient mice are quite different from each other. For example, CD4<sup>+</sup> T cell-specific <italic>Ezh2</italic> deletion has led to type 2 helper T cell (Th<sub>2</sub>)-prone immunity via the accumulation of memory Th<sub>2</sub> cells, which exacerbates allergic diseases (Table <xref rid=\"Tab3\" ref-type=\"table\">3</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR59\">59</xref></sup>. However, <italic>Bmi1</italic> and <italic>Mel18</italic> knockout mice are defective in Th<sub>2</sub> cell differentiation<sup><xref ref-type=\"bibr\" rid=\"CR71\">71</xref>,<xref ref-type=\"bibr\" rid=\"CR73\">73</xref></sup>. Furthermore, <italic>Bmi1</italic> knockout mice exhibit the enhanced apoptosis of memory Th<sub>2</sub> cells<sup><xref ref-type=\"bibr\" rid=\"CR69\">69</xref></sup>.</p></sec><sec id=\"Sec4\"><title>Current RNA-seq and chromatin immunoprecipitation (ChIP)-seq data for identifying the target loci of PcG proteins</title><p id=\"Par17\">The phenotypic analysis of transgenic mice in combination with RNA-seq and chromatin immunoprecipitation (ChIP)-seq analyses might be a good approach for identifying additional target loci of PcG proteins or the unique functions of each PcG protein paralog in specific immune cell types. Table <xref rid=\"Tab4\" ref-type=\"table\">4</xref> summarizes current gene chip and RNA-seq databases generated from specific cell types of transgenic mice or specific cell lines subjected to the inhibition of PcG function (<ext-link ext-link-type=\"uri\" xlink:href=\"http://www.ebi.ac.uk/arrayexpress/\">http://www.ebi.ac.uk/arrayexpress/</ext-link>). Most of the RNA-seq data were acquired from embryonic stem cells (ES cells), HSCs (LSK cells, LSK, Lin<sup>&#x02212;</sup>Sca-1<sup>+</sup>c-kit<sup>+</sup> cells), hematopoietic stem and progenitor cells (HSPC), and cancer cell lines including leukemia, multiple myeloma, sarcoma, ovarian tumor, and gastric cancer cell lines because the initial identification of PcG function emphasized the maintenance of self-renewal activity (Table <xref rid=\"Tab4\" ref-type=\"table\">4</xref>). To expand the collection of differentially expressed gene (DEG) data, RNA-seq analyses need to be performed using a broad range of immune cells, including B cells, monocytes, dendritic cells, mast cells, and polymorphonuclear cells. All DEGs identified in PcG-defective cells might not be direct targets of PcG proteins. Chip-seq data might be needed to verify whether these DEGs are direct targets of PcG proteins. Table <xref rid=\"Tab5\" ref-type=\"table\">5</xref> summarizes the current ChIP-seq databases for specific cell types (<ext-link ext-link-type=\"uri\" xlink:href=\"http://www.ebi.ac.uk/arrayexpress/\">http://www.ebi.ac.uk/arrayexpress/</ext-link>). The DNA-binding sites of most core subunits of PRC2, except for RBBP, and the core subunits of PRC1.2 (RING1B and MEL18) have been analyzed by ChIP-seq (Table <xref rid=\"Tab5\" ref-type=\"table\">5</xref>). The DNA-binding sites of some paralogs of CBX and Jarid2, a subunit of PRC2.2, have also been analyzed (Table <xref rid=\"Tab5\" ref-type=\"table\">5</xref>). However, most of the ChIP-seq data were acquired from stem cell lineages with few exceptions (Table <xref rid=\"Tab5\" ref-type=\"table\">5</xref>). Therefore, a broad range of cells need to be analyzed by Chip-seq using antibodies against the remainder of the PcG proteins, including RBBP, BMI-1, and PHC, to identify novel target genes repressed by PcG proteins.<table-wrap id=\"Tab4\"><label>Table 4</label><caption><p>RNA-seq or gene-chip data from loss or gain of function of each PcG subunit protein<sup>a, b</sup>.</p></caption><table frame=\"hsides\" rules=\"groups\"><thead><tr><th>Unit</th><th>Data ID</th><th>Target cell (cell line)</th><th>Functional alteration (method)</th></tr></thead><tbody><tr><td rowspan=\"2\">EZH1</td><td>E-GEOD-36288</td><td>LSK (Lin<sup>&#x02212;</sup>Sca-1<sup>+</sup>c-kit<sup>+</sup>) cells</td><td>Loss of function (KO)</td></tr><tr><td>E-GEOD-59090</td><td>CD34<sup>+</sup> hematopoietic stem and progenitor cell (HSPC)-derived proerythroblasts</td><td>Loss of function (shRNA)</td></tr><tr><td>EZH1/ EZH2</td><td>E-GEOD-62198</td><td>MLL-AF9 induced leukemia progenitor</td><td>Loss of function (inhibitor, UNC1999)</td></tr><tr><td rowspan=\"11\">EZH2</td><td>E-GEOD-59090</td><td>CD34<sup>+</sup> HSPC-derived proerythroblasts</td><td>Loss of function (shRNA)</td></tr><tr><td>E-GEOD-71870</td><td>Ovarian tumor (OC8 cell line)</td><td>Loss of function (shRNA)</td></tr><tr><td>E-GEOD-71671</td><td>Monocyte (THP-1 cell line)</td><td>Loss of function (inhibitor, GSK126)</td></tr><tr><td>E-MTAB-3227</td><td>Gastric cancer (MKN45 cell line)</td><td>Loss of function (siRNA)</td></tr><tr><td>E-GEOD-82072</td><td>Megakaryocyte-erythrocyte precursor</td><td>Loss of function (cKO; scl-cre)</td></tr><tr><td>E-GEOD-82073</td><td>Long-term-hematopoietic stem cell (LT-HSC)</td><td>Loss of function (cKO; scl-cre)</td></tr><tr><td>E-MTAB-2893</td><td>Chronic myeloid leukemia CD34<sup>+</sup> cells</td><td>Loss of function (inhibitor, GSK343)</td></tr><tr><td>E-MTAB-3552</td><td>Nonchronic myeloid leukemia CD34<sup>+</sup> cells</td><td>Loss of function (inhibitor, GSK343)</td></tr><tr><td>E-MTAB-5766</td><td>Acute myeloid leukemia cell (F-36P, MOLM-13, and OCI-M2 cell line)</td><td>Loss of function (shRNA)</td></tr><tr><td>E-MTAB-7739</td><td>Macroglobulinemia cell (RPCI-WM1 cell line)</td><td>Loss of function (inhibitor, Tazemetostat)</td></tr><tr><td>GSE101316</td><td>Mouse bone marrow-derived macrophage</td><td>Loss of function (KO)</td></tr><tr><td rowspan=\"6\">EED</td><td>E-GEOD-12982</td><td>Embryonic stem cell (ES cell)</td><td>Loss of function (KO)</td></tr><tr><td>E-GEOD-49305</td><td>ES cell</td><td>Loss of function (KO)</td></tr><tr><td>E-GEOD-62198</td><td>MLL-AF9 induced leukemia progenitor</td><td>Loss of function (shRNA)</td></tr><tr><td>E-GEOD-53508</td><td>ES cell</td><td>Loss of function (KO)</td></tr><tr><td>E-GEOD-53506</td><td>ES cell</td><td>Loss of function (KO)</td></tr><tr><td>E-GEOD-59090</td><td>CD34<sup>+</sup> HSPC-derived proerythroblasts</td><td>Loss of function (shRNA)</td></tr><tr><td rowspan=\"4\">SUZ12</td><td>E-GEOD-31354</td><td>ES cell</td><td>Loss of function (KO)</td></tr><tr><td>E-GEOD-53508</td><td>ES cell</td><td>Loss of function (Genetrap)</td></tr><tr><td>E-GEOD-59090</td><td>CD34<sup>+</sup> HSPC-derived proerythroblasts</td><td>Loss of function (shRNA)</td></tr><tr><td>E-GEOD-60808</td><td>HSPC</td><td>Loss of function (shRNA)</td></tr><tr><td>Jarid2</td><td>E-GEOD-60808</td><td>HSPC</td><td>Loss of function (shRNA)</td></tr><tr><td>RING1A</td><td>E-MTAB-5661</td><td>ES cell</td><td>Loss of function (KO)</td></tr><tr><td>RING1A/B</td><td>E-MTAB-5661</td><td>ES cell</td><td>Loss of function (KO, cKO)</td></tr><tr><td rowspan=\"3\">RING1B</td><td>E-GEOD-67868</td><td>ES cell</td><td>Loss of function (KO)</td></tr><tr><td>E-GEOD-69824</td><td>ES cell</td><td>Loss of function (KO)</td></tr><tr><td>E-GEOD-71007</td><td>Ewing&#x02019;s sarcoma (A4573, A673, ES1, and TC71 cell line)</td><td>Loss of function (shRNA)</td></tr><tr><td>Mel18</td><td>E-GEOD-67868</td><td>Embryonic stem cell</td><td>Loss of function (shRNA)</td></tr><tr><td rowspan=\"6\">Bmi1</td><td>E-GEOD-21912</td><td>Multiple myeloma(RPMI-8226 cell line)</td><td>Loss of function (shRNA)</td></tr><tr><td>E-GEOD-19796</td><td>HSPC</td><td>Loss of function (KO)</td></tr><tr><td>E-GEOD-20958</td><td>ES cell and HSC</td><td>Gain of function (overexpression)</td></tr><tr><td>E-GEOD-31086</td><td>Common myeloid progenitor</td><td>Loss of function (KO)</td></tr><tr><td>E-GEOD-54262</td><td>Erythroleukemia (K562 cell line) and Chronic myeloid leukemia</td><td>Loss of function (shRNA)</td></tr><tr><td>E-GEOD-71007</td><td>Ewing&#x02019;s sarcoma (A673 and TC71 cell line)</td><td>Loss of function (shRNA)</td></tr><tr><td>Cbx7</td><td>E-GEOD-34191</td><td>ES cell</td><td>Loss of function (shRNA)</td></tr><tr><td>PCGF1</td><td>E-GEOD-33280</td><td>HSPC</td><td>Loss of function (shRNA)</td></tr><tr><td>PCGF3/5</td><td>E-MTAB-5642</td><td>ES cell</td><td>Loss of function (KO)</td></tr></tbody></table><table-wrap-foot><p><italic>KO</italic> knockout mice, <italic>cKO</italic> conditional knockout mice.</p><p><sup>a</sup>Information in Table 4 was acquired from <ext-link ext-link-type=\"uri\" xlink:href=\"https://www.ebi.ac.uk/arrayexpress/\">https://www.ebi.ac.uk/arrayexpress/</ext-link>.</p><p><sup>b</sup>Each reference for Table 4 is contained within contents of each Data ID.</p></table-wrap-foot></table-wrap><table-wrap id=\"Tab5\"><label>Table 5</label><caption><p>ChIP-seq data of each PcG subunit protein<sup>a, b</sup>.</p></caption><table frame=\"hsides\" rules=\"groups\"><thead><tr><th>Unit</th><th>Data ID</th><th>Target cell (cell line)</th></tr></thead><tbody><tr><td>EZH1</td><td>E-GEOD-59090</td><td>CD34<sup>+</sup> hematopoietic stem and progenitor (HSPC)-derived proerythroblasts</td></tr><tr><td rowspan=\"18\">EZH2</td><td>E-GEOD-18776</td><td>Mouse embryonic stem cell (ES cell)</td></tr><tr><td>E-MTAB-1305</td><td>Human ES cell</td></tr><tr><td>E-GEOD-51079</td><td>In vitro cultured Th<sub>1</sub> and Th<sub>2</sub> cell</td></tr><tr><td>E-GEOD-42706</td><td>mouse resting B cell (CD43<sup>-</sup> B cell)</td></tr><tr><td>E-GEOD-49178</td><td>ES cell</td></tr><tr><td>E-GEOD-52300</td><td>Human liver cancer (HepG2 cell line)</td></tr><tr><td>E-GEOD-48518</td><td>induced pluripotent stem cell</td></tr><tr><td>E-GEOD-46536</td><td>ES cell</td></tr><tr><td>E-GEOD-53495</td><td>Human embryonic kidney cell (293T Rex cell line)</td></tr><tr><td>E-GEOD-57632</td><td>Multiple myeloma</td></tr><tr><td>E-GEOD-61586</td><td>Neural stem cell and glioma (SF7761 cell line)</td></tr><tr><td>E-MTAB-2002</td><td>Mouse ES cell</td></tr><tr><td>E-GEOD-47082</td><td>Mouse ES cell</td></tr><tr><td>E-GEOD-59090</td><td>CD34<sup>+</sup> HSPC-derived proerythroblasts</td></tr><tr><td>E-GEOD-70440</td><td>Mammary gland</td></tr><tr><td>E-GEOD-60160</td><td>Mouse ES cell</td></tr><tr><td>E-MTAB-6410</td><td>Chronic lymphocytic leukemia</td></tr><tr><td>GSE101320</td><td>Mouse bone marrow-derived macrophages</td></tr><tr><td rowspan=\"3\">EED</td><td>E-GEOD-61902</td><td>Spermatocyte</td></tr><tr><td>E-GEOD-59090</td><td>CD34<sup>+</sup> HSPC-derived proerythroblasts</td></tr><tr><td>E-MTAB-6165</td><td>Mouse ES cell</td></tr><tr><td rowspan=\"16\">SUZ12</td><td>E-GEOD-34483</td><td>Mouse ES cell</td></tr><tr><td>E-GEOD-42616</td><td>Mouse ES cell</td></tr><tr><td>E-GEOD-44286</td><td>Mouse ES cell</td></tr><tr><td>E-GEOD-52300</td><td>Human liver cancer (HepG2 cell line)</td></tr><tr><td>E-MTAB-2481</td><td>Mouse ES cell</td></tr><tr><td>E-GEOD-55698</td><td>ES cell</td></tr><tr><td>E-GEOD-52619</td><td>Mouse ES cell</td></tr><tr><td>E-GEOD-47528</td><td>Primary CD4<sup>+</sup> helper T cell</td></tr><tr><td>E-GEOD-58023</td><td>ES cell</td></tr><tr><td>E-GEOD-62437</td><td>MLL-AF9 transformed leukemia</td></tr><tr><td>E-GEOD-43915</td><td>Mouse E13.5 brain</td></tr><tr><td>E-GEOD-59090</td><td>CD34<sup>+</sup> HSPC-derived proerythroblasts</td></tr><tr><td>E-GEOD-57926</td><td>Mouse heart and embryonic fibroblast cell</td></tr><tr><td>E-GEOD-61148</td><td>Mouse thymocyte and thymic T cell tumor (ILC87 cell line)</td></tr><tr><td>E-GEOD-74330</td><td>Mouse ES cell</td></tr><tr><td>E-GEOD-83082</td><td>Mouse ES cell</td></tr><tr><td>Jarid2</td><td>E-GEOD-19708</td><td>Mouse ES cell</td></tr><tr><td rowspan=\"6\">RING1B</td><td>E-GEOD-23716</td><td>Mouse ES cell</td></tr><tr><td>E-GEOD-55698</td><td>ES cell</td></tr><tr><td>E-GEOD-43915</td><td>Mouse E13.5 brain</td></tr><tr><td>E-GEOD-67868</td><td>Mouse ES cell</td></tr><tr><td>E-GEOD-72164</td><td>Mouse ES cell</td></tr><tr><td>E-GEOD-74330</td><td>Mouse ES cell</td></tr><tr><td rowspan=\"2\">MEL18</td><td>E-GEOD-67868</td><td>Mouse ES cell</td></tr><tr><td>E-GEOD-74330</td><td>Mouse ES cell</td></tr><tr><td>CBX2</td><td>E-GEOD-29611</td><td>Erythroleukemia (K562 cell line)</td></tr><tr><td rowspan=\"4\">CBX3</td><td>E-GEOD-29611</td><td>Erythroleukemia (K562 cell line)</td></tr><tr><td>E-GEOD-32465</td><td>Human colon cancer (HCT116 cell line) and erythroleukemia (K562 cell line)</td></tr><tr><td>E-GEOD-28115</td><td>Human colon cancer (HCT116 cell line)</td></tr><tr><td>E-GEOD-44242</td><td>Mouse ES cell and induced pluripotent stem cell</td></tr><tr><td rowspan=\"3\">CBX7</td><td>E-GEOD-23716</td><td>Mouse ES cell</td></tr><tr><td>E-GEOD-42466</td><td>Mouse ES cell</td></tr><tr><td>E-GEOD-42706</td><td>mouse resting B cell (CD43<sup>-</sup> B cell)</td></tr><tr><td rowspan=\"2\">CBX8</td><td>E-GEOD-29611</td><td>Erythroleukemia (K562 cell line)</td></tr><tr><td>E-GEOD-54052</td><td>Mouse ES cell</td></tr></tbody></table><table-wrap-foot><p><sup>a</sup>Information in Table 5 was acquired from <ext-link ext-link-type=\"uri\" xlink:href=\"https://www.ebi.ac.uk/arrayexpress/\">https://www.ebi.ac.uk/arrayexpress/</ext-link>.</p><p><sup>b</sup>Each reference for Table 5 is contained within contents of each Data ID.</p></table-wrap-foot></table-wrap></p></sec><sec id=\"Sec5\"><title>Therapeutic agents for treating hematopoietic malignancies by inhibiting the activity of PcG proteins</title><p id=\"Par18\">Since the function of PcG proteins is important to maintain the self-renewal activity of stem cells, PcG proteins might act as oncogenes to facilitate tumorigenesis. In support of this idea, high expression of EZH2 has been observed in several hematopoietic malignancies, including myelodysplastic syndromes, acute myeloid leukemia, and various types of lymphomas<sup><xref ref-type=\"bibr\" rid=\"CR87\">87</xref>&#x02013;<xref ref-type=\"bibr\" rid=\"CR89\">89</xref></sup>. In particular, EZH2 deficiency in mice can inhibit leukemogenesis by decreasing the proliferation rate of leukemia<sup><xref ref-type=\"bibr\" rid=\"CR90\">90</xref></sup>. Consistent with these observations, the expression levels of canonical subunits of PRC1, including BMI1, CBX7, CBX8, and RING1A, are elevated in many hematopoietic-originating tumors<sup><xref ref-type=\"bibr\" rid=\"CR88\">88</xref>,<xref ref-type=\"bibr\" rid=\"CR91\">91</xref>,<xref ref-type=\"bibr\" rid=\"CR92\">92</xref></sup>. A mouse model involving <italic>Bmi1</italic>-deficient mice with transformed cells also supports the notion that BMI1 can act as an oncogene in some hematopoietic malignant cells<sup><xref ref-type=\"bibr\" rid=\"CR93\">93</xref></sup>. However, the loss of function of PcG proteins by mutation or deletion might also cause hematopoietic malignancies<sup><xref ref-type=\"bibr\" rid=\"CR91\">91</xref></sup>. In particular, defects in core subunits of PRC2, including EZH2, EED, and SUZ12, have been found in various acute lymphoblastic leukemia and myelodysplastic syndromes<sup><xref ref-type=\"bibr\" rid=\"CR94\">94</xref>&#x02013;<xref ref-type=\"bibr\" rid=\"CR97\">97</xref></sup>. Therefore, at least PRC2 can act as an oncogene or a tumor suppressor depending on the type of hematopoietic malignant cells involved<sup><xref ref-type=\"bibr\" rid=\"CR91\">91</xref></sup>. Further study is needed to define the mechanisms underlying the dual functions of these proteins in tumorigenesis.</p><p id=\"Par19\">Table <xref rid=\"Tab6\" ref-type=\"table\">6</xref> summarizes the inhibitors of PcG proteins applied to clinical trials in hematopoietic malignancies and other types of tumors. Major groups of inhibitors target EZH enzyme activity (Table <xref rid=\"Tab6\" ref-type=\"table\">6</xref>). Most EZH2 inhibitors undergoing clinical trials compete with SAM for binding to the SET domain<sup><xref ref-type=\"bibr\" rid=\"CR98\">98</xref></sup>. Among the competitive inhibitors of EZH, tazemetostat (EPZ-6438), an orally administered small chemical, has been applied to a broad range of malignant cell types, including lymphoma, sarcoma, mesothelioma, ovarian cancers and advanced solid tumors (Table <xref rid=\"Tab6\" ref-type=\"table\">6</xref>)<sup><xref ref-type=\"bibr\" rid=\"CR98\">98</xref></sup>. Other inhibitors of PcG proteins that are currently undergoing clinical trials target EED and BMI-1 activity (Table <xref rid=\"Tab6\" ref-type=\"table\">6</xref>). MAK683 is an allosteric EED inhibitor that drives conformational changes in the H3K27me3-binding pocket of EED upon binding<sup><xref ref-type=\"bibr\" rid=\"CR99\">99</xref></sup>. These conformational changes in EED further prevent the interaction between EED and EZH2, thus blocking H3K27me3<sup><xref ref-type=\"bibr\" rid=\"CR99\">99</xref></sup>. PTC596 is a BMI-1 inhibitor that can facilitate the degradation of BMI-1 by inducing the cyclin-dependent kinase 1-mediated biphosphorylation of the <italic>N</italic>-terminus of BMI-1<sup><xref ref-type=\"bibr\" rid=\"CR100\">100</xref></sup>.<table-wrap id=\"Tab6\"><label>Table 6</label><caption><p>Inhibitors of PcG proteins undergoing current clinical trials in malignant cells<sup>a</sup>.</p></caption><table frame=\"hsides\" rules=\"groups\"><thead><tr><th>Target polycomb subunit</th><th>Agent</th><th>Mode of action</th><th>NCT ID</th><th>Phase</th><th>Target tumor types</th><th>Status</th></tr></thead><tbody><tr><td rowspan=\"14\">EZH2</td><td rowspan=\"9\">Tazemetostat (EPZ-6438)</td><td rowspan=\"9\">S-adenosyl- <sub>L</sub>-methionine (SAM) competitive inhibitor</td><td>NCT02875548</td><td>II</td><td>Diffuse large B cell lymphoma</td><td>Recruiting</td></tr><tr><td>NCT02601950</td><td>II</td><td>INI (hSNF5; SMARCB1)-negative tumor and relapsed/refractory synovial sarcoma</td><td>Recruiting</td></tr><tr><td>NCT01897571</td><td>I/II</td><td>Advanced solid tumor and B cell lymphoma</td><td>Active, not recruiting</td></tr><tr><td>NCT02601937</td><td>I</td><td>INI (hSNF5; SMARCB1)-negative tumor and synovial sarcoma</td><td>Recruiting</td></tr><tr><td>NCT02860286</td><td>II</td><td>Malignant mesothelioma</td><td>Completed</td></tr><tr><td>NCT03213665</td><td>II</td><td>Relapsed/refractory advanced solid tumor and non-Hodgkin&#x02019;s lymphoma</td><td>Recruiting</td></tr><tr><td>NCT02889523</td><td>Ib/II</td><td>Diffuse large B cell lymphoma</td><td>Suspended</td></tr><tr><td>NCT03217253</td><td>I</td><td>Metastatic or unresectable solid tumor</td><td>Active,not recruiting</td></tr><tr><td>NCT03348631</td><td>II</td><td>Recurrent ovarian cancer</td><td>Suspended</td></tr><tr><td>GSK2816126</td><td>SAM competitive inhibitor</td><td>NCT02082977</td><td>I</td><td>Relapsed/refractory diffuse large B cell lymphoma and transformed follicular lymphoma</td><td>Terminated</td></tr><tr><td>CPI-1205</td><td>SAM competitive inhibitor</td><td>NCT02395601</td><td>I</td><td>B cell lymphoma</td><td>Completed</td></tr><tr><td>CPI-0209</td><td>Second generation inhibitor of EZH2</td><td>NCT04104776</td><td>I/II</td><td>Advanced solid tumor</td><td>Recruiting</td></tr><tr><td>PF-06821497</td><td>SAM competitive inhibitor</td><td>NCT03460977</td><td>I</td><td>Relapsed/refractory small cell lung cancer,castration-resistant prostate cancer, and follicular lymphoma</td><td>Recruiting</td></tr><tr><td>SHR2554</td><td>SAM competitive inhibitor</td><td>NCT03603951</td><td>I</td><td>Relapsed/refractory mature lymphoid neoplasm</td><td>Recruiting</td></tr><tr><td>EZH1 and EZH2</td><td>DS-3201b (Valemetostat tosylate)</td><td>SAM competitive inhibitor</td><td>NCT04102150</td><td>II</td><td>Relapsed/refractory adult T cell leukemia/lymphoma</td><td>Recruiting</td></tr><tr><td>EED</td><td>MAK683</td><td>Binds to EED and change overall shape of PRC2</td><td>NCT02900651</td><td>I/II</td><td>Diffuse large B cell lymphoma</td><td>Recruiting</td></tr><tr><td rowspan=\"4\">BMI1</td><td rowspan=\"4\">PTC-596</td><td rowspan=\"4\">Phosphorylation of BMI-1 at two <italic>N</italic>-terminal sites which leads to the degradation of BMI-1</td><td>NCT03605550</td><td>Ib</td><td>High grade glioma and diffuse intrinsic pontine glioma</td><td>Recruiting</td></tr><tr><td>NCT03206645</td><td>I</td><td>Ovarian cancer</td><td>Recruiting</td></tr><tr><td>NCT02404480</td><td>I</td><td>Advanced solid cancer</td><td>Completed</td></tr><tr><td>NCT03761095</td><td>I</td><td>Leiomyosarcoma</td><td>Recruiting</td></tr></tbody></table><table-wrap-foot><p><sup>a</sup>Information in Table 6 was acquired from <ext-link ext-link-type=\"uri\" xlink:href=\"https://clinicaltrials.gov/\">https://clinicaltrials.gov/</ext-link>.</p></table-wrap-foot></table-wrap></p></sec><sec id=\"Sec6\"><title>Conclusion and future prospects</title><p id=\"Par20\">In this review, we highlighted the structural diversity of mammalian PRC2 and PRC1 complexes related to their functional contribution to immune regulation. We also described currently available RNA-seq and ChIP-seq data that could be used to mine new target loci of PcG proteins. Finally, we listed the PcG inhibitors currently undergoing clinical trials. Many previous reports have demonstrated that PcG proteins are major chromatin modifiers that can modulate many biological processes by influencing specific gene repression, mainly using loss-of-function models.</p><p id=\"Par21\">Unfortunately, we still do not know how many different types of PRCs exist in nature due to structural heterogeneity caused by many paralogs and accessory proteins recruited by PRC complexes. We also do not know how each different PRC containing a particular paralog as a subunit contributes to the phenotype of a specific cell type. Solving these unknown issues might provide novel targets for PcG-mediated gene regulation and expand the range of PcG proteins considered as therapeutic targets to treat other human diseases in addition to cancer.</p></sec></body><back><fn-group><fn><p><bold>Publisher&#x02019;s note</bold> Springer Nature remains neutral with regard to jurisdictional claims in published maps and institutional affiliations.</p></fn></fn-group><ack><title>Acknowledgements</title><p>This work was supported by a National Research Foundation of Korea (NRF) grant funded by the Korean government (MSIT) (2019R1A2C2088046) and an intramural research program funded by a Korea University grant (2017).</p></ack><notes id=\"FPar1\" notes-type=\"COI-statement\"><title>Conflict of interest</title><p id=\"Par22\">The authors declare that they have no conflict of interest.</p></notes><ref-list id=\"Bib1\"><title>References</title><ref id=\"CR1\"><label>1.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Bannister</surname><given-names>AJ</given-names></name><name><surname>Kouzarides</surname><given-names>T</given-names></name></person-group><article-title>Regulation of chromatin by histone modifications</article-title><source>Cell Res.</source><year>2011</year><volume>21</volume><fpage>381</fpage><lpage>395</lpage><?supplied-pmid 21321607?><pub-id pub-id-type=\"pmid\">21321607</pub-id></element-citation></ref><ref id=\"CR2\"><label>2.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Greer</surname><given-names>EL</given-names></name><name><surname>Shi</surname><given-names>Y</given-names></name></person-group><article-title>Histone methylation: a dynamic mark in health, disease and inheritance</article-title><source>Nat. Rev. Genet.</source><year>2012</year><volume>13</volume><fpage>343</fpage><lpage>357</lpage><?supplied-pmid 22473383?><pub-id pub-id-type=\"pmid\">22473383</pub-id></element-citation></ref><ref id=\"CR3\"><label>3.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Sparmann</surname><given-names>A</given-names></name><name><surname>van Lohuizen</surname><given-names>M</given-names></name></person-group><article-title>Polycomb silencers control cell fate, development and cancer</article-title><source>Nat. Rev. Cancer</source><year>2006</year><volume>6</volume><fpage>846</fpage><lpage>856</lpage><?supplied-pmid 17060944?><pub-id pub-id-type=\"pmid\">17060944</pub-id></element-citation></ref><ref id=\"CR4\"><label>4.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Aranda</surname><given-names>S</given-names></name><name><surname>Mas</surname><given-names>G</given-names></name><name><surname>Di Croce</surname><given-names>L</given-names></name></person-group><article-title>Regulation of gene transcription by polycomb proteins</article-title><source>Sci. Adv.</source><year>2015</year><volume>1</volume><fpage>e1500737</fpage><?supplied-pmid 26665172?><pub-id pub-id-type=\"pmid\">26665172</pub-id></element-citation></ref><ref id=\"CR5\"><label>5.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Schwartz</surname><given-names>YB</given-names></name><name><surname>Pirrotta</surname><given-names>V</given-names></name></person-group><article-title>Ruled by ubiquitylation: a new order for polycomb recruitment</article-title><source>Cell Rep.</source><year>2014</year><volume>8</volume><fpage>321</fpage><lpage>325</lpage><?supplied-pmid 25061856?><pub-id pub-id-type=\"pmid\">25061856</pub-id></element-citation></ref><ref id=\"CR6\"><label>6.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Schuettengruber</surname><given-names>B</given-names></name><name><surname>Cavalli</surname><given-names>G</given-names></name></person-group><article-title>Recruitment of polycomb group complexes and their role in the dynamic regulation of cell fate choice</article-title><source>Development</source><year>2009</year><volume>136</volume><fpage>3531</fpage><lpage>3542</lpage><?supplied-pmid 19820181?><pub-id pub-id-type=\"pmid\">19820181</pub-id></element-citation></ref><ref id=\"CR7\"><label>7.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Connelly</surname><given-names>KE</given-names></name><name><surname>Dykhuizen</surname><given-names>EC</given-names></name></person-group><article-title>Compositional and functional diversity of canonical PRC1 complexes in mammals</article-title><source>Biochim. Biophys. Acta Gene Regul. Mech.</source><year>2017</year><volume>1860</volume><fpage>233</fpage><lpage>245</lpage><?supplied-pmid 28007606?><pub-id pub-id-type=\"pmid\">28007606</pub-id></element-citation></ref><ref id=\"CR8\"><label>8.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Oktaba</surname><given-names>K</given-names></name><etal/></person-group><article-title>Dynamic regulation by polycomb group protein complexes controls pattern formation and the cell cycle in <italic>Drosophila</italic></article-title><source>Dev. Cell</source><year>2008</year><volume>15</volume><fpage>877</fpage><lpage>889</lpage><?supplied-pmid 18993116?><pub-id pub-id-type=\"pmid\">18993116</pub-id></element-citation></ref><ref id=\"CR9\"><label>9.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Margueron</surname><given-names>R</given-names></name><etal/></person-group><article-title>Role of the polycomb protein EED in the propagation of repressive histone marks</article-title><source>Nature</source><year>2009</year><volume>461</volume><fpage>762</fpage><lpage>767</lpage><?supplied-pmid 19767730?><pub-id pub-id-type=\"pmid\">19767730</pub-id></element-citation></ref><ref id=\"CR10\"><label>10.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Chittock</surname><given-names>EC</given-names></name><name><surname>Latwiel</surname><given-names>S</given-names></name><name><surname>Miller</surname><given-names>TC</given-names></name><name><surname>M&#x000fc;ller</surname><given-names>CW</given-names></name></person-group><article-title>Molecular architecture of polycomb repressive complexes</article-title><source>Biochem. Soc. Trans.</source><year>2017</year><volume>45</volume><fpage>193</fpage><lpage>205</lpage><?supplied-pmid 28202673?><pub-id pub-id-type=\"pmid\">28202673</pub-id></element-citation></ref><ref id=\"CR11\"><label>11.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Laugesen</surname><given-names>A</given-names></name><name><surname>H&#x000f8;jfeldt</surname><given-names>JW</given-names></name><name><surname>Helin</surname><given-names>K</given-names></name></person-group><article-title>Molecular mechanisms directing PRC2 recruitment and H3K27 methylation</article-title><source>Mol. Cell</source><year>2019</year><volume>74</volume><fpage>8</fpage><lpage>18</lpage><?supplied-pmid 30951652?><pub-id pub-id-type=\"pmid\">30951652</pub-id></element-citation></ref><ref id=\"CR12\"><label>12.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Di Croce</surname><given-names>L</given-names></name><name><surname>Helin</surname><given-names>K</given-names></name></person-group><article-title>Transcriptional regulation by polycomb group proteins</article-title><source>Nat. Struct. Mol. Biol.</source><year>2013</year><volume>20</volume><fpage>1147</fpage><lpage>1155</lpage><?supplied-pmid 24096405?><pub-id pub-id-type=\"pmid\">24096405</pub-id></element-citation></ref><ref id=\"CR13\"><label>13.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Vidal</surname><given-names>M</given-names></name><name><surname>Starowicz</surname><given-names>K</given-names></name></person-group><article-title>Polycomb complexes PRC1 and their function in hematopoiesis</article-title><source>Exp. Hematol.</source><year>2017</year><volume>48</volume><fpage>12</fpage><lpage>31</lpage><?supplied-pmid 28087428?><pub-id pub-id-type=\"pmid\">28087428</pub-id></element-citation></ref><ref id=\"CR14\"><label>14.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Ferrari</surname><given-names>KJ</given-names></name><etal/></person-group><article-title>Polycomb-dependent H3K27me1 and H3K27me2 regulate active transcription and enhancer fidelity</article-title><source>Mol. Cell</source><year>2014</year><volume>53</volume><fpage>49</fpage><lpage>62</lpage><?supplied-pmid 24289921?><pub-id pub-id-type=\"pmid\">24289921</pub-id></element-citation></ref><ref id=\"CR15\"><label>15.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Sneeringer</surname><given-names>CJ</given-names></name><etal/></person-group><article-title>Coordinated activities of wild-type plus mutant EZH2 drive tumor-associated hypertrimethylation of lysine 27 on histone H3 (H3K27) in human B-cell lymphomas</article-title><source>Proc. Natl Acad. Sci. USA</source><year>2010</year><volume>107</volume><fpage>20980</fpage><lpage>20985</lpage><?supplied-pmid 21078963?><pub-id pub-id-type=\"pmid\">21078963</pub-id></element-citation></ref><ref id=\"CR16\"><label>16.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Zee</surname><given-names>BM</given-names></name><name><surname>Britton</surname><given-names>L-MP</given-names></name><name><surname>Wolle</surname><given-names>D</given-names></name><name><surname>Haberman</surname><given-names>DM</given-names></name><name><surname>Garcia</surname><given-names>BA</given-names></name></person-group><article-title>Origins and formation of histone methylation across the human cell cycle</article-title><source>Mol. Cell. Biol.</source><year>2012</year><volume>32</volume><fpage>2503</fpage><lpage>2514</lpage><?supplied-pmid 22547680?><pub-id pub-id-type=\"pmid\">22547680</pub-id></element-citation></ref><ref id=\"CR17\"><label>17.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Cao</surname><given-names>R</given-names></name><etal/></person-group><article-title>Role of histone H3 lysine 27 methylation in polycomb-group silencing</article-title><source>Science</source><year>2002</year><volume>298</volume><fpage>1039</fpage><lpage>1043</lpage><?supplied-pmid 12351676?><pub-id pub-id-type=\"pmid\">12351676</pub-id></element-citation></ref><ref id=\"CR18\"><label>18.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Cao</surname><given-names>R</given-names></name><name><surname>Zhang</surname><given-names>Y</given-names></name></person-group><article-title>SUZ12 is required for both the histone methyltransferase activity and the silencing function of the EED&#x02013;EZH2 complex</article-title><source>Mol. Cell</source><year>2004</year><volume>15</volume><fpage>57</fpage><lpage>67</lpage><?supplied-pmid 15225548?><pub-id pub-id-type=\"pmid\">15225548</pub-id></element-citation></ref><ref id=\"CR19\"><label>19.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Nekrasov</surname><given-names>M</given-names></name><name><surname>Wild</surname><given-names>B</given-names></name><name><surname>M&#x000fc;ller</surname><given-names>J</given-names></name></person-group><article-title>Nucleosome binding and histone methyltransferase activity of <italic>Drosophila</italic> PRC2</article-title><source>EMBO Rep.</source><year>2005</year><volume>6</volume><fpage>348</fpage><lpage>353</lpage><?supplied-pmid 15776017?><pub-id pub-id-type=\"pmid\">15776017</pub-id></element-citation></ref><ref id=\"CR20\"><label>20.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Gao</surname><given-names>Z</given-names></name><etal/></person-group><article-title>PCGF homologs, CBX proteins, and RYBP define functionally distinct PRC1 family complexes</article-title><source>Mol. Cell</source><year>2012</year><volume>45</volume><fpage>344</fpage><lpage>356</lpage><?supplied-pmid 22325352?><pub-id pub-id-type=\"pmid\">22325352</pub-id></element-citation></ref><ref id=\"CR21\"><label>21.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Junco</surname><given-names>SE</given-names></name><etal/></person-group><article-title>Structure of the polycomb group protein PCGF1 in complex with BCOR reveals basis for binding selectivity of PCGF homologs</article-title><source>Structure</source><year>2013</year><volume>21</volume><fpage>665</fpage><lpage>671</lpage><?supplied-pmid 23523425?><pub-id pub-id-type=\"pmid\">23523425</pub-id></element-citation></ref><ref id=\"CR22\"><label>22.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Rinn</surname><given-names>JL</given-names></name><etal/></person-group><article-title>Functional demarcation of active and silent chromatin domains in human HOX loci by noncoding RNAs</article-title><source>Cell</source><year>2007</year><volume>129</volume><fpage>1311</fpage><lpage>1323</lpage><?supplied-pmid 17604720?><pub-id pub-id-type=\"pmid\">17604720</pub-id></element-citation></ref><ref id=\"CR23\"><label>23.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Xu</surname><given-names>C</given-names></name><name><surname>Min</surname><given-names>J</given-names></name></person-group><article-title>Structure and function of WD40 domain proteins</article-title><source>Protein Cell</source><year>2011</year><volume>2</volume><fpage>202</fpage><lpage>214</lpage><?supplied-pmid 21468892?><pub-id pub-id-type=\"pmid\">21468892</pub-id></element-citation></ref><ref id=\"CR24\"><label>24.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Kuzmichev</surname><given-names>A</given-names></name><name><surname>Nishioka</surname><given-names>K</given-names></name><name><surname>Erdjument-Bromage</surname><given-names>H</given-names></name><name><surname>Tempst</surname><given-names>P</given-names></name><name><surname>Reinberg</surname><given-names>D</given-names></name></person-group><article-title>Histone methyltransferase activity associated with a human multiprotein complex containing the Enhancer of Zeste protein</article-title><source>Genes Dev.</source><year>2002</year><volume>16</volume><fpage>2893</fpage><lpage>2905</lpage><?supplied-pmid 12435631?><pub-id pub-id-type=\"pmid\">12435631</pub-id></element-citation></ref><ref id=\"CR25\"><label>25.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Davidovich</surname><given-names>C</given-names></name><name><surname>Cech</surname><given-names>TR</given-names></name></person-group><article-title>The recruitment of chromatin modifiers by long noncoding RNAs: lessons from PRC2</article-title><source>RNA</source><year>2015</year><volume>21</volume><fpage>2007</fpage><lpage>2022</lpage><?supplied-pmid 26574518?><pub-id pub-id-type=\"pmid\">26574518</pub-id></element-citation></ref><ref id=\"CR26\"><label>26.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Laugesen</surname><given-names>A</given-names></name><name><surname>Helin</surname><given-names>K</given-names></name></person-group><article-title>Chromatin repressive complexes in stem cells, development, and cancer</article-title><source>Cell Stem Cell</source><year>2014</year><volume>14</volume><fpage>735</fpage><lpage>751</lpage><?supplied-pmid 24905164?><pub-id pub-id-type=\"pmid\">24905164</pub-id></element-citation></ref><ref id=\"CR27\"><label>27.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Hauri</surname><given-names>S</given-names></name><etal/></person-group><article-title>A high-density map for navigating the human polycomb complexome</article-title><source>Cell Rep.</source><year>2016</year><volume>17</volume><fpage>583</fpage><lpage>595</lpage><?supplied-pmid 27705803?><pub-id pub-id-type=\"pmid\">27705803</pub-id></element-citation></ref><ref id=\"CR28\"><label>28.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Alekseyenko</surname><given-names>AA</given-names></name><name><surname>Gorchakov</surname><given-names>AA</given-names></name><name><surname>Kharchenko</surname><given-names>PV</given-names></name><name><surname>Kuroda</surname><given-names>MI</given-names></name></person-group><article-title>Reciprocal interactions of human C10orf12 and C17orf96 with PRC2 revealed by BioTAP-XL cross-linking and affinity purification</article-title><source>Proc. Natl Acad. Sci. USA</source><year>2014</year><volume>111</volume><fpage>2488</fpage><lpage>2493</lpage><?supplied-pmid 24550272?><pub-id pub-id-type=\"pmid\">24550272</pub-id></element-citation></ref><ref id=\"CR29\"><label>29.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Kloet</surname><given-names>SL</given-names></name><etal/></person-group><article-title>The dynamic interactome and genomic targets of polycomb complexes during stem-cell differentiation</article-title><source>Nat. Struct. Mol. Biol.</source><year>2016</year><volume>23</volume><fpage>682</fpage><lpage>690</lpage><?supplied-pmid 27294783?><pub-id pub-id-type=\"pmid\">27294783</pub-id></element-citation></ref><ref id=\"CR30\"><label>30.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Boulay</surname><given-names>G</given-names></name><name><surname>Rosnoblet</surname><given-names>C</given-names></name><name><surname>Gu&#x000e9;rardel</surname><given-names>C</given-names></name><name><surname>Angrand</surname><given-names>P-O</given-names></name><name><surname>Leprince</surname><given-names>D</given-names></name></person-group><article-title>Functional characterization of hPCL3 (human Polycomb-like 3) isoforms identifies them as components of distinct EZH2 protein complexes</article-title><source>Biochem. J.</source><year>2011</year><volume>434</volume><fpage>333</fpage><lpage>342</lpage><?supplied-pmid 21143197?><pub-id pub-id-type=\"pmid\">21143197</pub-id></element-citation></ref><ref id=\"CR31\"><label>31.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Ballar&#x000e9;</surname><given-names>C</given-names></name><etal/></person-group><article-title>Phf19 links methylated Lys36 of histone H3 to regulation of polycomb activity</article-title><source>Nat. Struct. Mol. Biol.</source><year>2012</year><volume>19</volume><fpage>1257</fpage><lpage>1265</lpage><?supplied-pmid 23104054?><pub-id pub-id-type=\"pmid\">23104054</pub-id></element-citation></ref><ref id=\"CR32\"><label>32.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Li</surname><given-names>H</given-names></name><etal/></person-group><article-title>Polycomb-like proteins link the PRC2 complex to CpG islands</article-title><source>Nature</source><year>2017</year><volume>549</volume><fpage>287</fpage><lpage>291</lpage><?supplied-pmid 28869966?><pub-id pub-id-type=\"pmid\">28869966</pub-id></element-citation></ref><ref id=\"CR33\"><label>33.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Perino</surname><given-names>M</given-names></name><etal/></person-group><article-title>MTF2 recruits polycomb repressive complex 2 by helical-shape-selective DNA binding</article-title><source>Nat. Genet.</source><year>2018</year><volume>50</volume><fpage>1002</fpage><lpage>1010</lpage><?supplied-pmid 29808031?><pub-id pub-id-type=\"pmid\">29808031</pub-id></element-citation></ref><ref id=\"CR34\"><label>34.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Conway</surname><given-names>E</given-names></name><etal/></person-group><article-title>A family of vertebrate-specific polycombs encoded by the LCOR/LCORL genes balance PRC2 subtype activities</article-title><source>Mol. Cell</source><year>2018</year><volume>70</volume><fpage>408</fpage><lpage>421</lpage><?supplied-pmid 29628311?><pub-id pub-id-type=\"pmid\">29628311</pub-id></element-citation></ref><ref id=\"CR35\"><label>35.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Beringer</surname><given-names>M</given-names></name><etal/></person-group><article-title>EPOP functionally links elongin and polycomb in pluripotent stem cells</article-title><source>Mol. Cell</source><year>2016</year><volume>64</volume><fpage>645</fpage><lpage>658</lpage><?supplied-pmid 27863225?><pub-id pub-id-type=\"pmid\">27863225</pub-id></element-citation></ref><ref id=\"CR36\"><label>36.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Kasinath</surname><given-names>V</given-names></name><etal/></person-group><article-title>Structures of human PRC2 with its cofactors AEBP2 and JARID2</article-title><source>Science</source><year>2018</year><volume>359</volume><fpage>940</fpage><lpage>944</lpage><?supplied-pmid 29348366?><pub-id pub-id-type=\"pmid\">29348366</pub-id></element-citation></ref><ref id=\"CR37\"><label>37.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Cooper</surname><given-names>S</given-names></name><etal/></person-group><article-title>Jarid2 binds mono-ubiquitylated H2A lysine 119 to mediate crosstalk between polycomb complexes PRC1 and PRC2</article-title><source>Nat. Commun.</source><year>2016</year><volume>7</volume><fpage>13661</fpage><?supplied-pmid 27892467?><pub-id pub-id-type=\"pmid\">27892467</pub-id></element-citation></ref><ref id=\"CR38\"><label>38.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>van den Boom</surname><given-names>V</given-names></name><etal/></person-group><article-title>Non-canonical PRC1. 1 targets active genes independent of H3K27me3 and is essential for leukemogenesis</article-title><source>Cell Rep.</source><year>2016</year><volume>14</volume><fpage>332</fpage><lpage>346</lpage><?supplied-pmid 26748712?><pub-id pub-id-type=\"pmid\">26748712</pub-id></element-citation></ref><ref id=\"CR39\"><label>39.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>de Napoles</surname><given-names>M</given-names></name><etal/></person-group><article-title>Polycomb group proteins Ring1A/B link ubiquitylation of histone H2A to heritable gene silencing and X inactivation</article-title><source>Dev. Cell</source><year>2004</year><volume>7</volume><fpage>663</fpage><lpage>676</lpage><?supplied-pmid 15525528?><pub-id pub-id-type=\"pmid\">15525528</pub-id></element-citation></ref><ref id=\"CR40\"><label>40.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Wang</surname><given-names>H</given-names></name><etal/></person-group><article-title>Role of histone H2A ubiquitination in polycomb silencing</article-title><source>Nature</source><year>2004</year><volume>431</volume><fpage>873</fpage><lpage>878</lpage><?supplied-pmid 15386022?><pub-id pub-id-type=\"pmid\">15386022</pub-id></element-citation></ref><ref id=\"CR41\"><label>41.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Srivastava</surname><given-names>A</given-names></name><name><surname>McGrath</surname><given-names>B</given-names></name><name><surname>Bielas</surname><given-names>SL</given-names></name></person-group><article-title>Histone H2A monoubiquitination in neurodevelopmental disorders</article-title><source>Trends Genet.</source><year>2017</year><volume>33</volume><fpage>566</fpage><lpage>578</lpage><?supplied-pmid 28669576?><pub-id pub-id-type=\"pmid\">28669576</pub-id></element-citation></ref><ref id=\"CR42\"><label>42.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Fischle</surname><given-names>W</given-names></name><etal/></person-group><article-title>Molecular basis for the discrimination of repressive methyl-lysine marks in histone H3 by polycomb and HP1 chromodomains</article-title><source>Genes Dev.</source><year>2003</year><volume>17</volume><fpage>1870</fpage><lpage>1881</lpage><?supplied-pmid 12897054?><pub-id pub-id-type=\"pmid\">12897054</pub-id></element-citation></ref><ref id=\"CR43\"><label>43.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Kaustov</surname><given-names>L</given-names></name><etal/></person-group><article-title>Recognition and specificity determinants of the human cbx chromodomains</article-title><source>J. Biol. Chem.</source><year>2011</year><volume>286</volume><fpage>521</fpage><lpage>529</lpage><?supplied-pmid 21047797?><pub-id pub-id-type=\"pmid\">21047797</pub-id></element-citation></ref><ref id=\"CR44\"><label>44.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Robinson</surname><given-names>AK</given-names></name><etal/></person-group><article-title>The growth-suppressive function of the polycomb group protein polyhomeotic is mediated by polymerization of its sterile alpha motif (SAM) domain</article-title><source>J. Biol. Chem.</source><year>2012</year><volume>287</volume><fpage>8702</fpage><lpage>8713</lpage><?supplied-pmid 22275371?><pub-id pub-id-type=\"pmid\">22275371</pub-id></element-citation></ref><ref id=\"CR45\"><label>45.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Kim</surname><given-names>CA</given-names></name><name><surname>Francis</surname><given-names>NJ</given-names></name></person-group><article-title>Chromatin: polycomb group SAMs unite</article-title><source>Curr. Biol.</source><year>2016</year><volume>26</volume><fpage>R710</fpage><lpage>R712</lpage><?supplied-pmid 27505242?><pub-id pub-id-type=\"pmid\">27505242</pub-id></element-citation></ref><ref id=\"CR46\"><label>46.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Levine</surname><given-names>SS</given-names></name><etal/></person-group><article-title>The core of the polycomb repressive complex is compositionally and functionally conserved in flies and humans</article-title><source>Mol. Cell. Biol.</source><year>2002</year><volume>22</volume><fpage>6070</fpage><lpage>6078</lpage><?supplied-pmid 12167701?><pub-id pub-id-type=\"pmid\">12167701</pub-id></element-citation></ref><ref id=\"CR47\"><label>47.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Bezsonova</surname><given-names>I</given-names></name><etal/></person-group><article-title>Ring1B contains a ubiquitin-like docking module for interaction with Cbx proteins</article-title><source>Biochemistry</source><year>2009</year><volume>48</volume><fpage>10542</fpage><lpage>10548</lpage><?supplied-pmid 19791798?><pub-id pub-id-type=\"pmid\">19791798</pub-id></element-citation></ref><ref id=\"CR48\"><label>48.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Wang</surname><given-names>R</given-names></name><etal/></person-group><article-title>Polycomb group targeting through different binding partners of RING1B C-terminal domain</article-title><source>Structure</source><year>2010</year><volume>18</volume><fpage>966</fpage><lpage>975</lpage><?supplied-pmid 20696397?><pub-id pub-id-type=\"pmid\">20696397</pub-id></element-citation></ref><ref id=\"CR49\"><label>49.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Taherbhoy</surname><given-names>AM</given-names></name><name><surname>Huang</surname><given-names>OW</given-names></name><name><surname>Cochran</surname><given-names>AG</given-names></name></person-group><article-title>BMI1-RING1B is an autoinhibited RING E3 ubiquitin ligase</article-title><source>Nat. Commun.</source><year>2015</year><volume>6</volume><fpage>7621</fpage><?supplied-pmid 26151332?><pub-id pub-id-type=\"pmid\">26151332</pub-id></element-citation></ref><ref id=\"CR50\"><label>50.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Farcas</surname><given-names>AM</given-names></name><etal/></person-group><article-title>KDM2B links the polycomb repressive complex 1 (PRC1) to recognition of CpG islands</article-title><source>Elife</source><year>2012</year><volume>1</volume><fpage>e00205</fpage><?supplied-pmid 23256043?><pub-id pub-id-type=\"pmid\">23256043</pub-id></element-citation></ref><ref id=\"CR51\"><label>51.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Blackledge</surname><given-names>NP</given-names></name><etal/></person-group><article-title>Variant PRC1 complex-dependent H2A ubiquitylation drives PRC2 recruitment and polycomb domain formation</article-title><source>Cell</source><year>2014</year><volume>157</volume><fpage>1445</fpage><lpage>1459</lpage><?supplied-pmid 24856970?><pub-id pub-id-type=\"pmid\">24856970</pub-id></element-citation></ref><ref id=\"CR52\"><label>52.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Mendenhall</surname><given-names>EM</given-names></name><etal/></person-group><article-title>GC-rich sequence elements recruit PRC2 in mammalian ES cells</article-title><source>PLoS Genet.</source><year>2010</year><volume>6</volume><fpage>e1001244</fpage><?supplied-pmid 21170310?><pub-id pub-id-type=\"pmid\">21170310</pub-id></element-citation></ref><ref id=\"CR53\"><label>53.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Brown</surname><given-names>JL</given-names></name><name><surname>Mucci</surname><given-names>D</given-names></name><name><surname>Whiteley</surname><given-names>M</given-names></name><name><surname>Dirksen</surname><given-names>M-L</given-names></name><name><surname>Kassis</surname><given-names>JA</given-names></name></person-group><article-title>The <italic>Drosophila</italic> polycomb group gene pleiohomeotic encodes a DNA binding protein with homology to the transcription factor YY1</article-title><source>Mol. Cell</source><year>1998</year><volume>1</volume><fpage>1057</fpage><lpage>1064</lpage><?supplied-pmid 9651589?><pub-id pub-id-type=\"pmid\">9651589</pub-id></element-citation></ref><ref id=\"CR54\"><label>54.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Atchison</surname><given-names>L</given-names></name><name><surname>Ghias</surname><given-names>A</given-names></name><name><surname>Wilkinson</surname><given-names>F</given-names></name><name><surname>Bonini</surname><given-names>N</given-names></name><name><surname>Atchison</surname><given-names>ML</given-names></name></person-group><article-title>Transcription factor YY1 functions as a PcG protein in vivo</article-title><source>EMBO J.</source><year>2003</year><volume>22</volume><fpage>1347</fpage><lpage>1358</lpage><?supplied-pmid 12628927?><pub-id pub-id-type=\"pmid\">12628927</pub-id></element-citation></ref><ref id=\"CR55\"><label>55.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Basu</surname><given-names>A</given-names></name><name><surname>Wilkinson</surname><given-names>FH</given-names></name><name><surname>Colavita</surname><given-names>K</given-names></name><name><surname>Fennelly</surname><given-names>C</given-names></name><name><surname>Atchison</surname><given-names>ML</given-names></name></person-group><article-title>YY1 DNA binding and interaction with YAF2 is essential for polycomb recruitment</article-title><source>Nucleic Acids Res.</source><year>2013</year><volume>42</volume><fpage>2208</fpage><lpage>2223</lpage><?supplied-pmid 24285299?><pub-id pub-id-type=\"pmid\">24285299</pub-id></element-citation></ref><ref id=\"CR56\"><label>56.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Hidalgo</surname><given-names>I</given-names></name><etal/></person-group><article-title>Ezh1 is required for hematopoietic stem cell maintenance and prevents senescence-like cell cycle arrest</article-title><source>Cell Stem Cell</source><year>2012</year><volume>11</volume><fpage>649</fpage><lpage>662</lpage><?supplied-pmid 23122289?><pub-id pub-id-type=\"pmid\">23122289</pub-id></element-citation></ref><ref id=\"CR57\"><label>57.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>O&#x02019;Carroll</surname><given-names>D</given-names></name><etal/></person-group><article-title>The polycomb-group Gene Ezh2 is required for early mouse development</article-title><source>Mol. Cell. Biol.</source><year>2001</year><volume>21</volume><fpage>4330</fpage><lpage>4336</lpage><?supplied-pmid 11390661?><pub-id pub-id-type=\"pmid\">11390661</pub-id></element-citation></ref><ref id=\"CR58\"><label>58.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Mochizuki-Kashio</surname><given-names>M</given-names></name><etal/></person-group><article-title>Dependency on the polycomb gene Ezh2 distinguishes fetal from adult hematopoietic stem cells</article-title><source>Blood</source><year>2011</year><volume>118</volume><fpage>6553</fpage><lpage>6561</lpage><?supplied-pmid 22042701?><pub-id pub-id-type=\"pmid\">22042701</pub-id></element-citation></ref><ref id=\"CR59\"><label>59.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Tumes</surname><given-names>DJ</given-names></name><etal/></person-group><article-title>The polycomb protein Ezh2 regulates differentiation and plasticity of CD4+ T helper type 1 and type 2 cells</article-title><source>Immunity</source><year>2013</year><volume>39</volume><fpage>819</fpage><lpage>832</lpage><?supplied-pmid 24238339?><pub-id pub-id-type=\"pmid\">24238339</pub-id></element-citation></ref><ref id=\"CR60\"><label>60.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Zhang</surname><given-names>Y</given-names></name><etal/></person-group><article-title>The polycomb repressive complex 2 governs life and death of peripheral T cells</article-title><source>Blood</source><year>2014</year><volume>124</volume><fpage>737</fpage><lpage>749</lpage><?supplied-pmid 24951427?><pub-id pub-id-type=\"pmid\">24951427</pub-id></element-citation></ref><ref id=\"CR61\"><label>61.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Gray</surname><given-names>SM</given-names></name><name><surname>Amezquita</surname><given-names>RA</given-names></name><name><surname>Guan</surname><given-names>T</given-names></name><name><surname>Kleinstein</surname><given-names>SH</given-names></name><name><surname>Kaech</surname><given-names>SM</given-names></name></person-group><article-title>Polycomb repressive complex 2-mediated chromatin repression guides effector CD8+ T cell terminal differentiation and loss of multipotency</article-title><source>Immunity</source><year>2017</year><volume>46</volume><fpage>596</fpage><lpage>608</lpage><?supplied-pmid 28410989?><pub-id pub-id-type=\"pmid\">28410989</pub-id></element-citation></ref><ref id=\"CR62\"><label>62.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Yin</surname><given-names>J</given-names></name><etal/></person-group><article-title>Ezh2 regulates differentiation and function of natural killer cells through histone methyltransferase activity</article-title><source>Proc. Natl Acad. Sci. USA</source><year>2015</year><volume>112</volume><fpage>15988</fpage><lpage>15993</lpage><?supplied-pmid 26668377?><pub-id pub-id-type=\"pmid\">26668377</pub-id></element-citation></ref><ref id=\"CR63\"><label>63.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Tumes</surname><given-names>D</given-names></name><etal/></person-group><article-title>Ezh2 controls development of natural killer T cells, which cause spontaneous asthma-like pathology</article-title><source>J. Allergy Clin. Immunol.</source><year>2019</year><volume>144</volume><fpage>549</fpage><lpage>560</lpage><?supplied-pmid 30851295?><pub-id pub-id-type=\"pmid\">30851295</pub-id></element-citation></ref><ref id=\"CR64\"><label>64.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Xie</surname><given-names>H</given-names></name><etal/></person-group><article-title>Polycomb repressive complex 2 regulates normal hematopoietic stem cell function in a developmental-stage-specific manner</article-title><source>Cell Stem Cell</source><year>2014</year><volume>14</volume><fpage>68</fpage><lpage>80</lpage><?supplied-pmid 24239285?><pub-id pub-id-type=\"pmid\">24239285</pub-id></element-citation></ref><ref id=\"CR65\"><label>65.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Simon</surname><given-names>C</given-names></name><etal/></person-group><article-title>A key role for EZH2 and associated genes in mouse and human adult T-cell acute leukemia</article-title><source>Genes Dev.</source><year>2012</year><volume>26</volume><fpage>651</fpage><lpage>656</lpage><?supplied-pmid 22431509?><pub-id pub-id-type=\"pmid\">22431509</pub-id></element-citation></ref><ref id=\"CR66\"><label>66.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Lee</surname><given-names>SC</given-names></name><etal/></person-group><article-title>Polycomb repressive complex 2 component Suz12 is required for hematopoietic stem cell function and lymphopoiesis</article-title><source>Blood</source><year>2015</year><volume>126</volume><fpage>167</fpage><lpage>175</lpage><?supplied-pmid 26036803?><pub-id pub-id-type=\"pmid\">26036803</pub-id></element-citation></ref><ref id=\"CR67\"><label>67.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Lessard</surname><given-names>J</given-names></name><name><surname>Sauvageau</surname><given-names>G</given-names></name></person-group><article-title>Bmi-1 determines the proliferative capacity of normal and leukaemic stem cells</article-title><source>Nature</source><year>2003</year><volume>423</volume><fpage>255</fpage><lpage>260</lpage><?supplied-pmid 12714970?><pub-id pub-id-type=\"pmid\">12714970</pub-id></element-citation></ref><ref id=\"CR68\"><label>68.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Park</surname><given-names>I-k</given-names></name><etal/></person-group><article-title>Bmi-1 is required for maintenance of adult self-renewing haematopoietic stem cells</article-title><source>Nature</source><year>2003</year><volume>423</volume><fpage>302</fpage><lpage>305</lpage><?supplied-pmid 12714971?><pub-id pub-id-type=\"pmid\">12714971</pub-id></element-citation></ref><ref id=\"CR69\"><label>69.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Yamashita</surname><given-names>M</given-names></name><etal/></person-group><article-title>Bmi1 regulates memory CD4 T cell survival via repression of the Noxa gene</article-title><source>J. Exp. Med.</source><year>2008</year><volume>205</volume><fpage>1109</fpage><lpage>1120</lpage><?supplied-pmid 18411339?><pub-id pub-id-type=\"pmid\">18411339</pub-id></element-citation></ref><ref id=\"CR70\"><label>70.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Miyazaki</surname><given-names>M</given-names></name><etal/></person-group><article-title>Thymocyte proliferation induced by pre-T cell receptor signaling is maintained through polycomb gene product Bmi-1-mediated Cdkn2a repression</article-title><source>Immunity</source><year>2008</year><volume>28</volume><fpage>231</fpage><lpage>245</lpage><?supplied-pmid 18275833?><pub-id pub-id-type=\"pmid\">18275833</pub-id></element-citation></ref><ref id=\"CR71\"><label>71.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Hosokawa</surname><given-names>H</given-names></name><etal/></person-group><article-title>Regulation of Th2 cell development by polycomb group gene bmi-1 through the stabilization of GATA3</article-title><source>J. Immunol.</source><year>2006</year><volume>177</volume><fpage>7656</fpage><lpage>7664</lpage><?supplied-pmid 17114435?><pub-id pub-id-type=\"pmid\">17114435</pub-id></element-citation></ref><ref id=\"CR72\"><label>72.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Iwama</surname><given-names>A</given-names></name><etal/></person-group><article-title>Enhanced self-renewal of hematopoietic stem cells mediated by the polycomb gene product Bmi-1</article-title><source>Immunity</source><year>2004</year><volume>21</volume><fpage>843</fpage><lpage>851</lpage><?supplied-pmid 15589172?><pub-id pub-id-type=\"pmid\">15589172</pub-id></element-citation></ref><ref id=\"CR73\"><label>73.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Kimura</surname><given-names>M</given-names></name><etal/></person-group><article-title>Regulation of Th2 cell differentiation by mel-18, a mammalian polycomb group gene</article-title><source>Immunity</source><year>2001</year><volume>15</volume><fpage>275</fpage><lpage>287</lpage><?supplied-pmid 11520462?><pub-id pub-id-type=\"pmid\">11520462</pub-id></element-citation></ref><ref id=\"CR74\"><label>74.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Ohta</surname><given-names>H</given-names></name><etal/></person-group><article-title>Polycomb group gene rae28 is required for sustaining activity of hematopoietic stem cells</article-title><source>J. Exp. Med.</source><year>2002</year><volume>195</volume><fpage>759</fpage><lpage>770</lpage><?supplied-pmid 11901201?><pub-id pub-id-type=\"pmid\">11901201</pub-id></element-citation></ref><ref id=\"CR75\"><label>75.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Kim</surname><given-names>JY</given-names></name><etal/></person-group><article-title>Defective long&#x02010;term repopulating ability in hematopoietic stem cells lacking the polycomb&#x02010;group gene rae28</article-title><source>Eur. J. Haematol.</source><year>2004</year><volume>73</volume><fpage>75</fpage><lpage>84</lpage><?supplied-pmid 15245505?><pub-id pub-id-type=\"pmid\">15245505</pub-id></element-citation></ref><ref id=\"CR76\"><label>76.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Bae</surname><given-names>J</given-names></name><etal/></person-group><article-title>Phc2 controls hematopoietic stem and progenitor cell mobilization from bone marrow by repressing Vcam1 expression</article-title><source>Nat. Commun.</source><year>2019</year><volume>10</volume><fpage>3496</fpage><?supplied-pmid 31375680?><pub-id pub-id-type=\"pmid\">31375680</pub-id></element-citation></ref><ref id=\"CR77\"><label>77.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Cor&#x000e9;</surname><given-names>N</given-names></name><etal/></person-group><article-title>Altered cellular proliferation and mesoderm patterning in polycomb-M33-deficient mice</article-title><source>Development</source><year>1997</year><volume>124</volume><fpage>721</fpage><lpage>729</lpage><?supplied-pmid 9043087?><pub-id pub-id-type=\"pmid\">9043087</pub-id></element-citation></ref><ref id=\"CR78\"><label>78.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Liu</surname><given-names>B</given-names></name><etal/></person-group><article-title>Cbx4 regulates the proliferation of thymic epithelial cells and thymus function</article-title><source>Development</source><year>2013</year><volume>140</volume><fpage>780</fpage><lpage>788</lpage><?supplied-pmid 23362346?><pub-id pub-id-type=\"pmid\">23362346</pub-id></element-citation></ref><ref id=\"CR79\"><label>79.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Tan</surname><given-names>J</given-names></name><etal/></person-group><article-title>CBX8, a polycomb group protein, is essential for MLL-AF9-induced leukemogenesis</article-title><source>Cancer Cell</source><year>2011</year><volume>20</volume><fpage>563</fpage><lpage>575</lpage><?supplied-pmid 22094252?><pub-id pub-id-type=\"pmid\">22094252</pub-id></element-citation></ref><ref id=\"CR80\"><label>80.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>B&#x000e9;guelin</surname><given-names>W</given-names></name><etal/></person-group><article-title>EZH2 and BCL6 cooperate to assemble CBX8-BCOR complex to repress bivalent promoters, mediate germinal center formation and lymphomagenesis</article-title><source>Cancer Cell</source><year>2016</year><volume>30</volume><fpage>197</fpage><lpage>213</lpage><?supplied-pmid 27505670?><pub-id pub-id-type=\"pmid\">27505670</pub-id></element-citation></ref><ref id=\"CR81\"><label>81.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Cal&#x000e9;s</surname><given-names>C</given-names></name><etal/></person-group><article-title>Inactivation of the polycomb group protein Ring1B unveils an antiproliferative role in hematopoietic cell expansion and cooperation with tumorigenesis associated with Ink4a deletion</article-title><source>Mol. Cell. Biol.</source><year>2008</year><volume>28</volume><fpage>1018</fpage><lpage>1028</lpage><?supplied-pmid 18039844?><pub-id pub-id-type=\"pmid\">18039844</pub-id></element-citation></ref><ref id=\"CR82\"><label>82.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Ikawa</surname><given-names>T</given-names></name><etal/></person-group><article-title>Conversion of T cells to B cells by inactivation of polycomb-mediated epigenetic suppression of the B-lineage program</article-title><source>Genes Dev.</source><year>2016</year><volume>30</volume><fpage>2475</fpage><lpage>2485</lpage><?supplied-pmid 27913604?><pub-id pub-id-type=\"pmid\">27913604</pub-id></element-citation></ref><ref id=\"CR83\"><label>83.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Shen</surname><given-names>X</given-names></name><etal/></person-group><article-title>EZH1 mediates methylation on histone H3 lysine 27 and complements EZH2 in maintaining stem cell identity and executing pluripotency</article-title><source>Mol. Cell</source><year>2008</year><volume>32</volume><fpage>491</fpage><lpage>502</lpage><?supplied-pmid 19026780?><pub-id pub-id-type=\"pmid\">19026780</pub-id></element-citation></ref><ref id=\"CR84\"><label>84.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Ezhkova</surname><given-names>E</given-names></name><etal/></person-group><article-title>EZH1 and EZH2 cogovern histone H3K27 trimethylation and are essential for hair follicle homeostasis and wound repair</article-title><source>Genes Dev.</source><year>2011</year><volume>25</volume><fpage>485</fpage><lpage>498</lpage><?supplied-pmid 21317239?><pub-id pub-id-type=\"pmid\">21317239</pub-id></element-citation></ref><ref id=\"CR85\"><label>85.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Jacobs</surname><given-names>JJ</given-names></name><name><surname>Kieboom</surname><given-names>K</given-names></name><name><surname>Marino</surname><given-names>S</given-names></name><name><surname>DePinho</surname><given-names>RA</given-names></name><name><surname>van Lohuizen</surname><given-names>M</given-names></name></person-group><article-title>The oncogene and polycomb-group gene bmi-1 regulates cell proliferation and senescence through the ink4a locus</article-title><source>Nature</source><year>1999</year><volume>397</volume><fpage>164</fpage><lpage>168</lpage><?supplied-pmid 9923679?><pub-id pub-id-type=\"pmid\">9923679</pub-id></element-citation></ref><ref id=\"CR86\"><label>86.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Lessard</surname><given-names>J</given-names></name><name><surname>Baban</surname><given-names>S</given-names></name><name><surname>Sauvageau</surname><given-names>G</given-names></name></person-group><article-title>Stage-specific expression of polycomb group genes in human bone marrow cells</article-title><source>Blood</source><year>1998</year><volume>91</volume><fpage>1216</fpage><lpage>1224</lpage><?supplied-pmid 9454751?><pub-id pub-id-type=\"pmid\">9454751</pub-id></element-citation></ref><ref id=\"CR87\"><label>87.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Lund</surname><given-names>K</given-names></name><name><surname>Adams</surname><given-names>PD</given-names></name><name><surname>Copland</surname><given-names>M</given-names></name></person-group><article-title>EZH2 in normal and malignant hematopoiesis</article-title><source>Leukemia</source><year>2014</year><volume>28</volume><fpage>44</fpage><lpage>49</lpage><?supplied-pmid 24097338?><pub-id pub-id-type=\"pmid\">24097338</pub-id></element-citation></ref><ref id=\"CR88\"><label>88.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Xu</surname><given-names>F</given-names></name><etal/></person-group><article-title>Overexpression of the EZH2, RING1 and BMI1 genes is common in myelodysplastic syndromes: relation to adverse epigenetic alteration and poor prognostic scoring</article-title><source>Ann. Hematol.</source><year>2011</year><volume>90</volume><fpage>643</fpage><lpage>653</lpage><?supplied-pmid 21125401?><pub-id pub-id-type=\"pmid\">21125401</pub-id></element-citation></ref><ref id=\"CR89\"><label>89.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Raaphorst</surname><given-names>FM</given-names></name><etal/></person-group><article-title>Coexpression of BMI-1 and EZH2 polycomb group genes in Reed-Sternberg cells of Hodgkin&#x02019;s disease</article-title><source>Am. J. Pathol.</source><year>2000</year><volume>157</volume><fpage>709</fpage><lpage>715</lpage><?supplied-pmid 10980109?><pub-id pub-id-type=\"pmid\">10980109</pub-id></element-citation></ref><ref id=\"CR90\"><label>90.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Tanaka</surname><given-names>S</given-names></name><etal/></person-group><article-title>Ezh2 augments leukemogenicity by reinforcing differentiation blockage in acute myeloid leukemia</article-title><source>Blood</source><year>2012</year><volume>120</volume><fpage>1107</fpage><lpage>1117</lpage><?supplied-pmid 22677129?><pub-id pub-id-type=\"pmid\">22677129</pub-id></element-citation></ref><ref id=\"CR91\"><label>91.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Takamatsu-Ichihara</surname><given-names>E</given-names></name><name><surname>Kitabayashi</surname><given-names>I</given-names></name></person-group><article-title>The roles of Polycomb group proteins in hematopoietic stem cells and hematological malignancies</article-title><source>Int. J. Hematol.</source><year>2016</year><volume>103</volume><fpage>634</fpage><lpage>642</lpage><?supplied-pmid 27086351?><pub-id pub-id-type=\"pmid\">27086351</pub-id></element-citation></ref><ref id=\"CR92\"><label>92.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Mihara</surname><given-names>K</given-names></name><etal/></person-group><article-title>Bmi-1 is useful as a novel molecular marker for predicting progression of myelodysplastic syndrome and patient prognosis</article-title><source>Blood</source><year>2006</year><volume>107</volume><fpage>305</fpage><lpage>308</lpage><?supplied-pmid 16160010?><pub-id pub-id-type=\"pmid\">16160010</pub-id></element-citation></ref><ref id=\"CR93\"><label>93.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Yuan</surname><given-names>J</given-names></name><etal/></person-group><article-title>Bmi1 is essential for leukemic reprogramming of myeloid progenitor cells</article-title><source>Leukemia</source><year>2011</year><volume>25</volume><fpage>1335</fpage><lpage>1343</lpage><?supplied-pmid 21527932?><pub-id pub-id-type=\"pmid\">21527932</pub-id></element-citation></ref><ref id=\"CR94\"><label>94.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Ntziachristos</surname><given-names>P</given-names></name><etal/></person-group><article-title>Genetic inactivation of the polycomb repressive complex 2 in T cell acute lymphoblastic leukemia</article-title><source>Nat. Med.</source><year>2012</year><volume>18</volume><fpage>298</fpage><lpage>301</lpage><?supplied-pmid 22237151?><pub-id pub-id-type=\"pmid\">22237151</pub-id></element-citation></ref><ref id=\"CR95\"><label>95.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Nikoloski</surname><given-names>G</given-names></name><etal/></person-group><article-title>Somatic mutations of the histone methyltransferase gene EZH2 in myelodysplastic syndromes</article-title><source>Nat. Genet.</source><year>2010</year><volume>42</volume><fpage>665</fpage><lpage>667</lpage><?supplied-pmid 20601954?><pub-id pub-id-type=\"pmid\">20601954</pub-id></element-citation></ref><ref id=\"CR96\"><label>96.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Ernst</surname><given-names>T</given-names></name><etal/></person-group><article-title>Inactivating mutations of the histone methyltransferase gene EZH2 in myeloid disorders</article-title><source>Nat. Genet.</source><year>2010</year><volume>42</volume><fpage>722</fpage><lpage>726</lpage><?supplied-pmid 20601953?><pub-id pub-id-type=\"pmid\">20601953</pub-id></element-citation></ref><ref id=\"CR97\"><label>97.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Score</surname><given-names>J</given-names></name><etal/></person-group><article-title>Inactivation of polycomb repressive complex 2 components in myeloproliferative and myelodysplastic/myeloproliferative neoplasms</article-title><source>Blood</source><year>2012</year><volume>119</volume><fpage>1208</fpage><lpage>1213</lpage><?supplied-pmid 22053108?><pub-id pub-id-type=\"pmid\">22053108</pub-id></element-citation></ref><ref id=\"CR98\"><label>98.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Stazi</surname><given-names>G</given-names></name><name><surname>Zwergel</surname><given-names>C</given-names></name><name><surname>Mai</surname><given-names>A</given-names></name><name><surname>Valente</surname><given-names>S</given-names></name></person-group><article-title>EZH2 inhibitors: a patent review (2014-2016)</article-title><source>Expert Opin. Ther. Pat.</source><year>2017</year><volume>27</volume><fpage>797</fpage><lpage>813</lpage><?supplied-pmid 28394193?><pub-id pub-id-type=\"pmid\">28394193</pub-id></element-citation></ref><ref id=\"CR99\"><label>99.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Qi</surname><given-names>W</given-names></name><etal/></person-group><article-title>An allosteric PRC2 inhibitor targeting the H3K27me3 binding pocket of EED</article-title><source>Nat. Chem. Biol.</source><year>2017</year><volume>13</volume><fpage>381</fpage><lpage>388</lpage><?supplied-pmid 28135235?><pub-id pub-id-type=\"pmid\">28135235</pub-id></element-citation></ref><ref id=\"CR100\"><label>100.</label><element-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Nishida</surname><given-names>Y</given-names></name><etal/></person-group><article-title>The novel BMI-1 inhibitor PTC596 downregulates MCL-1 and induces p53-independent mitochondrial apoptosis in acute myeloid leukemia progenitor cells</article-title><source>Blood Cancer J.</source><year>2017</year><volume>7</volume><fpage>e527</fpage><?supplied-pmid 28211885?><pub-id pub-id-type=\"pmid\">28211885</pub-id></element-citation></ref></ref-list></back></article>'"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response.content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a289219f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ast\n",
    "\n",
    "def calculate_spans(sentence, entities):\n",
    "    spans = []\n",
    "    for entity, etype in entities:\n",
    "        start = sentence.find(entity)\n",
    "        if start != -1:\n",
    "            end = start + len(entity)\n",
    "            spans.append([start, end, entity, etype])  # List instead of tuple\n",
    "    return spans\n",
    "\n",
    "def resolve_overlaps(spans):\n",
    "    spans.sort(key=lambda x: (x[1] - x[0]), reverse=True)  # Sort by span length\n",
    "    resolved = []\n",
    "    for span in spans:\n",
    "        overlap = False\n",
    "        for r in resolved:\n",
    "            if not (span[1] <= r[0] or span[0] >= r[1]):  # Check for overlap\n",
    "                overlap = True\n",
    "                break\n",
    "        if not overlap:\n",
    "            resolved.append(span)\n",
    "    return resolved\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9b68d7e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('../data/xxx7.tsv', sep = '\\t')\n",
    "df['new_ner'] = df.apply(lambda row: resolve_overlaps(calculate_spans(row['sentence'], ast.literal_eval(row['ner']))), axis=1)\n",
    "df.drop('ner', axis=1, inplace=True)\n",
    "df.rename(columns={'new_ner': 'ner'}, inplace=True)\n",
    "df.to_csv('../data/xxx6.tsv', sep = '\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "29974ca4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmcid</th>\n",
       "      <th>sentence</th>\n",
       "      <th>ner</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC4792959</td>\n",
       "      <td>A subunit of the oligosaccharyltransferase com...</td>\n",
       "      <td>[[17, 50, oligosaccharyltransferase complex, GO]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMC4792959</td>\n",
       "      <td>During plant pollen tube (PT) reception, gamet...</td>\n",
       "      <td>[[13, 24, pollen tube, GO]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMC4792959</td>\n",
       "      <td>ARU encodes the OST3/6 subunit of the oligosac...</td>\n",
       "      <td>[[38, 71, oligosaccharyltransferase complex, GO]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PMC4792959</td>\n",
       "      <td>Our results suggest that glycosylation pattern...</td>\n",
       "      <td>[[25, 38, glycosylation, EM]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMC4792959</td>\n",
       "      <td>show that ARTUMES regulates pollen tube recogn...</td>\n",
       "      <td>[[28, 39, pollen tube, GO]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        pmcid                                           sentence  \\\n",
       "0  PMC4792959  A subunit of the oligosaccharyltransferase com...   \n",
       "1  PMC4792959  During plant pollen tube (PT) reception, gamet...   \n",
       "2  PMC4792959  ARU encodes the OST3/6 subunit of the oligosac...   \n",
       "3  PMC4792959  Our results suggest that glycosylation pattern...   \n",
       "4  PMC4792959  show that ARTUMES regulates pollen tube recogn...   \n",
       "\n",
       "                                                 ner  \n",
       "0  [[17, 50, oligosaccharyltransferase complex, GO]]  \n",
       "1                        [[13, 24, pollen tube, GO]]  \n",
       "2  [[38, 71, oligosaccharyltransferase complex, GO]]  \n",
       "3                      [[25, 38, glycosylation, EM]]  \n",
       "4                        [[28, 39, pollen tube, GO]]  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rename entity names in file2.csv as per the provided mapping\n",
    "entity_rename_map = {\n",
    "    'Gene Ontology': 'GO',\n",
    "    'Experimental Methods': 'EM', \n",
    "#     'Accession Numbers': 'AN',\n",
    "#     'Resources': 'RS',\n",
    "    'Chemicals': 'CD'\n",
    "}\n",
    "\n",
    "file2_df = pd.read_csv('../data/xxx6.tsv', sep = '\\t')\n",
    "# Applying the renaming\n",
    "file2_df['ner'] = file2_df['ner'].apply(lambda x: eval(x))\n",
    "for row in file2_df['ner']:\n",
    "    for entity in row:\n",
    "        entity[3] = entity_rename_map.get(entity[3], entity[3])\n",
    "\n",
    "# Checking the changes\n",
    "file2_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "492d1a74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correcting the code to work with the 'ners' column, use sets to remove duplicates, \n",
    "# and applying specific handling for 'AN' type\n",
    "\n",
    "# Adjusting the safe_eval_list function to check if the input is a string\n",
    "def safe_eval_list(str_list):\n",
    "    if isinstance(str_list, str):\n",
    "        try:\n",
    "            return eval(str_list)\n",
    "        except SyntaxError:\n",
    "            return []\n",
    "    return str_list\n",
    "\n",
    "# Applying the adjusted renaming and extraction\n",
    "file2_df['ner'] = file2_df['ner'].apply(safe_eval_list)\n",
    "for row in file2_df['ner']:\n",
    "    for entity in row:\n",
    "        entity[3] = entity_rename_map.get(entity[3], entity[3])\n",
    "\n",
    "        \n",
    "def replace_substrings(input_str):\n",
    "    substrings_to_replace = ['*/', '18O ', '% ', 'μm', 'its ', 'are ',  'successful ', 'III ', '6 ', \n",
    "                             'their ', 'μm ', 'between ', 'some ', 'through ', 'mL ', 'while ','s layer', '50 ', \n",
    "                             'vs ', 'induces ', 'p53 ', 'promotes ', 'within ', 'precise ', 'nitrogen ', '—0', \n",
    "                             'other ', 'created ', 'basement ', 'best ', 'study ', 'than ', '94°C,', 'followed ', \n",
    "                             '°CAbbreviations:', '7500', 'simultaneous', 'represent', 'II ', '15 ', '10× ', \n",
    "                             'initiation ', 'all ', 'toward ', 'improve ', 'whose ', \n",
    "                             'influence ', 'convey', 'and/or ', 'acid ', \n",
    "                             'analyzed ', 'thus ', 'per ', 'affect ', 'associated ', 'increased' , \n",
    "                             'releases ', 'kill', 'Nonetheless, ', \n",
    "                            '“', 'against ', 'additional ', 'late ', 'reduced ', 'Distribution ', \n",
    "'giving ', 'maintain ', 'nighttime ', 'increase ', 'during ', '30 ' 'HIV ',\n",
    "'whether ', '- ', 'use ', 'Three ', 'Caused', 'block ', 'IL-8 ', 'No ', 'create ', '”'\n",
    "'Il-21 ', 'what ', 'base ', 'paid ', 'then ', 'without ', 'where ', '16100 ', '2 ', '325,000 ', 'underlying ', 'related ']\n",
    "    \n",
    "    for substring in substrings_to_replace:\n",
    "        input_str = input_str.replace(substring, '')\n",
    "    \n",
    "    return input_str\n",
    "\n",
    "\n",
    "# Extracting dictionary collection of set with AN, EM, RS, and GO\n",
    "extracted_entities = {'AN': set(), 'EM': set(), 'RS': set(), 'GO': set(), 'CD': set()}\n",
    "for row in file2_df['ner']:\n",
    "    for entity in row:\n",
    "        entity_type = entity[3]\n",
    "        if entity_type in extracted_entities:\n",
    "            # For 'AN', split with a space and take the last term\n",
    "            if entity_type == 'AN':\n",
    "                extracted_entities[entity_type].add(entity[2].split()[-1])\n",
    "            elif entity_type == 'RS' and 'database' in entity[2]:\n",
    "                extracted_entities[entity_type].add(entity[2].replace('database','').strip())\n",
    "            elif entity_type == 'GO' or entity_type == 'EM':\n",
    "                if 'membrane' in entity[2].lower() or 'behaviour' in entity[2].lower():\n",
    "                    extracted_entities[entity_type].add(entity[2].split()[-1])\n",
    "                else:\n",
    "                    replac_str = replace_substrings(entity[2]).strip()\n",
    "                    if replac_str and len(replac_str)>1:\n",
    "                        extracted_entities[entity_type].add(replac_str)\n",
    "            else:\n",
    "                extracted_entities[entity_type].add(entity[2])\n",
    "\n",
    "\n",
    "# Converting sets back to lists for easier viewing/usage\n",
    "extracted_entities = {k: list(v) for k, v in extracted_entities.items()}\n",
    "\n",
    "# extracted_entities\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e1ff2bfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# extracted_entities['RS'].extend(['NHGRI', 'GEO', 'Sequence Read Archive', 'SRA', 'KEGG', \n",
    "#                                  'Kyoto Encyclopedia of Genes and Genomes', 'IntAct62',\n",
    "#                                  'MINT63', 'AtPIN36', 'AtPID37', 'PAIR38', 'APID39', 'PRIN9', 'PHISTO', 'TAIR',\n",
    "#                                 'InParanoid', 'HPRD', 'Medline', 'EMBASE', 'PsycINFO', 'EconLit', 'HEED',\n",
    "#                                  'CRD', 'Health Economic Evaluations Database', 'HEED', 'NHS Economic Evaluation Database',\n",
    "#                                 'NHS EED', 'Database of Abstracts of Reviews of Effects', 'DARE', 'PROSITE',\n",
    "#                                 'CINAHL', 'Cochrane', 'ARAMEMNON', 'TrEMBL', 'PhosphoSitePlus', \n",
    "#                                  'UniProtKB/TrEMBL', 'SwissProt', 'miRWalk',\n",
    "#                                  'DAVID', 'SEER', 'Genebank','MSigDB', 'MGI', 'PsyGeNET', 'SFARI', 'DisGeNET',\n",
    "#                                 'GeoSymbio', 'TIGR', 'GenBank', 'Swiss-prot' ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8cc9efce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# new_accessions = list(set(['FN813615', 'FN813614', 'AM042701', 'AM042698', 'AM042697', 'AM042699', 'FN823053', \n",
    "#                   'FN995351', 'KY355383', 'ENSDARG000000', 'XP_694951', 'XP_694893', 'XP_686111', \n",
    "#                   'XP_686816', 'XP_689021', 'XP_693637', 'XP_694145', 'XP_693652', 'XP_691478', \n",
    "#                   'XP_689023', 'XP_698503', 'XP_690676', 'XP_689954', 'XP_693868', 'XP_688350', \n",
    "#                   'XP_697815', 'XP_687258', 'XP_690327', 'XP_687591', 'XP_691896', 'XP_689864', \n",
    "#                   'XP_690860', 'XP_694605', 'XP_694871', 'XP_696432', 'XP_698814', 'XP_699128', \n",
    "#                   'XP_697396', 'ENSGALG00000011535', 'XM_414969', 'XM_414012', 'XM_414012', 'Q5F3J4', \n",
    "#                   'ENSXETG00000022012', 'GENSCAN00000015124', 'AC157698', 'ENSXETG00000021586', \n",
    "#                   'scaffold_279.46c', 'scaffold_1247.5', 'scaffold_3599.1', 'scaffold_626.23', \n",
    "#                   'scaffold_279.46c', 'scaffold_626.24', 'scaffold_626.25', 'scaffold_279.46c', \n",
    "#                   'scaffold_279.46c', 'scaffold_111.48', 'ENSXETG00000018490', 'NEWSINFRUG00000144256', \n",
    "#                   'NEWSINFRUG00000134206', 'NEWSINFRUG00000138086', 'GENSCANSLICE00000000561', \n",
    "#                   'NEWSINFRUG00000148012', 'TC326097', 'TC353741', 'TC343155', 'EST487', 'MF468321', \n",
    "#                   'MF468322', 'MF468319', 'MF468318', 'MF468291', 'MF468293', 'MF468299', 'MF468290', \n",
    "#                   'MF468308', 'MF468291', 'MF468293', 'MF468294', 'MF468298', 'MF468309', 'MF468314', \n",
    "#                   'MF468295', 'MF468297', 'MF468325', 'MF468328', 'AAL56428', 'AAL56429', 'AAB05820', \n",
    "#                   'U39815', 'AF126830 ', 'AAL56430', 'AAL58703', 'AAL58704 ', 'rs4179590', 'rs4179590', \n",
    "#                   'rs4179589', 'rs51960690', 'rs50972129', 'rs4179517', 'rs4179515', 'rs48139623', \n",
    "#                   'rs32704753', 'rs4179510', 'rs4179504', 'rs4179702', 'rs31418354', 'rs4179917', \n",
    "#                   'rs4179913', 'rs45694900', 'rs4179908', 'Q1051R', 'rs4179907', 'rs4179728', \n",
    "#                   'rs4179729', 'rs4179745', 'ENSMUSSNP328538', 'rs32712110', 'rs4179843', 'rs4179872', \n",
    "#                   'rs31418356', 'rs32712642', 'PRJNA375720', 'AY157059', 'AJ000520', 'BA000042', \n",
    "#                   'ABE84009', 'ABE86610', 'AE017283', 'NM116255', 'CP000323', 'CP000653', 'AJ316032', \n",
    "#                   'JQ804931', 'GSE97544', 'PRJEB9815', 'GSM1334015', 'GSM1334016', 'GSM1334017', \n",
    "#                   'GSM1334018', 'GSM1334019', 'GSM1334020', 'KY355383', 'Q7RTR2', 'Q9Y239', 'Q7RTR2', \n",
    "#                   'Q86WI3', 'Q7RTR3', 'Q9C000', 'Q9NX02', 'Q96P20', 'Q96MN2', 'P59047', 'P59044', \n",
    "#                   'Q32MH8', 'Q86W28', 'Q86W27', 'Q86W26', 'P59045', 'P59046', 'Q86W25', 'Q86W24', \n",
    "#                   'Q9NPP4', 'Q13075', 'P33076', 'KT184354', 'DQ136185', 'AM492684a', 'EU339226', \n",
    "#                   'EU339209', 'EU339214', 'EU339217', 'Ynys1', 'AF387302a', 'EU339223', 'EU339216', \n",
    "#                   'WJ68', 'AY455805a', 'EU339224', 'EU339213', 'AY950676a', 'EU339231', 'EU304260a', \n",
    "#                   'EU339222', 'AY455808a', 'EU339230', 'EU339215', 'GSE24189', 'KJ598498.1', 'GSE9210', \n",
    "#                   'KF309066', 'MH238461', 'GSE43316', 'AY335912', 'AM902709', 'PRJNA236112', 'AM492684a', \n",
    "#                   'EU339226', 'EU339209', 'EU339214', 'EU339217', 'Ynys1', 'AF387302a', 'EU339223', \n",
    "#                   'EU339216', 'WJ68', 'AY455805a', 'EU339224', 'EU339213', 'AY950676a', 'EU339231', \n",
    "#                   'EU304260a', 'EU339222', 'AY455808a', 'EU339230', 'EU339215', 'EU792508', 'ACF19852', \n",
    "#                   'EU792509', 'ACF19853', 'EU792510', 'ACF19854', 'EU792511', 'ACF19855', 'AAC32826', \n",
    "#                   'ABK76312', 'AFJ95132', 'AAC48340', 'AF143817', 'AF041023', 'AB009305', 'AF126830', \n",
    "#                   'AAL56430', 'AAL58703', 'AAL58704', 'AAL56428', 'AAL56429', 'AAB07368', 'AAB05820', \n",
    "#                   'U39815', 'AF167707', 'AF167708', 'AF167709', 'AF167710', 'CAC28360', 'U25057', \n",
    "#                   'BAA14015', 'AAD01628', 'AAB26196', 'P54190', 'U29761', 'AAB53231', 'ACJ64718', \n",
    "#                   'AAB00227', 'SRP049689', 'EG10787', 'EG11221', 'EG12144', 'EG10379', 'EG10283', \n",
    "#                   'EG11448', 'EG10688', 'EG10489', 'EG10283', 'EG20173', 'EG10702', 'KX262676.1', \n",
    "#                   'KX352162', 'KF532967', 'NM_000484', 'KX553930', 'KX553934', 'AM117604', 'FN813615', \n",
    "#                   'FN813614', 'AM042701', 'AM042698', 'AM042697', 'AM042699', 'FN823053', 'FN995351', \n",
    "#                   'AM117602.1', 'NM-138554.1', 'P25297', 'PHO84', 'NP_001337.2', 'NP_003638', 'GSE50635','MF468290', 'SRR4011627', 'SRR4011628', 'HM019534', 'HM020374', 'FN813615',\n",
    "# 'FN813614', 'AM042701', 'AM042698', 'AM042697', 'AM042699', 'HM019533', 'HM020128', 'HM020130', 'HM020332',\n",
    "# 'HM202334', 'HM020374', 'COSM43967', 'COSM1640833', 'COSM562645', 'COSM44853', 'COSM99631']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "76ca0d16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "258"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(new_accessions)  #previously 299"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "eb1ca7c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "158\n",
      "416\n"
     ]
    }
   ],
   "source": [
    "# print(len(extracted_entities['AN']))\n",
    "# extracted_entities['AN'].extend(new_accessions)\n",
    "# print(len(extracted_entities['AN']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "252e5dbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open('../data/CD_terms_list.txt', 'w') as file:\n",
    "#     for term in list(set(extracted_entities['CD'])):\n",
    "#         file.write(term + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d25f2d4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "chem_to_remove = ['uc.134', 'uc','Yorkie','chemical substances','sec','Cuban','8690176','131127258',\n",
    "'CD','DS','steward','ZF-EST487-R3','-EST487','delta','n13131313','6429','sc-376167','QX-314','Cy3',\n",
    "'retina','gamma','singular','electron','1474310','1415180','HM5040',\n",
    "'singular','sec ','Nile Red ','91558 ','KK4824 ','910111213 ','23333435 ','678545556 ','torques ','dump',\n",
    "'torque','at a p','syber green','Short','Short chain fatty acids','2021222324','11192732','1001', '90th',\n",
    "'Gamma',  'Tech', 'fade', ' His-', '-pro-', 'mold', 'electton', 'Psi', 'radio', 'Radio', 'Cho',  'Fmoc', \n",
    " 'aaa', 'h groups', 'His-', '3At', 'a steroid', 'Cy3', 'Tris buffers', 'mega', \n",
    " 'Halogen', 'Ocn', 'an alcohol', 'fmoc', 'lime', '-Trp', '-Pro', '-His', 'Photon', ' suc', 'nape', \n",
    "'Texas red', 'Clo', ' pro-', 'polar amino acids', 'nonyl', 'Dapi', 'hydroxyl radicals', 'aromatic amino acid',\n",
    "  'Foscarnet', 'mannose', ' Cho', 'at a p', '3 ms', '3 at', 'bile salts', 'procure',  'thio',  'an aldehyde',               \n",
    "  'Dopa', 'limestones', ' sec', ' Pro-', 'Retina', 'e is', 'flash',   'molds', 'Flash', 'radical', 'torques']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "492566a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_list = [term for term in list(set(extracted_entities['CD'])) if term not in chem_to_remove]\n",
    "# filtered_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "cd968049",
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open('../data/CD_terms_list.txt', 'w') as file:\n",
    "#     for term in list(set(filtered_list)):\n",
    "#         file.write(term + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "ff2d3907",
   "metadata": {},
   "outputs": [],
   "source": [
    "extracted_entities['GO'].remove('part')\n",
    "# extracted_entities['RS'].remove('HPA')\n",
    "extracted_entities['CD'] = filtered_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "6811c7f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pandas as pd\n",
    "\n",
    "# def enrich_entities(file_path, new_entity_dict):\n",
    "#     # Load the CSV file\n",
    "#     df = pd.read_csv(file_path, names=['sentence', 'ner'])\n",
    "\n",
    "#     # Function to find if there is an overlap with existing spans or if the term is a substring\n",
    "#     def is_valid_entity(sentence, new_span, existing_spans, current_entity_type):\n",
    "#         start_idx, end_idx = new_span\n",
    "#         # Check for overlap with existing spans and allow 'AN' entities to override 'CD'\n",
    "#         for span, span_type in existing_spans:\n",
    "#             if not (end_idx <= span[0] or start_idx >= span[1]):\n",
    "#                 # Allow 'AN' to override 'CD'\n",
    "#                 if current_entity_type == 'AN' and span_type == 'CD':\n",
    "#                     continue\n",
    "#                 return False\n",
    "#         # Check if the entity is not part of a larger word (to avoid substrings)\n",
    "#         if (start_idx > 0 and sentence[start_idx - 1].isalnum()) or (end_idx < len(sentence) and sentence[end_idx].isalnum()):\n",
    "#             return False\n",
    "#         return True\n",
    "\n",
    "#     # Function to add new entity annotations\n",
    "#     def add_new_entities(sentence, entities, entity_dict):\n",
    "#         existing_spans = [(tuple(map(int, entity[:2])), entity[3]) for entity in entities if entity != 'None']  # Tuple of span and type\n",
    "#         for entity_type, terms in entity_dict.items():\n",
    "#             for term in terms:\n",
    "#                 start_idx = sentence.find(term)\n",
    "#                 while start_idx != -1:\n",
    "#                     end_idx = start_idx + len(term)\n",
    "#                     new_span = (start_idx, end_idx)\n",
    "#                     if is_valid_entity(sentence, new_span, existing_spans, entity_type):\n",
    "#                         entities = [entity for entity in entities if not (entity[0] == start_idx and entity[3] == 'CD')]  # Remove overridden 'CD'\n",
    "#                         entities.append([start_idx, end_idx, term, entity_type])\n",
    "#                         existing_spans.append((new_span, entity_type))\n",
    "#                     start_idx = sentence.find(term, start_idx + 1)\n",
    "#         return [entity for entity in entities if entity != 'None']  # Exclude 'None' if other entities are present\n",
    "\n",
    "#     # Enriching the entity annotations\n",
    "#     for i, row in df.iterrows():\n",
    "#         sentence = row['sentence']\n",
    "#         entities = eval(row['ner']) if row['ner'] != \"['None']\" else ['None']\n",
    "#         df.at[i, 'ner'] = add_new_entities(sentence, entities, new_entity_dict)\n",
    "\n",
    "#     return df\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "def enrich_entities(file_path, new_entity_dict, chem_to_remove):\n",
    "    # Load the CSV file\n",
    "    df = pd.read_csv(file_path, names=['sentence', 'ner'])\n",
    "\n",
    "    # Function to find if there is an overlap with existing spans or if the term is a substring\n",
    "    def is_valid_entity(sentence, new_span, existing_spans, current_entity_type):\n",
    "        start_idx, end_idx = new_span\n",
    "        # Check for overlap with existing spans\n",
    "        for span, span_type in existing_spans:\n",
    "            if not (end_idx <= span[0] or start_idx >= span[1]):\n",
    "                return False\n",
    "        # Check if the entity is not part of a larger word (to avoid substrings)\n",
    "        if (start_idx > 0 and sentence[start_idx - 1].isalnum()) or (end_idx < len(sentence) and sentence[end_idx].isalnum()):\n",
    "            return False\n",
    "        return True\n",
    "\n",
    "    # Function to add new entity annotations\n",
    "    def add_new_entities(sentence, entities, entity_dict):\n",
    "        existing_spans = [(tuple(map(int, entity[:2])), entity[3]) for entity in entities if entity != 'None']  # Tuple of span and type\n",
    "        for entity_type, terms in entity_dict.items():\n",
    "            for term in terms:\n",
    "                start_idx = sentence.find(term)\n",
    "                while start_idx != -1:\n",
    "                    end_idx = start_idx + len(term)\n",
    "                    new_span = (start_idx, end_idx)\n",
    "                    if is_valid_entity(sentence, new_span, existing_spans, entity_type):\n",
    "                        entities.append([start_idx, end_idx, term, entity_type])\n",
    "                        existing_spans.append((new_span, entity_type))\n",
    "                    start_idx = sentence.find(term, start_idx + 1)\n",
    "\n",
    "        # Remove 'CD' entities if they are in chem_to_remove list\n",
    "        entities = [entity for entity in entities if not (entity[2] in chem_to_remove and entity[3] == 'CD')]\n",
    "        return [entity for entity in entities if entity != 'None']  # Exclude 'None' if other entities are present\n",
    "\n",
    "    # Enriching the entity annotations\n",
    "    for i, row in df.iterrows():\n",
    "        sentence = row['sentence']\n",
    "        entities = eval(row['ner']) if row['ner'] != \"['None']\" else ['None']\n",
    "        df.at[i, 'ner'] = add_new_entities(sentence, entities, new_entity_dict)\n",
    "\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e96b9a3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_sub_span(sub_span_range, full_spans_range):\n",
    "    # Check if a sub span is present in full span\n",
    "    if sub_span_range[0] in range(full_spans_range[0], full_spans_range[1]):\n",
    "        return sub_span_range\n",
    "\n",
    "def convert2IOB(text_data, ner_tags):\n",
    "    if 'None' in ner_tags:\n",
    "        # Split text into words and return with 'O' tags if there are no NER tags\n",
    "        split_text = tokenizer.tokenize(text_data)\n",
    "        return zip(split_text, ['O'] * len(split_text))\n",
    "\n",
    "    # Tokenize text and create a list of token spans\n",
    "    tokens = tokenizer.tokenize(text_data)\n",
    "    current_pos = 0\n",
    "    token_spans = []\n",
    "    for token in tokens:\n",
    "        start_pos = text_data.find(token, current_pos)\n",
    "        end_pos = start_pos + len(token)\n",
    "        token_spans.append((start_pos, end_pos))\n",
    "        current_pos = end_pos\n",
    "\n",
    "    # Initialize IOB tags as 'O'\n",
    "    iob_tags = ['O'] * len(tokens)\n",
    "\n",
    "    # Process each NER tag\n",
    "    for start, end, entity, entity_type in sorted(ner_tags, key=lambda x: x[0]):\n",
    "        for i, span in enumerate(token_spans):\n",
    "            if start <= span[0] < end:\n",
    "                if start == span[0]:\n",
    "                    # Beginning of entity\n",
    "                    iob_tags[i] = 'B-' + entity_type\n",
    "                else:\n",
    "                    # Inside of entity\n",
    "                    iob_tags[i] = 'I-' + entity_type\n",
    "\n",
    "    return zip(tokens, iob_tags)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b8a6658d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import pathlib\n",
    "from nltk.tokenize import WordPunctTokenizer, wordpunct_tokenize\n",
    "tokenizer = WordPunctTokenizer()\n",
    "\n",
    "def convert_to_IOB_format(data_file, output_folder,filename):\n",
    "    # Read the CSV file\n",
    "    df = pd.read_csv(data_file)\n",
    "\n",
    "    # Define the output path\n",
    "    result_path = os.path.join(output_folder, filename)\n",
    "    pathlib.Path(output_folder).mkdir(parents=True, exist_ok=True)\n",
    "    \n",
    "    with open(result_path, 'w', newline='\\n') as f1:\n",
    "        train_writer = csv.writer(f1, delimiter='\\t', lineterminator='\\n')\n",
    "        \n",
    "        for index, row in df.iterrows():\n",
    "            text = row['sentence']\n",
    "            ner_tags = eval(row['ner']) if row['ner'] != \"['None']\" else ['None']\n",
    "            tagged_tokens = convert2IOB(text, ner_tags)\n",
    "            for each_token in tagged_tokens:\n",
    "                train_writer.writerow(list(each_token))\n",
    "            train_writer.writerow('')\n",
    "\n",
    "output_folder = '../data/enriched_IOB/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "662184ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example usage\n",
    "filename = 'test.csv'\n",
    "\n",
    "file_path = '../data/CD_GP_DS_OG_test/clean_CD_GP_DS_OG_'+filename\n",
    "new_df = enrich_entities(file_path, extracted_entities, chem_to_remove)\n",
    "new_df.to_csv(file_path.replace('clean_','enriched_2_'), index=False)\n",
    "\n",
    "data_file = '../data/CD_GP_DS_OG_test/enriched_1_CD_GP_DS_OG_'+filename\n",
    "convert_to_IOB_format(data_file, output_folder,filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "19d90801",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c91451ad",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e78bc07",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b85cce2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e94d5e5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "206c2cea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42387535",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b2307dd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dece993e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is a XML aware sentenciser for JATS DTD. Sentence splitting rules are based on SSSplit.\n",
    "# Developed by Shyamasree Saha, 2021\n",
    "\n",
    "import argparse\n",
    "import sys\n",
    "import regex as re\n",
    "import io\n",
    "import gzip\n",
    "from bs4 import BeautifulSoup\n",
    "import lxml\n",
    "from collections import defaultdict\n",
    "from tqdm import tqdm\n",
    "\n",
    "# sys.setrecursionlimit(2500)\n",
    "# print(sys.getrecursionlimit())\n",
    "## defining all the patterns\n",
    "bad_white_space = \"(\\t|\\r|\\n)\"\n",
    "space_btw_tags = \"> <\"\n",
    "single_fig_graphic = \"<FIGURE/>|<graphic.*?/>\"  # replace with \"\"\n",
    "fig_stop = \"\\\\.<fig \"  # replace with \". <fig \"\n",
    "table_wrap_stop = \"\\\\.<table-wrap \"  # replace with \". <table-wrap \"\n",
    "allowedAttrChars = \"[\\\\w-_\\u2013\\\\.\\\\(\\\\)\\\\[\\\\]]\"\n",
    "refSCIgeneral = \"(?:(?:(?:\\\\d+[,\\u2013])*\\\\d+)?(?:</italic>)?<REF(?:(?: ID=\\\"(?:\\\\w-)?\\\\w+?(?: \\\\w+?)*\\\")|(?: \" \\\n",
    "                \"REFID=\\\"\\\\w+?\\\")|(?: text=\\\"(?:refs?\\\\.(?: )?)?(?:(?:\\\\d+[,�\\u2013\\u002D])*\\\\d+)?\\\\w*?\\\")|(?: \" \\\n",
    "                \"TYPE=\\\"\\\\w+?\\\"))*(?: ?/>|>(?:(?:refs?\\\\.(?: )?)?(?:(?:(?:\\\\d+(?:<italic>|<i>)?\\\\w?(?:</italic>|</i>)?)|(?:(\" \\\n",
    "                \"?:<italic>|<i>)?\\\\w(?:</italic>|</i>)?[-,\\u2013])|(?:(?:<italic>|<i>)?\\\\w(?:</italic>|</i>)?[-,\\u2013]))*(?:(?:\\\\d+(\" \\\n",
    "                \"?:<italic>|<i>)?\\\\w?(?:</italic>|</i>)?)|(?:(?:<italic>|<i>)?\\\\w?(?:</italic>|</i>)?)))?</REF>))|(?:<xref(?: \" \\\n",
    "                \"id=\\\"(?:\\\\w-)?\\\\w+?(?: \\\\w+?)*\\\")?(?: ref-type=\\\"(\" \\\n",
    "                \"?:aff|app|author-notes|bibr|boxed-text|chem|contrib|corresp|disp-formula|fig|fn|kwd|list|plate|ref\" \\\n",
    "                \"|scheme|sec|statement|supplementary-material|table|table-fn|other)\\\")?(?: rid=\\\"(?:[\" \\\n",
    "                \"-\\u2013\\\\w])*\\\")?>(?:(?:<sup>)?[-\\u2013,\\\\d]+(?:</sup>))</xref>))\"\n",
    "# refFootnote = \"<SUP TYPE=\\\"FOOTNOTE_MARKER\\\" ID=\\\"[^\\\"]+?\\\"/>\"\n",
    "atLeastOneRefSCI = \"(\" + refSCIgeneral + \"+\" + \"\\\\)?|\\\\))\"  # there may be a bracket after the\n",
    "# reference\n",
    "capturePunctuation = \"(\\\\.|\\\\?|(?<!\\\\(\\\\w{1,15})\\\\!)\"\n",
    "abbreviations = '((\\\\(|\\\\[| |>|^)(al|Am|Angew|approx|Biochim|Biophys|ca|cf|Chem|Co|conc|Dr|Drs|Corp|Ed|no|No|Sci|Nat' \\\n",
    "                '|Rep|e\\\\.g|eg|p\\\\.p\\\\.m|Engl|eq|eqns?|exp|Rs|Figs?|Labs?|Dr|etc|Calc|i\\\\.e|ie|Inc|Int|Lett|Ltd|p|p\\\\.a' \\\n",
    "                '|Phys|Prof|prot|refs?|Rev|sect|st|vs|(?<!(?:</SB>|(?:\\\\d ?(<italic>?))|<italic>))(?:(?:(?:[A-Z]|[' \\\n",
    "                'a-z])?\\\\.) ?[a-z])|(?<!(?:</SB>|(?:\\\\d ?(<italic>?))|<italic>))(?:(?:(?:[A-Z]|[a-z])\\\\.)? ?(?:(?:[A-Z]|[' \\\n",
    "                'a-z])\\\\.)? ?[A-Z])|(?<!(?:</SB>|(?:\\\\d ?(<italic>?))|<italic>))(?:(?:(?:[A-Z]|[a-z])\\\\.)? ?(?:(?:[A-Z]|[' \\\n",
    "                'a-z])\\\\.) ?[a-z])|(?<=(?:(?<!(?:</SB>|(?:\\\\d)))(?:(?: |\\\\(|^)<italic>)))(?: ?\\\\w{1,10})| ?\\\\. ?\\\\. ?|\\\\())'\n",
    "# moving all references inside the punctuation so that the sentence splitting is easier\n",
    "refSentence = \"(.*?)\" + capturePunctuation + atLeastOneRefSCI + \"(\\\\s?(?:</P>)?)\"\n",
    "abbrevs = \".*?(\" + abbreviations + \"(?:|</i>|</italic>)?)\" + \"$\"\n",
    "capitals = \"[A-Z0-9]\"  # caps and numbers may begin a sentence\n",
    "punctuation = \"(?:\\\\.|\\\\?|(?<!\\\\(\\\\w{1,15})\\\\!)\"  # .?!     \"(?:\\\\.|\\\\?|(?<!\\\\(\\\\w{1,15}))\\\\!|\\\\:(?=<list.+>))\"\n",
    "optPunctuation = punctuation + \"??\"\n",
    "endEquation = \"</EQN>\"\n",
    "# String endPara1 = \"(</P>|</ABSTRACT>)\"\n",
    "endPara = \"(</p>|</list>)\"\n",
    "beginPara = \"<p>\"\n",
    "# notFollowFigTab = \"(?!(<fig)|(<table-wrap))\";\n",
    "optStartQuote = \"['\\\"\\u201C]?\"\n",
    "optCloseQuote = \"['\\\"\\u201D]?\"\n",
    "optReferenceSCI = refSCIgeneral + \"*\"\n",
    "beginFirstSentence = \"^\"\n",
    "endLastSentence = \"$\"\n",
    "openHeader = \"(<HEADER/>|<label>)\"\n",
    "wholeHeader = \"((<BODY>)?(<DIV( DEPTH=\\\"\\\\d+\\\")?>)?(<HEADER( HEADER_MARKER=\\\"\" + allowedAttrChars + \\\n",
    "              \"+?\\\")?>.*?</HEADER>|<HEADER/>)|<TITLE>)\"\n",
    "optOpenHeader = openHeader + \"?\"\n",
    "\n",
    "eqn = \"<EQN( ID=\\\"\" + allowedAttrChars + \"+?\\\")?( TYPE=\\\"\" + allowedAttrChars + \"+?\\\")?>\"\n",
    "xref = \"<xref( .+)?>\"\n",
    "listTag = \"<list.+?>\"\n",
    "# manyStartTags = \"(\" + eqn + \"|\" + xref + \"|<BODY>|<DIV( DEPTH=\\\"\\d+\\\")?>|<P>|<B>|<IT>)*\"\n",
    "manyStartTags = \"(\" + xref + \"|\" + listTag + \"|<p.+?>|<license-p>|<title>|<caption>|<sec.+?>|<disp-quote>|<supplementary-material.+?>|<boxed-text.+?>|<list list-type=\\\"\\\\w{2,20}\\\">|<list-item>|<article-title>|<abstract.+>|<AbstractText.+>|<statement>|<def>|<fig .+>)*\"  # shs listTag on 13/10/09, <ABSTRACT>|<abstract.+> on 09/11/09\n",
    "optEndTags = \"(</label>|</boxed-text>|</list>|</list-item>|</statement>|</copyright-statement>)?\"\n",
    "# <list.+?> added to optEndTags1\n",
    "optEndTags1 = \"(</caption>|</boxed-text>|<list.+?>|</list>|</list-item>|</statement\" \\\n",
    "              \">|</inline-supplementary-material>|<inline-supplementary-material/>|</related-article>|</related\" \\\n",
    "              \"-object>|</address>|</alternatives>|</array>|</boxed-text>|</chem-struct-wrap>|</fig>|</fig-group\" \\\n",
    "              \">|<graphic.+?/>|</media>|</preformat>|</supplementary-material>|</table-wrap-foot>|</table-wrap\" \\\n",
    "              \">|</table-wrap-group>|</disp-formula>|</disp-formula-group>|</element-citation>|</mixed-citation\" \\\n",
    "              \">|</nlm-citation>|</bold>|</italic>|</monospace>|</overline>|</overline-start>|</overline-end>|</roman\" \\\n",
    "              \">|</sans-serif>|</sc>|</strike>|</underline>|</underline-start>|</underline-end>|</award-id>|</funding\" \\\n",
    "              \"-source>|</open-access>|</chem-struct>|<inline-formula/>|<inline-graphic/> \" \\\n",
    "              \"|</private-char>|</def-list>|</list>|</tex-math>|</mml:math>|</abbrev>|</milestone-end>|</milestone\" \\\n",
    "              \"-start>|</named-content>|</styled-content>|</ack> \" \\\n",
    "              \"|</disp-quote>|</speech>|</statement>|</verse-group>|</fn>|</target>|</xref>|</ref>|</sub>|</sup>\" \\\n",
    "              \">|</def>)*\"\n",
    "# </P> added on 23-11-10 by shs\n",
    "endTags = \"(</licence-p>|</italic>|</sup>|</bold>|</article-title>|</title>|</abstract>|<disp-formula>|</disp-quote>|</supplementary-material>|</boxed-text>|</list>|</list-item>|</statement>)\"\n",
    "manyEndTags = endTags + \"*\"\n",
    "# String sentenceTerminator = \"(?>\" +endPara+ \"|\"+ endEquation +\"|\" + \"(?<!(\\ |>)refs?)\"+punctuation+\")\"\n",
    "endParaOrEq = \"(\" + endPara + \"|\" + endEquation + \") ?\"  # shs notFollowedBy added\n",
    "formatting = \"(<B>|<IT>|<i>|<SP>|<italic>|<BOLD>|<bold>|<sup>|<disp-formula>|<named-content.+?>)\"\n",
    "puncNoAbbrv = \"(?<!\" + abbreviations + \"(</italic>)?)\" + punctuation + \" \"\n",
    "greekLetters = \"[\\u0370-\\u03FF\\u1F00-\\u1FFF]\"\n",
    "pAttr = \"<p.+?>\"\n",
    "sentenceCommencer = \"(?>\" + beginPara + \"|\" + pAttr + \"|\" + \"Fig(s)?\\\\.\" + \"|\" + \"fig(s)?\\\\.\" + \"|\" + capitals + \"|\" + \\\n",
    "                    formatting + \"|\" + \"\\\\[|\\\\(|\" + greekLetters + \"|\\u007C)\"\n",
    "equationCommencer = \"(\" + eqn + \".)\"\n",
    "commencer = \"(\" + sentenceCommencer + \"|\" + equationCommencer + \")\"\n",
    "noSpaceReqLookahead = manyStartTags + optOpenHeader + optStartQuote + commencer\n",
    "nocapsParaLookAhead = \"( ?<P>)\"\n",
    "startSentence = manyStartTags + optStartQuote + commencer\n",
    "# For matching the end of the previous sentence\n",
    "sentenceFigLookbehind = \"(?<=(?<!\" + abbreviations + punctuation + \")((\" + endParaOrEq + \")|(\" + puncNoAbbrv + \")|(\" \\\n",
    "                        + optPunctuation + optEndTags + endTags + \" ?)))\"\n",
    "\n",
    "# for matching the start of a sentence following a header\n",
    "headerLookahead = \"(?=(?:\" + manyStartTags + optOpenHeader + optStartQuote + commencer + \"))\"\n",
    "# (Headerstuff | ((normal sentence | firstsentence) sentenceContent, (punctuation|endEquation), optionalEndings,\n",
    "# lookahead))\n",
    "# String figLookAhead = \"\";\n",
    "Figure = \"<fig ?.+?</fig>\"\n",
    "tableWrap = \"<table-wrap.+?</table-wrap>|<table-wrap-foot.+?</table-wrap-foot>\"\n",
    "title = \"((?:<title/>)|(?:<title.+?</title>))\"  # \"<title.+?</title>\"\n",
    "# String absBodyLookBehind = \"((?:<abstract>)|(?:<abstract.+?>)|(?:<body>))\"\n",
    "secLookBehind = \"((?:<sec>)|(?:<sec.+?>))?\"  # <sec> can be followed by<body>.\n",
    "supplimentbehind = \"<supplementary-material.+?</supplementary-material>\"\n",
    "refList = \"<ref-list.+?</ref-list>\"\n",
    "boxed = \"(?:<boxed-text.+?>)(?:<caption>)?\"\n",
    "\n",
    "\n",
    "# def get_blocks_from_xml_string(xml_string, document_flag):\n",
    "#     sub_file_blocks = []\n",
    "#     xml = '<?xml version=\"1.0\" encoding=\"UTF-8\"?>'\n",
    "#     doc_str = '<!DOCTYPE '\n",
    "#     start_str = '<article '\n",
    "#     start_str2 = '<articles><article '\n",
    "#     doc_str_flag = 0\n",
    "\n",
    "#     lines = xml_string.split('\\n')\n",
    "#     for line in lines:\n",
    "#         if document_flag == 'f':\n",
    "#             if line.startswith(start_str) or line.startswith(start_str2):\n",
    "#                 sub_file_blocks.append(line.replace('^<articles>', ''))\n",
    "#                 doc_str_flag = 0\n",
    "#             elif line.startswith(doc_str):\n",
    "#                 doc_str_flag = 1\n",
    "#             else:\n",
    "#                 if doc_str_flag != 1:\n",
    "#                     sub_file_blocks[-1] += line.strip().replace('</articles>$', '')\n",
    "#         elif document_flag == 'a':\n",
    "#             start_str1 = '<article>'\n",
    "#             start_str2 = '<articles>'\n",
    "#             if line.startswith(start_str1):\n",
    "#                 sub_file_blocks.append(line)\n",
    "#             elif line.startswith(start_str2):\n",
    "#                 continue\n",
    "#             else:\n",
    "#                 if len(sub_file_blocks) > 0:\n",
    "#                     sub_file_blocks[-1] += line\n",
    "#         else:\n",
    "#             print('ERROR: unknown document type :' + document_flag)\n",
    "#             return []\n",
    "\n",
    "#     return sub_file_blocks\n",
    "\n",
    "def get_blocks_from_xml_string(xml_string, document_flag):\n",
    "    sub_file_blocks = []\n",
    "    xml = '<?xml version=\"1.0\" encoding=\"UTF-8\"?>'\n",
    "    doc_str = '<!DOCTYPE '\n",
    "    start_str = '<article '\n",
    "    start_str2 = '<articles><article '\n",
    "    doc_str_flag = 0\n",
    "\n",
    "    lines = xml_string.split('\\n')\n",
    "    for line in lines:\n",
    "        if document_flag == 'f':\n",
    "            if line.startswith(start_str) or line.startswith(start_str2):\n",
    "                sub_file_blocks.append(line.replace('^<articles>', ''))\n",
    "                doc_str_flag = 0\n",
    "            elif line.startswith(doc_str):\n",
    "                doc_str_flag = 1\n",
    "            else:\n",
    "                if doc_str_flag != 1:\n",
    "                    if sub_file_blocks:\n",
    "                        sub_file_blocks[-1] += line.strip().replace('</articles>$', '')\n",
    "        elif document_flag == 'a':\n",
    "            start_str1 = '<article>'\n",
    "            start_str2 = '<articles>'\n",
    "            if line.startswith(start_str1):\n",
    "                sub_file_blocks.append(line)\n",
    "            elif line.startswith(start_str2):\n",
    "                continue\n",
    "            else:\n",
    "                if sub_file_blocks:\n",
    "                    sub_file_blocks[-1] += line\n",
    "        else:\n",
    "            print('ERROR: unknown document type :' + document_flag)\n",
    "            return []\n",
    "\n",
    "    return sub_file_blocks\n",
    "\n",
    "\n",
    "def sentence_split(clear_string, sid):\n",
    "    # clear_string = str(xml_item)\n",
    "    #print(clear_string)\n",
    "    # Apply all replace statements\n",
    "    clear_string = re.sub(bad_white_space, '', clear_string)\n",
    "    clear_string = re.sub(space_btw_tags, '><', clear_string)\n",
    "    clear_string = re.sub(single_fig_graphic, '', clear_string)\n",
    "    clear_string = re.sub(fig_stop, '. <fig ', clear_string)\n",
    "    clear_string = re.sub(table_wrap_stop, '. <table-wrap ', clear_string)\n",
    "    clear_string = re.sub(\"</xref>, <xref\", \"</xref><xref\", clear_string)\n",
    "    # print(refSentence[740:])\n",
    "    refSentence_pattern = re.compile(refSentence, flags=0)\n",
    "    matches = re.finditer(refSentence_pattern, clear_string)\n",
    "    # print(matches)\n",
    "    swapped_string = \"\"\n",
    "    ref_last = None\n",
    "    for refm in matches:\n",
    "        a = refm.group(1)  # sentence\n",
    "        b = refm.group(2)  # punctuation\n",
    "        c = refm.group(3)  # reference\n",
    "        # print(\"a:\" + a + \"\\n b:\" + b + \"\\n c:\" + c + \"\\n\")\n",
    "        abbrevm = re.search(abbrevs, a)\n",
    "        if abbrevm:\n",
    "            swapped_string = swapped_string + a  # don't change order\n",
    "            swapped_string = swapped_string + b\n",
    "            swapped_string = swapped_string + c\n",
    "        else:\n",
    "            # for a I use match instead of search because match can search through newline etc. Whereas for c,\n",
    "            # I am looking for a number at the beginning of a sentence, so it does not matter.\n",
    "            if re.search('[0-9]]$', a) and re.search('^[1-9]]', c) and b == \".\":\n",
    "                # printl(\"a ends with number and c starts with number, in character by character comparison\")\n",
    "                swapped_string = swapped_string + a  # don't change order\n",
    "                swapped_string = swapped_string + b\n",
    "                swapped_string = swapped_string + c\n",
    "            else:\n",
    "                # print(\"order swapping\")\n",
    "                swapped_string = swapped_string + a  # sentence\n",
    "                swapped_string = swapped_string + c  # reference or bracket\n",
    "                swapped_string = swapped_string + b + \" \"  # punctuation\n",
    "        ref_last = refm\n",
    "    if ref_last:\n",
    "        end_bit = clear_string[ref_last.end():]\n",
    "    else:\n",
    "        end_bit = clear_string[len(swapped_string):]\n",
    "    # swapped_string = re.sub(\"</xref><xref\", \"</xref>, <xref\", swapped_string)\n",
    "    # print('swapped string')\n",
    "    # print(swapped_string)\n",
    "    # print('endbit')\n",
    "    swapped_string = swapped_string + end_bit\n",
    "    clear_string = swapped_string\n",
    "    clear_string = re.sub(\">\\\\.<\", \">. <\", clear_string)\n",
    "    # print('\\n\\n\\n Clear String:\\n')\n",
    "    # print(clear_string)\n",
    "    sentence_pattern = \"(\" + sentenceFigLookbehind + \"(\" + secLookBehind + title + \")+\" + \")|\" + \\\n",
    "                       \"(\" + sentenceFigLookbehind + \"(\" + boxed + secLookBehind + title + \"?)\" + optEndTags1 + \")|\" + \\\n",
    "                       \"(\" + sentenceFigLookbehind + wholeHeader + headerLookahead + \")|\" + \\\n",
    "                       \"(\" + sentenceFigLookbehind + \"(((\" + tableWrap + \")+(\" + Figure + \")+)|((\" + Figure + \")+(\" + \\\n",
    "                       tableWrap + \")+)|(\" + tableWrap + \")+|(\" + Figure + \")+)\" + optEndTags + \")|\" \\\n",
    "                       + \"(\" + sentenceFigLookbehind + \"(((\" + supplimentbehind + \")+\" + optEndTags + \")|((\" + \\\n",
    "                       supplimentbehind + \")+\" + secLookBehind + title + \")))|\" \\\n",
    "                       + \"(\" + sentenceFigLookbehind + \"(\" + refList + \")+\" + optEndTags1 + \")|\" \\\n",
    "                       + \"(((\" + sentenceFigLookbehind + \"\" + startSentence + \\\n",
    "                       \")|\" + beginFirstSentence + \"|\" + beginPara + \")\" + \\\n",
    "                       \"(.*?)(Fig(s)?\\\\..+?)*?\" + \\\n",
    "                       \"(((?<!(\" + endEquation + \" ?|\" + puncNoAbbrv + \" ?|\" + endPara + \" ?))\" \\\n",
    "                       + \"(?=(?:\" + nocapsParaLookAhead + \")))|\" + \\\n",
    "                       \"((?>\" + endEquation + \" ?|\" + puncNoAbbrv + \" ?|\" + endPara + \" ?)\" \\\n",
    "                       + optCloseQuote + optReferenceSCI + manyEndTags + \" ?\" \\\n",
    "                       + \"(?=(?:\" + noSpaceReqLookahead + \"|\" + nocapsParaLookAhead + \"|\\n| *$)))|\" + endLastSentence \\\n",
    "                       + \"))\"\n",
    "    somethingFound = 0\n",
    "    sentences = []\n",
    "    # print('before finding sentences\\n')\n",
    "    for m in re.finditer(sentence_pattern, clear_string):\n",
    "        somethingFound = 1\n",
    "        # print('M TUPLE:')\n",
    "        # print(m.group(0))\n",
    "        # print('\\n\\n')\n",
    "        sentences.append(m.group(0))\n",
    "    if somethingFound == 0:\n",
    "        print('No Sentence found')\n",
    "    # else:\n",
    "    #    print(sentences)\n",
    "    refSentenceRev = re.compile(\"(.*?)\" + atLeastOneRefSCI + capturePunctuation + \"(\\\\s?(?:</p>)?)\\\\Z\")\n",
    "    count = 0\n",
    "    newSentences = []\n",
    "    for s in sentences:\n",
    "        # print(\"in sentences: \" + str(count) + \": \" + s + \"\\n\")\n",
    "        refmRev = re.match(refSentenceRev, s)\n",
    "        # print(\"refSentenceRev:\"+refSentenceRev)\n",
    "        # if sentence finishes with reference + punctuation, swap the two over\n",
    "        if refmRev:\n",
    "            # print(\"ref found\")\n",
    "            a = refmRev.group(1)  # sentence\n",
    "            b = refmRev.group(2)  # reference or bracket or both\n",
    "            c = refmRev.group(3)  # punctuation\n",
    "            d = refmRev.group(4)  # space and/or </P>\n",
    "            # print(\"sentence \" + str(count) + \": \" + s+ \"A: \" + a + \"\\n B: \" + b + \"\\n C: \"+ c +\"\\n\", s, a, b, c)\n",
    "            ns = \"\"\n",
    "            # if (b.equals(\")\")) {\n",
    "            if b.endswith(\n",
    "                    \")\"):  # { // made it endsWith instead of equals on 28/03/11 coz if reference is within bracket we do not want to swap punctuation, it worked, 31,39). is no longer becoming 31,.39) , why square bracket is not considered in bracket?????\n",
    "                # #println(\"b ends with )\")\n",
    "                ns = ns + a  # sentence\n",
    "                ns = ns + b  # bracket\n",
    "                ns = ns + c  # punctuation\n",
    "                ns = ns + d  # space\n",
    "            else:\n",
    "                # print(\"b does not end with )\")\n",
    "                if not b.startswith(\"<\"):\n",
    "                    index = b.index(\"<\")\n",
    "                    # print(\"index:\"+index)\n",
    "                    b1 = b[0:index]  # end index exclusive\n",
    "                    # print(\"B1:\"+b1)\n",
    "                    # print(\"b at 0th position\"+b.toCharArray()[0])\n",
    "                    b2 = b[:index]\n",
    "                    a = a + b1\n",
    "                    b = b2\n",
    "                ns = ns + a  # sentence\n",
    "                ns = ns + c  # punctuation\n",
    "                ns = ns + b  # reference\n",
    "                ns = ns + d  # space\n",
    "            # print(\"sentence \" + str(count) + \": \" + s + \"A: \" + a + \"\\n B: \" + b + \"\\n C: \" + c + \"\\n\", s, a, b, c)\n",
    "            newSentences.append(ns)\n",
    "            # print(\"Sentence found: \" + str(count))\n",
    "        else:\n",
    "            # print(\"sentence added s:\"+s)\n",
    "            if s.replace(\" +\", \"\") != \"\":\n",
    "                newSentences.append(s)\n",
    "        # print(\"not if nor else : count=\"+count)\n",
    "        count = count + 1\n",
    "    sentences = newSentences\n",
    "    # Post-processing sentence array to move XML tags outside the sentences\n",
    "    # ppSentenceTag = \"(<s sid=\\\"\\\\d+\\\">)\"\n",
    "    ppStartTags = \"((?:<abstract>|<body>|<abstracttext(?:.+?)>|<sec(?:.+?)>|<div(\\\\sDEPTH=\\\"\\\\d+\\\")?>|<p>|<p.+>|<disp-quote>|<list.+?>|<list-item>|<abstract.+?>|<statement>|<boxed-text.+?>|<boxed-text.+?><graphic.+?>)*)\"\n",
    "    # added overall brackets and ?: 15/6/09 mal //shs <sec> <abstract.+?>\n",
    "    ppSentence = \"(.+?)\"\n",
    "    ppEndTags = \"((?:</abstract>|</body>|</abstract-title>|</sec>|</div>|</p>|</disp-quote>|</list>|</list-item>|<boxed-text.+?><graphic.+?></graphic></boxed-text>|</graphic>|</boxed-text>|</statement>|<list.+?>|<list-item>)*\\\\s?)\\\\Z\"\n",
    "    pp = re.compile(ppStartTags + ppSentence + ppEndTags)\n",
    "    ppHeader = re.compile(\".*?<caption.+?\", flags=re.I)\n",
    "    ppTitle = re.compile(\".*?<title.*?>|.*?<title>\", flags=re.I)\n",
    "    ppFig = re.compile(\"<fig\\\\s?.+?>|<table-wrap.+?>\", flags=re.I)\n",
    "    # ppCap = re.compile(\"</caption.+?>|</fn.+?>\",Pattern.CASE_INSENSITIVE); //shs comment : seems it can't find the tag without .+?\n",
    "    ppCap = re.compile(\"</caption.+?>|</fn.+?>\", flags=re.I)\n",
    "    paraAttr = re.compile(\"<p.+></p>|<p>\", flags=re.I)\n",
    "    suppli = re.compile(\"<supplementary-material.+?>\", flags=re.I)\n",
    "    # named = re.compile(\"<named-content.+?>\", flags=re.I)\n",
    "    # boxed_text = re.compile(\"<boxed-text.+?>\", flags=re.I)\n",
    "    sent_id = sid\n",
    "    # Vector<Element> allSentences = new Vector<Element>();\n",
    "    # print(\"before sentences loop.\")\n",
    "    nsentences = []\n",
    "    # print('before filling nsentences')\n",
    "    for s in sentences:\n",
    "        # print(s)\n",
    "        ppm = re.search(pp, s)\n",
    "        if ppm:\n",
    "            # print(\"Sentence \" + str(sent_id) + \" matches the post- processing tags\")\n",
    "            one = \"\"\n",
    "            if ppm.group(1):\n",
    "                one = ppm.group(1)\n",
    "            two = ppm.group(3)\n",
    "            three = \"\"\n",
    "            if ppm.group(4):\n",
    "                three = ppm.group(4)\n",
    "            # print(\"Group \\nSent id is: \" + str(sent_id) + \" Group 1 is: \" + ppm.group(1) + \" Group 2 is: \" + ppm.group(2) + \"Group 3 is: \" +ppm.group(3) + \"Group 4 is: \" + ppm.group(4))\n",
    "            # print(\"one:\" + str(one) + \", two: \" + str(two) + \", three: \" + str(three))\n",
    "            mHead = re.search(ppHeader, s)\n",
    "            mTitle = re.search(ppTitle, s)\n",
    "            mfig = re.search(ppFig, s)\n",
    "            mCap = re.search(ppCap, s)\n",
    "            mParaAttr = re.search(paraAttr, s)\n",
    "            mSuppli = re.search(suppli, s)\n",
    "            # mNamed = re.search(named, s)\n",
    "            # mBoxed = re.search(boxed_text, s)\n",
    "            # s.insertChild(titleNodes.get(titleNodes.size()-1).getValue(), 0);\n",
    "            if not (mHead or mTitle or mfig or mCap or mParaAttr or mSuppli):\n",
    "                # print('head, title, mfig etc')\n",
    "                # print(one + \"<sent id=\\\"\" + str(sent_id) + \"\\\">\" + two + \"</sent>\" + three)\n",
    "                nsentences.append(\n",
    "                    str(one) + \"<SENT id=\\\"\" + str(sent_id) + \"\\\"><plain>\" + str(two) + \"</plain></SENT>\" + str(three))\n",
    "                # print(\"Sentence \" + str(sent_id) + \": \" + s)\n",
    "                # print(\"after s tag:\"+one + \"<s sid=\\\"\" + id+ \"\\\">\" + two + \"</s>\" + three);\n",
    "                # if re.search(ppFig, two):\n",
    "                #    print(\"two contains fig\")\n",
    "                # print(two)\n",
    "                sent_id = sent_id + 1\n",
    "            else:\n",
    "                # this is a header, mHead matches\n",
    "                # print(\"\\n\\n\\nHeader : \"+s+\"\\n\\n \")\n",
    "                '''\n",
    "                if mHead:\n",
    "                    # print(\"after s tag:\" + one + \"<sent id=\\\"\" + str(sent_id) + \"\\\">\" + two + \"</sent>\" + three)\n",
    "                    print(\"Head matched.\")\n",
    "                if mTitle:\n",
    "                    # print(\"after s tag:\" + one + \"<sent id=\\\"\" + str(sent_id) + \"\\\">\" + two + \"</sent>\" + three)\n",
    "                    print(\"Title matched.\")\n",
    "                if mfig:\n",
    "                    # print(\"after s tag:\" + one + \"<sent id=\\\"\" + str(sent_id) + \"\\\">\" + two + \"</sent>\" + three)\n",
    "                    print(\"ppFig matched.\")\n",
    "                if mCap:\n",
    "                    # print(\"after s tag:\" + one + \"<sent id=\\\"\" + str(sent_id) + \"\\\">\" + two + \"</sent>\" + three)\n",
    "                    print(\"mCap matched.\")\n",
    "                if mParaAttr:\n",
    "                    # print(\"after s tag:\" + one + \"<sent id=\\\"\" + str(sent_id) + \"\\\">\" + two + \"</sent>\" + three)\n",
    "                    print(\"ParaAttr matched.\")\n",
    "                if mSuppli:\n",
    "                    # print(\"after s tag:\" + one + \"<sent id=\\\"\" + str(sent_id) + \"\\\">\" + two + \"</sent>\" + three)\n",
    "                    print(\"mSuppli matched.\")\n",
    "                '''\n",
    "                # print('No head, title, mfig etc')\n",
    "                # print(s)\n",
    "                nsentences.append(s)  # (\"<SENT id=\\\"\" + str(sent_id) + \"\\\">\" + two + \"</SENT>\")\n",
    "\n",
    "        else:\n",
    "            print(\"Sentence \" + str(sent_id) + \"-- \" + str(s) + \" -- didn't match!\")\n",
    "    # print('\\n\\n\\n Clear String:\\n')\n",
    "    # print(clear_string)\n",
    "    # print(\"N sentences\")\n",
    "    # print(nsentences)\n",
    "    clear_string = \"\".join(nsentences)\n",
    "    clear_string = re.sub(\"</xref><xref\", \"</xref>, <xref\", clear_string)\n",
    "    clear_string = re.sub(\"</SENT> <SENT \", \"</SENT><SENT \", clear_string)\n",
    "    return sent_id, clear_string\n",
    "\n",
    "\n",
    "# return StringInt(finalbuffer,sent_id)\n",
    "def create_tag(soup, sid):\n",
    "    sent_tag = soup.new_tag('SENT')\n",
    "    sent_tag['id'] = sid\n",
    "    return sent_tag\n",
    "\n",
    "\n",
    "def reference_sents(ref_list, sent_id):\n",
    "    # references are special sentences.\n",
    "    # print('in ref_list sentence')\n",
    "    n_sentences = \"\"\n",
    "    for ch in ref_list.children:\n",
    "        ch_str = str(ch)\n",
    "        if ch.name == 'ref':\n",
    "            sub_text = ''\n",
    "            for gch in ch.children:\n",
    "                # if gch.name in ['citation-alternatives', 'element-citation', 'mixed-citation', 'nlm-citation',\n",
    "                # 'note', 'x']:\n",
    "                sub_text = sub_text + \" \" + \" \".join([d.string for d in gch.descendants if d.string])\n",
    "                # print('sub_text')\n",
    "                # print(sub_text)\n",
    "            sent_id, n_sentences_method = sentence_split(sub_text, sent_id)\n",
    "            n_sentences = n_sentences + n_sentences_method\n",
    "        elif ch.name in [\"sec\", \"fig\", \"statement\", \"div\", \"boxed-text\", \"list\", \"list-item\", \"disp-quote\", \"speech\",\n",
    "                         \"fn-group\", \"fn\", \"def-list\", \"def-item\", \"def\", \"ack\", \"array\", \"table-wrap\", \"table\",\n",
    "                         \"tbody\", \"caption\", \"answer\", \"sec-meta\", \"glossary\", \"question\", \"question-wrap\", \"x\"]:\n",
    "            # print(\"in first elseif\")\n",
    "            sent_id, n_sentences_method = call_sentence_tags(ch, sent_id)\n",
    "            n_sentences = n_sentences + ch_str[:ch_str.find('>') + 1] + n_sentences_method + \"</\" + gch.name + \">\"\n",
    "        else:\n",
    "            n_sentences = n_sentences + ch_str\n",
    "    return sent_id, n_sentences\n",
    "\n",
    "\n",
    "def process_p_tag(gch, sent_id):\n",
    "    n_sentences = \"\"\n",
    "    p_children = gch.contents\n",
    "    gch_str = str(gch)\n",
    "    if len(p_children) == 1 and (not p_children[0].string) and (p_children[0].name in [\"ext-link\", \"e-mail\", \"uri\", \"inline-supplementary-material\",\n",
    "                                           \"related-article\", \"related-object\", \"address\", \"alternatives\", \"array\",\n",
    "                                           \"funding-source\", \"inline-graphic\"]):\n",
    "        n_sentences = n_sentences + gch_str\n",
    "    else:\n",
    "        # print('P tag:\\n')\n",
    "        # print(p_tag)\n",
    "        # print(p_tag[3:-4])\n",
    "        # check if p contains graphic and boxed-text without text inside.\n",
    "        sent_id, n_sentences_method = sentence_split(gch_str[gch_str.find('>')+1:-4], sent_id)\n",
    "        # print(n_sentences_method)\n",
    "        # print(sub_sent)\n",
    "        n_sentences = n_sentences + gch_str[:gch_str.find('>') + 1] + n_sentences_method + \"</\" + gch.name + \">\"\n",
    "    return sent_id, n_sentences\n",
    "\n",
    "\n",
    "def call_sentence_tags(ch, sent_id):\n",
    "    #print('In call_sentence_tags')\n",
    "    # print()\n",
    "    n_sentences = \"\"\n",
    "    for gch in ch.children:\n",
    "        # print(gch.name)\n",
    "        gch_str = str(gch)\n",
    "        if gch.name in ['article-title', 'title', 'subtitle', 'trans-title', 'trans-subtitle', 'alt-title', 'label', 'td', 'th']:\n",
    "            # assumption is that these tags only contain one sentence.\n",
    "            # print(\"GCH\")\n",
    "            # print(str(gch))\n",
    "            if gch.find('p', recursive=False):\n",
    "                sent_id, n_sentences_method = call_sentence_tags(gch, sent_id)\n",
    "                n_sentences = n_sentences + gch_str[:gch_str.find('>') + 1] + n_sentences_method + \"</\" + gch.name + \">\"\n",
    "            else:\n",
    "                sub_text = [str(d) for d in gch.children]\n",
    "                # print(sub_text)\n",
    "                n_sentences = n_sentences + gch_str[:gch_str.find('>') + 1] + '<SENT id=\"' + str(\n",
    "                    sent_id) + '\"><plain>' + \"\".join(sub_text) + \"</plain></SENT></\" + gch.name + \">\"\n",
    "                sent_id = sent_id + 1\n",
    "        elif gch.name in [\"sec\", \"fig\", \"statement\", \"div\", \"boxed-text\", \"list\", \"list-item\", \"disp-quote\", \"speech\",\n",
    "                          \"fn-group\", \"fn\", \"def-list\", \"def-item\", \"def\", \"ack\", \"array\", \"table-wrap\", \"table\",\n",
    "                          \"tbody\", \"thead\", \"tr\", \"caption\", \"answer\", \"sec-meta\", \"glossary\", \"question\", \"question-wrap\"]:\n",
    "            # print(\"in first elseif : sec etc string\")\n",
    "            # print(gch_str)\n",
    "            sent_id, n_sentences_method = call_sentence_tags(gch, sent_id)\n",
    "            n_sentences = n_sentences + gch_str[:gch_str.find('>') + 1] + n_sentences_method + \"</\" + gch.name + \">\"\n",
    "            # print(gch_str)\n",
    "            # print(gch_str[:gch_str.find('>')+1]) # works, tag.name will\n",
    "            # print(\"n sentences method\")\n",
    "            # print(n_sentences_method)\n",
    "            # print(gch_str[:gch_str.find('>')+1] + \"\".join(n_sentences_method) + \"</\" + gch.name + \">\")\n",
    "            # may need to do further changes\n",
    "        elif gch.name == 'p':\n",
    "            sent_id, n_sentences_method = process_p_tag(gch, sent_id)\n",
    "            n_sentences = n_sentences + n_sentences_method\n",
    "        else:\n",
    "            # convert the xml into string and concat\n",
    "            n_sentences = n_sentences + gch_str\n",
    "\n",
    "    return sent_id, n_sentences\n",
    "\n",
    "\n",
    "# def process_front(front):\n",
    "#     #print('In process_front')\n",
    "#     sent_id = 1\n",
    "#     n_sentences = \"\"\n",
    "#     if front.find('article-meta'):\n",
    "#         # print('In article-meta')\n",
    "#         art_meta = front.find('article-meta')\n",
    "#         # print(art_meta)\n",
    "#         # directly using 0th item because according to the JATS DTD it should appears precisely once.\n",
    "\n",
    "#         for ch in art_meta.find_all(recursive=False):\n",
    "#             ch_str = str(ch)\n",
    "#             if ch.name in ['title-group', 'supplement', 'supplementary-material', 'abstract', 'trans-abstract',\n",
    "#                            'kwd-group', 'funding-group']:\n",
    "#                 sent_id, n_sentences_method = call_sentence_tags(ch, sent_id)\n",
    "#                 n_sentences = n_sentences + ch_str[:ch_str.find('>') + 1] + n_sentences_method + \"</\" + ch.name + \">\"\n",
    "#             else:\n",
    "#                 n_sentences = n_sentences + ch_str\n",
    "#         #print('\\n\\nend of process_front')\n",
    "#     return sent_id, n_sentences\n",
    "\n",
    "\n",
    "# def process_body(body, sent_id):\n",
    "#     n_sentences_parts = []\n",
    "#     for ch in body.find_all(recursive=False):\n",
    "#         if ch.name == 'p':\n",
    "#             sent_id, n_sentences_method = process_p_tag(ch, sent_id)\n",
    "#             n_sentences_parts.append(n_sentences_method)\n",
    "#         elif ch.name in ['sec', 'ack', 'alternatives', 'array', 'preformat', 'fig', 'fig-group', 'question-wrap',\n",
    "#                          'question-wrap-group', 'list', 'table-wrap-group', 'table-wrap', 'display-formula',\n",
    "#                          'display-formula-group', 'def-list', 'list', 'supplementary-material', 'kwd-group',\n",
    "#                          'funding-group', 'statement', 'fig']:\n",
    "#             sent_id, n_sentences_method = call_sentence_tags(ch, sent_id)\n",
    "#             tag_open = str(ch)[:str(ch).find('>') + 1]\n",
    "#             tag_close = \"</\" + ch.name + \">\"\n",
    "#             n_sentences_parts.append(tag_open + n_sentences_method + tag_close)\n",
    "#         else:\n",
    "#             n_sentences_parts.append(str(ch))\n",
    "    \n",
    "#     n_sentences = ''.join(n_sentences_parts)\n",
    "#     return sent_id, n_sentences\n",
    "\n",
    "# def process_back(back, sent_id):\n",
    "#     # sent_id = 0\n",
    "#     #print('In process_back')\n",
    "#     n_sentences = \"\"\n",
    "#     for ch in back.find_all(recursive=False):\n",
    "#         ch_str = str(ch)\n",
    "#         if ch.name in ['sec', 'p', 'ack', 'alternatives', 'array', 'preformat', 'fig', 'fig-group' 'question-wrap',\n",
    "#              'question-wrap-group', 'list', 'table-wrap-group', 'table-wrap', 'display-formula',\n",
    "#              'display-formula-group', 'def-list', 'list', 'supplementary-material', 'kwd-group',\n",
    "#              'funding-group', 'statement', 'ref-list', 'glossary']:\n",
    "#             # print(ch.name)\n",
    "#             if ch.name == 'ref-list':\n",
    "#                 #print('\\n\\n\\n in ref-list')\n",
    "#                 sent_id, n_sentences_method = reference_sents(ch, sent_id)\n",
    "#             else:\n",
    "#                 sent_id, n_sentences_method = call_sentence_tags(ch, sent_id)\n",
    "#             n_sentences = n_sentences + ch_str[:ch_str.find('>') + 1] + n_sentences_method + \"</\" + ch.name + \">\"\n",
    "#         else:\n",
    "#             n_sentences = n_sentences + ch_str\n",
    "#     #print('\\n\\nend of process_body')\n",
    "#     return sent_id, n_sentences\n",
    "\n",
    "\n",
    "# def process_full_text(each_file):  \n",
    "#     # replacing body tag with orig_body to stop BeautifulSoup from removing this tag.\n",
    "#     each_file = each_file.replace('<body>', '<orig_body>')\n",
    "#     each_file = each_file.replace('<body ', '<orig_body ')\n",
    "#     each_file = each_file.replace('</body>', '</orig_body>')\n",
    "#     try:\n",
    "#         xml_soup = BeautifulSoup(each_file, 'lxml')\n",
    "#         # BeautifulSoup adds an extra html and body tag. removing those 2 tags.\n",
    "#         if xml_soup.html:\n",
    "#             # print('html unwrap')\n",
    "#             xml_soup.html.unwrap()\n",
    "#         if xml_soup.body:\n",
    "#             # print('body unwrap')\n",
    "#             xml_soup.body.unwrap()\n",
    "#         # print(xml_soup.find('front'))\n",
    "#         if xml_soup.find('orig_body'):\n",
    "#             # print('orig_body find')\n",
    "#             xml_soup.find('orig_body').name = 'body'\n",
    "#         sent_id = 1\n",
    "#         xml_string = None\n",
    "#         # Access regions of interests for sentence splitting.\n",
    "#         last_tag=None\n",
    "#         if xml_soup.article.find('front'):\n",
    "#             index = str(xml_soup).index('<front>')\n",
    "#             # print(\"\\n\\nFront index: \" + str(index))\n",
    "#             xml_string = str(xml_soup)[:index] + \"<front>\"\n",
    "#             # print(\"Pre-front\")\n",
    "#             # print(xml_string)\n",
    "#             sent_id, n_sentences_front = process_front(xml_soup.article.find('front'))\n",
    "#             xml_string = xml_string + n_sentences_front + \"</front>\"\n",
    "#             last_tag = \"</front>\"\n",
    "#         if xml_soup.article.find('body'):\n",
    "#             if xml_string == None:\n",
    "#                 index = str(xml_soup).index('<body>')\n",
    "#                 # print(\"\\n\\nBody index: \" + str(index))\n",
    "#                 xml_string = str(xml_soup)[:index]\n",
    "#                 # print(\"Pre-body\")\n",
    "#                 # print(xml_string)\n",
    "#             sent_id, n_sentences_body = process_body(xml_soup.article.find('body'), sent_id)\n",
    "#             xml_string = xml_string + \"<body>\" + n_sentences_body + \"</body>\"\n",
    "#             last_tag = \"</body>\"\n",
    "#         if xml_soup.article.find('back'):\n",
    "#             if xml_string == None:\n",
    "#                 index = str(xml_soup).index('<back>')\n",
    "#                 # print(\"\\n\\nBack index: \" + str(index))\n",
    "#                 xml_string = str(xml_soup)[:index]\n",
    "#                 # print(\"Pre-back\")\n",
    "#                 # print(xml_string)\n",
    "#             sent_id, n_sentences_back = process_back(xml_soup.article.find('back'), sent_id)\n",
    "#             xml_string = xml_string + \"<back>\" + n_sentences_back + \"</back>\"\n",
    "#             '''    \n",
    "#                 print('\\n\\n\\nFront sentences:\\n')\n",
    "#                 print(n_sentences_front)\n",
    "#                 print('\\n\\n\\nBody sentences:\\n')\n",
    "#                 print(n_sentences_body)\n",
    "#                 print('\\n\\n\\nBack sentences:\\n')\n",
    "#                 print(n_sentences_back)\n",
    "#             '''\n",
    "#             last_tag = \"</back>\"\n",
    "#         if last_tag:\n",
    "#             index_last_tag = str(xml_soup).index(last_tag)\n",
    "#             xml_string = xml_string + str(xml_soup)[index_last_tag + len(last_tag):]\n",
    "#             return xml_string\n",
    "#         else:\n",
    "#             print(\"No FRONT, BODY BACK.\")\n",
    "#             xml_string = str(xml_soup)\n",
    "#             return xml_string\n",
    "\n",
    "#     except Exception as e:\n",
    "#         print(e)\n",
    "#         return \"\"  # Return an empty string or handle the exception appropriately\n",
    "\n",
    "def process_body(body, sent_id):\n",
    "    n_sentences_parts = []\n",
    "    for ch in body.find_all(recursive=False):\n",
    "        if ch.name == 'p':\n",
    "            sent_id, n_sentences_method = process_p_tag(ch, sent_id)\n",
    "            n_sentences_parts.append(n_sentences_method)\n",
    "        elif ch.name in ['sec', 'ack', 'alternatives', 'array', 'preformat', 'fig', 'fig-group', 'question-wrap',\n",
    "                         'question-wrap-group', 'list', 'table-wrap-group', 'table-wrap', 'display-formula',\n",
    "                         'display-formula-group', 'def-list', 'list', 'supplementary-material', 'kwd-group',\n",
    "                         'funding-group', 'statement', 'fig']:\n",
    "            sent_id, n_sentences_method = call_sentence_tags(ch, sent_id)\n",
    "            tag_open = str(ch)[:str(ch).find('>') + 1]\n",
    "            tag_close = \"</\" + ch.name + \">\"\n",
    "            n_sentences_parts.append(tag_open + n_sentences_method + tag_close)\n",
    "        else:\n",
    "            n_sentences_parts.append(str(ch))\n",
    "    \n",
    "    n_sentences = ''.join(n_sentences_parts)\n",
    "    return sent_id, n_sentences\n",
    "\n",
    "def process_back(back, sent_id):\n",
    "    n_sentences_parts = []\n",
    "    for ch in back.find_all(recursive=False):\n",
    "        if ch.name in ['sec', 'p', 'ack', 'alternatives', 'array', 'preformat', 'fig', 'fig-group', 'question-wrap',\n",
    "                       'question-wrap-group', 'list', 'table-wrap-group', 'table-wrap', 'display-formula',\n",
    "                       'display-formula-group', 'def-list', 'list', 'supplementary-material', 'kwd-group',\n",
    "                       'funding-group', 'statement', 'ref-list', 'glossary']:\n",
    "            if ch.name == 'ref-list':\n",
    "                sent_id, n_sentences_method = reference_sents(ch, sent_id)\n",
    "            else:\n",
    "                sent_id, n_sentences_method = call_sentence_tags(ch, sent_id)\n",
    "            tag_open = str(ch)[:str(ch).find('>') + 1]\n",
    "            tag_close = \"</\" + ch.name + \">\"\n",
    "            n_sentences_parts.append(tag_open + n_sentences_method + tag_close)\n",
    "        else:\n",
    "            n_sentences_parts.append(str(ch))\n",
    "    \n",
    "    n_sentences = ''.join(n_sentences_parts)\n",
    "    return sent_id, n_sentences\n",
    "\n",
    "def process_front(front):\n",
    "    sent_id = 1\n",
    "    n_sentences_parts = []\n",
    "    if front.find('article-meta'):\n",
    "        art_meta = front.find('article-meta')\n",
    "        for ch in art_meta.find_all(recursive=False):\n",
    "            if ch.name in ['title-group', 'supplement', 'supplementary-material', 'abstract', 'trans-abstract',\n",
    "                           'kwd-group', 'funding-group']:\n",
    "                sent_id, n_sentences_method = call_sentence_tags(ch, sent_id)\n",
    "                tag_open = str(ch)[:str(ch).find('>') + 1]\n",
    "                tag_close = \"</\" + ch.name + \">\"\n",
    "                n_sentences_parts.append(tag_open + n_sentences_method + tag_close)\n",
    "            else:\n",
    "                n_sentences_parts.append(str(ch))\n",
    "    \n",
    "    n_sentences = ''.join(n_sentences_parts)\n",
    "    return sent_id, n_sentences\n",
    "\n",
    "\n",
    "def process_full_text(each_file):  \n",
    "    # Replacing body tag with orig_body to prevent BeautifulSoup from removing this tag.\n",
    "    each_file = each_file.replace('<body>', '<orig_body>')\n",
    "    each_file = each_file.replace('<body ', '<orig_body ')\n",
    "    each_file = each_file.replace('</body>', '</orig_body>')\n",
    "\n",
    "    try:\n",
    "        xml_soup = BeautifulSoup(each_file, 'lxml')\n",
    "\n",
    "        # BeautifulSoup adds an extra html and body tag. Removing those tags.\n",
    "        if xml_soup.html:\n",
    "            xml_soup.html.unwrap()\n",
    "        if xml_soup.body:\n",
    "            xml_soup.body.unwrap()\n",
    "\n",
    "        # Renaming 'orig_body' back to 'body'\n",
    "        if xml_soup.find('orig_body'):\n",
    "            xml_soup.find('orig_body').name = 'body'\n",
    "\n",
    "        processed_parts = []\n",
    "        sent_id = 1\n",
    "\n",
    "        for part in ['front', 'body', 'back']:\n",
    "            tag = xml_soup.article.find(part)\n",
    "            if tag:\n",
    "                if part == 'front':\n",
    "                    sent_id, processed_content = process_front(tag)\n",
    "                elif part == 'body':\n",
    "                    sent_id, processed_content = process_body(tag, sent_id)\n",
    "                elif part == 'back':\n",
    "                    sent_id, processed_content = process_back(tag, sent_id)\n",
    "                \n",
    "                processed_parts.append(f'<{part}>{processed_content}</{part}>')\n",
    "            else:\n",
    "                processed_parts.append(f'<{part}></{part}>')  # Empty tag if not found\n",
    "\n",
    "        return ''.join(processed_parts)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(e)\n",
    "        return \"\"  # Return an empty string or handle the exception appropriately\n",
    "\n",
    "    \n",
    "    \n",
    "def process_abstract_text(gch, sent_id, gch_flag):\n",
    "    n_sentences = \"\"\n",
    "    p_children = gch.contents\n",
    "    gch_str = str(gch)\n",
    "    sent_id, n_sentences_method = sentence_split(gch_str[gch_str.find('>') + 1:-(len(\"</\" + gch.name + \">\"))], sent_id)\n",
    "    #print(\"n sentences in process anbstract text\")\n",
    "    #print(n_sentences_method)\n",
    "    # print(sub_sent)\n",
    "    if gch_flag==1:\n",
    "        n_sentences = n_sentences + gch_str[:gch_str.find('>') + 1] + n_sentences_method + \"</\" + gch.name + \">\"\n",
    "    else:\n",
    "        n_sentences = n_sentences + n_sentences_method\n",
    "    return sent_id, n_sentences\n",
    "\n",
    "\n",
    "def process_abstracts(each_file, out_file):\n",
    "    sent_id = 1\n",
    "    try:\n",
    "        xml_soup = BeautifulSoup(each_file, 'lxml')\n",
    "        # BeautifulSoup adds an extra html and body tag. removing those 2 tags.\n",
    "        if xml_soup.html:\n",
    "            # print('html unwrap')\n",
    "            xml_soup.html.unwrap()\n",
    "        if xml_soup.body:\n",
    "            # print('body unwrap')\n",
    "            xml_soup.body.unwrap()\n",
    "        n_sentences = \"\"\n",
    "        #print(str(xml_soup))\n",
    "        target_tags = xml_soup.article.find_all([\"abstract\"])\n",
    "        if len(target_tags)>0:\n",
    "            # abstract_tags = xml_soup.pubmedarticle.medlinecitation.find_all([\"abstract\", \"otherabstract\"], recursive=False)\n",
    "            prev = None\n",
    "            xml_str = str(xml_soup)\n",
    "            #print(target_tags)\n",
    "            if target_tags:\n",
    "                # print(\"targets found\")\n",
    "                # print(abstract_tags)\n",
    "                ch_tag_end = 0\n",
    "                prev_start_index=0\n",
    "                prev=None\n",
    "                for ch in target_tags:\n",
    "                    #print(ch.name)\n",
    "                    ch_str = str(ch)\n",
    "                    match = re.search(\"<\" + ch.name +\"[^>]*?>\", xml_str)\n",
    "                    if match:\n",
    "                        #print(match.group())\n",
    "                        ch_tag_end = match.end()\n",
    "                        #print(\"ch_tag_end:\" + str(ch_tag_end))\n",
    "                    if prev==None:\n",
    "                        n_sentences = xml_str[:ch_tag_end]\n",
    "                        prev = ch\n",
    "                    else:\n",
    "                        # find end of previous ch\n",
    "                        #print(\"previous name:\" + prev.name)\n",
    "                        match = re.search(\"</\"+prev.name+\">\", xml_str)\n",
    "                        if match:\n",
    "                            prev_start_index = match.start()\n",
    "                        # print(\"prev_start:\" + str(prev_start_index))\n",
    "                        # print(match)\n",
    "                        n_sentences = n_sentences + xml_str[prev_start_index:ch_tag_end]\n",
    "                        prev=ch\n",
    "                    # print(\"before gchildren loop\")\n",
    "                    # print(n_sentences)\n",
    "                    # print(\"\\nEND of n_sentences\")\n",
    "                    if ch.name == \"title\":\n",
    "                        #print(\"article title\")\n",
    "                        sent_id, n_sentences_method = process_abstract_text(ch, sent_id, 0)\n",
    "                        #print(n_sentences_method)\n",
    "                        n_sentences = n_sentences + n_sentences_method\n",
    "                    else:\n",
    "                        for gch in ch.find_all(recursive=False):\n",
    "                            gch_str = str(gch)\n",
    "                            if gch.name=='title' or gch.name=='p':\n",
    "                                #print(\"abstracttext\")\n",
    "                                sent_id, n_sentences_method = process_abstract_text(gch, sent_id, 1)\n",
    "                                #print(n_sentences_method)\n",
    "                                n_sentences = n_sentences + n_sentences_method\n",
    "                            else:\n",
    "                                n_sentences = n_sentences + gch_str\n",
    "                if prev:\n",
    "                    match = re.search(\"</\"+prev.name+\">\", xml_str)\n",
    "                    if match:\n",
    "                        prev_start_index = match.start()\n",
    "                    n_sentences = n_sentences + xml_str[prev_start_index:]\n",
    "                else:\n",
    "                    print(\"\\nERROR: Prev should not be none\")\n",
    "            else:\n",
    "                print(\"No Abstract or Other Abstract\\n\")\n",
    "                n_sentences=xml_str\n",
    "        with open(out_file, 'a') as out:\n",
    "            out.write(n_sentences)\n",
    "    except Exception as e:\n",
    "        print(e)\n",
    "    \n",
    "\n",
    "def process_each_article(xml_content, document_flag):\n",
    "    files_list = get_blocks_from_xml_string(xml_content, document_flag)\n",
    "    processed_content = \"\"\n",
    "    \n",
    "    if document_flag == 'a':\n",
    "        processed_content += \"<articles>\\n\"\n",
    "    \n",
    "    for each_file in tqdm(files_list):\n",
    "        if document_flag == 'f':\n",
    "            processed_content += process_full_text(each_file)\n",
    "        elif document_flag == 'a':\n",
    "            print(\"abstract\")\n",
    "            processed_content += process_abstracts(each_file)\n",
    "        else:\n",
    "            print(\"Wrong document Flag\")\n",
    "            return \"\"\n",
    "\n",
    "    if document_flag == 'a':\n",
    "        processed_content += \"</articles>\\n\"\n",
    "\n",
    "    return processed_content\n",
    "\n",
    "# # Example usage\n",
    "# if __name__ == '__main__':\n",
    "#     parser = argparse.ArgumentParser(\n",
    "#         description='This script will process XML articles in JAT-Archive DTD and do sentence splitting')\n",
    "#     parser.add_argument(\"-x\", \"--xml\", required=True, help=\"XML content as string\", metavar=\"XML_STRING\")\n",
    "#     parser.add_argument(\"-d\", \"--document\", required=True, help=\"Document type flag\", metavar=\"FLAG\")\n",
    "#     args = parser.parse_args()\n",
    "\n",
    "#     modified_xml = process_each_article(args.xml, args.document)\n",
    "#     print(modified_xml)\n",
    "#     print(\"Sentenciser Finished!\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "4468b325",
   "metadata": {},
   "outputs": [],
   "source": [
    "from bs4 import BeautifulSoup\n",
    "import lxml\n",
    "from collections import defaultdict\n",
    "from tqdm import tqdm\n",
    "import random\n",
    "import sys, io, re, os\n",
    "import argparse\n",
    "from collections import defaultdict\n",
    "import re\n",
    "\n",
    "\n",
    "titleMapsBody = {\n",
    "    'INTRO': ['introduction', 'background', 'related literature', 'literature review', 'objective', 'aim ', 'purpose of this study', 'study (purpose|aim|aims)', '(\\d)+\\. (purpose|aims|aim)', '(aims|aim|purpose) of the study', '(the|drug|systematic|book) review', 'review of literature', 'related work', 'recent advance'],\n",
    "    'METHODS': ['supplement', 'methods and materials', 'method', 'material', 'experimental procedure', 'implementation', 'methodology', 'treatment', 'statistical analysis', \"experimental\", '(\\d)+\\. experimental$', 'experimental (section|evaluation|design|approach|protocol|setting|set up|investigation|detail|part|pespective|tool)', \"the study\", '(\\d)+\\. the study$', \"protocol\", \"protocols\", 'study protocol', 'construction and content', 'experiment (\\d)+', '^experiments$', 'analysis', 'utility', 'design', '(\\d)+\\. theory$', \"theory\", 'theory and ', 'theory of '],\n",
    "    'RESULTS': ['result', 'finding', 'diagnosis'],\n",
    "    'DISCUSS': ['discussion', 'management of', '(\\d)+\\. management', 'safety and tolerability', 'limitations', 'perspective', 'commentary', '(\\d)+\\. comment'],\n",
    "    'CONCL': ['conclusion', 'key message', 'future', 'summary', 'recommendation', 'implications for clinical practice','concluding remark'],\n",
    "    'CASE': ['case study report', 'case report', 'case presentation', 'case description', 'case (\\d)+', '(\\d)+\\. case', 'case summary', 'case history'],\n",
    "    'ACK_FUND': ['funding', 'acknowledgement', 'acknowledgment', 'financial disclosure'],\n",
    "    'AUTH_CONT': ['author contribution', 'authors\\' contribution', 'author\\'s contribution'],\n",
    "    'COMP_INT': ['competing interest', 'conflict of interest', 'conflicts of interest', 'disclosure', 'decleration'],\n",
    "    'ABBR': ['abbreviation'],\n",
    "    'SUPPL': ['supplemental data', 'supplementary file', 'supplemental file', 'supplementary data', 'supplementary figure', 'supplemental figure', 'supporting information', 'supplemental file', 'supplemental material', 'supplementary material', 'supplement material', 'additional data files', 'supplemental information', 'supplementary information', 'supplemental information', 'supporting information', 'supplemental table', 'supplementary table', 'supplement table', 'supplementary material', 'supplemental material', 'supplement material', 'supplementary video']\n",
    "}\n",
    "\n",
    "titleExactMapsBody = {\n",
    "    'INTRO': [\"aim\", \"aims\", \"purpose\", \"purposes\", \"purpose/aim\", \"purpose of study\", \"review\", \"reviews\", \"minireview\"],\n",
    "    'METHODS': [\"experimental\", \"the study\", \"protocol\", \"protocols\"],\n",
    "    'DISCUSS': [\"management\", \"comment\", \"comments\"],\n",
    "    'CASE': [\"case\", \"cases\"]\n",
    "}\n",
    "\n",
    "titleMapsBack = {\n",
    "    'REF': ['reference', 'literature cited', 'references', 'bibliography'],\n",
    "    'ACK_FUND': ['funding', 'acknowledgement', 'acknowledgment', 'aknowledgement', 'acknowlegement', 'open access', 'financial support', 'grant', 'author note', 'financial disclosure'],\n",
    "    'ABBR': ['abbreviation', 'glossary'],\n",
    "    'COMP_INT': ['competing interest', 'conflict of interest', 'conflicts of interest', 'disclosure', 'decleration', 'conflits', 'interest'],\n",
    "    'SUPPL': ['supplementary', 'supporting information', 'supplemental', 'web extra material'],\n",
    "    'APPENDIX': ['appendix', 'appendices'],\n",
    "    'AUTH_CONT': ['author', 'contribution']\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "def categorize_section(title):\n",
    "    # Lowercase the title for consistency\n",
    "    title_lower = title.lower()\n",
    "\n",
    "    # INTRO\n",
    "    if any(re.search(pattern, title_lower) for pattern in ['introduction', 'background', 'literature review', 'objective', 'purpose of this study', 'systematic review', 'recent advance']):\n",
    "        return 'INTRO'\n",
    "\n",
    "    # METHODS\n",
    "    # Enhanced with NLP logic: If the title contains 'experiment' or 'study', categorize as METHODS\n",
    "    if 'experiment' in title_lower or 'study' in title_lower:\n",
    "        return 'METHODS'\n",
    "    if any(re.search(pattern, title_lower) for pattern in ['methods and materials', 'methodology', 'experimental procedure', 'implementation', 'treatment', 'statistical analysis', 'experimental design', 'study protocol', 'design theory']):\n",
    "        return 'METHODS'\n",
    "\n",
    "    # RESULTS\n",
    "    if any(re.search(pattern, title_lower) for pattern in ['result', 'finding', 'diagnosis']):\n",
    "        return 'RESULTS'\n",
    "\n",
    "    # DISCUSS\n",
    "    if any(re.search(pattern, title_lower) for pattern in ['discussion', 'management of', 'safety and tolerability', 'limitations', 'perspective', 'commentary']):\n",
    "        return 'DISCUSS'\n",
    "\n",
    "    # CONCL\n",
    "    if any(re.search(pattern, title_lower) for pattern in ['conclusion', 'key message', 'future direction', 'summary', 'recommendation', 'implications for clinical practice', 'concluding remark']):\n",
    "        return 'CONCL'\n",
    "\n",
    "    # CASE\n",
    "    if any(re.search(pattern, title_lower) for pattern in ['case study report', 'case presentation', 'case description', 'case history']):\n",
    "        return 'CASE'\n",
    "\n",
    "    # ACK_FUND\n",
    "    if any(re.search(pattern, title_lower) for pattern in ['funding', 'acknowledgement', 'financial disclosure']):\n",
    "        return 'ACK_FUND'\n",
    "\n",
    "    # AUTH_CONT\n",
    "    if any(re.search(pattern, title_lower) for pattern in ['author contribution']):\n",
    "        return 'AUTH_CONT'\n",
    "\n",
    "    # COMP_INT\n",
    "    if any(re.search(pattern, title_lower) for pattern in ['competing interest', 'conflict of interest', 'disclosure']):\n",
    "        return 'COMP_INT'\n",
    "\n",
    "    # ABBR\n",
    "    if any(re.search(pattern, title_lower) for pattern in ['abbreviation', 'glossary']):\n",
    "        return 'ABBR'\n",
    "\n",
    "    # SUPPL\n",
    "    if any(re.search(pattern, title_lower) for pattern in ['supplemental data', 'supplementary material', 'additional data files', 'supporting information']):\n",
    "        return 'SUPPL'\n",
    "\n",
    "    # If no category matches, return 'OTHER'\n",
    "    return 'OTHER'\n",
    "\n",
    "# Assuming soup is a BeautifulSoup object of the parsed XML/HTML document\n",
    "\n",
    "def extract_sections_and_sentences(soup, titleMapsBody, titleExactMapsBody, titleMapsBack):\n",
    "    sections_sentences = defaultdict(list)\n",
    "    \n",
    "    section_order = []\n",
    "    \n",
    "    # Function to identify section type based on title\n",
    "    def identify_section_type(title, secFlag):\n",
    "        title_lower = title.lower()\n",
    "\n",
    "        if secFlag == 'body':\n",
    "            title_map = titleMapsBody\n",
    "            exact_title_map = titleExactMapsBody\n",
    "        elif secFlag == 'back':\n",
    "            title_map = titleMapsBack\n",
    "            exact_title_map = None\n",
    "        else:\n",
    "            return 'OTHER'\n",
    "\n",
    "        # Check for exact match first\n",
    "        if exact_title_map:\n",
    "            for key, patterns in exact_title_map.items():\n",
    "                if title_lower in patterns:\n",
    "                    return key\n",
    "\n",
    "        # Check for partial match\n",
    "        for key, patterns in title_map.items():\n",
    "            if any(re.search(pattern, title_lower) for pattern in patterns):\n",
    "                return key\n",
    "\n",
    "        # Default to 'OTHER' if no match found\n",
    "        return 'OTHER'\n",
    "\n",
    "     # Function to extract and categorize sections\n",
    "    def extract_and_categorize_sections(soup_section, secFlag):\n",
    "        for sec in soup_section.find_all('sec', recursive=False):\n",
    "            if sec.title:\n",
    "                section_type = identify_section_type(sec.title.text, secFlag)\n",
    "                section_order.append(section_type)\n",
    "                for p in sec.find_all('p', recursive=False):\n",
    "                    sections_sentences[section_type].append(p.text.strip())\n",
    "                    \n",
    "    # Extract from body and back sections\n",
    "    if soup.body:\n",
    "        extract_and_categorize_sections(soup.body, 'body')\n",
    "    if soup.back:\n",
    "        extract_and_categorize_sections(soup.back, 'back')\n",
    "\n",
    "    # Apply logic for 'OTHER' sections\n",
    "    for index, section in enumerate(section_order):\n",
    "        if section == 'OTHER':\n",
    "            # If it follows 'INTRO', classify as 'METHODS'\n",
    "            if index > 0 and section_order[index - 1] == 'INTRO':\n",
    "                sections_sentences['METHODS'].extend(sections_sentences['OTHER'])\n",
    "                section_order[index] = 'METHODS'\n",
    "            # If it precedes 'DISCUSS', classify as 'RESULTS'\n",
    "            elif index < len(section_order) - 1 and section_order[index + 1] == 'DISCUSS':\n",
    "                sections_sentences['RESULTS'].extend(sections_sentences['OTHER'])\n",
    "                section_order[index] = 'RESULTS'\n",
    "\n",
    "    # Remove the reclassified 'OTHER' sections\n",
    "    if 'OTHER' in sections_sentences:\n",
    "        del sections_sentences['OTHER']\n",
    "\n",
    "    return sections_sentences\n",
    "\n",
    "\n",
    "\n",
    "def process_each_file(filename):\n",
    "    each_file = each_file.replace('<body>', '<orig_body>')\n",
    "    each_file = each_file.replace('<body ', '<orig_body ')\n",
    "    each_file = each_file.replace('</body>', '</orig_body>')\n",
    "    xml_soup = BeautifulSoup(each_file, 'lxml')\n",
    "    xml_soup.html.unwrap()\n",
    "    xml_soup.body.unwrap()\n",
    "    if xml_soup.find('orig_body'):\n",
    "        #print(\"orig_body found\")\n",
    "        xml_soup.find('orig_body').name = 'body'\n",
    "\n",
    "    extract_sections_and_sentences(xml_soup)\n",
    "\n",
    "\n",
    "def extract_section_titles(xml_content):\n",
    "    # Parse the XML/HTML content\n",
    "    # Find all 'sec' tags and extract their 'title' tags\n",
    "    section_titles = [sec.find('title').get_text(strip=True) for sec in soup.find_all('sec') if sec.find('title')]\n",
    "\n",
    "\n",
    "    return section_titles\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "5423dbd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "# pmcid = 'PMC8080698'\n",
    "\n",
    "pmcid = 'PMC1501050'\n",
    "\n",
    "url = f\"https://www.ebi.ac.uk/europepmc/webservices/rest/{pmcid}/fullTextXML\"\n",
    "response = requests.get(url)\n",
    "\n",
    "if response.status_code == 200:\n",
    "    xml_content = response.content.decode('utf-8')\n",
    "#     processed_xml = process_each_article(xml_content, document_flag='f')\n",
    "    processed_xml = get_blocks_from_xml_string(xml_content, document_flag='f')\n",
    "#     soup = BeautifulSoup(response.content, 'xml')\n",
    "#     p_tags = soup.find_all('p')\n",
    "#     p_texts = [tag.get_text() for tag in p_tags]\n",
    "#     # Using nlp.pipe for batch processing\n",
    "#     for doc in nlp.pipe(p_texts):\n",
    "#         sentences = [sent.text.strip() for sent in doc.sents]\n",
    "#         segmented_sentences.extend(sentences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "f6479781",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<article article-type=\"product-review\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><front><journal-meta><journal-id journal-id-type=\"nlm-ta\">BMC Bioinformatics</journal-id><journal-title>BMC Bioinformatics</journal-title><issn pub-type=\"epub\">1471-2105</issn><publisher><publisher-name>BioMed Central</publisher-name><publisher-loc>London</publisher-loc></publisher></journal-meta><article-meta><article-id pub-id-type=\"pmcid\">1501050</article-id><article-id pub-id-type=\"publisher-id\">1471-2105-7-194</article-id><article-id pub-id-type=\"pmid\">16600052</article-id><article-id pub-id-type=\"doi\">10.1186/1471-2105-7-194</article-id><article-categories><subj-group subj-group-type=\"heading\"><subject>Software</subject></subj-group></article-categories><title-group><article-title>Mining gene expression data by interpreting principal components</article-title></title-group><contrib-group><contrib contrib-type=\"author\" corresp=\"yes\" id=\"A1\"><name><surname>Roden</surname><given-names>Joseph C</given-names></name><xref ref-type=\"aff\" rid=\"I1\">1</xref><email>joe.roden@jpl.nasa.gov</email></contrib><contrib contrib-type=\"author\" id=\"A2\"><name><surname>King</surname><given-names>Brandon W</given-names></name><xref ref-type=\"aff\" rid=\"I2\">2</xref><email>kingb@caltech.edu</email></contrib><contrib contrib-type=\"author\" id=\"A3\"><name><surname>Trout</surname><given-names>Diane</given-names></name><xref ref-type=\"aff\" rid=\"I2\">2</xref><email>diane@caltech.edu</email></contrib><contrib contrib-type=\"author\" id=\"A4\"><name><surname>Mortazavi</surname><given-names>Ali</given-names></name><xref ref-type=\"aff\" rid=\"I2\">2</xref><email>alim@caltech.edu</email></contrib><contrib contrib-type=\"author\" id=\"A5\"><name><surname>Wold</surname><given-names>Barbara J</given-names></name><xref ref-type=\"aff\" rid=\"I2\">2</xref><email>woldb@caltech.edu</email></contrib><contrib contrib-type=\"author\" id=\"A6\"><name><surname>Hart</surname><given-names>Christopher E</given-names></name><xref ref-type=\"aff\" rid=\"I2\">2</xref><email>hart@caltech.edu</email></contrib></contrib-group><aff id=\"I1\"><label>1</label>Jet Propulsion Laboratory, California Institute of Technology, Pasadena, USA</aff><aff id=\"I2\"><label>2</label>Division of Biology, California Institute of Technology, Pasadena, USA</aff><pub-date pub-type=\"collection\"><year>2006</year></pub-date><pub-date pub-type=\"epub\"><day>7</day><month>4</month><year>2006</year></pub-date><volume>7</volume><fpage>194</fpage><lpage>194</lpage><ext-link ext-link-type=\"uri\" xlink:href=\"http://www.biomedcentral.com/1471-2105/7/194\"/><history><date date-type=\"received\"><day>3</day><month>7</month><year>2005</year></date><date date-type=\"accepted\"><day>7</day><month>4</month><year>2006</year></date></history><permissions><copyright-statement>Copyright © 2006 Roden et al; licensee BioMed Central Ltd.</copyright-statement><copyright-year>2006</copyright-year><copyright-holder>Roden et al; licensee BioMed Central Ltd.</copyright-holder><license license-type=\"open-access\" xlink:href=\"http://creativecommons.org/licenses/by/2.0\"><p>This is an Open Access article distributed under the terms of the Creative Commons Attribution License (<ext-link ext-link-type=\"uri\" xlink:href=\"http://creativecommons.org/licenses/by/2.0\"/>), which permits unrestricted use, distribution, and reproduction in any medium, provided the original work is properly cited.</p><!--<rdf xmlns=\"http://web.resource.org/cc/\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\" xmlns:dc=\"http://purl.org/dc/elements/1.1\" xmlns:dcterms=\"http://purl.org/dc/terms\"><Work xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:dcterms=\"http://purl.org/dc/terms/\" rdf:about=\"\"><license rdf:resource=\"http://creativecommons.org/licenses/by/2.0\"/><dc:type rdf:resource=\"http://purl.org/dc/dcmitype/Text\"/><dc:author>RodenCJosephjoe.roden@jpl.nasa.gov</dc:author><dc:title>Mining gene expression data by interpreting principal components</dc:title><dc:date>2006</dc:date><dcterms:bibliographicCitation>BMC Bioinformatics 7(1): 194-. (2006)</dcterms:bibliographicCitation><dc:identifier type=\"sici\">1471-2105(2006)7:1&#x0003c;194&#x0003e;</dc:identifier><dcterms:isPartOf>urn:ISSN:1471-2105</dcterms:isPartOf><License rdf:about=\"http://creativecommons.org/licenses/by/2.0\"><permits rdf:resource=\"http://web.resource.org/cc/Reproduction\" xmlns=\"\"/><permits rdf:resource=\"http://web.resource.org/cc/Distribution\" xmlns=\"\"/><requires rdf:resource=\"http://web.resource.org/cc/Notice\" xmlns=\"\"/><requires rdf:resource=\"http://web.resource.org/cc/Attribution\" xmlns=\"\"/><permits rdf:resource=\"http://web.resource.org/cc/DerivativeWorks\" xmlns=\"\"/></License></Work></rdf>--></license></permissions><abstract><sec><title>Background</title><p>There are many methods for analyzing microarray data that group together genes having similar patterns of expression over all conditions tested. However, in many instances the biologically important goal is to identify relatively small sets of genes that share coherent expression across only some conditions, rather than all or most conditions as required in traditional clustering; e.g. genes that are highly up-regulated and/or down-regulated similarly across only a subset of conditions. Equally important is the need to learn which conditions are the decisive ones in forming such gene sets of interest, and how they relate to diverse conditional covariates, such as disease diagnosis or prognosis.</p></sec><sec><title>Results</title><p>We present a method for automatically identifying such candidate sets of biologically relevant genes using a combination of principal components analysis and information theoretic metrics. To enable easy use of our methods, we have developed a data analysis package that facilitates visualization and subsequent data mining of the independent sources of significant variation present in gene microarray expression datasets (or in any other similarly structured high-dimensional dataset). We applied these tools to two public datasets, and highlight sets of genes most affected by specific subsets of conditions (e.g. tissues, treatments, samples, etc.). Statistically significant associations for highlighted gene sets were shown via global analysis for Gene Ontology term enrichment. Together with covariate associations, the tool provides a basis for building testable hypotheses about the biological or experimental causes of observed variation.</p></sec><sec><title>Conclusion</title><p>We provide an unsupervised data mining technique for diverse microarray expression datasets that is distinct from major methods now in routine use. In test uses, this method, based on publicly available gene annotations, appears to identify numerous sets of biologically relevant genes. It has proven especially valuable in instances where there are many diverse conditions (10's to hundreds of different tissues or cell types), a situation in which many clustering and ordering algorithms become problematic. This approach also shows promise in other topic domains such as multi-spectral imaging datasets.</p></sec></abstract></article-meta></front><body><sec><title>Background</title><p>Bioinformatics has placed much emphasis on using various unsupervised clustering techniques as a means to understand the information present in gene microarray expression datasets. Clustering techniques produce a rich taxonomy of results by defining groups of genes that act more or less similarly across a number of experimental conditions. The diverse approaches to clustering genes by expression levels include k-means [<xref ref-type=\"bibr\" rid=\"B1\">1</xref>], self-organizing maps [<xref ref-type=\"bibr\" rid=\"B2\">2</xref>], hierarchical algorithms [<xref ref-type=\"bibr\" rid=\"B3\">3</xref>,<xref ref-type=\"bibr\" rid=\"B4\">4</xref>] and probabilistic models [<xref ref-type=\"bibr\" rid=\"B5\">5</xref>]. Some approaches permit clustering of the conditions as well [<xref ref-type=\"bibr\" rid=\"B6\">6</xref>-<xref ref-type=\"bibr\" rid=\"B8\">8</xref>]. Based on co-expression, genes that comprise individual expression clusters are often postulated to be co-regulated, and to the extent that this hypothesis is correct in any specific biological situation, the gene cluster definitions can offer key insights into gene regulatory network (GRN) structure and function.</p><p>Another common data mining task is to try to identify small sets of genes that can serve as effective predictors of disease diagnosis or prognosis. While clustering at its best is good at finding sets of genes that are similarly expressed across all conditions within a dataset, many issues (e.g. selection of K, stochastic effects, and \"noise\" from large numbers of genes that change little over most of the conditions) can prevent clustering from successfully highlighting small groups of interestingly co-expressed genes [<xref ref-type=\"bibr\" rid=\"B9\">9</xref>,<xref ref-type=\"bibr\" rid=\"B10\">10</xref>]. This often encountered problem is addressed in part by hierarchical phylogenetic ordering algorithms such as average linkage in Mike Eisen's cluster program [<xref ref-type=\"bibr\" rid=\"B3\">3</xref>], but the information biologists seek regarding shared sub-patterns of co-expression can be obscured by both algorithmic and visualization constraints. The algorithmic limitations in hierarchical clustering confound and \"cover\" the presence and organization of smaller and more specific gene groups that are similar across only a subset of conditions within the larger dataset. In any case, biologists generally subjectively define a cluster of genes from such phylogenetic trees based largely on human pattern recognition. Finally, nothing inherent in the clustering approach helps to direct a biologist to which cluster is interesting or relevant. Instead, biologists generally take the path of focussing on a group of genes exhibiting a pattern of expression that supports a specific hypothesis, or search for a known gene or genes of interest within a cluster to form an explanation for others in the cluster.</p><p>Support vector machines have been shown to be useful for identifying small sets of related and predictive genes [<xref ref-type=\"bibr\" rid=\"B11\">11</xref>-<xref ref-type=\"bibr\" rid=\"B14\">14</xref>], but represent a supervised learning approach which requires one to first define a set of known positive examples, a set of known negative examples, and a specific covariate to predict. We wanted an unsupervised algorithm that would help us to find relationships and structure in the data that is more specific than what clustering algorithms usually deliver, yet is hypothesis independent. We found it efficient and useful to use as an independent starting tool a very direct approach based on principal components analysis (PCA, see Methods section). A virtue is that this approach is computationally efficient for very large datasets, especially compared with most clustering algorithms, but is also applicable to much smaller ones. It allows one to directly explore each of the independent and diverse sources of variation present within a gene expression dataset and to subsequently identify the specific genes that vary the most, together with the conditions in which they vary.</p><p>Unlike conventional clustering and ordering algorithms, this PCA based approach permits a gene to be highlighted and grouped as influential in multiple condition sets, whereas in cluster membership a gene is typically assigned to one unique cluster. The \"single cluster assignment\" quality of traditional clustering and ordering algorithms is problematical because it tends to hide commonality of expression that is restricted to a small, interesting, and often entirely unpredicted subgroups of tissues, cell types, treatments or other condition types. This situation, perhaps because of inherent properties of gene network structure, will arise increasingly as the number and diversity of conditions represented in expression datasets increases.</p><p>The clustering method of Barkai et al. [<xref ref-type=\"bibr\" rid=\"B15\">15</xref>,<xref ref-type=\"bibr\" rid=\"B16\">16</xref>] addresses this issue of multiple membership in a different way, by using randomly-selected gene sets to iteratively search for and refine self-consistent groups. Their approach, which is related to PCA through singular value decomposition (SVD), also permits genes to be assigned to multiple \"expression modules.\" In contrast to the method presented here, there is no provision for correlating modules with covariate data.</p><p>The use of principal components analysis presented here differs from other recent applications in gene expression analysis. PCA is most commonly used in as a means of dimensionality reduction prior to clustering [<xref ref-type=\"bibr\" rid=\"B7\">7</xref>,<xref ref-type=\"bibr\" rid=\"B17\">17</xref>] or prior to classification [<xref ref-type=\"bibr\" rid=\"B18\">18</xref>,<xref ref-type=\"bibr\" rid=\"B19\">19</xref>]. It is also used to visualize or confirm clustering results [<xref ref-type=\"bibr\" rid=\"B19\">19</xref>-<xref ref-type=\"bibr\" rid=\"B21\">21</xref>]. In contrast, our use of PCA aims to find, then examine, and where possible, generate hypotheses to explain individual principal components. In this manner, we build on the observations by Hilsenbeck [<xref ref-type=\"bibr\" rid=\"B22\">22</xref>] Raychaudhuri et. al. [<xref ref-type=\"bibr\" rid=\"B23\">23</xref>] who used PCA to gain insight into the underlying factors present in the Chu et. al. yeast sporulation experiments [<xref ref-type=\"bibr\" rid=\"B24\">24</xref>]. Wall et. al. [<xref ref-type=\"bibr\" rid=\"B25\">25</xref>] introduced a novel use of singular value decomposition (SVD) for gene expression analysis that identifies non-exclusive gene groups, and Selaru et. al. [<xref ref-type=\"bibr\" rid=\"B26\">26</xref>] illustrated the potential of PCA to detect molecular phenotypic bases that correspond to relevant clinical or biological features of human tumors. Their approach identifies a subset of principal components that correlate well with known covariates. Here we introduce methods that extend beyond producing gene groups and observing a few principal components. Our methods provide a path for systematically analyzing each principal component by identifying the genes most influential in defining a particular principal component and the conditions in which those influential genes vary significantly. Finally, we describe methods which aim to explain each principal component's observed variance in terms of the condition variables deemed most likely to be driving the variance. We also introduce a software package that implements these methods. These methods and our software package provide automated and objective way of doing what a biologist naturally tries to do through inspection and pattern recognition.</p></sec><sec><title>Implementation</title><p>We have developed a Python package to implement the PCA interpretation capability described in detail in the Methodology section. This PCA analysis package has been added to CompClust developed previously [<xref ref-type=\"bibr\" rid=\"B9\">9</xref>,<xref ref-type=\"bibr\" rid=\"B27\">27</xref>]. The combined packages allow one to cluster, classify and visualize numeric datasets that have discrete or numeric annotations (referred to as labelings, or labelled datasets), and to compare labelings with confusion matrices and metrics such as normalized mutual information (NMI) [<xref ref-type=\"bibr\" rid=\"B28\">28</xref>]. This PCA analysis tool (including the complete results for the dataset analysis described in the Results section) has also been made accessible through the CompClustWeb web-based interface [<xref ref-type=\"bibr\" rid=\"B29\">29</xref>]. Our software makes use of data manipulation and graphical plotting using the matplotlib package [<xref ref-type=\"bibr\" rid=\"B30\">30</xref>], and the statistics are generated using the rPy package [<xref ref-type=\"bibr\" rid=\"B31\">31</xref>] and Gary Strangman's Python stats package [<xref ref-type=\"bibr\" rid=\"B32\">32</xref>].</p><p>The web-based front-end permits users to get a complete report on the interpretation of each principal component, including interactive PCA projection plots with the principal component's extreme genes (PCEGs) highlighted; ranked lists of the PCEGs with detailed annotations; interactive significance-ordered gene set trajectory plots that permit users to drill down to the individual gene level; similarly ordered condition reports ordered by expression difference and grouped by significance, including covariate info (with significantly correlated covariates highlighted); and finally a report of any suggestive covariates that are well correlated with the significant column grouping, including the confusion matrices and/or plots of statistics scores to back up the conclusions. All principal component analysis and results generation is implemented in a Python package so that analyses of large datasets can be executed in a batch mode rather than through the graphical interface. Further, the software that implements the CompClustWeb interface is provided within the CompClust package, so a software developer can create his or her own CompClustWeb server to review results of their PCA interpretation.</p></sec><sec><title>Results</title><sec><title>Application to microarray expression data, Case 1: GNF human data</title><p>We obtained gene expression data from the Genomics Institute of the Novartis Research Foundation (\"GNF\") Gene Expression Database via their SymAtlas web site [<xref ref-type=\"bibr\" rid=\"B33\">33</xref>,<xref ref-type=\"bibr\" rid=\"B34\">34</xref>]. The dataset is a challenge for most clustering algorithms because it contains 158 tissue samples hybridized to two Affymetrix microarray chips: U133A and GNF1H. The dataset combines the measurements of these chips to provide a total of 33,689 unique probe identities across the 158 tissue samples. Expression data are signal intensities estimated by Affymetrix Microarray Suite v5. For our analysis we used the log base 2 of the expression signal, and included data for all tissues and probes (noting that absent and present calls were not provided with the signal intensities). We applied our principal components analysis tool to generate interpretations for each of this dataset's 158 principal components.</p><p>As detailed in the Methodology section, for each principal component we identified a set of gene probes occupying the high and low extremes of that principal component's axis (we refer to these as principal component extreme genes, or PCEGs). One can adjust parameters to recover smaller and larger numbers of PCEGs per component by specifying either a likelihood threshold or an explicit number of PCEGs. The PCEGs are those probes having the most highly weighted values for that principal component, selected because they stand out from the others, they influence the principal component's direction, and thus they warrant further investigation. We selected probes with likelihoods less than <italic>extremeThresh </italic>= 0.00001, which yielded on average 20 low and 20 high extreme genes per principal component, though the sets sizes do vary considerably (μ = 18.9, σ = 17.2). Next we identified the tissues in which the high PCEGs showed significantly different expression than the low PCEGs. Visualizations produced include scatter plots of the extreme genes in PCA sub-spaces (PC <italic>N</italic>-1vs. PC <italic>N</italic>), and extreme gene trajectories in original tissue order as well as with tissues ordered by decreasing difference of mean of high PCEGs and mean of low PCEGs. The latter trajectory plot emphasizes how the extreme genes for a principal component show a pattern of expression that imposes a partitioning of tissues. It is left to human interpretation to examine the extreme genes and the tissue partitioning exposed by each principal component, and thus to build hypotheses that attach meaning to the sources of variation. The percentage variance explained by the top 50 principal components is provided in Table <xref ref-type=\"table\" rid=\"T1\">1</xref>. Example results for two illustrative principal components, PC7 and PC21, are shown in Figures <xref ref-type=\"fig\" rid=\"F1\">1</xref> through <xref ref-type=\"fig\" rid=\"F4\">4</xref> and in Tables <xref ref-type=\"table\" rid=\"T2\">2</xref> through <xref ref-type=\"table\" rid=\"T5\">5</xref>, and are discussed below. The complete analysis results generated for PC7 are provided as an example in the supplemental files: [see <xref ref-type=\"supplementary-material\" rid=\"S1\">Additional file 1</xref>], [see <xref ref-type=\"supplementary-material\" rid=\"S2\">Additional file 2</xref>], [see <xref ref-type=\"supplementary-material\" rid=\"S3\">Additional file 3</xref>], [see <xref ref-type=\"supplementary-material\" rid=\"S4\">Additional file 4</xref>], [see <xref ref-type=\"supplementary-material\" rid=\"S5\">Additional file 5</xref>], [see <xref ref-type=\"supplementary-material\" rid=\"S6\">Additional file 6</xref>]. Our supplemental materials web site [<xref ref-type=\"bibr\" rid=\"B35\">35</xref>] contains the complete collection of PCA interpretation results generated for these dataset for all principal components, as well as results of a comparable analysis done at <italic>extremeThresh </italic>= 0.001 which yielded larger PCEG sets.</p><p>We addressed the question of biological and statistical significance of PCs and the sets of extreme genes identified. Each set of high and low extreme genes from each principal component was tested for Gene Ontology (GO) statistical enrichment when compared to the human GO annotations from NCBI's loc2go dataset using the hypergeometric to calculate the p-value of each GO term. Terms that were enriched for a particular PC at 1% significance threshold and that were still significant following a Bonferroni correction for multiple hypothesis testing as described in [<xref ref-type=\"bibr\" rid=\"B36\">36</xref>] are reported as enriched (see Table <xref ref-type=\"table\" rid=\"T6\">6</xref>). 26 of the top 42 PCEG lists, derived using a stringent cut-off of 0.00001, produced significant GO enrichments; no PCEG sets beyond PC42 showed significant enrichment. As discussed below, many of the significant results showed obvious biological coherence and relationships to the specific samples associated with the PC of origin. This argues that PCs containing less than 1% of the total variation in this dataset are still relevant and point to coherent and important gene sets and their related samples.</p><p>Relationships between extreme gene sets and the corresponding sets of driving samples could be discerned for many PCs. The Methodology section presents a way to additionally correlate each principal component's sample partitioning with any available sample covariates. Although some human sample covariate information is provided in our test case, the GNF human expression dataset is not amenable to this additional layer of analysis because multiple subject's RNA samples were pooled prior to amplification and array hybridization. However, a second publicly available dataset with rich covariate information is presented below.</p></sec><sec><title>Application to microarray expression data, case 2: human diabetes</title><p>We acquired a Human diabetes expression dataset [<xref ref-type=\"bibr\" rid=\"B37\">37</xref>] from the Broad Institute Cancer Program dataset repository [<xref ref-type=\"bibr\" rid=\"B38\">38</xref>] along with the corresponding phenotype covariate data, and applied the filtering step as they described to produce a set containing 10,983 probes across 43 samples. Tissue samples were skeletal muscle biopsies from 3 diagnosis groups: normal glucose tolerance (NGT, <italic>n </italic>= 17); impaired glucose tolerance (IGT, <italic>n </italic>= 8); and Type 2 diabetes mellitus (DM2, <italic>n </italic>= 18). We used our PCA interpretation software to perform an unsupervised analysis of the DM2 vs. NGT subset (as that subset is comparable to the previous published result). As described in the Methodology section, PCEG sets were determined using an <italic>extremeThresh </italic>likelihood threshold of 0.001, which yielded about 50 high and 50 low extreme genes per principal component. For each principal component <italic>N</italic>, samples were partitioned in to <italic>UP</italic><sub><italic>N</italic></sub>, <italic>FLAT</italic><sub><italic>N </italic></sub>and <italic>DOWN</italic><sub><italic>N </italic></sub>sample sets on the basis of PC-<italic>N </italic>extreme high and extreme low expression differences. The supplemental materials contain PCA interpretation results for all 35 principal components, as well as results of a comparable analysis done at <italic>extremeThresh </italic>= 0.0001 which yielded smaller PCEG sets (see [<xref ref-type=\"bibr\" rid=\"B35\">35</xref>]).</p><p>This dataset contains more than 50 covariates, which provides the opportunity to interpret each principal component by searching for covariates that correlate well with expression patterns in the PCEG sets. As described in the Methodology section, we asked if any of the covariate annotations are well correlated with the partitioning of samples into <italic>UP</italic><sub><italic>N</italic></sub>, <italic>FLAT</italic><sub><italic>N </italic></sub>and <italic>DOWN</italic><sub><italic>N </italic></sub>sets. Covariate distributions were compared across different partitions (when sufficient data was available) and any significant trends identified were recorded (see Table <xref ref-type=\"table\" rid=\"T7\">7</xref>). For covariates identified as significantly correlated with a principal component's sample partitioning, covariate distribution plots were generated to further investigate and evaluate the apparent relationship. For example, Figure <xref ref-type=\"fig\" rid=\"F5\">5</xref> illustrates that PC14's <italic>UP<sub>14</sub></italic>, <italic>FLAT<sub>14 </sub></italic>and <italic>DOWN<sub>14 </sub></italic>sample partitions appear to be significantly related to two covariate measurements: Insulin_0 (<italic>sig </italic>= 0.0010); and Type2b_(%) (<italic>sig</italic>= 0.0077). The Pearson's correlation between the mean expression for the PC14EG-high set and Insulin_0 and Type2b_(%) covariates are <italic>r </italic>= 0.411 and <italic>r </italic>= 0.467 respectively.</p></sec></sec><sec><title>Discussion</title><p>This PCA-based data-mining tool highlights specific patterns of expression and associates them in a convenient way with the genes and samples responsible for those patterns. Some associations in the first few principal components (PCs) of the GNF set reflect major features in the data that are expected. This includes the global high and low constitutive expression profiles of PC1 (67% of variance in the GNF dataset). A component similar to this is often the first or second PC in Affymetrix array datasets. GNF PC3, in contrast, highlighted brain/neuronal tissues, which we expected in this dataset because there are many more samples from brain regions than from any other tissue, and there are thousands of genes that are expressed in a general brain pattern. The GO enrichment analysis associated with PC3lowEGs confirmed this impression by identifying neurogenesis, central nervous system, and synaptic terms as significantly enriched for PC3EG (Table <xref ref-type=\"table\" rid=\"T6\">6</xref>).</p><p>We asked if principal components that account individually for small fractions of variation in the data are likely to be significant. Conventional practice generally ignores principal components accounting for a few percent, or less, of total variation, on the assumption that such minor components are most likely dominated by noise. We believe this assumption, that all of the PC's accounting for small fractions of data should be ignored because they are artefact, to be wrong in the context of our analysis. We believe this, in part, because in our analysis we find many PCEGs for minor components are statistically enriched (Table <xref ref-type=\"table\" rid=\"T6\">6</xref>). Further, computational experiments using randomized data fail to produce any significant enrichments. This GO enrichment analysis, it should be noted, tends to underestimate the fraction of gene sets that are significant. This is a known artefact in instances where the input gene number (here those passing the selected rank sum threshold) is small, and its effects are exacerbated by the fact that GO annotations for human are still very much in a building phase. Many genes that will eventually be associated with GO terms are not yet entered. This means that reducing the threshold modestly, and therefore increasing the gene number, can uncover additional significant GO term enrichment in some of these PCs. For complete results of GNF analysis at <italic>p </italic>&lt; = 0.001 see our supplemental materials web site [<xref ref-type=\"bibr\" rid=\"B35\">35</xref>].</p><p>Viewed from a biological perspective, this PCA mining revealed several different classes of relationships. GNF PC21 is a good example of a component that highlights a coherent gene set and its corresponding tissues, many of which would also be grouped together by conventional clustering algorithms. This is true even though PC21 accounts for just 0.25% of total variation. The PC21EG-low set (defined at <italic>p </italic>&lt; = 0.00001) was enriched in a statistically significant way for five different GO terms (Table <xref ref-type=\"table\" rid=\"T6\">6</xref>), and these terms (myogenesis, muscle contraction, etc.) tell a simple and internally consistent story about muscle development and function. The top tissue samples driving PC21-low are skeletal muscle, tongue (also composed largely of striated muscle), heart and thyroid (which includes a population of myoid cells). Most top-ranked genes in this PCEG set are so specific for striated muscle that they would also appear together in conventional clusterings, although many clustering approaches become technically problematic with datasets of this size. However the PC21EG-low list differs from a conventional muscle cluster because it also includes some genes that are partly associated with muscle and partly associated with other tissues. PP1R1A, a regulatory subunit of protein phosphatase1, is such a gene. A role for it in striated muscle is suggested based on its coherent presence in tongue, skeletal and cardiac samples, even though it might well not have been seen in this light by standard clustering.</p><p>The second example is GNF PC7, which accounts for 0.81% of variation. It illustrates a different kind of biological relationship that more strongly distinguishes results of PCA mining from classical clustering. Top extreme genes associated with PC7 by inspection turn out to be a \"who's who\" of extracellular matrix components (a specific subset of fibronectins, collagens, laminins plus matrix associated proteins like <italic>MFAP5</italic>, <italic>MGP</italic>, <italic>LUM</italic>; regulatory molecules that mediate stability and function of those matrix components (thrombospondin, <italic>SPARC</italic>, <italic>ADAMTS1</italic>, <italic>Plod2</italic>); and matrix associated signalling and matrix associated signal modulators (insulin like growth factor binding proteins 7, 8 and 10; <italic>Sema3c</italic>). GO analysis confirms what inspection of the top PC7EGs suggested: namely that a set of extracellular matrix components are expressed in these driving tissues. It is instructive to look at the individual expression profiles for these genes directly at the GNF website and also in aggregate, as represented in the tissue (conditions) list in Table <xref ref-type=\"table\" rid=\"T2\">2</xref>. The most prominent contributing tissues associated with high expression of these genes are informative specifically because they <underline>do not</underline> constitute a group that would have been selected <italic>a priori </italic>as a coherent set based on known tissue function or shared developmental origin.</p><p>This is useful because a biologist interrogating the GNF database would not likely have constructed a query combining adipocytes, smooth muscle, bronchial epithelium, nor would one expect traditional clustering algorithms to place these genes so close to each other as to catalyze the same observation. Similarly, conventional ordering algorithms would not have placed them adjacent to each other because other parts of their expression profiles, containing different genes than those in PC7, would dominate their positions. And concerning genes, they have, in addition to commonalities of expression highlighted by PC7, differences from each other in additional diverse tissues not highlighted by PC7. The PCA grouping gives impetus and a necessary starting gene list to search for one or more factors or regulatory RNAs with a similar expression pattern, or to search for a shared and perhaps evolutionarily conserved cis-acting DNA sequence motifs. It is unlikely that these working hypotheses would have been arrived at easily by widely used methods of gene expression analysis.</p><p>The diabetes dataset offered us an opportunity to add more value to principal component interpretations by searching for covariates that appear correlated at some significance level. The relationships highlighted between covariates and principal components are suggestive, but not conclusive by themselves. Rather, they provide hypotheses that a researcher may wish to further investigate. While we have not delved deeply into this dataset, we believe that a number of the principal components are highlighting a number of meaningful sources of variation present. It is not clear whether this set should exhibit the same proportion of meaningful principal components as the GNF dataset, as by design the NGT vs. GM2 dataset does not contain substantial diversity of samples. Likewise, the selection of covariates was focused on a narrow set of measurements selected to be indicative of diabetes status, and so many covariates are redundant. We anticipate this tool will be maximally useful in cases where datasets are rich in both sample complexity and diversity of covariates.</p></sec><sec><title>Conclusion</title><p>Results presented above show that this PCA-mining approach can guide a user to biologically significant observations that both complement and reinforce those from conventional clustering analysis. The software package and web interface make this style of microarray analysis straightforward and accessible.</p><p>We have applied this to four additional microarray datasets (as yet unpublished) and to one multi-spectral imaging dataset. In each case we found the interpretations that the tool presented to be useful. In general, it seems that the top few principal components identify very broad characteristics of the data. Digging to the deeper components that comprise smaller but more particular substructure in the data leads to more subtle but often meaningful observations, many being complementary to standard clustering.</p><p>With respect to the top few components, PC1 is usually the approximate diagonal through the sample/condition space, explaining the overall variation in absolute expression level. For some other datasets we have noticed that the top few PCs can also highlight effects of preprocessing normalization steps or global data quality issues. This means they do not necessarily expose the most important biological variation. Thus, in one microarray dataset not shown here, PC2 was found to be extremely well correlated with a measure of quality of samples, as reflected by the percent of Affymetrix probes called present. Given this evidence of data quality effects comprising a major source of variation over the entire dataset, one might be motivated to remove the major offending conditions, and then repeat the PCA interpretation on the remaining conditions (columns). The idea is that an independent source of variation might be obscured by more dominant signals or noise present in the data from the offending condition.</p><p>Our experience thus far leads us to think that this PCA interpretation method will contribute to microarray expression analysis, as one part of a panel of methods that are sensitive to different features in a dataset, such as sample number, gene number, and distribution of variation across the samples. The PCA method should be especially useful for large, complex datasets that offer rich variation among many samples. What is certain is that there are almost always multiple sources of variation in a dataset and that in any specific study their nature and relative strength is informative, whether the origin is an easily-understood biologic one, a technical one, or a poorly-understood but nonetheless biologically pertinent one.</p><p>We are continuing to explore ways to improve our methodology and software package. We anticipate further advances will come with software infrastructure improvements to permit covariate analyses of both column (sample) covariates and row (gene) covariates. The CompClust dataset labeling capability [<xref ref-type=\"bibr\" rid=\"B27\">27</xref>] allows a user to attach diverse and numerous labelings to rows or columns. For example we can pull in additional row (gene probe) annotations such as Gene Ontology (GO) functional groups. Beyond explicitly comparing the NMI significance of specific row partitionings for discrete covariates, we plan to add routines to CompClust to automatically indicate when a group of genes are found to be enriched in specific GO categories (as was done in our analyses above), and more generally to handle large, discrete, multi-valued distributions of values. Our use of NMI treats discrete covariates as discrete random variables that can have at most a single value per condition, and so does not optimally address issue of multi-valued discrete random variables (e.g. GNF data has covariate \"concomitant medications\" with values like \"aspirin\", \"tylenol\", and \"aspirin &amp; tylenol\"). We are considering more elaborate extensions of mutual information or alternatives that might be able to take further advantage of such multi-valued entries.</p></sec><sec><title>Availability and requirements</title><p>The PCA interpretation software is implemented as one component of the CompClust Python package [<xref ref-type=\"bibr\" rid=\"B9\">9</xref>,<xref ref-type=\"bibr\" rid=\"B27\">27</xref>], which is freely available for non-commercial use. The software capability is also accessible through the CompClustWeb web-based interface [<xref ref-type=\"bibr\" rid=\"B29\">29</xref>]. The software that implements the web application is also included within the CompClust software distribution.</p><p>Project name: the PCA interpretation component of CompClust</p><p>Project home page: <ext-link ext-link-type=\"uri\" xlink:href=\"http://woldlab.caltech.edu/compclust\"/></p><p>Operating system(s): platform independent (Windows, Linux, Mac OS X)</p><p>Programming language: Python</p><p>Other requirements: Python 2.3 or higher (and some free Python packages)</p><p>License: MLX Public License 1.0 (non-commercial use allowed)</p><p>Use by non-academics: licence needed</p><p>We recommend that interested researchers use the web-based application, CompClustWeb, from any platform to review the PCA interpretation results for the GNF human gene expression and Broad Institute human diabetes expression data sets. We have written a CompClust PCA interpretation tutorial that demonstrates how to use CompClust's programming interface to generate PCA interpretations. Following the tutorial requires that CompClust and its Python prerequisites be installed. We created an easy-to-use CompClustShell installer for Windows that provides everything needed. For other operating systems (e.g. Linux &amp; OS X) we recommend that a software developer or system administrator help with the CompClust source code installation. We are working to simplify installation and plan to provide user-friendly installers for other operating systems in the near future.</p></sec><sec sec-type=\"methods\"><title>Methodology</title><p>We have developed the following algorithm for identifying and analyzing multiple independent sources of variance present within multi-dimensional sample datasets, in particular those that are produced by gene microarray expression experiments. The overall approach can be summarized as follows: 1) perform principal components analysis of the dataset; for each principal component: 2) identify the most extreme gene probes (those with the highest or lowest weighting) for that principal component; 3) identify and group any conditions in which those extreme probes vary significantly; 4) identify any condition covariates that correlate well with the condition grouping. By extending the interpretation of each principal component from extreme genes (rows) to ordered groups of significant conditions (columns) and further to identifying statistically significant correlations with column covariates, we attempt to make full use of the available data, in an objective and data-driven way, to analyze and provide meaningful interpretations of the diverse sources of variation present within the dataset.</p><sec><title>Determine the principal components of the dataset</title><p>Our dataset <italic>D </italic>consists of <italic>nc </italic>columns (e.g. tissue samples or conditions) and <italic>nr </italic>row vectors (e.g. gene probes), each row vector <italic>x<sub>i </sub></italic>∈ ℜ<sup><italic>nc </italic></sup>where <italic>i </italic>∈ [1, <italic>nr</italic>]. Such a dataset is usually represented as a two-dimensional <italic>nr </italic>× <italic>nc </italic>matrix (where <italic>nr </italic>&gt; <italic>nc</italic>). The dataset may optionally have <italic>nk </italic>supplemental covariate annotations <italic>C </italic>associated with each row or column. Each annotation <italic>C</italic><sub><italic>k </italic></sub>where <italic>k </italic>∈ [1, <italic>nk</italic>] can be either discrete (e.g. sex) or continuous (e.g. age), to permit the association of one discrete value per column (e.g. values male or female), or one continuous value per column (e.g. values 12, 16, or 42).</p><p>Our procedure starts by employing principal components analysis (PCA) to sequentially identify a series of new basis vectors or axes <italic>PC</italic><sub>1</sub>, <italic>PC</italic><sub>2</sub>, ...<italic>PC</italic><sub><italic>nc </italic></sub>in the high-dimensional column space ℜ<sup><italic>n </italic></sup>that are each aligned sequentially to capture the most as-yet unexplained variance. This is accomplished by applying the numeric procedure singular value decomposition (SVD) to the covariance matrix of <italic>D</italic>, <italic>cov(D)</italic>, to produce the decomposition <italic>cov</italic>(<italic>D</italic>) = <italic>USV<sup>T </sup></italic>that contains the eigenvectors of <italic>cov</italic>(<italic>D</italic>) in the columns of <italic>U </italic>and eigenvalues in the diagonal of <italic>S </italic>such that the eigenvalues are sorted by descending size. Each covariance eigenvector, or principal component <italic>PC</italic><sub>1</sub>, <italic>PC</italic><sub>2</sub>, ...<italic>PC</italic><sub><italic>nc</italic></sub>, explains a fraction of the total variance contained in the dataset, and each principal component <italic>PC</italic><sub><italic>n+1 </italic></sub>is orthogonal to the previous principal component <italic>PC</italic><sub><italic>n</italic></sub>. such that they define the basis of a new vector space <italic>P</italic>. These results are made available to the users in the form of <italic>nc </italic>plots, one for each of the principal component vectors, as well as a plot of the singular values contained in the diagonal of <italic>S </italic>to indicate the relative amount of variance each component explains.</p></sec><sec><title>Identify extreme gene probes for each principal component</title><p>Next, we project each data point <italic>x<sub>i </sub></italic>(corresponding to a gene probe, or row vector) into the new coordinate system by <italic>P </italic>= <italic>DU</italic>, effectively rotating the entire data point set <italic>D </italic>into the new principal component axes space, producing the rotated data set <italic>P</italic>. Each data point <italic>p<sub>i </sub></italic>in the rows of <italic>P </italic>corresponding to <italic>x<sub>i </sub></italic>has a coordinate for each principal component axis that describes where the data point <italic>p<sub>i </sub></italic>lies when projected along each axis <italic>PC</italic><sub>1</sub>, <italic>PC</italic><sub>1</sub>, ...<italic>PC</italic><sub><italic>nc</italic></sub>. For each principal component <italic>PC</italic><sub><italic>n </italic></sub>(<italic>n </italic>∈ [1, <italic>nc</italic>]) we select a set of data points from each end of that principal component axis- these are the extreme points for <italic>PC</italic><sub><italic>n</italic></sub>, called the principal component extreme genes, or PCEGs for convenience. The PCEGs can be identified and ranked in one of two ways: by identifying points having a low probability (<italic>p </italic>&lt; = <italic>extremeThresh</italic>) of belonging to a Gaussian fit to the distribution of points along the <italic>PC</italic><sub><italic>n </italic></sub>axis, or by taking a fixed number of <italic>nExtreme </italic>points at each tail of the distribution. <italic>H</italic><sub><italic>n </italic></sub>is the resulting set of data points having the highest coordinate values for <italic>PC</italic><sub><italic>n</italic></sub>, and <italic>L</italic><sub><italic>n </italic></sub>is the resulting set of data points having the lowest coordinate values for <italic>PC</italic><sub><italic>n</italic></sub>. These high and low extreme gene point sets are informative in and of themselves because they represent the most extreme of the data points along a principal axis of variation. As such, the high and low PCEG sets are some of the primary outputs generated by our procedure. We use the term \"extreme\" in a very general sense, in that the points stand out because they are far from the main distribution. We do not mean to imply that such points are either biologically relevant or nuisance data that should be removed; rather we are interested in these points in an unbiased way. By further analyzing their pattern of expression in the original axes we hope to gain a better understanding of their possible biologic significance.</p></sec><sec><title>Identify significant conditions for each principal component</title><p>The extreme gene points comprising <italic>H</italic><sub><italic>n </italic></sub>are located near one edge of the high-dimensional cloud of points, and points <italic>L</italic><sub><italic>n </italic></sub>are near the opposite edge. Thus, points <italic>H</italic><sub><italic>n </italic></sub>are likely to have coordinate values that are maximally different from points <italic>L</italic><sub><italic>n </italic></sub>in a subset of the original column space coordinate system. Our procedure next seeks to identify in which of the original columns (the original axes or dimensions) we find the greatest difference of values for points <italic>H</italic><sub><italic>n </italic></sub>versus points <italic>L</italic><sub><italic>n</italic></sub>. We do this by comparing distributions of values in <italic>H</italic><sub><italic>nj </italic></sub>versus <italic>L</italic><sub><italic>nj </italic></sub>for each of the original columns <italic>j </italic>(<italic>j </italic>∈ [1, <italic>nc</italic>]). A two-sided Wilcoxon rank sum test is used to estimate the likelihood that these two sets of values are drawn from the same distribution [<xref ref-type=\"bibr\" rid=\"B39\">39</xref>]; the resulting p-values for each column are used to rank order and group the columns, rather than as actual probabilities. Columns having a likelihood less than a user-defined significance level <italic>test1Thresh </italic>are identified and placed into one of two column sets: <italic>UP</italic><sub><italic>n </italic></sub>for those where column <italic>j </italic>has <italic>mean</italic>(<italic>H</italic><sub><italic>nj</italic></sub>) &gt; <italic>mean</italic>(<italic>L</italic><sub><italic>nj</italic></sub>), and <italic>DOWN</italic><sub><italic>n </italic></sub>for those columns <italic>j </italic>where <italic>mean</italic>(<italic>H</italic><sub><italic>nj</italic></sub>) &lt;<italic>mean</italic>(<italic>L</italic><sub><italic>nj</italic></sub>). Remaining columns that do not show significant variation are placed in the column set <italic>FLAT</italic><sub><italic>n</italic></sub>. The column sets are also meaningful outputs of our procedure, as <italic>UP</italic><sub><italic>n </italic></sub>and <italic>DOWN</italic><sub><italic>n </italic></sub>describe the groups of columns in which the extreme genes <italic>H</italic><sub><italic>n </italic></sub>and <italic>L</italic><sub><italic>n </italic></sub>vary significantly. Our procedure can output these columns and column sets in various orders simply as an aid to human interpretation, including: original column order; grouped by set and within set ordered by the Wilcoxon p-value significance; ordered by mean difference, <italic>mean</italic>(<italic>H</italic><sub><italic>nj</italic></sub>) – <italic>mean</italic>(<italic>L</italic><sub><italic>nj</italic></sub>); or ordered by the eigenvector column loading. Taken together, the PCEG point sets and significant column sets should provide valuable insight to researchers wishing to interpret each of the sources of variation identified by the principal components procedure.</p></sec><sec><title>Interpret each principal component using covariate annotations</title><p>When provided additional covariate annotations <italic>C</italic>, the procedure seeks to determine which, if any of the annotations <italic>C</italic><sub><italic>k </italic></sub>are well correlated with the partitioning of columns into the sets {<italic>UP</italic><sub><italic>n</italic></sub>, <italic>FLAT</italic><sub><italic>n</italic></sub>, <italic>DOWN</italic><sub><italic>n</italic></sub>}. A discrete annotation <italic>C</italic><sub><italic>k </italic></sub>containing <italic>m </italic>unique values <italic>V<sub>1</sub></italic>, <italic>V<sub>2</sub></italic>, ... <italic>V</italic><sub><italic>m </italic></sub>also defines a partitioning of the columns {<italic>KV<sub>1</sub></italic>, <italic>KV<sub>2</sub></italic>, ..., <italic>KV</italic><sub><italic>m</italic></sub>} where <italic>KV<sub>1 </sub></italic>is the set of columns that share the value <italic>V<sub>1</sub></italic>, <italic>KV<sub>2 </sub></italic>are those that share value <italic>V<sub>2</sub></italic>, and so on. An information theoretic measure known as normalized mutual information (NMI) [<xref ref-type=\"bibr\" rid=\"B28\">28</xref>] describes the degree to which two discrete random variables share information. When there is high mutual information, knowing the value of one of the variables should be useful predictor of the other variable. (See [<xref ref-type=\"bibr\" rid=\"B9\">9</xref>] for a description of the merits of NMI in terms of clustering and understanding the relationships between clusterings.) We construct the <italic>3 x m </italic>confusion matrix to compare the {<italic>UP</italic><sub><italic>n</italic></sub>, <italic>FLAT</italic><sub><italic>n</italic></sub>, <italic>DOWN</italic><sub><italic>n</italic></sub>} column partitioning with the {<italic>KV<sub>1</sub></italic>, <italic>KV<sub>2</sub></italic>, ..., <italic>KV</italic><sub><italic>m</italic></sub>} partitioning and calculate an NMI score between the partitionings. Because the usual NMI score is not symmetric (i.e. NMI(<italic>r</italic>,<italic>c</italic>) ≠ NMI(<italic>c</italic>,<italic>r</italic>)), we use a variant that we refer to as the average NMI score, which is simply the average of the NMI of the confusion matrix and the NMI of the transpose of the confusion matrix. Those covariates <italic>C</italic><sub><italic>k </italic></sub>having an average NMI greater than a user defined threshold <italic>nmiThresh </italic>are added to the set of significant covariate annotations <italic>A</italic><sub><italic>n</italic></sub>.</p><p>We apply a different approach when evaluating the <italic>C</italic><sub><italic>k </italic></sub>that are continuous covariates. We need to assess whether each principal component's {<italic>UP</italic><sub><italic>n</italic></sub>, <italic>FLAT</italic><sub><italic>n</italic></sub>, <italic>DOWN</italic><sub><italic>n</italic></sub>} column partitioning correlates with each <italic>C</italic><sub><italic>k </italic></sub>distribution of values. We can separately score three different partitioning schemes, <italic>UP</italic><sub><italic>n </italic></sub>vs. <italic>DOWN</italic><sub><italic>n</italic></sub>, {<italic>UP</italic><sub><italic>n</italic></sub>+<italic>FLAT</italic><sub><italic>n</italic></sub>} vs. <italic>DOWN</italic><sub><italic>n</italic></sub>, and <italic>UP</italic><sub><italic>n </italic></sub>vs.{<italic>FLAT</italic><sub><italic>n</italic></sub>+<italic>DOWN</italic><sub><italic>n</italic></sub>}, by determining if <italic>C</italic><sub><italic>k</italic></sub>'s value distributions differ significantly across the partition. E.g. does <italic>C</italic><sub><italic>k </italic></sub>within <italic>UP</italic><sub><italic>n </italic></sub>have a different distribution than <italic>C</italic><sub><italic>k </italic></sub>within <italic>DOWN</italic><sub><italic>n</italic></sub>? Next, does <italic>C</italic><sub><italic>k </italic></sub>within {<italic>UP</italic><sub><italic>n </italic></sub>+ <italic>FLAT</italic><sub><italic>n</italic></sub>} have a different distribution than <italic>C</italic><sub><italic>k </italic></sub>within <italic>DOWN</italic><sub><italic>n</italic></sub>? Finally, does <italic>C</italic><sub><italic>k </italic></sub>within <italic>UP</italic><sub><italic>n </italic></sub>have a different distribution than <italic>C</italic><sub><italic>k </italic></sub>within {<italic>FLAT</italic><sub><italic>n</italic></sub>+<italic>DOWN</italic><sub><italic>n</italic></sub>}? For each partition scheme we again use a two-tailed Wilcoxon rank sum test, including the small sample adjustments when sample size is less than 10, to determine whether the covariate's value distributions on each side of the partitioning differ significantly from each other. A <italic>minSetSize </italic>parameter can be specified as desired to reduce false positives when set sizes are very small, e.g. when comparing a distribution of 2 values vs. 7 values. Thus we calculate the Wilcoxon p-value for three partitionings of columns: <italic>UP</italic><sub><italic>n </italic></sub>vs. <italic>DOWN</italic><sub><italic>n</italic></sub>, {<italic>UP</italic><sub><italic>n</italic></sub>+<italic>FLAT</italic><sub><italic>n</italic></sub>} vs. <italic>DOWN</italic><sub><italic>n</italic></sub>, and <italic>UP</italic><sub><italic>n </italic></sub>vs. {<italic>FLAT</italic><sub><italic>n</italic></sub>+<italic>DOWN</italic><sub><italic>n</italic></sub>}. Those covariates having a p-value less than the user defined threshold <italic>test2Thresh </italic>for any of the three partitionings are added to the set of significant covariate annotations <italic>A</italic><sub><italic>n</italic></sub>.</p><p>Upon completion of the covariate analysis, covariates in the set <italic>A</italic><sub><italic>n</italic></sub>, that previously met user-controlled significance thresholds are reported by the software. Covariate reports provide the following information: For discrete-valued covariates, corresponding confusion matrices and average NMI scores are reported; for continuous-valued covariates, the three <italic>Wilcoxon </italic>p-values are reported together with supporting plots illustrating the covariate distributions. There is no guarantee that any covariates will be significantly related to a principal component. Conversely, spurious relationships might be reported, especially in the case of small numbers of samples due to small column partitions. The tool simply points to those covariates related to a principal component that also satisfy a user-controlled significance threshold. It is up to the investigator to consider these hypotheses and to confirm the interesting ones through further investigation.</p></sec><sec><title>Terminating condition</title><p>We have shown that some large-scale expression datasets have biologically pertinent structure that is revealed by deep PC analysis that goes well beyond the first few principal components. However there are limits to the depth of mining and these limits depend on both size and character of the dataset. In all cases, the last principal component is not free to seek a source of variation because it must be orthogonal to all prior <italic>nc-1 </italic>components. To some degree that is also true of some portion of latter principal components that explain ever-diminishing fractions of the variance. We suggest that a natural terminating condition exists: When a principal component cannot find any columns in which the extreme gene sets show significant differences, there is no need to proceed to subsequent principal components. We observe that this condition is often not met because the extreme genes are typically differentially expressed in at least a few of the original columns (the original axes or dimensions), even for the most minor principal components. We also observe that variants of a dataset (e.g. representative column subsets) can affect the relative ordering, but not the existence, of multiple factors or sources of variation that are reflected in the minor principal component regime. We may therefore choose to investigate all principal components, but do so with the expectation that minor principal components will describe increasingly subtle sources of variation, which can, and often do, include noisy processes inherent in the data source.</p></sec></sec><sec><title>Authors' contributions</title><p>JR and CH conceived the methodology of exhaustively analyzing and interpreting principal components, in particular how to identify extreme genes and significant conditions, and how to automate correlating these conditions with covariates to aid interpretation. JR, BK, DT and CH carried out the software development. JR carried out the initial PCA interpretation studies and drafted the manuscript. BK and DT performed additional PCA analyses and results interpretation. AM performed the gene set GO term enrichment analysis. BW conceived of the GNF dataset interpretation study, participated in its design and results interpretation, and helped to draft the manuscript. All authors read and approved the final manuscript.</p></sec><sec sec-type=\"supplementary-material\"><title>Supplementary Material</title><supplementary-material content-type=\"local-data\" id=\"S1\"><caption><title>Additional File 1</title><p>The supplemental files provided with this publication are only a representative set of those generated by the PCA interpretation software. The complete collections of PCA interpretation results for both the GNF and diabetes datasets are provided as a supplement to this publication at [<xref ref-type=\"bibr\" rid=\"B35\">35</xref>].</p><p>Tab-delimited text file listing the high and low extreme genes for PC<sub>7</sub>, including PC<sub>7 </sub>coefficient and additional GNF gene annotations: ProbeId, Name, Aliases, Description, Function and Protein Families.</p></caption><media mime-subtype=\"plain\" mimetype=\"text\" xlink:href=\"1471-2105-7-194-S1.txt\"><caption><p>Click here for file</p></caption></media></supplementary-material><supplementary-material content-type=\"local-data\" id=\"S2\"><caption><title>Additional File 2</title><p>The supplemental files provided with this publication are only a representative set of those generated by the PCA interpretation software. The complete collections of PCA interpretation results for both the GNF and diabetes datasets are provided as a supplement to this publication at [<xref ref-type=\"bibr\" rid=\"B35\">35</xref>].</p><p>Tab-delimited text file listing the conditions that are up, flat and down for PC<sub>7</sub>, ordered by decreasing difference of means.</p></caption><media mime-subtype=\"plain\" mimetype=\"text\" xlink:href=\"1471-2105-7-194-S2.txt\"><caption><p>Click here for file</p></caption></media></supplementary-material><supplementary-material content-type=\"local-data\" id=\"S3\"><caption><title>Additional File 3</title><p>The supplemental files provided with this publication are only a representative set of those generated by the PCA interpretation software. The complete collections of PCA interpretation results for both the GNF and diabetes datasets are provided as a supplement to this publication at [<xref ref-type=\"bibr\" rid=\"B35\">35</xref>].</p><p>Trajectory plot of the PC<sub>7 </sub>eigenvector, or \"eigen-condition\".</p></caption><media mime-subtype=\"plain\" mimetype=\"text\" xlink:href=\"1471-2105-7-194-S3.png\"><caption><p>Click here for file</p></caption></media></supplementary-material><supplementary-material content-type=\"local-data\" id=\"S4\"><caption><title>Additional File 4</title><p>The supplemental files provided with this publication are only a representative set of those generated by the PCA interpretation software. The complete collections of PCA interpretation results for both the GNF and diabetes datasets are provided as a supplement to this publication at [<xref ref-type=\"bibr\" rid=\"B35\">35</xref>].</p><p>Scatter plot of gene probe expression levels projected onto PC<sub>6 </sub>vs. PC<sub>7 </sub>space. The PC<sub>7 </sub>high and low extreme gene sets are highlighted in red and blue colors, respectively.</p></caption><media mime-subtype=\"plain\" mimetype=\"text\" xlink:href=\"1471-2105-7-194-S4.png\"><caption><p>Click here for file</p></caption></media></supplementary-material><supplementary-material content-type=\"local-data\" id=\"S5\"><caption><title>Additional File 5</title><p>The supplemental files provided with this publication are only a representative set of those generated by the PCA interpretation software. The complete collections of PCA interpretation results for both the GNF and diabetes datasets are provided as a supplement to this publication at [<xref ref-type=\"bibr\" rid=\"B35\">35</xref>].</p><p>Gene trajectory plots for PC<sub>7 </sub>high and low extreme gene sets with tissues in the order in which the original data were provided.</p></caption><media mime-subtype=\"plain\" mimetype=\"text\" xlink:href=\"1471-2105-7-194-S5.png\"><caption><p>Click here for file</p></caption></media></supplementary-material><supplementary-material content-type=\"local-data\" id=\"S6\"><caption><title>Additional File 6</title><p>The supplemental files provided with this publication are only a representative set of those generated by the PCA interpretation software. The complete collections of PCA interpretation results for both the GNF and diabetes datasets are provided as a supplement to this publication at [<xref ref-type=\"bibr\" rid=\"B35\">35</xref>].</p><p>Gene trajectory plots for PC<sub>7 </sub>high and low extreme gene sets with tissues ordered by decreasing mean differences, and thus grouped by significance (up group at left, flat group in middle and low group at right).</p></caption><media mime-subtype=\"plain\" mimetype=\"text\" xlink:href=\"1471-2105-7-194-S6.png\"><caption><p>Click here for file</p></caption></media></supplementary-material></sec></body><back><ack><sec><title>Acknowledgements</title><p>This work was supported in part by grants to BJW from the Department of Energy and the National Cancer Institute's Director's Challenge program. Additional support was provided by the NASA Office of Biological and Physical Research (OBPR) program. We also acknowledge Eric Mjolsness for discussions at the earliest phases of this research, and Ken McCue for additional discussions. We acknowledge that the GNF gene microarray expression data presented herein was obtained from Genomics Institute of the Novartis Research Foundation, and is © 2003–2005 GNF. We acknowledge that the diabetes expression data presented herein was obtained from the Broad Institute's Cancer Program dataset repository.</p></sec></ack><ref-list><ref id=\"B1\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Tavazoie</surname><given-names>S</given-names></name><name><surname>Hughes</surname><given-names>JD</given-names></name><name><surname>Campbell</surname><given-names>MJ</given-names></name><name><surname>Cho</surname><given-names>RJ</given-names></name><name><surname>Church</surname><given-names>GM</given-names></name></person-group><article-title>Systematic determination of genetic network architecture</article-title><source>Nat Genet</source><year>1999</year><volume>22</volume><fpage>281</fpage><lpage>285</lpage><pub-id pub-id-type=\"pmid\">10391217</pub-id><pub-id pub-id-type=\"doi\">10.1038/10343</pub-id></citation></ref><ref id=\"B2\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Tamayo</surname><given-names>P</given-names></name><name><surname>Slonim</surname><given-names>D</given-names></name><name><surname>Mesirov</surname><given-names>J</given-names></name><name><surname>Zhu</surname><given-names>Q</given-names></name><name><surname>Kitareewan</surname><given-names>S</given-names></name><name><surname>Dmitrovsky</surname><given-names>E</given-names></name><name><surname>Lander</surname><given-names>E</given-names></name><name><surname>Golub</surname><given-names>T</given-names></name></person-group><article-title>Interpreting patterns of gene expression with self-organizing maps: Methods and application to hematopoietic differentiation</article-title><source>Proc Natl Acad Sci USA</source><year>1999</year><volume>96</volume><fpage>2907</fpage><lpage>2912</lpage><pub-id pub-id-type=\"pmid\">10077610</pub-id><pub-id pub-id-type=\"doi\">10.1073/pnas.96.6.2907</pub-id></citation></ref><ref id=\"B3\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Eisen</surname><given-names>M</given-names></name><name><surname>Spellman</surname><given-names>P</given-names></name><name><surname>Brown</surname><given-names>P</given-names></name><name><surname>Botstein</surname><given-names>D</given-names></name></person-group><article-title>Cluster analysis and display of genome-wide expression patterns</article-title><source>Proc Natl Acad Sci USA</source><year>1998</year><volume>95</volume><fpage>14863</fpage><lpage>14868</lpage><pub-id pub-id-type=\"pmid\">9843981</pub-id><pub-id pub-id-type=\"doi\">10.1073/pnas.95.25.14863</pub-id></citation></ref><ref id=\"B4\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Wang</surname><given-names>R</given-names></name><name><surname>Scharenbroich</surname><given-names>L</given-names></name><name><surname>Hart</surname><given-names>C</given-names></name><name><surname>Wold</surname><given-names>B</given-names></name><name><surname>Mjolsness</surname><given-names>E</given-names></name></person-group><article-title>Clustering analysis of microarray gene expression data by splitting algorithm</article-title><source>J Parallel Distrib Comput</source><year>2003</year><volume>63</volume><fpage>692</fpage><lpage>706</lpage><pub-id pub-id-type=\"doi\">10.1016/S0743-7315(03)00085-6</pub-id></citation></ref><ref id=\"B5\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Yeung</surname><given-names>KY</given-names></name><name><surname>Fraley</surname><given-names>C</given-names></name><name><surname>Murua</surname><given-names>A</given-names></name><name><surname>Raftery</surname><given-names>AE</given-names></name><name><surname>Ruzzo</surname><given-names>WL</given-names></name></person-group><article-title>Model-based clustering and data transformations for gene expression data</article-title><source>Bioinformatics</source><year>2001</year><volume>17</volume><fpage>977</fpage><lpage>987</lpage><pub-id pub-id-type=\"pmid\">11673243</pub-id><pub-id pub-id-type=\"doi\">10.1093/bioinformatics/17.10.977</pub-id></citation></ref><ref id=\"B6\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Alon</surname><given-names>U</given-names></name><name><surname>Barkai</surname><given-names>N</given-names></name><name><surname>Notterman</surname><given-names>DA</given-names></name><name><surname>Gish</surname><given-names>K</given-names></name><name><surname>Ybarra</surname><given-names>S</given-names></name><name><surname>Mack</surname><given-names>D</given-names></name><name><surname>Levine</surname><given-names>AJ</given-names></name></person-group><article-title>Broad patterns of gene expression revealed by clustering analysis of tumor and normal colon tissues probed by oligonucleotide arrays</article-title><source>Proc Natl Acad Sci USA</source><year>1999</year><volume>96</volume><fpage>6745</fpage><lpage>6750</lpage><pub-id pub-id-type=\"pmid\">10359783</pub-id><pub-id pub-id-type=\"doi\">10.1073/pnas.96.12.6745</pub-id></citation></ref><ref id=\"B7\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Quackenbush</surname><given-names>J</given-names></name></person-group><article-title>Computational Analysis of Microarray Data</article-title><source>Nature Reviews Genetics</source><year>2001</year><volume>2</volume><fpage>418</fpage><lpage>427</lpage><pub-id pub-id-type=\"pmid\">11389458</pub-id><pub-id pub-id-type=\"doi\">10.1038/35076576</pub-id></citation></ref><ref id=\"B8\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Slonim</surname><given-names>DK</given-names></name></person-group><article-title>From patterns to pathways: gene expression data analysis comes of age</article-title><source>Nat Genet</source><year>2002</year><volume>32</volume><fpage>502</fpage><lpage>8</lpage><pub-id pub-id-type=\"pmid\">12454645</pub-id><pub-id pub-id-type=\"doi\">10.1038/ng1033</pub-id></citation></ref><ref id=\"B9\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Hart</surname><given-names>CE</given-names></name><name><surname>Sharenbroich</surname><given-names>L</given-names></name><name><surname>Bornstein</surname><given-names>BJ</given-names></name><name><surname>Trout</surname><given-names>D</given-names></name><name><surname>King</surname><given-names>B</given-names></name><name><surname>Mjolsness</surname><given-names>E</given-names></name><name><surname>Wold</surname><given-names>BJ</given-names></name></person-group><article-title>A Mathematical and computational framework for quantitative comparison and integration of large scale gene expression data</article-title><source>Nucleic Acids Research</source><volume>33</volume><fpage>2580</fpage><lpage>2594</lpage><comment>2005, May 10</comment><pub-id pub-id-type=\"pmid\">15886390</pub-id><pub-id pub-id-type=\"doi\">10.1093/nar/gki536</pub-id></citation></ref><ref id=\"B10\"><citation citation-type=\"book\"><person-group person-group-type=\"author\"><name><surname>Hart</surname><given-names>CE</given-names></name></person-group><source>Inferring Genetic Regulatory Network Structure: Integrative Analysis of Genome-Scale Data</source><year>2005</year><publisher-name>PhD Thesis, California Institute of Technology</publisher-name></citation></ref><ref id=\"B11\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Brown</surname><given-names>MPS</given-names></name><name><surname>Grundy</surname><given-names>WN</given-names></name><name><surname>Lin</surname><given-names>D</given-names></name><name><surname>Cristianini</surname><given-names>N</given-names></name><name><surname>Sugnet</surname><given-names>CW</given-names></name><name><surname>Furey</surname><given-names>TS</given-names></name><name><surname>Ares</surname><given-names>M</given-names></name><name><surname>Haussler</surname><given-names>D</given-names></name></person-group><article-title>Knowledge-based analysis of microarray gene expression data by using support vector machines</article-title><source>Proc Natl Acad Sci USA</source><volume>97</volume><fpage>262</fpage><lpage>267</lpage><comment>2000, January 4</comment><pub-id pub-id-type=\"pmid\">10618406</pub-id><pub-id pub-id-type=\"doi\">10.1073/pnas.97.1.262</pub-id></citation></ref><ref id=\"B12\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Mjolsness</surname><given-names>E</given-names></name><name><surname>DeCoste</surname><given-names>D</given-names></name></person-group><article-title>Machine learning for science: state of the art and future prospects</article-title><source>Science</source><volume>293</volume><fpage>2051</fpage><lpage>2055</lpage><comment>2001 Sep 14</comment><pub-id pub-id-type=\"pmid\">11557883</pub-id><pub-id pub-id-type=\"doi\">10.1126/science.293.5537.2051</pub-id></citation></ref><ref id=\"B13\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Ramaswamy</surname><given-names>S</given-names></name><name><surname>Tamayo</surname><given-names>P</given-names></name><name><surname>Rifkin</surname><given-names>R</given-names></name><name><surname>Mukherjee</surname><given-names>S</given-names></name><name><surname>Yeang</surname><given-names>CH</given-names></name><name><surname>Angelo</surname><given-names>M</given-names></name><name><surname>Ladd</surname><given-names>C</given-names></name><name><surname>Reich</surname><given-names>M</given-names></name><name><surname>Latulippe</surname><given-names>E</given-names></name><name><surname>Mesirov</surname><given-names>JP</given-names></name><name><surname>Poggio</surname><given-names>T</given-names></name><name><surname>Gerald</surname><given-names>W</given-names></name><name><surname>Loda</surname><given-names>M</given-names></name><name><surname>Lander</surname><given-names>ES</given-names></name><name><surname>Golub</surname><given-names>TR</given-names></name></person-group><article-title>Multiclass cancer diagnosis using tumor gene expression signatures</article-title><source>PNAS</source><fpage>15149</fpage><lpage>15154</lpage><comment>2001, Dec 18</comment><pub-id pub-id-type=\"pmid\">11742071</pub-id></citation></ref><ref id=\"B14\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Tothill</surname><given-names>RW</given-names></name><name><surname>Kowalczyk</surname><given-names>A</given-names></name><name><surname>Rischin</surname><given-names>D</given-names></name><name><surname>Bousioutas</surname><given-names>A</given-names></name><name><surname>Haviv</surname><given-names>I</given-names></name><name><surname>van Laar</surname><given-names>RK</given-names></name><name><surname>Waring</surname><given-names>PM</given-names></name><name><surname>Zalcberg</surname><given-names>J</given-names></name><name><surname>Ward</surname><given-names>R</given-names></name><name><surname>Biankin</surname><given-names>AV</given-names></name><name><surname>Sutherland</surname><given-names>RL</given-names></name><name><surname>Henshall</surname><given-names>SM</given-names></name><name><surname>Fong</surname><given-names>K</given-names></name><name><surname>Pollack</surname><given-names>JR</given-names></name><name><surname>Bowtell</surname><given-names>DDL</given-names></name><name><surname>Holloway</surname><given-names>AJ</given-names></name></person-group><article-title>An expression-based site of origin diagnostic method designed for clinical application to cancer of unknown origin</article-title><source>Cancer Research</source><volume>65</volume><fpage>4031</fpage><lpage>4040</lpage><comment>2005, May 15</comment><pub-id pub-id-type=\"pmid\">15899792</pub-id><pub-id pub-id-type=\"doi\">10.1158/0008-5472.CAN-04-3617</pub-id></citation></ref><ref id=\"B15\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Ihmels</surname><given-names>J</given-names></name><name><surname>Friedlander</surname><given-names>G</given-names></name><name><surname>Bergmann</surname><given-names>S</given-names></name><name><surname>Sarig</surname><given-names>O</given-names></name><name><surname>Ziv</surname><given-names>Y</given-names></name><name><surname>Barkai</surname><given-names>N</given-names></name></person-group><article-title>Revealing modular organization in the yeast transcriptional network</article-title><source>Nat Genet</source><year>2002</year><volume>31</volume><fpage>370</fpage><lpage>377</lpage><pub-id pub-id-type=\"pmid\">12134151</pub-id></citation></ref><ref id=\"B16\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Bergmann</surname><given-names>S</given-names></name><name><surname>Ihmels</surname><given-names>J</given-names></name><name><surname>Barkai</surname><given-names>N</given-names></name></person-group><article-title>Iterative signature algorithm for the analysis of large-scale gene expression data</article-title><source>Phys Rev E Stat Nonlin Soft Matter Phys</source><year>2003</year><volume>67</volume><fpage>031902</fpage><pub-id pub-id-type=\"pmid\">12689096</pub-id></citation></ref><ref id=\"B17\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Yeung</surname><given-names>KY</given-names></name><name><surname>Ruzzo</surname><given-names>WL</given-names></name></person-group><article-title>Principal component analysis for clustering gene expression data</article-title><source>Bioinformatics</source><year>2001</year><volume>17</volume><fpage>763</fpage><lpage>774</lpage><pub-id pub-id-type=\"pmid\">11590094</pub-id><pub-id pub-id-type=\"doi\">10.1093/bioinformatics/17.9.763</pub-id></citation></ref><ref id=\"B18\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Khan</surname><given-names>J</given-names></name><name><surname>Wei</surname><given-names>JS</given-names></name><name><surname>Ringner</surname><given-names>M</given-names></name><name><surname>Saal</surname><given-names>LH</given-names></name><name><surname>Ladanyi</surname><given-names>M</given-names></name><name><surname>Westermann</surname><given-names>F</given-names></name><name><surname>Berthold</surname><given-names>F</given-names></name><name><surname>Schwab</surname><given-names>M</given-names></name><name><surname>Antonescu</surname><given-names>CR</given-names></name><name><surname>Peterson</surname><given-names>C</given-names></name><name><surname>Meltzer</surname><given-names>PS</given-names></name></person-group><article-title>Classification and diagnostic prediction of concers using gene expression profiling and artificial reural networks</article-title><source>Nat Med</source><year>2001</year><fpage>673</fpage><lpage>679</lpage><pub-id pub-id-type=\"pmid\">11385503</pub-id><pub-id pub-id-type=\"doi\">10.1038/89044</pub-id></citation></ref><ref id=\"B19\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Nguyen</surname><given-names>D</given-names></name><name><surname>Rocke</surname><given-names>D</given-names></name></person-group><article-title>Tumor classification by partial least squares using microarray gene expression data</article-title><source>Bioinformatics</source><year>2002</year><volume>18</volume><fpage>39</fpage><lpage>50</lpage><pub-id pub-id-type=\"pmid\">11836210</pub-id><pub-id pub-id-type=\"doi\">10.1093/bioinformatics/18.1.39</pub-id></citation></ref><ref id=\"B20\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Wen</surname><given-names>X</given-names></name><name><surname>Fuhrman</surname><given-names>S</given-names></name><name><surname>Michaels</surname><given-names>GS</given-names></name><name><surname>Carr</surname><given-names>DB</given-names></name><name><surname>Smith</surname><given-names>S</given-names></name><name><surname>Barker</surname><given-names>JL</given-names></name><name><surname>Somogyi</surname><given-names>R</given-names></name></person-group><article-title>Large-scale temporal gene expression mapping of central nervous system development</article-title><source>Proc Natl Acad Sci U S A</source><volume>95</volume><fpage>334</fpage><lpage>339</lpage><comment>1998, January 6</comment><pub-id pub-id-type=\"pmid\">9419376</pub-id><pub-id pub-id-type=\"doi\">10.1073/pnas.95.1.334</pub-id></citation></ref><ref id=\"B21\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Sturn</surname><given-names>A</given-names></name><name><surname>Quackenbush</surname><given-names>J</given-names></name><name><surname>Trajanoski</surname><given-names>Z</given-names></name></person-group><article-title>Genesis: cluster analysis of microarray data</article-title><source>Bioinformatics application note</source><year>2002</year><volume>18</volume><fpage>207</fpage><lpage>208</lpage></citation></ref><ref id=\"B22\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Hilsenbeck</surname><given-names>SG</given-names></name><name><surname>Friedrichs</surname><given-names>WE</given-names></name><name><surname>Schiff</surname><given-names>R</given-names></name><name><surname>O'Connell</surname><given-names>P</given-names></name><name><surname>Hansen</surname><given-names>RK</given-names></name><name><surname>Osborne</surname><given-names>CK</given-names></name><name><surname>Fuqua</surname><given-names>SAW</given-names></name></person-group><article-title>Statistical Analysis of Array Expression Data as Applied to the Problem of Tamoxifen Resistance</article-title><source>J Natl Cancer Institute</source><year>1999</year><volume>91</volume><fpage>453</fpage><lpage>459</lpage><pub-id pub-id-type=\"doi\">10.1093/jnci/91.5.453</pub-id></citation></ref><ref id=\"B23\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Raychaudhuri</surname><given-names>S</given-names></name><name><surname>Stuart</surname><given-names>JM</given-names></name><name><surname>Altman</surname><given-names>RB</given-names></name></person-group><article-title>Principal Components Analysis to Summarize Microarray Experiments: Application to Sporulation Time Series</article-title><source>Pac Symp Biocomput</source><year>2000</year><fpage>455</fpage><lpage>466</lpage><pub-id pub-id-type=\"pmid\">10902193</pub-id></citation></ref><ref id=\"B24\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Chu</surname><given-names>S</given-names></name><name><surname>DeRisi</surname><given-names>J</given-names></name><name><surname>Eisen</surname><given-names>M</given-names></name><name><surname>Mulholland</surname><given-names>J</given-names></name><name><surname>Botstein</surname><given-names>D</given-names></name><name><surname>Brown</surname><given-names>PO</given-names></name><name><surname>Herskowitz</surname><given-names>I</given-names></name></person-group><article-title>The transcriptional program of sporulation in budding yeast</article-title><source>Science</source><year>1998</year><volume>282</volume><fpage>699</fpage><lpage>705</lpage><pub-id pub-id-type=\"pmid\">9784122</pub-id><pub-id pub-id-type=\"doi\">10.1126/science.282.5389.699</pub-id></citation></ref><ref id=\"B25\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Wall</surname><given-names>ME</given-names></name><name><surname>Dyck</surname><given-names>PA</given-names></name><name><surname>Brettin</surname><given-names>TS</given-names></name></person-group><article-title>SVDMAN – Singular value decomposition analysis of microarray data</article-title><source>Bioinformatics</source><year>2001</year><volume>17</volume><fpage>566</fpage><lpage>568</lpage><pub-id pub-id-type=\"pmid\">11395437</pub-id><pub-id pub-id-type=\"doi\">10.1093/bioinformatics/17.6.566</pub-id></citation></ref><ref id=\"B26\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Selaru</surname><given-names>FM</given-names></name><name><surname>Yin</surname><given-names>J</given-names></name><name><surname>Olaru</surname><given-names>A</given-names></name><name><surname>Mori</surname><given-names>Y</given-names></name><name><surname>Xu</surname><given-names>Y</given-names></name><name><surname>Epstein</surname><given-names>SH</given-names></name><name><surname>Sato</surname><given-names>F</given-names></name><name><surname>Deacu</surname><given-names>E</given-names></name><name><surname>Wang</surname><given-names>S</given-names></name><name><surname>Sterian</surname><given-names>A</given-names></name><name><surname>Fulton</surname><given-names>A</given-names></name><name><surname>Abraham</surname><given-names>JM</given-names></name><name><surname>Shibata</surname><given-names>D</given-names></name><name><surname>Baquet</surname><given-names>C</given-names></name><name><surname>Stass</surname><given-names>SA</given-names></name><name><surname>Meltzer</surname><given-names>SJ</given-names></name></person-group><article-title>An Unsupervised Approach to Identify Molecular Phenotypic Components Influencing Breast Cancer Features</article-title><source>Cancer Research</source><fpage>1584</fpage><lpage>1588</lpage><comment>2004, March 1</comment><pub-id pub-id-type=\"pmid\">14996713</pub-id></citation></ref><ref id=\"B27\"><citation citation-type=\"other\"><article-title>The CompClust software package</article-title><ext-link ext-link-type=\"uri\" xlink:href=\"http://woldlab.caltech.edu/compclust\"/></citation></ref><ref id=\"B28\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Forbes</surname><given-names>AD</given-names></name></person-group><article-title>Classification-algorithm evaluation: five performance measures based on confusion matrices</article-title><source>J Clin Monit</source><year>1995</year><volume>11</volume><fpage>189</fpage><lpage>206</lpage><pub-id pub-id-type=\"pmid\">7623060</pub-id><pub-id pub-id-type=\"doi\">10.1007/BF01617722</pub-id></citation></ref><ref id=\"B29\"><citation citation-type=\"other\"><article-title>The CompClustWeb software demonstration</article-title><ext-link ext-link-type=\"uri\" xlink:href=\"http://woldlab.caltech.edu/publications/pca-bmc-2005/demo\"/></citation></ref><ref id=\"B30\"><citation citation-type=\"other\"><article-title>Matplotlib/pylab – matlab style python plotting (plots, graphs, charts)</article-title><ext-link ext-link-type=\"uri\" xlink:href=\"http://matplotlib.sourceforge.net\"/></citation></ref><ref id=\"B31\"><citation citation-type=\"other\"><article-title>RPy home page</article-title><ext-link ext-link-type=\"uri\" xlink:href=\"http://rpy.sourceforge.net\"/></citation></ref><ref id=\"B32\"><citation citation-type=\"other\"><article-title>Gary Strangman's Python Modules</article-title><ext-link ext-link-type=\"uri\" xlink:href=\"http://www.nmr.mgh.harvard.edu/Neural_Systems_Group/gary/python.html\"/></citation></ref><ref id=\"B33\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><collab>HG_U133A/GNF1H and GNF1M Tissue Atlas Datasets</collab><name><surname>Su</surname><given-names>AI</given-names></name><name><surname>Wiltshire</surname><given-names>T</given-names></name><name><surname>Batalov</surname><given-names>S</given-names></name><name><surname>Lapp</surname><given-names>H</given-names></name><name><surname>Ching</surname><given-names>KA</given-names></name><name><surname>Block</surname><given-names>D</given-names></name><name><surname>Zhang</surname><given-names>J</given-names></name><name><surname>Soden</surname><given-names>R</given-names></name><name><surname>Hayakawa</surname><given-names>M</given-names></name><name><surname>Kreiman</surname><given-names>G</given-names></name><name><surname>Cooke</surname><given-names>MP</given-names></name><name><surname>Walker</surname><given-names>JR</given-names></name><name><surname>Hogenesch</surname><given-names>JB</given-names></name></person-group><article-title>A gene atlas of the mouse and human protein-encoding transcriptomes</article-title><source>Proc Natl Acad Sci USA</source><volume>101</volume><fpage>6062</fpage><lpage>6067</lpage><comment>2004, Apr 20</comment><pub-id pub-id-type=\"pmid\">15075390</pub-id></citation></ref><ref id=\"B34\"><citation citation-type=\"other\"><article-title>The GNF SymAtlas web application</article-title><ext-link ext-link-type=\"uri\" xlink:href=\"http://symatlas.gnf.org/SymAtlas\"/></citation></ref><ref id=\"B35\"><citation citation-type=\"other\"><article-title>Supplemental materials web site</article-title><ext-link ext-link-type=\"uri\" xlink:href=\"http://woldlab.caltech.edu/publications/pca-bmc-2005\"/></citation></ref><ref id=\"B36\"><citation citation-type=\"other\"><article-title>Mortazavi and Wold, in preparation</article-title></citation></ref><ref id=\"B37\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Mootha</surname><given-names>VK</given-names></name><name><surname>Lindgren</surname><given-names>CM</given-names></name><name><surname>Eriksson</surname><given-names>KF</given-names></name><name><surname>Subramanian</surname><given-names>A</given-names></name><name><surname>Sihag</surname><given-names>S</given-names></name><name><surname>Lehar</surname><given-names>J</given-names></name><name><surname>Puigserver</surname><given-names>P</given-names></name><name><surname>Carlsson</surname><given-names>E</given-names></name><name><surname>Ridderstrale</surname><given-names>M</given-names></name><name><surname>Laurila</surname><given-names>E</given-names></name><name><surname>Houstis</surname><given-names>N</given-names></name><name><surname>Daly</surname><given-names>MJ</given-names></name><name><surname>Patterson</surname><given-names>N</given-names></name><name><surname>Mesirov</surname><given-names>JP</given-names></name><name><surname>Golub</surname><given-names>TR</given-names></name><name><surname>Tamayo</surname><given-names>P</given-names></name><name><surname>Spiegelman</surname><given-names>B</given-names></name><name><surname>Lander</surname><given-names>ES</given-names></name><name><surname>Hirschhorn</surname><given-names>JN</given-names></name><name><surname>Altshuler</surname><given-names>D</given-names></name><name><surname>Groop</surname><given-names>LC</given-names></name></person-group><article-title>PGC-1alpha-responsive genes involved in oxidative phosphorylation are coordinately downregulated in human diabetes</article-title><source>Nat Genet</source><year>2003</year><volume>34</volume><fpage>267</fpage><lpage>273</lpage><pub-id pub-id-type=\"pmid\">12808457</pub-id><pub-id pub-id-type=\"doi\">10.1038/ng1180</pub-id></citation></ref><ref id=\"B38\"><citation citation-type=\"other\"><article-title>Broad Institute Cancer Program dataset repository</article-title><ext-link ext-link-type=\"uri\" xlink:href=\"http://www.broad.mit.edu/cgi-bin/cancer/datasets.cgi\"/></citation></ref><ref id=\"B39\"><citation citation-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Troyanskaya</surname><given-names>OG</given-names></name><name><surname>Garber</surname><given-names>ME</given-names></name><name><surname>Brown</surname><given-names>PO</given-names></name><name><surname>Botstein</surname><given-names>D</given-names></name><name><surname>Altman</surname><given-names>RB</given-names></name></person-group><article-title>Nonparametric methods for identifying differentially expressed genes in microarray data</article-title><source>Bioinformatics</source><year>2002</year><volume>18</volume><fpage>1454</fpage><lpage>1461</lpage><pub-id pub-id-type=\"pmid\">12424116</pub-id><pub-id pub-id-type=\"doi\">10.1093/bioinformatics/18.11.1454</pub-id></citation></ref></ref-list><sec sec-type=\"display-objects\"><title>Figures and Tables</title><fig id=\"F1\" position=\"float\"><label>Figure 1</label><caption><p><bold>GNF PC7 High and Low Extreme Gene Probes</bold>. Scatter plot of N = 33689 probe expression levels projected onto PC<sub>6 </sub>vs. PC<sub>7 </sub>space with high and low extreme gene sets <italic>H<sub>7 </sub></italic>(red points, <italic>n </italic>= 88) and <italic>L<sub>7 </sub></italic>(blue points, <italic>n </italic>= 20); extreme genes selected at <italic>extremeThresh </italic>= 0.00001. The extreme genes in <italic>H<sub>7 </sub></italic>are listed in Table 2.</p></caption><graphic xlink:href=\"1471-2105-7-194-1\"/></fig><fig id=\"F2\" position=\"float\"><label>Figure 2</label><caption><p><bold>GNF PC7 Extreme Gene Trajectory Plot with Tissues Ordered by Significance</bold>. Trajectory plots for high and low extreme gene sets <italic>H<sub>7 </sub></italic>(red, <italic>n </italic>= 88) and <italic>L<sub>7 </sub></italic>(blue, <italic>n </italic>= 20) with tissues ordered by decreasing <italic>mean</italic>(<italic>H<sub>7</sub></italic>) – <italic>mean</italic>(<italic>L<sub>7</sub></italic>), and thus grouped by significance (<italic>UP<sub>7 </sub></italic>group at left, <italic>FLAT<sub>7 </sub></italic>group in middle and <italic>LOW<sub>7 </sub></italic>group at right) at <italic>test1Thresh </italic>= 0.05. Table 3 lists the tissues within <italic>UP<sub>7 </sub></italic>and <italic>DOWN<sub>7 </sub></italic>that occur at each end of this plot.</p></caption><graphic xlink:href=\"1471-2105-7-194-2\"/></fig><fig id=\"F3\" position=\"float\"><label>Figure 3</label><caption><p><bold>GNF PC21 High and Low Extreme Gene Probes</bold>. Scatter plot of N = 33689 probe expression levels projected onto PC<sub>20 </sub>vs. PC<sub>21 </sub>space with high and low extreme gene sets <italic>H<sub>21 </sub></italic>(red points, <italic>n </italic>= 37) and <italic>L<sub>21 </sub></italic>(blue points, <italic>n </italic>= 49); extreme genes selected at <italic>extremeThresh </italic>= 0.00001. The extreme genes in <italic>L<sub>21 </sub></italic>are listed in Table 4.</p></caption><graphic xlink:href=\"1471-2105-7-194-3\"/></fig><fig id=\"F4\" position=\"float\"><label>Figure 4</label><caption><p><bold>GNF PC21 Extreme Gene Trajectory Plot with Tissues Ordered by Significance</bold>. Trajectory plots for high and low extreme gene sets <italic>H<sub>21 </sub></italic>(red, <italic>n </italic>= 37) and <italic>L<sub>21 </sub></italic>(blue, <italic>n </italic>= 49) with tissues ordered by decreasing <italic>mean</italic>(<italic>H<sub>21</sub></italic>) – <italic>mean</italic>(<italic>L<sub>21</sub></italic>), and thus grouped by significance (<italic>UP<sub>21 </sub></italic>group at left, <italic>FLAT<sub>21 </sub></italic>group in middle and <italic>LOW<sub>21 </sub></italic>group at right) at <italic>test1Thresh </italic>= 0.05. Table 5 lists the tissues within <italic>UP<sub>21 </sub></italic>and <italic>DOWN<sub>21 </sub></italic>that occur at each end of this plot.</p></caption><graphic xlink:href=\"1471-2105-7-194-4\"/></fig><fig id=\"F5\" position=\"float\"><label>Figure 5</label><caption><p><bold>Diabetes PC14 Sample Partitioning is Correlated with Certain Covariates</bold>. When sufficient covariate data is available (a number of the measurements are missing for certain covariates), covariate distributions are compared across partitions and significant differences are reported (as in Table 7). When a covariate is identified as significantly correlated with a principal component's sample partitioning, covariate distribution plots can be generated to further investigate and evaluate the apparent relationship. For example, the diabetes dataset's PC14 extreme genes partition the samples into <italic>UP<sub>14 </sub></italic>(<italic>n </italic>= 8), <italic>FLAT<sub>14 </sub></italic>(<italic>n </italic>= 17) and <italic>DOWN<sub>14 </sub></italic>(<italic>n </italic>= 10) based on their expression patterns. PC14's <italic>UP<sub>14 </sub></italic>vs. {<italic>FLAT<sub>14</sub></italic>+<italic>DOWN<sub>14</sub></italic>} sample partitioning appears related to the Insulin_0 measure (<italic>sig </italic>= 0.001), and the {<italic>UP<sub>14</sub></italic>+<italic>FLAT<sub>14</sub></italic>} vs. <italic>DOWN<sub>14 </sub></italic>partitioning appears related to Type2b_(%) (<italic>sig </italic>= 0.008). The mean expression of the PC14EG-high genes appears modestly correlated with Insulin_0 (<italic>r </italic>= 0.411) and with Type2b_(%) (<italic>r </italic>= 0.467). <italic>UP<sub>14 </sub></italic>samples are in red, <italic>FLAT<sub>14 </sub></italic>are black, and <italic>DOWN<sub>14 </sub></italic>are blue.</p></caption><graphic xlink:href=\"1471-2105-7-194-5\"/></fig><table-wrap id=\"T1\" position=\"float\"><label>Table 1</label><caption><p>Variance Explained By Principal Components. Table 1 lists the percentage of variance in the GNF human tissue microarray data explained by each principal component. The first 10 components explain 80.36% of the total variance. Principal components 11 through 158 each explain less than 0.5% of the total variance, but combined explain almost 20%.</p></caption><table frame=\"hsides\" rules=\"groups\"><thead><tr><td align=\"center\"><bold>Principal Component</bold></td><td align=\"center\"><bold>Percentage of Variance</bold></td></tr></thead><tbody><tr><td align=\"center\">1</td><td align=\"center\">67.62</td></tr><tr><td align=\"center\">2</td><td align=\"center\">4.66</td></tr><tr><td align=\"center\">3</td><td align=\"center\">2.04</td></tr><tr><td align=\"center\">4</td><td align=\"center\">1.41</td></tr><tr><td align=\"center\">5</td><td align=\"center\">1.07</td></tr><tr><td align=\"center\">6</td><td align=\"center\">0.90</td></tr><tr><td align=\"center\">7</td><td align=\"center\">0.81</td></tr><tr><td align=\"center\">8</td><td align=\"center\">0.69</td></tr><tr><td align=\"center\">9</td><td align=\"center\">0.63</td></tr><tr><td align=\"center\">10</td><td align=\"center\">0.54</td></tr><tr><td align=\"center\">11</td><td align=\"center\">0.46</td></tr><tr><td align=\"center\">12</td><td align=\"center\">0.42</td></tr><tr><td align=\"center\">13</td><td align=\"center\">0.37</td></tr><tr><td align=\"center\">14</td><td align=\"center\">0.35</td></tr><tr><td align=\"center\">15</td><td align=\"center\">0.32</td></tr><tr><td align=\"center\">16</td><td align=\"center\">0.31</td></tr><tr><td align=\"center\">17</td><td align=\"center\">0.30</td></tr><tr><td align=\"center\">18</td><td align=\"center\">0.29</td></tr><tr><td align=\"center\">19</td><td align=\"center\">0.26</td></tr><tr><td align=\"center\">20</td><td align=\"center\">0.26</td></tr><tr><td align=\"center\">21</td><td align=\"center\">0.25</td></tr><tr><td align=\"center\">22</td><td align=\"center\">0.24</td></tr><tr><td align=\"center\">23</td><td align=\"center\">0.23</td></tr><tr><td align=\"center\">24</td><td align=\"center\">0.22</td></tr><tr><td align=\"center\">25</td><td align=\"center\">0.22</td></tr><tr><td align=\"center\">26</td><td align=\"center\">0.22</td></tr><tr><td align=\"center\">27</td><td align=\"center\">0.21</td></tr><tr><td align=\"center\">28</td><td align=\"center\">0.20</td></tr><tr><td align=\"center\">29</td><td align=\"center\">0.20</td></tr><tr><td align=\"center\">30</td><td align=\"center\">0.19</td></tr><tr><td align=\"center\">31</td><td align=\"center\">0.19</td></tr><tr><td align=\"center\">32</td><td align=\"center\">0.19</td></tr><tr><td align=\"center\">33</td><td align=\"center\">0.18</td></tr><tr><td align=\"center\">34</td><td align=\"center\">0.18</td></tr><tr><td align=\"center\">35</td><td align=\"center\">0.18</td></tr><tr><td align=\"center\">36</td><td align=\"center\">0.18</td></tr><tr><td align=\"center\">37</td><td align=\"center\">0.17</td></tr><tr><td align=\"center\">38</td><td align=\"center\">0.16</td></tr><tr><td align=\"center\">39</td><td align=\"center\">0.16</td></tr><tr><td align=\"center\">40</td><td align=\"center\">0.16</td></tr><tr><td align=\"center\">41</td><td align=\"center\">0.16</td></tr><tr><td align=\"center\">42</td><td align=\"center\">0.16</td></tr><tr><td align=\"center\">43</td><td align=\"center\">0.16</td></tr><tr><td align=\"center\">44</td><td align=\"center\">0.15</td></tr><tr><td align=\"center\">45</td><td align=\"center\">0.15</td></tr><tr><td align=\"center\">46</td><td align=\"center\">0.15</td></tr><tr><td align=\"center\">47</td><td align=\"center\">0.15</td></tr><tr><td align=\"center\">48</td><td align=\"center\">0.15</td></tr><tr><td align=\"center\">49</td><td align=\"center\">0.14</td></tr><tr><td align=\"center\">50</td><td align=\"center\">0.14</td></tr></tbody></table></table-wrap><table-wrap id=\"T2\" position=\"float\"><label>Table 2</label><caption><p>PC7 High Extreme Genes. Detailed info for (<italic>n </italic>= 88) extreme genes in PC<sub>7 </sub>set <italic>H<sub>7</sub></italic>, ordered most extreme first, including common name, description, and associated Gene Ontology terms (provided with the GNF dataset). These extreme genes are shown as red points in Figure 1.</p></caption><table frame=\"hsides\" rules=\"groups\"><thead><tr><td align=\"left\"><bold>PC-7 Value</bold></td><td align=\"left\"><bold>Name</bold></td><td align=\"left\"><bold>Description</bold></td><td align=\"left\"><bold>Function</bold></td></tr></thead><tbody><tr><td align=\"left\">18.63</td><td align=\"left\"><italic>COL1A2</italic></td><td align=\"left\">collagen, type I, alpha 2</td><td align=\"left\">(GO:0006817) phosphate transport; (GO:0001501) skeletal development; (GO:0008147) structural constituent of bone;</td></tr><tr><td align=\"left\">17.89</td><td align=\"left\"><italic>LUM</italic></td><td align=\"left\">lumican</td><td align=\"left\">(GO:0005518) collagen binding; (GO:0030199) collagen fibril organization; (GO:0005207) extracellular matrix glycoprotein; (GO:0005201) extracellular matrix structural constituent; (GO:0005203) proteoglycan; (GO:0007601) visual perception;</td></tr><tr><td align=\"left\">15.48</td><td align=\"left\"><italic>TPM2</italic></td><td align=\"left\">tropomyosin 2 (beta)</td><td align=\"left\">(GO:0003779) actin binding; (GO:0007517) muscle development; (GO:0008307) structural constituent of muscle;</td></tr><tr><td align=\"left\">15.30</td><td align=\"left\"><italic>IGFBP7</italic></td><td align=\"left\">insulin-like growth factor binding protein 7</td><td align=\"left\">(GO:0005520) insulin-like growth factor binding; (GO:0008285) negative regulation of cell proliferation; (GO:0001558) regulation of cell growth;</td></tr><tr><td align=\"left\">14.54</td><td align=\"left\"><italic>CAV1</italic></td><td align=\"left\">caveolin 1, caveolae protein, 22 kDa</td><td align=\"left\">(GO:0005198) structural molecule activity; (GO:0008181) tumor suppressor;</td></tr><tr><td align=\"left\">14.08</td><td align=\"left\"><italic>COL3A1</italic></td><td align=\"left\">collagen, type III, alpha 1 (Ehlers-Danlos syndrome type IV, autosomal dominant)</td><td align=\"left\">(GO:0008015) circulation; (GO:0005202) collagen; (GO:0005201) extracellular matrix structural constituent; (GO:0007397) histogenesis and organogenesis; (GO:0009887) organogenesis; (GO:0006817) phosphate transport;</td></tr><tr><td align=\"left\">13.84</td><td align=\"left\"><italic>KCTD12</italic></td><td align=\"left\">potassium channel tetramerisation domain containing 12</td><td align=\"left\">(GO:0006813) potassium ion transport; (GO:0005515) protein binding; (GO:0005249) voltage-gated potassium channel activity;</td></tr><tr><td align=\"left\">13.83</td><td align=\"left\"><italic>COL1A1</italic></td><td align=\"left\">collagen, type I, alpha 1</td><td align=\"left\">(GO:0005202) collagen; (GO:0005201) extracellular matrix structural constituent; (GO:0007605) perception of sound; (GO:0006817) phosphate transport; (GO:0001501) skeletal development; (GO:0008147) structural constituent of bone;</td></tr><tr><td align=\"left\">13.71</td><td align=\"left\"><italic>MGP</italic></td><td align=\"left\">matrix Gla protein</td><td align=\"left\">(GO:0005509) calcium ion binding; (GO:0005201) extracellular matrix structural constituent; (GO:0007048) oncogenesis; (GO:0007605) perception of sound; (GO:0008147) structural constituent of bone;</td></tr><tr><td align=\"left\">13.57</td><td align=\"left\"><italic>COL3A1</italic></td><td align=\"left\">collagen, type III, alpha 1 (Ehlers-Danlos syndrome type IV, autosomal dominant)</td><td align=\"left\">(GO:0008015) circulation; (GO:0005202) collagen; (GO:0005201) extracellular matrix structural constituent; (GO:0007397) histogenesis and organogenesis; (GO:0009887) organogenesis; (GO:0006817) phosphate transport;</td></tr><tr><td align=\"left\">13.25</td><td align=\"left\"><italic>CALD1</italic></td><td align=\"left\">caldesmon 1</td><td align=\"left\">(GO:0003779) actin binding; (GO:0005516) calmodulin binding; (GO:0006936) muscle contraction; (GO:0007517) muscle development; (GO:0017022) myosin binding; (GO:0005523) tropomyosin binding</td></tr><tr><td align=\"left\">13.19</td><td align=\"left\"><italic>MYL9</italic></td><td align=\"left\">myosin, light polypeptide 9, regulatory</td><td align=\"left\">(GO:0005509) calcium ion binding; (GO:0008307) structural constituent of muscle</td></tr><tr><td align=\"left\">13.12</td><td align=\"left\"><italic>FN1</italic></td><td align=\"left\">fibronectin 1</td><td align=\"left\">(GO:0006953) acute-phase response; (GO:0007155) cell adhesion; (GO:0016477) cell migration; (GO:0005518) collagen binding; (GO:0005201) extracellular matrix structural constituent; (GO:0008201) heparin binding; (GO:0009611) response to wounding</td></tr><tr><td align=\"left\">12.71</td><td align=\"left\"><italic>PLK2</italic></td><td align=\"left\">polo-like kinase 2 (Drosophila)</td><td align=\"left\">(GO:0005524) ATP binding; (GO:0043123) positive regulation of I-kappaB kinase/NF-kappaB cascade; (GO:0006468) protein amino acid phosphorylation; (GO:0004674) protein serine/threonine kinase activity; (GO:0004871) signal transducer activity; (GO:0016740) transferase activity</td></tr><tr><td align=\"left\">12.59</td><td align=\"left\"><italic>CALD1</italic></td><td align=\"left\">caldesmon 1</td><td align=\"left\">(GO:0003779) actin binding; (GO:0005516) calmodulin binding; (GO:0006936) muscle contraction; (GO:0007517) muscle development; (GO:0017022) myosin binding; (GO:0005523) tropomyosin binding</td></tr><tr><td align=\"left\">12.54</td><td align=\"left\"><italic>CTGF</italic></td><td align=\"left\">connective tissue growth factor</td><td align=\"left\">(GO:0006259) DNA metabolism; (GO:0007155) cell adhesion; (GO:0005194) cell adhesion molecule activity; (GO:0008151) cell growth and/or maintenance; (GO:0006928) cell motility; (GO:0008201) heparin binding; (GO:0005520) insulin-like growth factor binding; (GO:0005515) protein binding; (GO:0001558) regulation of cell growth; (GO:0009611) response to wounding</td></tr><tr><td align=\"left\">12.39</td><td align=\"left\"><italic>D2S448</italic></td><td align=\"left\">Melanoma associated gene</td><td align=\"left\">(GO:0006955) immune response; (GO:0005152) interleukin-1 receptor antagonist activity; (GO:0004601) peroxidase activity</td></tr><tr><td align=\"left\">12.28</td><td align=\"left\"><italic>MFAP5</italic></td><td align=\"left\">microfibrillar associated protein 5</td><td align=\"left\">(GO:0005201) extracellular matrix structural constituent</td></tr><tr><td align=\"left\">12.24</td><td align=\"left\"><italic>SEMA3C</italic></td><td align=\"left\">sema domain, immunoglobulin domain (Ig), short basic domain, secreted, (semaphorin) 3C</td><td align=\"left\">(GO:0008151) cell growth and/or maintenance; (GO:0007275) development; (GO:0009315) drug resistance; (GO:0006955) immune response; (GO:0042493) response to drug</td></tr><tr><td align=\"left\">12.20</td><td align=\"left\"><italic>ADAMTS1</italic></td><td align=\"left\">a disintegrin-like and metalloprotease (reprolysin type) with thrombospondin type 1 motif, 1</td><td align=\"left\">(GO:0008201) heparin binding; (GO:0016787) hydrolase activity; (GO:0005178) integrin binding; (GO:0007229) integrin-mediated signaling pathway; (GO:0004222) metalloendopeptidase activity; (GO:0008237) metallopeptidase activity; (GO:0008285) negative regulation of cell proliferation; (GO:0006508) proteolysis and peptidolysis; (GO:0008270) zinc ion binding</td></tr><tr><td align=\"left\">11.99</td><td align=\"left\"><italic>SPARC</italic></td><td align=\"left\">secreted protein, acidic, cysteine-rich (osteonectin)</td><td align=\"left\">(GO:0005509) calcium ion binding; (GO:0005518) collagen binding</td></tr><tr><td align=\"left\">11.76</td><td align=\"left\"><italic>F3</italic></td><td align=\"left\">coagulation factor III (thromboplastin, tissue factor)</td><td align=\"left\">(GO:0007596) blood coagulation; (GO:0003801) blood coagulation factor activity; (GO:0004896) hematopoietin/interferon-class (D200-domain) cytokine receptor activity; (GO:0006955) immune response; (GO:0004872) receptor activity; (GO:0004888) transmembrane receptor activity</td></tr><tr><td align=\"left\">11.71</td><td align=\"left\"><italic>PLS3</italic></td><td align=\"left\">plastin 3 (T isoform)</td><td align=\"left\">(GO:0003779) actin binding; (GO:0005509) calcium ion binding</td></tr><tr><td align=\"left\">11.70</td><td align=\"left\"><italic>THBS1</italic></td><td align=\"left\">thrombospondin 1</td><td align=\"left\">(GO:0007596) blood coagulation; (GO:0005509) calcium ion binding; (GO:0007155) cell adhesion; (GO:0005194) cell adhesion molecule activity; (GO:0006928) cell motility; (GO:0007275) development; (GO:0004866) endopeptidase inhibitor activity; (GO:0008201) heparin binding; (GO:0007399) neurogenesis; (GO:0005515) protein binding; (GO:0004871) signal transducer activity; (GO:0005198) structural molecule activity</td></tr><tr><td align=\"left\">11.69</td><td align=\"left\"><italic>PLOD2</italic></td><td align=\"left\">procollagen-lysine, 2-oxoglutarate 5-dioxygenase (lysine hydroxylase) 2</td><td align=\"left\">(GO:0016491) oxidoreductase activity; (GO:0016702) oxidoreductase activity\\, acting on single donors with incorporation of molecular oxygen\\, incorporation of two atoms of oxygen; (GO:0008475) procollagen-lysine 5-dioxygenase activity; (GO:0019538) protein metabolism; (GO:0006464) protein modification</td></tr><tr><td align=\"left\">11.37</td><td align=\"left\"><italic>CAV1</italic></td><td align=\"left\">caveolin 1, caveolae protein, 22 kDa</td><td align=\"left\">(GO:0005198) structural molecule activity; (GO:0008181) tumor suppressor</td></tr><tr><td align=\"left\">11.36</td><td align=\"left\"><italic>gnf1h04130_x_at</italic></td><td align=\"left\">None</td><td align=\"left\">None</td></tr><tr><td align=\"left\">11.28</td><td align=\"left\"><italic>FN1</italic></td><td align=\"left\">fibronectin 1</td><td align=\"left\">(GO:0006953) acute-phase response; (GO:0007155) cell adhesion; (GO:0016477) cell migration; (GO:0005518) collagen binding; (GO:0005201) extracellular matrix structural constituent; (GO:0008201) heparin binding; (GO:0009611) response to wounding</td></tr><tr><td align=\"left\">11.18</td><td align=\"left\"><italic>SMARCA1</italic></td><td align=\"left\">SWI/SNF related, matrix associated, actin dependent regulator of chromatin, subfamily a, member 1</td><td align=\"left\">(GO:0005524) ATP binding; (GO:0008026) ATP-dependent helicase activity; (GO:0003677) DNA binding; (GO:0004002) adenosinetriphosphatase; (GO:0006338) chromatin remodeling; (GO:0004386) helicase activity; (GO:0045182) translation regulator activity</td></tr><tr><td align=\"left\">11.08</td><td align=\"left\"><italic>TAGLN</italic></td><td align=\"left\">transgelin</td><td align=\"left\">(GO:0007517) Muscle development</td></tr><tr><td align=\"left\">10.95</td><td align=\"left\"><italic>COL4A1</italic></td><td align=\"left\">collagen, type IV, alpha 1</td><td align=\"left\">(GO:0003677) DNA binding; (GO:0005202) collagen; (GO:0005201) extracellular matrix structural constituent; (GO:0006817) phosphate transport</td></tr><tr><td align=\"left\">10.90</td><td align=\"left\"><italic>CAV2</italic></td><td align=\"left\">caveolin 2</td><td align=\"left\">(GO:0008181) tumor suppressor</td></tr><tr><td align=\"left\">10.90</td><td align=\"left\"><italic>TPM1</italic></td><td align=\"left\">tropomyosin 1 (alpha)</td><td align=\"left\">(GO:0003779) actin binding; (GO:0007517) muscle development; (GO:0008016) regulation of heart rate; (GO:0005200) structural constituent of cytoskeleton; (GO:0008307) structural constituent of muscle</td></tr><tr><td align=\"left\">10.85</td><td align=\"left\"><italic>GJA1</italic></td><td align=\"left\">gap junction protein, alpha 1, 43 kDa (connexin 43)</td><td align=\"left\">(GO:0007267) cell-cell signaling; (GO:0015285) connexon channel activity; (GO:0007507) heart development; (GO:0015075) ion transporter activity; (GO:0006936) muscle contraction; (GO:0007605) perception of sound; (GO:0043123) positive regulation of I-kappaB kinase/NF-kappaB cascade; (GO:0004871) signal transducer activity; (GO:0006832) small molecule transport; (GO:0006810) transport</td></tr><tr><td align=\"left\">10.82</td><td align=\"left\"><italic>COL1A2</italic></td><td align=\"left\">collagen, type I, alpha 2</td><td align=\"left\">(GO:0005202) collagen; (GO:0005201) extracellular matrix structural constituent; (GO:0006817) phosphate transport; (GO:0001501) skeletal development; (GO:0008147) structural constituent of bone</td></tr><tr><td align=\"left\">10.61</td><td align=\"left\"><italic>FXR1</italic></td><td align=\"left\">fragile X mental retardation, autosomal homolog 1</td><td align=\"left\">(GO:0003677) DNA binding; (GO:0003723) RNA binding; (GO:0006915) apoptosis; (GO:0003676) nucleic acid binding; (GO:0006913) nucleocytoplasmic transport</td></tr><tr><td align=\"left\">10.60</td><td align=\"left\"><italic>COL6A3</italic></td><td align=\"left\">collagen, type VI, alpha 3</td><td align=\"left\">cell adhesion (GO:0007155);cell adhesion molecule activity (GO:0005194);collagen (GO:0005202);extracellular matrix structural constituent (GO:0005201);muscle development (GO:0007517);phosphate transport (GO:0006817);protein binding (GO:0005515);serine-type endopeptidase inhibitor activity (GO:0004867)</td></tr><tr><td align=\"left\">10.52</td><td align=\"left\"><italic>CYR61</italic></td><td align=\"left\">cysteine-rich, angiogenic inducer, 61</td><td align=\"left\">cell adhesion (GO:0007155);cell proliferation (GO:0008283);chemotaxis (GO:0006935);embryogenesis and morphogenesis (GO:0007345);heparin binding (GO:0008201);insulin-like growth factor binding (GO:0005520);morphogenesis (GO:0009653);regulation of cell growth (GO:0001558)</td></tr><tr><td align=\"left\">10.50</td><td align=\"left\"><italic>10-Sep</italic></td><td align=\"left\">septin 10</td><td align=\"left\">GTP binding (GO:0005525)</td></tr><tr><td align=\"left\">10.44</td><td align=\"left\"><italic>IL1R1</italic></td><td align=\"left\">interleukin 1 receptor, type I</td><td align=\"left\">cell surface receptor linked signal transduction (GO:0007166);immune response (GO:0006955);inflammatory response (GO:0006954);interleukin-1 receptor activity (GO:0004908);interleukin-1\\, Type I\\, activating receptor activity (GO:0004909);signal transducer activity (GO:0004871);transmembrane receptor activity (GO:0004888)</td></tr><tr><td align=\"left\">10.37</td><td align=\"left\"><italic>WBP5</italic></td><td align=\"left\">WW domain binding protein 1</td><td align=\"left\">DNA binding (GO:0003677)</td></tr><tr><td align=\"left\">10.35</td><td align=\"left\"><italic>LAMB1</italic></td><td align=\"left\">laminin, beta 1</td><td align=\"left\">cell adhesion (GO:0007155);cell adhesion molecule activity (GO:0005194);protein binding (GO:0005515);structural molecule activity (GO:0005198)</td></tr><tr><td align=\"left\">10.33</td><td align=\"left\"><italic>EMP1</italic></td><td align=\"left\">epithelial membrane protein 1</td><td align=\"left\">cell death (GO:0008219);cell growth (GO:0016049);cell proliferation (GO:0008283);development (GO:0007275);oncogenesis (GO:0007048)</td></tr><tr><td align=\"left\">10.23</td><td align=\"left\"><italic>NRP1</italic></td><td align=\"left\">neuropilin 1</td><td align=\"left\">axon guidance (GO:0007411);cell adhesion (GO:0007155);cell-cell signaling (GO:0007267);histogenesis and organogenesis (GO:0007397);neurogenesis (GO:0007399);organogenesis (GO:0009887);positive regulation of cell proliferation (GO:0008284);receptor activity (GO:0004872);signal transduction (GO:0007165);vascular endothelial growth factor receptor activity (GO:0005021)</td></tr><tr><td align=\"left\">10.12</td><td align=\"left\"><italic>COL3A1</italic></td><td align=\"left\">collagen, type III, alpha 1 (Ehlers-Danlos syndrome type IV, autosomal dominant)</td><td align=\"left\">circulation (GO:0008015);collagen (GO:0005202);extracellular matrix structural constituent (GO:0005201);histogenesis and organogenesis (GO:0007397);organogenesis (GO:0009887);phosphate transport (GO:0006817)</td></tr><tr><td align=\"left\">10.10</td><td align=\"left\"><italic>CALD1</italic></td><td align=\"left\">caldesmon 1</td><td align=\"left\">actin binding (GO:0003779);calmodulin binding (GO:0005516);muscle contraction (GO:0006936);muscle development (GO:0007517);myosin binding (GO:0017022);tropomyosin binding (GO:0005523)</td></tr><tr><td align=\"left\">10.10</td><td align=\"left\"><italic>THBS1</italic></td><td align=\"left\">thrombospondin 1</td><td align=\"left\">blood coagulation (GO:0007596);calcium ion binding (GO:0005509);cell adhesion (GO:0007155);cell adhesion molecule activity (GO:0005194);cell motility (GO:0006928);development (GO:0007275);endopeptidase inhibitor activity (GO:0004866);heparin binding (GO:0008201);neurogenesis (GO:0007399);protein binding (GO:0005515);signal transducer activity (GO:0004871);structural molecule activity (GO:0005198)</td></tr><tr><td align=\"left\">10.05</td><td align=\"left\"><italic>FN1</italic></td><td align=\"left\">fibronectin 1</td><td align=\"left\">acute-phase response (GO:0006953);cell adhesion (GO:0007155);cell migration (GO:0016477);collagen binding (GO:0005518);extracellular matrix structural constituent (GO:0005201);heparin binding (GO:0008201);response to wounding (GO:0009611)</td></tr><tr><td align=\"left\">10.03</td><td align=\"left\"><italic>SPARC</italic></td><td align=\"left\">secreted protein, acidic, cysteine-rich (osteonectin)</td><td align=\"left\">calcium ion binding (GO:0005509);collagen binding (GO:0005518)</td></tr><tr><td align=\"left\">10.02</td><td align=\"left\"><italic>IGFBP7</italic></td><td align=\"left\">insulin-like growth factor binding protein 7</td><td align=\"left\">insulin-like growth factor binding (GO:0005520);negative regulation of cell proliferation (GO:0008285);regulation of cell growth (GO:0001558)</td></tr><tr><td align=\"left\">9.99</td><td align=\"left\"><italic>IL6ST</italic></td><td align=\"left\">interleukin 6 signal transducer (gp130, oncostatin M receptor)</td><td align=\"left\">cell surface receptor linked signal transduction (GO:0007166);gp130 (GO:0004898);immune response (GO:0006955);interleukin-6 receptor activity (GO:0004915);oncostatin-M receptor activity (GO:0004924);receptor activity (GO:0004872);signal transduction (GO:0007165)</td></tr><tr><td align=\"left\">9.98</td><td align=\"left\"><italic>LAMB1</italic></td><td align=\"left\">laminin, beta 1</td><td align=\"left\">cell adhesion (GO:0007155);cell adhesion molecule activity (GO:0005194);protein binding (GO:0005515);structural molecule activity (GO:0005198)</td></tr><tr><td align=\"left\">9.94</td><td align=\"left\"><italic>TAZ</italic></td><td align=\"left\">transcriptional co-activator with PDZ-binding motif (TAZ)</td><td align=\"left\">transcription coactivator activity (GO:0003713)</td></tr><tr><td align=\"left\">9.93</td><td align=\"left\"><italic>DCN</italic></td><td align=\"left\">decorin</td><td align=\"left\">chondroitin sulfate/dermatan sulfate proteoglycan (GO:0005205);histogenesis and organogenesis (GO:0007397);organogenesis (GO:0009887)</td></tr><tr><td align=\"left\">9.93</td><td align=\"left\"><italic>ANXA1</italic></td><td align=\"left\">annexin A1</td><td align=\"left\">calcium ion binding (GO:0005509);calcium-dependent phospholipid binding (GO:0005544);cell motility (GO:0006928);cell surface receptor linked signal transduction (GO:0007166);inflammatory response (GO:0006954);lipid metabolism (GO:0006629);phospholipase A2 inhibitor activity (GO:0019834);phospholipase inhibitor activity (GO:0004859);phospholipid binding (GO:0005543);receptor binding (GO:0005102)</td></tr><tr><td align=\"left\">9.89</td><td align=\"left\"><italic>MYH11</italic></td><td align=\"left\">myosin, heavy polypeptide 11, smooth muscle</td><td align=\"left\">ATP binding (GO:0005524);actin binding (GO:0003779);calmodulin binding (GO:0005516);cell growth and/or maintenance (GO:0008151);motor activity (GO:0003774);muscle development (GO:0007517);protein amino acid alkylation (GO:0008213);striated muscle contraction (GO:0006941)</td></tr><tr><td align=\"left\">9.89</td><td align=\"left\"><italic>EFEMP1</italic></td><td align=\"left\">EGF-containing fibulin-like extracellular matrix protein 1</td><td align=\"left\">calcium ion binding (GO:0005509);visual perception (GO:0007601)</td></tr><tr><td align=\"left\">9.88</td><td align=\"left\"><italic>SPUVE</italic></td><td align=\"left\">protease, serine, 23</td><td align=\"left\">chymotrypsin activity (GO:0004263);hydrolase activity (GO:0016787);proteolysis and peptidolysis (GO:0006508);trypsin activity (GO:0004295)</td></tr><tr><td align=\"left\">9.87</td><td align=\"left\"><italic>Hs.514018</italic></td><td align=\"left\">CDNA: FLJ22209 fis, clone HRC01496</td><td align=\"left\">None</td></tr><tr><td align=\"left\">9.80</td><td align=\"left\"><italic>FN1</italic></td><td align=\"left\">fibronectin 1</td><td align=\"left\">acute-phase response (GO:0006953);cell adhesion (GO:0007155);cell migration (GO:0016477);collagen binding (GO:0005518);extracellular matrix structural constituent (GO:0005201);heparin binding (GO:0008201);response to wounding (GO:0009611)</td></tr><tr><td align=\"left\">9.77</td><td align=\"left\"><italic>COL16A1</italic></td><td align=\"left\">collagen, type XVI, alpha 1</td><td align=\"left\">cell adhesion (GO:0007155);collagen (GO:0005202);extracellular matrix structural constituent (GO:0005201);phosphate transport (GO:0006817);pregnancy (GO:0007565)</td></tr><tr><td align=\"left\">9.73</td><td align=\"left\"><italic>SNX7</italic></td><td align=\"left\">sorting nexin 7</td><td align=\"left\">intracellular protein transport (GO:0006886);intracellular signaling cascade (GO:0007242);protein transporter activity (GO:0008565)</td></tr><tr><td align=\"left\">9.69</td><td align=\"left\"><italic>AHR</italic></td><td align=\"left\">aryl hydrocarbon receptor</td><td align=\"left\">DNA binding (GO:0003677);apoptosis (GO:0006915);ligand-dependent nuclear receptor activity (GO:0004879);response to stress (GO:0006950);response to xenobiotic stimulus (GO:0009410);signal transduction (GO:0007165);transcription factor activity (GO:0003700);transcription from Pol II promoter (GO:0006366)</td></tr><tr><td align=\"left\">9.57</td><td align=\"left\"><italic>COL6A1</italic></td><td align=\"left\">collagen, type VI, alpha 1</td><td align=\"left\">DNA binding (GO:0003677);cell adhesion (GO:0007155);cell adhesion molecule activity (GO:0005194);collagen (GO:0005202);extracellular matrix structural constituent (GO:0005201);histogenesis and organogenesis (GO:0007397);molecular_function unknown (GO:0005554);phosphate transport (GO:0006817);protein binding (GO:0005515)</td></tr><tr><td align=\"left\">9.55</td><td align=\"left\"><italic>KIAA0992</italic></td><td align=\"left\">palladin</td><td align=\"left\">amino acid metabolism (GO:0006520)</td></tr><tr><td align=\"left\">9.54</td><td align=\"left\"><italic>COL5A2</italic></td><td align=\"left\">collagen, type V, alpha 2</td><td align=\"left\">cell growth and/or maintenance (GO:0008151);collagen (GO:0005202);extracellular matrix glycoprotein (GO:0005207);extracellular matrix structural constituent (GO:0005201);phosphate transport (GO:0006817)</td></tr><tr><td align=\"left\">9.53</td><td align=\"left\"><italic>FBN1</italic></td><td align=\"left\">fibrillin 1 (Marfan syndrome)</td><td align=\"left\">calcium ion binding (GO:0005509);development (GO:0007275);extracellular matrix structural constituent (GO:0005201);skeletal development (GO:0001501);visual perception (GO:0007601)</td></tr><tr><td align=\"left\">9.51</td><td align=\"left\"><italic>PAM</italic></td><td align=\"left\">peptidylglycine alpha-amidating monooxygenase</td><td align=\"left\">electron transporter activity (GO:0005489);monooxygenase activity (GO:0004497);peptide amidation (GO:0001519);peptidylglycine monooxygenase activity (GO:0004504);protein modification (GO:0006464)</td></tr><tr><td align=\"left\">9.50</td><td align=\"left\"><italic>LOC92912</italic></td><td align=\"left\">hypothetical protein LOC92912</td><td align=\"left\">ligase activity (GO:0016874);ubiquitin conjugating enzyme activity (GO:0004840);ubiquitin cycle (GO:0006512);ubiquitin-protein ligase activity (GO:0004842)</td></tr><tr><td align=\"left\">9.47</td><td align=\"left\"><italic>COL6A2</italic></td><td align=\"left\">collagen, type VI, alpha 2</td><td align=\"left\">cell-cell adhesion (GO:0016337);collagen (GO:0005202);extracellular matrix organization and biogenesis (GO:0030198);extracellular matrix structural constituent (GO:0005201);muscle development (GO:0007517);phosphate transport (GO:0006817);protein binding, bridging (GO:0030674)</td></tr><tr><td align=\"left\">9.45</td><td align=\"left\"><italic>PTX1</italic></td><td align=\"left\">PTX1 protein</td><td align=\"left\">None</td></tr><tr><td align=\"left\">9.41</td><td align=\"left\"><italic>KIAA0992</italic></td><td align=\"left\">palladin</td><td align=\"left\">amino acid metabolism (GO:0006520)</td></tr><tr><td align=\"left\">9.33</td><td align=\"left\"><italic>CSPG2</italic></td><td align=\"left\">chondroitin sulfate proteoglycan 2 (versican)</td><td align=\"left\">calcium ion binding (GO:0005509);cell adhesion (GO:0007155);chondroitin sulfate proteoglycan (GO:0005204);development (GO:0007275);glycosaminoglycan binding (GO:0005539);heterophilic cell adhesion (GO:0007157);hyaluronic acid binding (GO:0005540);lectin (GO:0005530);proteoglycan (GO:0005203);sugar binding (GO:0005529)</td></tr><tr><td align=\"left\">9.32</td><td align=\"left\"><italic>CSPG2</italic></td><td align=\"left\">chondroitin sulfate proteoglycan 2 (versican)</td><td align=\"left\">calcium ion binding (GO:0005509);cell adhesion (GO:0007155);chondroitin sulfate proteoglycan (GO:0005204);development (GO:0007275);glycosaminoglycan binding (GO:0005539);heterophilic cell adhesion (GO:0007157);hyaluronic acid binding (GO:0005540);lectin (GO:0005530);proteoglycan (GO:0005203);sugar binding (GO:0005529)</td></tr><tr><td align=\"left\">9.30</td><td align=\"left\"><italic>SERPINH1</italic></td><td align=\"left\">serine (or cysteine) proteinase inhibitor, clade H (heat shock protein 47), member 1, (collagen binding protein 1)</td><td align=\"left\">collagen binding (GO:0005518);heat shock protein activity (GO:0003773);response to stress (GO:0006950);serine-type endopeptidase inhibitor activity (GO:0004867);serpin (GO:0004868)</td></tr><tr><td align=\"left\">9.23</td><td align=\"left\"><italic>TM4SF6</italic></td><td align=\"left\">transmembrane 4 superfamily member 6</td><td align=\"left\">cell adhesion molecule activity (GO:0005194);cell motility (GO:0006928);positive regulation of I-kappaB kinase/NF-kappaB cascade (GO:0043123);signal transducer activity (GO:0004871)</td></tr><tr><td align=\"left\">9.20</td><td align=\"left\"><italic>PTRF</italic></td><td align=\"left\">polymerase I and transcript release factor</td><td align=\"left\">None</td></tr><tr><td align=\"left\">9.19</td><td align=\"left\"><italic>LATS2</italic></td><td align=\"left\">LATS, large tumor suppressor, homolog 2 (Drosophila)</td><td align=\"left\">ATP binding (GO:0005524);protein amino acid phosphorylation (GO:0006468);protein serine/threonine kinase activity (GO:0004674);transferase activity (GO:0016740);tumor suppressor (GO:0008181)</td></tr><tr><td align=\"left\">9.12</td><td align=\"left\"><italic>CTBP2</italic></td><td align=\"left\">C-terminal binding protein 2</td><td align=\"left\">L-serine biosynthesis (GO:0006564);negative regulation of cell proliferation (GO:0008285);oxidoreductase activity (GO:0016491);oxidoreductase activity, acting on the CH-OH group of donors, NAD or NADP as acceptor (GO:0016616);tumor suppressor (GO:0008181);viral replication (GO:0008166)</td></tr><tr><td align=\"left\">9.06</td><td align=\"left\"><italic>ITGAV</italic></td><td align=\"left\">integrin, alpha V (vitronectin receptor, alpha polypeptide, antigen CD51)</td><td align=\"left\">cell adhesion (GO:0007155);cell adhesion receptor activity (GO:0004895);cell-matrix adhesion (GO:0007160);integrin-mediated signaling pathway (GO:0007229);protein binding (GO:0005515)</td></tr><tr><td align=\"left\">9.05</td><td align=\"left\"><italic>RCN2</italic></td><td align=\"left\">reticulocalbin 2, EF-hand calcium binding domain</td><td align=\"left\">calcium ion binding (GO:0005509);protein binding (GO:0005515);tumor suppressor (GO:0008181)</td></tr><tr><td align=\"left\">8.98</td><td align=\"left\"><italic>CPD</italic></td><td align=\"left\">carboxypeptidase D</td><td align=\"left\">carboxypeptidase A activity (GO:0004182);carboxypeptidase D activity (GO:0004187);carboxypeptidase activity (GO:0004180);hydrolase activity (GO:0016787);proteolysis and peptidolysis (GO:0006508);zinc ion binding (GO:0008270)</td></tr><tr><td align=\"left\">8.98</td><td align=\"left\"><italic>FLJ21174</italic></td><td align=\"left\">hypothetical protein FLJ21174</td><td align=\"left\">None</td></tr><tr><td align=\"left\">8.96</td><td align=\"left\"><italic>MGC5395</italic></td><td align=\"left\">hypothetical protein MGC5395</td><td align=\"left\">intracellular signaling cascade (GO:0007242);neurogenesis (GO:0007399);protein binding (GO:0005515)</td></tr><tr><td align=\"left\">8.96</td><td align=\"left\"><italic>MFAP2</italic></td><td align=\"left\">microfibrillar-associated protein 2</td><td align=\"left\">extracellular matrix glycoprotein (GO:0005207);extracellular matrix structural constituent (GO:0005201)</td></tr><tr><td align=\"left\">8.94</td><td align=\"left\"><italic>CRIM1</italic></td><td align=\"left\">cysteine-rich motor neuron 1</td><td align=\"left\">cysteine-type endopeptidase activity (GO:0004197);insulin-like growth factor binding (GO:0005520);insulin-like growth factor receptor activity (GO:0005010);neurogenesis (GO:0007399);proteolysis and peptidolysis (GO:0006508);regulation of cell growth (GO:0001558);serine-type endopeptidase inhibitor activity (GO:0004867)</td></tr><tr><td align=\"left\">8.92</td><td align=\"left\"><italic>ALDH1A3</italic></td><td align=\"left\">aldehyde dehydrogenase 1 family, member A3</td><td align=\"left\">alcohol metabolism (GO:0006066);aldehyde dehydrogenase [NAD(P)+] activity (GO:0004030);aldehyde dehydrogenase activity (GO:0004028);lipid metabolism (GO:0006629);metabolism (GO:0008152);oxidoreductase activity (GO:0016491)</td></tr><tr><td align=\"left\">8.92</td><td align=\"left\"><italic>MGC34132</italic></td><td align=\"left\">hypothetical protein MGC34132</td><td align=\"left\">None</td></tr></tbody></table></table-wrap><table-wrap id=\"T3\" position=\"float\"><label>Table 3</label><caption><p>PC7 Tissue Partitioning and Ordering. Partitioning and ordering of tissues into sets <italic>UP<sub>7</sub></italic>(<italic>n </italic>= 51) and <italic>DOWN<sub>7</sub></italic>(<italic>n </italic>= 48) sets found to have significant expression differences for <italic>H<sub>7 </sub></italic>and <italic>L<sub>7 </sub></italic>at <italic>test1Thresh </italic>= 0.05. Tissues within groups are ordered by decreasing abs (<italic>mean </italic>(<italic>H<sub>7</sub></italic>)–<italic>mean</italic>(<italic>L<sub>7</sub></italic>)), which has the effect of placing the most significantly affected tissues at the top of each list. The most significant tissues in <italic>UP<sub>7 </sub></italic>are at the left of Figure 2, and the most significant conditions in <italic>DOWN<sub>7 </sub></italic>are at the right of Figure 2.</p></caption><table frame=\"hsides\" rules=\"groups\"><thead><tr><td align=\"left\"><bold>UP7</bold></td><td align=\"left\"><bold>DOWN7</bold></td></tr></thead><tbody><tr><td align=\"left\">SmoothMuscle</td><td align=\"left\">PB-CD19+Bcells</td></tr><tr><td align=\"left\">SmoothMuscle</td><td align=\"left\">PB-CD19+Bcells</td></tr><tr><td align=\"left\">CardiacMyocytes</td><td align=\"left\">PB-BDCA4+Dentritic_Cells</td></tr><tr><td align=\"left\">ADIPOCYTE</td><td align=\"left\">PB-BDCA4+Dentritic_Cells</td></tr><tr><td align=\"left\">ADIPOCYTE</td><td align=\"left\">lymphomaburkittsRaji</td></tr><tr><td align=\"left\">CardiacMyocytes</td><td align=\"left\">lymphomaburkittsDaudi</td></tr><tr><td align=\"left\">TestisGermCell</td><td align=\"left\">lymphomaburkittsDaudi</td></tr><tr><td align=\"left\">bronchialepithelialcells</td><td align=\"left\">lymphomaburkittsRaji</td></tr><tr><td align=\"left\">bronchialepithelialcells</td><td align=\"left\">PB-CD56+NKCells</td></tr><tr><td align=\"left\">PLACENTA</td><td align=\"left\">bonemarrow</td></tr><tr><td align=\"left\">PLACENTA</td><td align=\"left\">bonemarrow</td></tr><tr><td align=\"left\">TestisInterstitial</td><td align=\"left\">PB-CD56+NKCells</td></tr><tr><td align=\"left\">UterusCorpus</td><td align=\"left\">leukemiapromyelocytic(hl60)</td></tr><tr><td align=\"left\">Uterus</td><td align=\"left\">721_B_lymphoblasts</td></tr><tr><td align=\"left\">FetalThyroid</td><td align=\"left\">Tonsil</td></tr><tr><td align=\"left\">OlfactoryBulb</td><td align=\"left\">Tonsil</td></tr><tr><td align=\"left\">TestisLeydigCell</td><td align=\"left\">WHOLEBLOOD</td></tr><tr><td align=\"left\">UterusCorpus</td><td align=\"left\">721_B_lymphoblasts</td></tr><tr><td align=\"left\">FetalThyroid</td><td align=\"left\">leukemiapromyelocytic(hl60)</td></tr><tr><td align=\"left\">atrioventricularnode</td><td align=\"left\">WHOLEBLOOD</td></tr><tr><td align=\"left\">TestisInterstitial</td><td align=\"left\">thymus</td></tr><tr><td align=\"left\">TestisLeydigCell</td><td align=\"left\">PB-CD8+Tcells</td></tr><tr><td align=\"left\">TestisSeminiferousTubule</td><td align=\"left\">PB-CD8+Tcells</td></tr><tr><td align=\"left\">atrioventricularnode</td><td align=\"left\">thymus</td></tr><tr><td align=\"left\">DRG</td><td align=\"left\">lymphnode</td></tr><tr><td align=\"left\">Fetallung</td><td align=\"left\">Heart</td></tr><tr><td align=\"left\">Uterus</td><td align=\"left\">lymphnode</td></tr><tr><td align=\"left\">Ciliaryganglion</td><td align=\"left\">PB-CD14+Monocytes</td></tr><tr><td align=\"left\">TestisGermCell</td><td align=\"left\">PB-CD4+Tcells</td></tr><tr><td align=\"left\">AdrenalCortex</td><td align=\"left\">PB-CD4+Tcells</td></tr><tr><td align=\"left\">Fetallung</td><td align=\"left\">PB-CD14+Monocytes</td></tr><tr><td align=\"left\">Ciliaryganglion</td><td align=\"left\">BM-CD71+EarlyErythroid</td></tr><tr><td align=\"left\">DRG</td><td align=\"left\">BM-CD34+</td></tr><tr><td align=\"left\">TONGUE</td><td align=\"left\">Liver</td></tr><tr><td align=\"left\">SuperiorCervicalGanglion</td><td align=\"left\">Heart</td></tr><tr><td align=\"left\">Ovary</td><td align=\"left\">BM-CD71+EarlyErythroid</td></tr><tr><td align=\"left\">TestisSeminiferousTubule</td><td align=\"left\">BM-CD34+</td></tr><tr><td align=\"left\">Skin</td><td align=\"left\">BM-CD105+Endothelial</td></tr><tr><td align=\"left\">OlfactoryBulb</td><td align=\"left\">salivarygland</td></tr><tr><td align=\"left\">TONGUE</td><td align=\"left\">leukemialymphoblastic(molt4)</td></tr><tr><td align=\"left\">Ovary</td><td align=\"left\">Liver</td></tr><tr><td align=\"left\">TrigeminalGanglion</td><td align=\"left\">Lung</td></tr><tr><td align=\"left\">AdrenalCortex</td><td align=\"left\">Lung</td></tr><tr><td align=\"left\">PancreaticIslets</td><td align=\"left\">leukemialymphoblastic(molt4)</td></tr><tr><td align=\"left\">Skin</td><td align=\"left\">BM-CD105+Endothelial</td></tr><tr><td align=\"left\">Fetalbrain</td><td align=\"left\">BM-CD33+Myeloid</td></tr><tr><td align=\"left\">TrigeminalGanglion</td><td align=\"left\">salivarygland</td></tr><tr><td align=\"left\">Fetalbrain</td><td align=\"left\">BM-CD33+Myeloid</td></tr><tr><td align=\"left\">Amygdala</td><td/></tr><tr><td align=\"left\">SuperiorCervicalGanglion</td><td/></tr><tr><td align=\"left\">PrefrontalCortex</td><td/></tr></tbody></table></table-wrap><table-wrap id=\"T4\" position=\"float\"><label>Table 4</label><caption><p>PC21 Low Extreme Genes. Detailed info for (<italic>n </italic>= 49) extreme genes in PC<sub>21 </sub>set <italic>L<sub>21</sub></italic>, ordered most extreme first, including common name, description, and associated Gene Ontology terms (provided with the GNF dataset). These extreme genes are shown as blue points in Figure 3.</p></caption><table frame=\"hsides\" rules=\"groups\"><thead><tr><td align=\"left\"><bold>PC-21 Value</bold></td><td align=\"left\"><bold><italic>Name</italic></bold></td><td align=\"left\"><bold>Description</bold></td><td align=\"left\"><bold>Function</bold></td></tr></thead><tbody><tr><td align=\"left\">-11.74</td><td align=\"left\"><italic>Ckm</italic></td><td align=\"left\">creatine kinase, muscle</td><td align=\"left\">creatine kinase activity (GO:0004111); transferase activity, transferring phosphorus-containing groups (GO:0016772)</td></tr><tr><td align=\"left\">-11.61</td><td align=\"left\"><italic>ACTA1</italic></td><td align=\"left\">actin, alpha 1, skeletal muscle</td><td align=\"left\">motor activity (GO:0003774);muscle contraction (GO:0006936);muscle development (GO:0007517);structural constituent of cytoskeleton (GO:0005200)</td></tr><tr><td align=\"left\">-10.27</td><td align=\"left\"><italic>MYH7</italic></td><td align=\"left\">myosin, heavy polypeptide 7, cardiac muscle, beta</td><td align=\"left\">ATP binding (GO:0005524);actin binding (GO:0003779);calmodulin binding (GO:0005516);microfilament motor activity (GO:0000146);motor activity (GO:0003774);muscle contraction (GO:0006936);muscle development (GO:0007517);protein amino acid alkylation (GO:0008213);striated muscle contraction (GO:0006941);structural constituent of muscle (GO:0008307)</td></tr><tr><td align=\"left\">-10.01</td><td align=\"left\"><italic>TPM1</italic></td><td align=\"left\">tropomyosin 1 (alpha)</td><td align=\"left\">actin binding (GO:0003779);muscle development (GO:0007517);regulation of heart rate (GO:0008016);structural constituent of cytoskeleton (GO:0005200);structural constituent of muscle (GO:0008307)</td></tr><tr><td align=\"left\">-9.10</td><td align=\"left\"><italic>MYL1</italic></td><td align=\"left\">myosin, light polypeptide 1, alkali; skeletal, fast</td><td align=\"left\">calcium ion binding (GO:0005509);muscle development (GO:0007517);structural constituent of muscle (GO:0008307)</td></tr><tr><td align=\"left\">-9.10</td><td align=\"left\"><italic>TNNC1</italic></td><td align=\"left\">troponin C, slow</td><td align=\"left\">calcium ion binding (GO:0005509);muscle development (GO:0007517)</td></tr><tr><td align=\"left\">-8.92</td><td align=\"left\"><italic>PPP1R1A</italic></td><td align=\"left\">protein phosphatase 1, regulatory (inhibitor) subunit 1A</td><td align=\"left\">glycogen metabolism (GO:0005977);protein phosphatase inhibitor activity (GO:0004864);signal transduction (GO:0007165);type 1 serine/threonine specific protein phosphatase inhibitor activity (GO:0004865)</td></tr><tr><td align=\"left\">-8.79</td><td align=\"left\"><italic>TNNC2</italic></td><td align=\"left\">troponin C2, fast</td><td align=\"left\">calcium ion binding (GO:0005509);muscle development (GO:0007517)</td></tr><tr><td align=\"left\">-8.71</td><td align=\"left\"><italic>TNNT1</italic></td><td align=\"left\">troponin T1, skeletal, slow</td><td align=\"left\">muscle development (GO:0007517);tropomyosin binding (GO:0005523)</td></tr><tr><td align=\"left\">-8.70</td><td align=\"left\"><italic>KRT14</italic></td><td align=\"left\">keratin 14 (epidermolysis bullosa simplex, Dowling-Meara, Koebner)</td><td align=\"left\">biological_process unknown (GO:0000004);structural constituent of cytoskeleton (GO:0005200);structural constituent of epidermis (GO:0030280)</td></tr><tr><td align=\"left\">-8.67</td><td align=\"left\"><italic>TTID</italic></td><td align=\"left\">titin immunoglobulin domain protein (myotilin)</td><td align=\"left\">muscle contraction (GO:0006936);protein binding- (GO:0005515);structural constituent of muscle (GO:0008307)</td></tr><tr><td align=\"left\">-8.49</td><td align=\"left\"><italic>HUMMLC2B</italic></td><td align=\"left\">myosin light chain 2</td><td align=\"left\">calcium ion binding (GO:0005509);structural constituent of muscle (GO:0008307)</td></tr><tr><td align=\"left\">-8.35</td><td align=\"left\"><italic>TRIM</italic></td><td align=\"left\">T-cell receptor interacting molecule</td><td align=\"left\">DNA binding (GO:0003677);cellular defense response (GO:0006968);signal transduction (GO:0007165);transmembrane receptor protein tyrosine kinase adaptor protein activity (GO:0005068)</td></tr><tr><td align=\"left\">-8.20</td><td align=\"left\"><italic>TTN</italic></td><td align=\"left\">titin</td><td align=\"left\">ATP binding (GO:0005524);calmodulin binding (GO:0005516);cation transporter activity (GO:0008324);hematopoietin/interferon-class (D200-domain) cytokine receptor activity (GO:0004896);muscle development (GO:0007517);myosin binding (GO:0017022);protein amino acid phosphorylation (GO:0006468);protein serine/threonine kinase activity (GO:0004674);regulation of actin filament length (GO:0030832);somatic muscle development (GO:0007525);striated muscle contraction (GO:0006941);structural constituent of muscle (GO:0008307);structural molecule activity (GO:0005198);transferase activity (GO:0016740)</td></tr><tr><td align=\"left\">-7.91</td><td align=\"left\"><italic>CSRP3</italic></td><td align=\"left\">cysteine and glycine-rich protein 3 (cardiac LIM protein)</td><td align=\"left\">None</td></tr><tr><td align=\"left\">-7.66</td><td align=\"left\"><italic>MB</italic></td><td align=\"left\">myoglobin</td><td align=\"left\">electron transporter activity (GO:0005489);globin (GO:0001524);oxygen transport (GO:0015671);oxygen transporter activity (GO:0005344);transport (GO:0006810)</td></tr><tr><td align=\"left\">-7.46</td><td align=\"left\"><italic>ENO3</italic></td><td align=\"left\">enolase 3, (beta, muscle)</td><td align=\"left\">glycolysis (GO:0006096);lyase activity (GO:0016829);magnesium ion binding (GO:0000287);phosphopyruvate hydratase activity (GO:0004634)</td></tr><tr><td align=\"left\">-7.32</td><td align=\"left\"><italic>NTRK2</italic></td><td align=\"left\">neurotrophic tyrosine kinase, receptor, type 2</td><td align=\"left\">ATP binding (GO:0005524);kinase activity (GO:0016301);neurogenesis (GO:0007399);neurotrophin TRKB receptor activity (GO:0005015);neurotrophin binding (GO:0043121);protein amino acid phosphorylation (GO:0006468);receptor activity (GO:0004872);transferase activity (GO:0016740);transmembrane receptor protein tyrosine kinase activity (GO:0004714);transmembrane receptor protein tyrosine kinase signaling pathway (GO:0007169)</td></tr><tr><td align=\"left\">-7.28</td><td align=\"left\"><italic>MYL2</italic></td><td align=\"left\">myosin, light polypeptide 2, regulatory, cardiac, slow</td><td align=\"left\">calcium ion binding (GO:0005509);muscle development (GO:0007517);structural constituent of muscle (GO:0008307)</td></tr><tr><td align=\"left\">-7.27</td><td align=\"left\"><italic>TF</italic></td><td align=\"left\">Transferring</td><td align=\"left\">ferric iron binding (GO:0008199);iron ion binding (GO:0005506);iron ion homeostasis (GO:0006879);iron ion transport (GO:0006826);transport (GO:0006810)</td></tr><tr><td align=\"left\">-7.16</td><td align=\"left\"><italic>ACSL1</italic></td><td align=\"left\">acyl-CoA synthetase long-chain family member 1</td><td align=\"left\">digestion (GO:0007586);fatty acid metabolism (GO:0006631);ligase activity (GO:0016874);long-chain-fatty-acid-CoA ligase activity (GO:0004467);magnesium ion binding (GO:0000287);metabolism (GO:0008152)</td></tr><tr><td align=\"left\">-7.10</td><td align=\"left\"><italic>MYBPC1</italic></td><td align=\"left\">myosin binding protein C, slow type</td><td align=\"left\">actin binding (GO:0003779);cell adhesion (GO:0007155);cell adhesion molecule activity (GO:0005194);muscle development (GO:0007517);protein binding (GO:0005515);striated muscle contraction (GO:0006941);structural constituent of muscle (GO:0008307)</td></tr><tr><td align=\"left\">-7.07</td><td align=\"left\"><italic>S100A2</italic></td><td align=\"left\">S100 calcium binding protein A2</td><td align=\"left\">biological_process unknown (GO:0000004);calcium ion binding (GO:0005509)</td></tr><tr><td align=\"left\">-7.07</td><td align=\"left\"><italic>PYGM</italic></td><td align=\"left\">phosphorylase, glycogen; muscle (McArdle syndrome, glycogen storage disease type V)</td><td align=\"left\">amino acid metabolism (GO:0006520);carbohydrate metabolism (GO:0005975);glycogen metabolism (GO:0005977);glycogen phosphorylase activity (GO:0008184);transferase activity, transferring glycosyl groups (GO:0016757)</td></tr><tr><td align=\"left\">-6.99</td><td align=\"left\"><italic>ACTN2</italic></td><td align=\"left\">actinin, alpha 2</td><td align=\"left\">actin binding (GO:0003779);calcium ion binding (GO:0005509);protein binding (GO:0005515);structural constituent of muscle (GO:0008307)</td></tr><tr><td align=\"left\">-6.76</td><td align=\"left\"><italic>MYBPC2</italic></td><td align=\"left\">myosin binding protein C, fast type</td><td align=\"left\">actin binding (GO:0003779);cell adhesion (GO:0007155);cell adhesion molecule activity (GO:0005194);muscle development (GO:0007517);protein binding (GO:0005515);striated muscle contraction (GO:0006941);structural constituent of muscle (GO:0008307)</td></tr><tr><td align=\"left\">-6.71</td><td align=\"left\"><italic>UCP2</italic></td><td align=\"left\">uncoupling protein 2 (mitochondrial, proton carrier)</td><td align=\"left\">binding (GO:0005488);mitochondrial transport (GO:0006839);proton transport (GO:0015992);small molecule transport (GO:0006832);transport (GO:0006810);transporter activity (GO:0005215);uncoupling protein activity (GO:0015302)</td></tr><tr><td align=\"left\">-6.63</td><td align=\"left\"><italic>NEB</italic></td><td align=\"left\">nebulin</td><td align=\"left\">actin binding (GO:0003779);muscle development (GO:0007517);regulation of actin filament length (GO:0030832);somatic muscle development (GO:0007525);structural constituent of muscle (GO:0008307)</td></tr><tr><td align=\"left\">-6.60</td><td align=\"left\"><italic>IDH2</italic></td><td align=\"left\">isocitrate dehydrogenase 2 (NADP+), mitochondrial</td><td align=\"left\">carbohydrate metabolism (GO:0005975);glyoxylate cycle (GO:0006097);isocitrate dehydrogenase (NADP+) activity (GO:0004450);main pathways of carbohydrate metabolism (GO:0006092);metabolism (GO:0008152);oxidoreductase activity (GO:0016491);tricarboxylic acid cycle (GO:0006099)</td></tr><tr><td align=\"left\">-6.47</td><td align=\"left\"><italic>CSTA</italic></td><td align=\"left\">cystatin A (stefin A)</td><td align=\"left\">cysteine protease inhibitor activity (GO:0004869);endopeptidase inhibitor activity (GO:0004866)</td></tr><tr><td align=\"left\">-6.45</td><td align=\"left\"><italic>MYH2</italic></td><td align=\"left\">myosin, heavy polypeptide 2, skeletal muscle, adult</td><td align=\"left\">ATP binding (GO:0005524);actin binding (GO:0003779);calmodulin binding (GO:0005516);microfilament motor activity (GO:0000146);motor activity (GO:0003774);muscle contraction (GO:0006936);muscle development (GO:0007517);muscle motor activity (GO:0003776);protein amino acid alkylation (GO:0008213);striated muscle contraction (GO:0006941)</td></tr><tr><td align=\"left\">-6.42</td><td align=\"left\"><italic>KRT4</italic></td><td align=\"left\">keratin 4</td><td align=\"left\">cytoskeleton organization and biogenesis (GO:0007010);structural constituent of cytoskeleton (GO:0005200);structural molecule activity (GO:0005198)</td></tr><tr><td align=\"left\">-6.40</td><td align=\"left\"><italic>CLIC6</italic></td><td align=\"left\">chloride intracellular channel 6</td><td align=\"left\">chloride transport (GO:0006821);ion transport (GO:0006811);voltage-gated chloride channel activity (GO:0005247)</td></tr><tr><td align=\"left\">-6.27</td><td align=\"left\"><italic>SPRR1A</italic></td><td align=\"left\">small proline-rich protein 1A</td><td align=\"left\">structural molecule activity (GO:0005198)</td></tr><tr><td align=\"left\">-6.25</td><td align=\"left\"><italic>TNNI2</italic></td><td align=\"left\">troponin I, skeletal, fast</td><td align=\"left\">actin binding (GO:0003779);muscle development (GO:0007517)</td></tr><tr><td align=\"left\">-6.22</td><td align=\"left\"><italic>TNNI1</italic></td><td align=\"left\">troponin I, skeletal, slow</td><td align=\"left\">actin binding (GO:0003779);muscle development (GO:0007517);tropomyosin binding (GO:0005523)</td></tr><tr><td align=\"left\">-6.17</td><td align=\"left\"><italic>LDB3</italic></td><td align=\"left\">LIM domain binding 3</td><td align=\"left\">electron transport (GO:0006118);electron transporter activity (GO:0005489);protein binding (GO:0005515)</td></tr><tr><td align=\"left\">-6.14</td><td align=\"left\"><italic>CRYAB</italic></td><td align=\"left\">crystallin, alpha B</td><td align=\"left\">chaperone activity (GO:0003754);muscle contraction (GO:0006936);protein folding (GO:0006457);structural constituent of eye lens (GO:0005212);visual perception (GO:0007601)</td></tr><tr><td align=\"left\">-6.12</td><td align=\"left\"><italic>HFL1</italic></td><td align=\"left\">H factor (complement)-like 1</td><td align=\"left\">plasma protein (GO:0005209)</td></tr><tr><td align=\"left\">-6.15</td><td align=\"left\"><italic>S100A1</italic></td><td align=\"left\">S100 calcium binding protein A1</td><td align=\"left\">calcium ion binding (GO:0005509);cell communication (GO:0007154);intracellular signaling cascade (GO:0007242);neurogenesis (GO:0007399);protein binding (GO:0005515);zinc ion binding (GO:0008270)</td></tr><tr><td align=\"left\">-6.09</td><td align=\"left\"><italic>HSPB8</italic></td><td align=\"left\">heat shock 22 kDa protein 8</td><td align=\"left\">biological_process unknown (GO:0000004);heat shock protein activity (GO:0003773);protein serine/threonine kinase activity (GO:0004674);transferase activity (GO:0016740)</td></tr><tr><td align=\"left\">-5.93</td><td align=\"left\"><italic>PPFIBP2</italic></td><td align=\"left\">PTPRF interacting protein, binding protein 2 (liprin beta 2)</td><td align=\"left\">None</td></tr><tr><td align=\"left\">-5.88</td><td align=\"left\"><italic>GHR</italic></td><td align=\"left\">growth hormone receptor</td><td align=\"left\">endocytosis (GO:0006897);growth hormone receptor activity (GO:0004903);growth pattern (GO:0007150);receptor activity (GO:0004872);skeletal development (GO:0001501)</td></tr><tr><td align=\"left\">-5.85</td><td align=\"left\"><italic>ACSL1</italic></td><td align=\"left\">acyl-CoA synthetase long-chain family member 1</td><td align=\"left\">digestion (GO:0007586);fatty acid metabolism (GO:0006631);ligase activity (GO:0016874);long-chain-fatty-acid-CoA ligase activity (GO:0004467);magnesium ion binding (GO:0000287);metabolism (GO:0008152)</td></tr><tr><td align=\"left\">-5.84</td><td align=\"left\"><italic>MGC42493</italic></td><td align=\"left\">hypothetical protein MGC42493</td><td align=\"left\">nucleic acid binding (GO:0003676);zinc ion binding (GO:0008270)</td></tr><tr><td align=\"left\">-5.73</td><td align=\"left\"><italic>SCD</italic></td><td align=\"left\">stearoyl-CoA desaturase (delta-9-desaturase)</td><td align=\"left\">fatty acid biosynthesis (GO:0006633);iron ion binding (GO:0005506);oxidoreductase activity (GO:0016491);stearoyl-CoA 9-desaturase activity (GO:0004768)</td></tr><tr><td align=\"left\">-5.67</td><td align=\"left\"><italic>hIAN2</italic></td><td align=\"left\">human immune associated nucleotide 2</td><td align=\"left\">GTP binding (GO:0005525)</td></tr><tr><td align=\"left\">-5.65</td><td align=\"left\"><italic>PDE4DIP</italic></td><td align=\"left\">phosphodiesterase 4D interacting protein (myomegalin)</td><td align=\"left\">None</td></tr><tr><td align=\"left\">-5.63</td><td align=\"left\"><italic>HBA2</italic></td><td align=\"left\">hemoglobin, alpha 2</td><td align=\"left\">oxygen transport (GO:0015671);oxygen transporter activity (GO:0005344);protein binding (GO:0005515);transport (GO:0006810)</td></tr></tbody></table></table-wrap><table-wrap id=\"T5\" position=\"float\"><label>Table 5</label><caption><p>PC21 Tissue Partitioning and Ordering. Partitioning and ordering of tissues into sets <italic>UP<sub>21</sub></italic>(<italic>n </italic>= 8) and <italic>DOWN<sub>21 </sub></italic>(<italic>n </italic>= 20) sets found to have significant expression differences for <italic>H<sub>21 </sub></italic>and <italic>L<sub>21 </sub></italic>at <italic>test1Thresh </italic>= 0.05. Tissues within groups are ordered by decreasing abs(<italic>mean </italic>(<italic>H<sub>21</sub></italic>)–<italic>mean</italic>(<italic>L<sub>21</sub></italic>)), which has the effect of placing the most significantly affected tissues at the top of each list. The most significant tissues in <italic>UP<sub>21 </sub></italic>are at the left of Figure 4, and the most significant conditions in <italic>DOWN<sub>21 </sub></italic>are at the right of Figure 4.</p></caption><table frame=\"hsides\" rules=\"groups\"><thead><tr><td align=\"left\"><bold>UP21</bold></td><td align=\"left\"><bold>DOWN21</bold></td></tr></thead><tbody><tr><td align=\"left\">Pancreas</td><td align=\"left\">TONGUE</td></tr><tr><td align=\"left\">Pancreas</td><td align=\"left\">SkeletalMuscle</td></tr><tr><td align=\"left\">PancreaticIslets</td><td align=\"left\">TONGUE</td></tr><tr><td align=\"left\">PancreaticIslets</td><td align=\"left\">SkeletalMuscle</td></tr><tr><td align=\"left\">Bonemarrow</td><td align=\"left\">Thyroid</td></tr><tr><td align=\"left\">Bonemarrow</td><td align=\"left\">Heart</td></tr><tr><td align=\"left\">BM-CD34+</td><td align=\"left\">Thyroid</td></tr><tr><td align=\"left\">BM-CD34+</td><td align=\"left\">FetalThyroid</td></tr><tr><td/><td align=\"left\">Heart</td></tr><tr><td/><td align=\"left\">FetalThyroid</td></tr><tr><td/><td align=\"left\">Testis</td></tr><tr><td/><td align=\"left\">721_B_lymphoblasts</td></tr><tr><td/><td align=\"left\">Testis</td></tr><tr><td/><td align=\"left\">ADIPOCYTE</td></tr><tr><td/><td align=\"left\">721_B_lymphoblasts</td></tr><tr><td/><td align=\"left\">ADIPOCYTE</td></tr><tr><td/><td align=\"left\">Thalamus</td></tr><tr><td/><td align=\"left\">Caudatenucleus</td></tr><tr><td/><td align=\"left\">Spinalcord</td></tr><tr><td/><td align=\"left\">Prostate</td></tr></tbody></table></table-wrap><table-wrap id=\"T6\" position=\"float\"><label>Table 6</label><caption><p>GNF human expression principal components having extreme gene sets enriched in GO categories. Each of the GNF human expression PCEG sets, derived using an <italic>extremeThresh </italic>of 0.00001, was tested for Gene Ontology (GO) statistical enrichment. Terms that were enriched for a particular PC at 1% significance threshold are reported. 26 of the top 42 principal components produced extreme gene lists having significant GO term enrichment.</p></caption><table frame=\"hsides\" rules=\"groups\"><thead><tr><td align=\"left\"><bold>PCEG Set</bold></td><td align=\"left\"><bold>GO Description (p-value)</bold></td></tr></thead><tbody><tr><td align=\"left\">01 low</td><td align=\"left\">GO:0006412 protein biosynthesis (4.51e-07)</td></tr><tr><td align=\"left\">03 high</td><td align=\"left\">GO:0005201 extracellular matrix structural constituent (1.43e-08)</td></tr><tr><td align=\"left\">03 low</td><td align=\"left\">GO:0007399 neurogenesis (5.05e-09)<break/>GO:0007268 synaptic transmission (9.6e-09)<break/>GO:0007417 central nervous system development (1.31e-08)<break/>GO:0016820 hydrolase activity, acting on acid anhydrides, catalyzing transmembrane movement of substances (4.85e-07)</td></tr><tr><td align=\"left\">04 low</td><td align=\"left\">GO:0005201 extracellular matrix structural constituent (5.82e-07)</td></tr><tr><td align=\"left\">05 high</td><td align=\"left\">GO:0007067 mitosis (3.43e-17)<break/>GO:0051301 cell division (4.45e-17)<break/>GO:0048015 phosphoinositide-mediated signaling (5.07e-13)<break/>GO:0007049 cell cycle (1.79e-09)<break/>GO:0006260 DNA replication (5.37e-08)<break/>GO:0004674 protein serine/threonine kinase activity (1.09e-07)<break/>GO:0005524 ATP binding (1.29e-07)<break/>GO:0005634 nucleus (1.93e-07)<break/>GO:0006468 protein amino acid phosphorylation (1.01e-06)</td></tr><tr><td align=\"left\">05 low</td><td align=\"left\">GO:0045012 MHC class II receptor activity (8.59e-08)<break/>GO:0006968 cellular defense response (5.07e-07)<break/>GO:0006955 immune response (6.9e-07)</td></tr><tr><td align=\"left\">07 high</td><td align=\"left\">GO:0005201 extracellular matrix structural constituent (3.27e-20)<break/>GO:0005578 extracellular matrix (sensu Metazoa) (2.94e-15)<break/>GO:0005581 collagen (1.76e-13)<break/>GO:0006817 phosphate transport (2.79e-12)<break/>GO:0007517 muscle development (8.27e-10)<break/>GO:0005518 collagen binding (1.1e-07)<break/>GO:0008201 heparin binding (8.6e-07)<break/>GO:0007155 cell adhesion (1.55e-06)<break/>GO:0005509 calcium ion binding (2.16e-06)</td></tr><tr><td align=\"left\">08 high</td><td align=\"left\">GO:0007283 spermatogenesis (3.91e-10)</td></tr><tr><td align=\"left\">11 low</td><td align=\"left\">GO:0042742 defense response to bacteria (1.09e-06)</td></tr><tr><td align=\"left\">13 high</td><td align=\"left\">GO:0019884 antigen presentation, exogenous antigen (3.23e-16)<break/>GO:0019886 antigen processing, exogenous antigen via MHC class II (5.1e-16)<break/>GO:0045012 MHC class II receptor activity (5.1e-16)<break/>GO:0006955 immune response (1.28e-15)</td></tr><tr><td align=\"left\">13 low</td><td align=\"left\">GO:0006955 immune response (9.9e-16)<break/>GO:0042110 T cell activation (1.15e-10)<break/>GO:0007166 cell surface receptor linked signal transduction (1.64e-10)<break/>GO:0004888 transmembrane receptor activity (1.25e-09)<break/>GO:0006968 cellular defense response (2.05e-07)<break/>GO:0019735 antimicrobial humoral response (sensu Vertebrata) (4.81e-07)<break/>GO:0005515 protein binding (1.4e-06)</td></tr><tr><td align=\"left\">14 low</td><td align=\"left\">GO:0005615 extracellular space (4.94e-24)<break/>GO:0005319 lipid transporter activity (5.77e-11)<break/>GO:0007596 blood coagulation (6.05e-11)<break/>GO:0006953 acute-phase response (1.47e-10)<break/>GO:0004867 serine-type endopeptidase inhibitor activity (6.14e-10)<break/>GO:0004263 chymotrypsin activity (4.83e-09)<break/>GO:0004295 trypsin activity (7.86e-09)<break/>GO:0006810 transport (1.71e-08)<break/>GO:0016042 lipid catabolism (6.48e-08)<break/>GO:0005576 extracellular region (8.91e-08)<break/>GO:0008201 heparin binding (1.05e-07)<break/>GO:0006508 proteolysis and peptidolysis (4e-07)<break/>GO:0006869 lipid transport (1.14e-06)</td></tr><tr><td align=\"left\">15 high</td><td align=\"left\">GO:0005200 structural constituent of cytoskeleton (3.17e-12)<break/>GO:0005882 intermediate filament (1.99e-10)<break/>GO:0008544 epidermis development (4.44e-10)<break/>GO:0007517 muscle development (9.24e-10)</td></tr><tr><td align=\"left\">15 low</td><td align=\"left\">GO:0008201 heparin binding (1.71e-07)<break/>GO:0005201 extracellular matrix structural constituent (3.95e-07)</td></tr><tr><td align=\"left\">16 low</td><td align=\"left\">GO:0005179 hormone activity (4e-10)<break/>GO:0005576 extracellular region (1.51e-07)<break/>GO:0007565 pregnancy (1.57e-07)</td></tr><tr><td align=\"left\">17 high</td><td align=\"left\">GO:0006955 immune response (6.06e-07)</td></tr><tr><td align=\"left\">18 high</td><td align=\"left\">GO:0005882 intermediate filament (1.47e-11)<break/>GO:0005200 structural constituent of cytoskeleton (5.67e-10)<break/>GO:0005615 extracellular space (2.74e-08)<break/>GO:0008544 epidermis development (1.75e-07)<break/>GO:0005319 lipid transporter activity (9.61e-07)<break/>GO:0005198 structural molecule activity (1.72e-06)</td></tr><tr><td align=\"left\">19 high</td><td align=\"left\">GO:0005200 structural constituent of cytoskeleton (2.05e-07)</td></tr><tr><td align=\"left\">20 high</td><td align=\"left\">GO:0005179 hormone activity (1.88e-08)</td></tr><tr><td align=\"left\">21 low</td><td align=\"left\">GO:0007517 muscle development (7.01e-27)<break/>GO:0008307 structural constituent of muscle (4.15e-23)<break/>GO:0003779 actin binding (6.2e-14)<break/>GO:0006941 striated muscle contraction (5.69e-10)<break/>GO:0005859 muscle myosin (1.99e-08)</td></tr><tr><td align=\"left\">22 high</td><td align=\"left\">GO:0016042 lipid catabolism (9.41e-08)<break/>GO:0004263 chymotrypsin activity (1.94e-06)</td></tr><tr><td align=\"left\">23 high</td><td align=\"left\">GO:0019825 oxygen binding (1.9e-08)<break/>GO:0005344 oxygen transporter activity (2.67e-07)<break/>GO:0015671 oxygen transport (3.8e-07)</td></tr><tr><td align=\"left\">24 low</td><td align=\"left\">GO:0007517 muscle development (8.02e-15)<break/>GO:0008307 structural constituent of muscle (2.11e-12)<break/>GO:0003779 actin binding (3.61e-09)<break/>GO:0005319 lipid transporter actiGO:0005856 cytoskeleton (1e-06)</td></tr><tr><td align=\"left\">26 high</td><td align=\"left\">GO:0005576 extracellular region (3.7e-10)<break/>GO:0005319 lipid transporter activity (2.31e-08)<break/>GO:0006869 lipid transport (9.09e-08)</td></tr><tr><td align=\"left\">26 low</td><td align=\"left\">GO:0008307 structural constituent of muscle (1.19e-08)<break/>GO:0005344 oxygen transporter activity (2.67e-07)<break/>GO:0015671 oxygen transport (3.8e-07)</td></tr><tr><td align=\"left\">29 high</td><td align=\"left\">GO:0005615 extracellular space (1.8e-06)</td></tr><tr><td align=\"left\">32 low</td><td align=\"left\">GO:0005576 extracellular region (3.34e-15)<break/>GO:0005615 extracellular space (6.56e-07)<break/>GO:0005179 hormone activity (2.15e-06)</td></tr><tr><td align=\"left\">35 low</td><td align=\"left\">GO:0007585 respiratory gaseous exchange (4e-09)</td></tr><tr><td align=\"left\">39 high</td><td align=\"left\">GO:0005576 extracellular region (2.35e-07)</td></tr><tr><td align=\"left\">40 high</td><td align=\"left\">GO:0005179 hormone activity (8.22e-08)<break/>GO:0005576 extracellular region (9.39e-07)</td></tr><tr><td align=\"left\">42 high</td><td align=\"left\">GO:0005179 hormone activity (8.22e-08)<break/>GO:0005576 extracellular region (9.39e-07)</td></tr></tbody></table></table-wrap><table-wrap id=\"T7\" position=\"float\"><label>Table 7</label><caption><p>Diabetes dataset covariates identified as correlated with principal components. Results of a search for covariates with value distributions having significant differences in either <italic>UP</italic><sub><italic>n </italic></sub>vs. <italic>DOWN</italic><sub><italic>n</italic></sub>, {<italic>UP</italic><sub><italic>n</italic></sub>+<italic>FLAT</italic><sub><italic>n</italic></sub>} vs. <italic>DOWN</italic><sub><italic>n</italic></sub>, or <italic>UP</italic><sub><italic>n </italic></sub>vs. {<italic>FLAT</italic><sub><italic>n</italic></sub>+<italic>DOWN</italic><sub><italic>n</italic></sub>} are shown below. An \"X\" indicates the covariate's values varied significantly in the corresponding principal component's sample partitioning. 11 (out of 53) covariates were identified at a 1% significance level (<italic>minSetSize </italic>= 5) in 10 of the first 20 principal components. To investigate further, corresponding plots of covariate value distributions within sample partitions can be generated, such as the covariate plots for PC<sub>14 </sub>(see Figure 5).</p></caption><table frame=\"hsides\" rules=\"groups\"><thead><tr><td/><td align=\"left\" colspan=\"20\">Principal Component Number</td></tr><tr><td/><td colspan=\"20\"><hr/></td></tr><tr><td/><td align=\"left\">1</td><td align=\"left\">2</td><td align=\"left\">3</td><td align=\"left\">4</td><td align=\"left\">5</td><td align=\"left\">6</td><td align=\"left\">7</td><td align=\"left\">8</td><td align=\"center\">9</td><td align=\"center\">10</td><td align=\"center\">11</td><td align=\"center\">12</td><td align=\"center\">13</td><td align=\"center\">14</td><td align=\"center\">15</td><td align=\"center\">16</td><td align=\"center\">17</td><td align=\"center\">18</td><td align=\"center\">19</td><td align=\"center\">20</td></tr></thead><tbody><tr><td align=\"left\">Cap_(mm2)</td><td/><td align=\"left\">X</td><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/></tr><tr><td align=\"left\">Cap__Type2b_(mean_n)</td><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td align=\"center\">X</td><td/><td/><td/><td/><td/><td/><td/><td/><td/></tr><tr><td align=\"left\">Centroid_(Using_34_OXPHOS-CR_Genes)</td><td/><td/><td/><td/><td/><td/><td/><td align=\"left\">X</td><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/></tr><tr><td align=\"left\">Insulin_0</td><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td align=\"center\">X</td><td/><td/><td/><td/><td/><td/></tr><tr><td align=\"left\">Patient_#</td><td/><td/><td/><td align=\"left\">X</td><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/></tr><tr><td align=\"left\">Type1_Min_Area_(um2)</td><td/><td/><td/><td/><td/><td/><td/><td/><td align=\"center\">X</td><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/></tr><tr><td align=\"left\">Type2a_(n)</td><td/><td/><td/><td/><td/><td/><td align=\"left\">X</td><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/></tr><tr><td align=\"left\">Type2a_Min_Area_(um2)</td><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td align=\"center\">X</td><td/><td/><td/><td/></tr><tr><td align=\"left\">Type2b_(%)</td><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td align=\"center\">X</td><td/><td/><td/><td/><td/><td/></tr><tr><td align=\"left\">Type2b_(n)</td><td/><td/><td align=\"left\">X</td><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/></tr><tr><td align=\"left\">UQCRB_(209065_at)</td><td/><td/><td/><td/><td/><td align=\"left\">X</td><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/></tr></tbody></table></table-wrap></sec></back></article>"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "soup = BeautifulSoup(processed_xml[0], 'xml')\n",
    "soup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b995d460",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Background',\n",
       " 'Results',\n",
       " 'Conclusion',\n",
       " 'Background',\n",
       " 'Implementation',\n",
       " 'Results',\n",
       " 'Application to microarray expression data, Case 1: GNF human data',\n",
       " 'Application to microarray expression data, case 2: human diabetes',\n",
       " 'Discussion',\n",
       " 'Conclusion',\n",
       " 'Availability and requirements',\n",
       " 'Methodology',\n",
       " 'Determine the principal components of the dataset',\n",
       " 'Identify extreme gene probes for each principal component',\n",
       " 'Identify significant conditions for each principal component',\n",
       " 'Interpret each principal component using covariate annotations',\n",
       " 'Terminating condition',\n",
       " \"Authors' contributions\",\n",
       " 'Supplementary Material',\n",
       " 'Acknowledgements',\n",
       " 'Figures and Tables']"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dd = extract_section_titles(soup)\n",
    "dd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4abfd34e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Introduction',\n",
       " 'Structural heterogeneity related to the function of PcG proteins',\n",
       " 'The role of PcG proteins in immune regulation',\n",
       " 'Current RNA-seq and chromatin immunoprecipitation (ChIP)-seq data for identifying the target loci of PcG proteins',\n",
       " 'Therapeutic agents for treating hematopoietic malignancies by inhibiting the activity of PcG proteins',\n",
       " 'Conclusion and future prospects']"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dd = extract_section_titles(soup)\n",
    "dd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "6eec3c15",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████| 1/1 [00:45<00:00, 45.05s/it]\n"
     ]
    }
   ],
   "source": [
    "sent_xml = process_each_article(processed_xml[0], document_flag='f')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e044a3aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['INTRO', 'CONCL', 'METHODS'])"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dd.keys()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
